// @adobe/aio-commerce-lib-app/dist/cjs/management/index.cjs
/**
 * @license
 *
 * Copyright 2026 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
const require_management = require('../management-Dehz9qSJ.cjs');

exports.createInitialInstallationState = require_management.createInitialInstallationState;
exports.createInitialUninstallationState = require_management.createInitialUninstallationState;
exports.defineCustomInstallationStep = require_management.defineCustomInstallationStep;
exports.isCompletedState = require_management.isCompletedState;
exports.isFailedState = require_management.isFailedState;
exports.isInProgressState = require_management.isInProgressState;
exports.isSucceededState = require_management.isSucceededState;
exports.runInstallation = require_management.runInstallation;
exports.runUninstallation = require_management.runUninstallation;
exports.runValidation = require_management.runValidation;
