// @adobe/aio-commerce-lib-app/dist/cjs/validate-oAOpyFyb.cjs
/**
 * @license
 *
 * Copyright 2026 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */

const require_chunk = require('./chunk-Cozy-1ZF.cjs');
const require_schemas = require('./schemas-FngDuIeo.cjs');
const require_webhooks = require('./webhooks-tbZy9G3p.cjs');
let _adobe_aio_commerce_lib_core_error = require("@adobe/aio-commerce-lib-core/error");
let _adobe_aio_commerce_lib_config = require("@adobe/aio-commerce-lib-config");
let valibot = require("valibot");
valibot = require_chunk.__toESM(valibot, 1);

//#region source/config/schema/business-configuration.ts
/**
* Check if config has business config.
* @param config - The configuration to check.
*/
function hasBusinessConfig(config) {
	return config.businessConfig !== void 0;
}
/**
* Check if config has business config schema.
* @param config - The configuration to check.
*/
function hasBusinessConfigSchema(config) {
	return config.businessConfig?.schema !== void 0 && config.businessConfig.schema.length > 0;
}

//#endregion
//#region source/config/schema/metadata.ts
const MAX_ID_LENGTH = 100;
const MAX_DESCRIPTION_LENGTH = 255;
const MAX_DISPLAY_NAME_LENGTH = 50;
const NUMERIC_IDENTIFIER = "(0|[1-9]\\d*)";
const SEMVER_REGEX = new RegExp(`^${NUMERIC_IDENTIFIER}\\.${NUMERIC_IDENTIFIER}\\.${NUMERIC_IDENTIFIER}$`);
function nonEmptyString(fieldName) {
	return valibot.pipe(valibot.string(`Expected a string for the ${fieldName}`), valibot.nonEmpty(`The ${fieldName} must not be empty`));
}
/** The schema for the metadata of the application. */
const MetadataSchema = valibot.object({
	id: valibot.pipe(require_schemas.alphaNumericOrHyphenSchema("application id (metadata.id)"), valibot.maxLength(MAX_ID_LENGTH, `The application id must not be longer than ${MAX_ID_LENGTH} characters`)),
	displayName: valibot.pipe(nonEmptyString("application display name"), valibot.maxLength(MAX_DISPLAY_NAME_LENGTH, `The application display name must not be longer than ${MAX_DISPLAY_NAME_LENGTH} characters`)),
	description: valibot.pipe(nonEmptyString("metadata description"), valibot.maxLength(MAX_DESCRIPTION_LENGTH, `The metadata description must not be longer than ${MAX_DESCRIPTION_LENGTH} characters`)),
	version: valibot.pipe(nonEmptyString("version"), valibot.regex(SEMVER_REGEX, "The version must follow semantic versioning (semver) format: Major.Minor.Patch (e.g., '1.0.0', '2.3.1')"))
});
/**
* Check if config has metadata.
* @param config - The configuration to check.
*/
function hasMetadata(config) {
	return config.metadata !== void 0;
}

//#endregion
//#region source/config/schema/domains.ts
/** The individual validatable domains of the app config. */
const CommerceAppConfigSchemas = {
	metadata: MetadataSchema,
	businessConfig: _adobe_aio_commerce_lib_config.SchemaBusinessConfig,
	eventing: require_webhooks.EventingSchema,
	installation: require_webhooks.InstallationSchema,
	webhooks: require_webhooks.WebhooksSchema,
	"businessConfig.schema": valibot.unwrap(_adobe_aio_commerce_lib_config.SchemaBusinessConfig.entries.schema),
	"eventing.commerce": valibot.unwrap(require_webhooks.EventingSchema.entries.commerce),
	"eventing.external": valibot.unwrap(require_webhooks.EventingSchema.entries.external),
	"installation.customInstallationSteps": valibot.unwrap(require_webhooks.InstallationSchema.entries.customInstallationSteps),
	adminUiSdk: require_webhooks.AdminUiSdkSchema
};
/**
* Get the config domains that are present in the config.
* @param config - The configuration to check.
*/
function getConfigDomains(config) {
	const withCommerceEvents = require_webhooks.hasCommerceEvents(config);
	const withExternalEvents = require_webhooks.hasExternalEvents(config);
	const domains = {
		metadata: hasMetadata(config),
		businessConfig: hasBusinessConfig(config),
		eventing: withCommerceEvents || withExternalEvents,
		installation: require_webhooks.hasCustomInstallation(config),
		webhooks: require_webhooks.hasWebhooks(config),
		adminUiSdk: require_webhooks.hasAdminUiSdk(config),
		"businessConfig.schema": hasBusinessConfigSchema(config),
		"eventing.commerce": withCommerceEvents,
		"eventing.external": withExternalEvents,
		"installation.customInstallationSteps": require_webhooks.hasCustomInstallationSteps(config)
	};
	const domainsList = Object.entries(domains).filter(([_, value]) => value).map(([key]) => key);
	return new Set(domainsList);
}
/**
* Check if the config has a specific domain.
* @param config - The configuration to check.
* @param domain - The domain to check.
*/
function hasConfigDomain(config, domain) {
	return getConfigDomains(config).has(domain);
}

//#endregion
//#region source/config/schema/app.ts
/** biome-ignore-all lint/performance/noBarrelFile: We want to have all the schema exports in one file. */
/** The schema used to validate the commerce app config file. */
const CommerceAppConfigSchema = valibot.looseObject({
	metadata: MetadataSchema,
	businessConfig: valibot.optional(_adobe_aio_commerce_lib_config.SchemaBusinessConfig),
	eventing: valibot.optional(require_webhooks.EventingSchema),
	adminUiSdk: valibot.optional(require_webhooks.AdminUiSdkSchema),
	installation: valibot.optional(require_webhooks.InstallationSchema),
	webhooks: valibot.optional(require_webhooks.WebhooksSchema)
});
/**
* Returns true if the given config domains require installation, false otherwise.
* @param domains - The config domains to check.
*/
function requiresInstallationFromDomains(domains) {
	return [
		"installation.customInstallationSteps",
		"eventing.commerce",
		"eventing.external",
		"webhooks",
		"adminUiSdk"
	].some((feature) => domains.has(feature));
}
/**
* Returns true if the given config requires installation, false otherwise.
* @param config - The commerce app config to check.
*/
function requiresInstallation(config) {
	return requiresInstallationFromDomains(getConfigDomains(config));
}

//#endregion
//#region source/config/lib/validate.ts
const commerceAppConfigDomainsSchema = valibot.picklist(Object.keys(CommerceAppConfigSchemas));
/**
* Validates a complete commerce app configuration object against the schema.
*
* @param config - The configuration object to validate.
* @returns The validated and typed configuration output model.
*
* @throws {CommerceSdkValidationError} If the configuration is invalid, with
* detailed validation issues included.
*
* @example
* ```typescript
* const config = {
*   businessConfiguration: {
*     // ... configuration data
*   }
* };
*
* try {
*   const validatedConfig = validateCommerceAppConfig(config);
*   // Use validatedConfig safely
* } catch (error) {
*   if (error instanceof CommerceSdkValidationError) {
*     console.error('Validation failed:', error.display());
*   }
* }
* ```
*/
function validateCommerceAppConfig(config) {
	const validatedConfig = valibot.safeParse(CommerceAppConfigSchema, config);
	if (!validatedConfig.success) throw new _adobe_aio_commerce_lib_core_error.CommerceSdkValidationError("Invalid commerce app config", { issues: validatedConfig.issues });
	return validatedConfig.output;
}
/**
* Validates a specific domain configuration within the commerce app config.
*
* This function validates only a specific domain's configuration rather than
* the entire commerce app configuration object. It first validates that the
* domain name is valid, then validates the configuration data against the
* schema for that specific domain.
*
* @template T - The type of the domain, constrained to valid domain names.
*
* @param config - The domain configuration object to validate.
* @param domain - The name of the domain to validate (e.g., 'businessConfiguration').
* @returns The validated and typed configuration for the specified domain.
*
* @throws {CommerceSdkValidationError} If the domain name is invalid or if the
* configuration doesn't match the domain's schema.
*
* @example
* ```typescript
* const businessConfig = {
*   fields: [
*     {
*       name: 'category',
*       type: 'dropdown',
*       // ... field configuration
*     }
*   ]
* };
*
* try {
*   const validatedConfig = validateCommerceAppConfigDomain(
*     businessConfig,
*     'businessConfig'
*   );
*   // Use validatedConfig safely
* } catch (error) {
*   if (error instanceof CommerceSdkValidationError) {
*     console.error('Domain validation failed:', error.issues);
*   }
* }
* ```
*/
function validateCommerceAppConfigDomain(config, domain) {
	const domainSchema = valibot.safeParse(commerceAppConfigDomainsSchema, domain);
	if (!domainSchema.success) throw new _adobe_aio_commerce_lib_core_error.CommerceSdkValidationError("Invalid commerce app config domain", { issues: domainSchema.issues });
	const domainConfigSchema = CommerceAppConfigSchemas[domain];
	const validatedConfig = valibot.safeParse(domainConfigSchema, config);
	if (!validatedConfig.success) throw new _adobe_aio_commerce_lib_core_error.CommerceSdkValidationError(`Invalid commerce app config: ${domain}`, { issues: validatedConfig.issues });
	return validatedConfig.output;
}

//#endregion
Object.defineProperty(exports, 'CommerceAppConfigSchemas', {
  enumerable: true,
  get: function () {
    return CommerceAppConfigSchemas;
  }
});
Object.defineProperty(exports, 'getConfigDomains', {
  enumerable: true,
  get: function () {
    return getConfigDomains;
  }
});
Object.defineProperty(exports, 'hasBusinessConfig', {
  enumerable: true,
  get: function () {
    return hasBusinessConfig;
  }
});
Object.defineProperty(exports, 'hasBusinessConfigSchema', {
  enumerable: true,
  get: function () {
    return hasBusinessConfigSchema;
  }
});
Object.defineProperty(exports, 'hasConfigDomain', {
  enumerable: true,
  get: function () {
    return hasConfigDomain;
  }
});
Object.defineProperty(exports, 'hasMetadata', {
  enumerable: true,
  get: function () {
    return hasMetadata;
  }
});
Object.defineProperty(exports, 'requiresInstallation', {
  enumerable: true,
  get: function () {
    return requiresInstallation;
  }
});
Object.defineProperty(exports, 'requiresInstallationFromDomains', {
  enumerable: true,
  get: function () {
    return requiresInstallationFromDomains;
  }
});
Object.defineProperty(exports, 'validateCommerceAppConfig', {
  enumerable: true,
  get: function () {
    return validateCommerceAppConfig;
  }
});
Object.defineProperty(exports, 'validateCommerceAppConfigDomain', {
  enumerable: true,
  get: function () {
    return validateCommerceAppConfigDomain;
  }
});
