// @adobe/aio-commerce-lib-app/dist/cjs/webhooks-tbZy9G3p.cjs
/**
 * @license
 *
 * Copyright 2026 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */

const require_chunk = require('./chunk-Cozy-1ZF.cjs');
const require_schemas = require('./schemas-FngDuIeo.cjs');
let valibot = require("valibot");
valibot = require_chunk.__toESM(valibot, 1);

//#region source/config/schema/admin-ui-sdk.ts
const SANDBOX_VALUES = [
	"allow-downloads",
	"allow-modals",
	"allow-popups"
];
function isSandboxValue(value) {
	return SANDBOX_VALUES.includes(value);
}
const SandboxSchema = valibot.pipe(valibot.string("Expected a string value for \"sandbox\""), valibot.check((val) => val.split(" ").every(isSandboxValue), `sandbox must contain only single-space-separated values from: ${SANDBOX_VALUES.map((t) => `"${t}"`).join(", ")}`));
const ColumnTypeSchema = valibot.picklist([
	"boolean",
	"date",
	"float",
	"integer",
	"string"
]);
const ColumnAlignSchema = valibot.picklist([
	"left",
	"right",
	"center"
]);
const ViewButtonLevelSchema = valibot.picklist([
	-1,
	0,
	1
]);
const MassActionConfirmSchema = valibot.object({
	title: valibot.optional(require_schemas.nonEmptyStringValueSchema("confirm title")),
	message: valibot.optional(require_schemas.nonEmptyStringValueSchema("confirm message"))
});
const ViewButtonConfirmSchema = valibot.object({ message: valibot.optional(require_schemas.nonEmptyStringValueSchema("confirm message")) });
const iframeActionEntries = {
	displayIframe: valibot.optional(require_schemas.booleanValueSchema("displayIframe")),
	timeout: valibot.optional(require_schemas.positiveNumberValueSchema("timeout")),
	sandbox: valibot.optional(SandboxSchema)
};
const GridColumnPropertySchema = valibot.object({
	label: require_schemas.nonEmptyStringValueSchema("column label"),
	columnId: require_schemas.nonEmptyStringValueSchema("column ID"),
	type: ColumnTypeSchema,
	align: ColumnAlignSchema
});
const GridColumnsSchema = valibot.object({
	data: valibot.object({ meshId: require_schemas.nonEmptyStringValueSchema("mesh ID") }),
	properties: valibot.pipe(valibot.array(GridColumnPropertySchema), valibot.minLength(1, "At least one grid column property is required"))
});
const massActionBaseEntries = {
	actionId: require_schemas.nonEmptyStringValueSchema("mass action ID"),
	label: require_schemas.nonEmptyStringValueSchema("mass action label"),
	title: valibot.optional(require_schemas.nonEmptyStringValueSchema("mass action page title")),
	confirm: valibot.optional(MassActionConfirmSchema),
	path: require_schemas.nonEmptyStringValueSchema("mass action path"),
	...iframeActionEntries
};
const sandboxDisplayIframeCheck = valibot.forward(valibot.partialCheck([["sandbox"], ["displayIframe"]], (input) => input.sandbox === void 0 || input.displayIframe === true, "sandbox is only relevant when displayIframe is set to true"), ["sandbox"]);
function withSandboxDisplayIframeCheck(schema) {
	return valibot.pipe(schema, sandboxDisplayIframeCheck);
}
function createMassActionSchema(variantEntries) {
	return withSandboxDisplayIframeCheck(valibot.strictObject({
		...massActionBaseEntries,
		...variantEntries
	}));
}
const OrderMassActionSchema = createMassActionSchema({ selectionLimit: valibot.optional(require_schemas.positiveNumberValueSchema("selectionLimit")) });
const ProductMassActionSchema = createMassActionSchema({ productSelectLimit: valibot.optional(require_schemas.positiveNumberValueSchema("productSelectLimit")) });
const CustomerMassActionSchema = createMassActionSchema({ customerSelectLimit: valibot.optional(require_schemas.positiveNumberValueSchema("customerSelectLimit")) });
const OrderViewButtonSchema = withSandboxDisplayIframeCheck(valibot.object({
	buttonId: require_schemas.nonEmptyStringValueSchema("view button ID"),
	label: require_schemas.nonEmptyStringValueSchema("view button label"),
	confirm: valibot.optional(ViewButtonConfirmSchema),
	path: require_schemas.nonEmptyStringValueSchema("view button path"),
	level: valibot.optional(ViewButtonLevelSchema),
	sortOrder: valibot.optional(require_schemas.positiveNumberValueSchema("sortOrder")),
	...iframeActionEntries
}));
const CustomFeeSchema = valibot.object({
	id: require_schemas.nonEmptyStringValueSchema("custom fee ID"),
	label: require_schemas.nonEmptyStringValueSchema("custom fee label"),
	value: valibot.number("Custom fee value must be a number"),
	orderMinimumAmount: valibot.optional(valibot.number("orderMinimumAmount must be a number")),
	applyFeeOnLastInvoice: valibot.optional(require_schemas.booleanValueSchema("applyFeeOnLastInvoice")),
	applyFeeOnLastCreditMemo: valibot.optional(require_schemas.booleanValueSchema("applyFeeOnLastCreditMemo"))
});
const OrderExtensionPointsSchema = valibot.object({
	massActions: valibot.optional(valibot.array(OrderMassActionSchema)),
	gridColumns: valibot.optional(GridColumnsSchema),
	viewButtons: valibot.optional(valibot.array(OrderViewButtonSchema)),
	customFees: valibot.optional(valibot.array(CustomFeeSchema))
});
const ProductExtensionPointsSchema = valibot.object({
	massActions: valibot.optional(valibot.array(ProductMassActionSchema)),
	gridColumns: valibot.optional(GridColumnsSchema)
});
const CustomerExtensionPointsSchema = valibot.object({
	massActions: valibot.optional(valibot.array(CustomerMassActionSchema)),
	gridColumns: valibot.optional(GridColumnsSchema)
});
const MassActionBannerSchema = valibot.object({
	actionId: require_schemas.nonEmptyStringValueSchema("mass action ID"),
	successMessage: valibot.optional(require_schemas.nonEmptyStringValueSchema("success message")),
	errorMessage: valibot.optional(require_schemas.nonEmptyStringValueSchema("error message"))
});
const OrderViewButtonBannerSchema = valibot.object({
	buttonId: require_schemas.nonEmptyStringValueSchema("view button ID"),
	successMessage: valibot.optional(require_schemas.nonEmptyStringValueSchema("success message")),
	errorMessage: valibot.optional(require_schemas.nonEmptyStringValueSchema("error message"))
});
const BannerNotificationSchema = valibot.object({
	massActions: valibot.optional(valibot.object({
		order: valibot.optional(valibot.array(MassActionBannerSchema)),
		product: valibot.optional(valibot.array(MassActionBannerSchema)),
		customer: valibot.optional(valibot.array(MassActionBannerSchema))
	})),
	orderViewButtons: valibot.optional(valibot.array(OrderViewButtonBannerSchema))
});
const MenuItemSchema = valibot.object({
	id: require_schemas.nonEmptyStringValueSchema("menu item ID"),
	title: valibot.optional(require_schemas.nonEmptyStringValueSchema("menu item title")),
	parent: valibot.optional(require_schemas.nonEmptyStringValueSchema("menu item parent")),
	sortOrder: valibot.optional(valibot.number()),
	isSection: valibot.optional(require_schemas.booleanValueSchema("isSection")),
	sandbox: valibot.optional(SandboxSchema)
});
/**
* Schema for the Admin UI SDK registration parameters (for the `adminUiSdk.registration` config section).
* @see https://developer.adobe.com/commerce/extensibility/admin-ui-sdk/extension-points/ for more details.
*/
const AdminUiSdkRegistrationSchema = valibot.object({
	menuItems: valibot.optional(valibot.array(MenuItemSchema)),
	order: valibot.optional(OrderExtensionPointsSchema),
	product: valibot.optional(ProductExtensionPointsSchema),
	customer: valibot.optional(CustomerExtensionPointsSchema),
	bannerNotification: valibot.optional(BannerNotificationSchema)
});
/**
* Schema for Admin UI SDK configuration.
* @experimental
*/
const AdminUiSdkSchema = valibot.object({ registration: AdminUiSdkRegistrationSchema });
/**
* Check if config has Admin UI SDK registration configuration.
* @experimental
*/
function hasAdminUiSdk(config) {
	return config.adminUiSdk !== void 0 && config.adminUiSdk.registration !== void 0;
}

//#endregion
//#region source/config/schema/eventing.ts
const MAX_DESCRIPTION_LENGTH$1 = 255;
const MAX_LABEL_LENGTH = 100;
const MAX_KEY_LENGTH = 50;
const MAX_EVENT_NAME_LENGTH = 180;
/**
* Regex for Commerce event names that must start with "plugin." or "observer."
* followed by one or more dot-separated lowercase segments containing letters
* and underscores only.
* Examples: "observer.order_placed", "plugin.sales.api.order_management.place"
*/
const COMMERCE_EVENT_NAME_REGEX = /^(?:plugin|observer)\.[a-z_]+(?:\.[a-z_]+)*$/;
/**
* Regex for external event names.
* Allows word characters (letters, digits, underscore), hyphens, underscores, and dots.
* Examples: "external_event", "webhook.received", "my-event_123"
*/
const EXTERNAL_EVENT_NAME_REGEX = /^[\w\-_.]+$/;
/**
* Regex for field names according to XSD fieldName pattern.
* Field name can either contain only [a-zA-Z0-9_\-\.\[\]] or be set to *.
*/
const FIELD_NAME_REGEX = /^([a-zA-Z0-9_\-.[\]]+|\*)$/;
/**
* Regex for Adobe I/O Events API text fields (label, description).
* Valid characters per API: letters, numbers, spaces, underscores, hyphens,
* dots, colons, parentheses, commas, @, and /.
*/
const IO_EVENTS_TEXT_REGEX = /^[a-zA-Z0-9 _.:()\-,@/]+$/;
/**
* Schema for Commerce event names.
* Validates that the event name starts with "plugin." or "observer."
* followed by one or more dot-separated lowercase segments containing letters
* and underscores only.
*/
function commerceEventNameSchema() {
	return valibot.pipe(require_schemas.nonEmptyStringValueSchema("event name"), valibot.regex(COMMERCE_EVENT_NAME_REGEX, "Event name must start with \"plugin.\" or \"observer.\" followed by one or more dot-separated lowercase segments containing letters and underscores only (e.g., \"observer.order_placed\", \"plugin.sales.api.order_management.place\")"), valibot.maxLength(MAX_EVENT_NAME_LENGTH, `The event name must not be longer than ${MAX_EVENT_NAME_LENGTH} characters`));
}
/**
* Schema for external event names.
* Validates that the event name contains only word characters (letters, digits, underscore),
* hyphens, underscores, and dots.
*/
function externalEventNameSchema() {
	return valibot.pipe(require_schemas.nonEmptyStringValueSchema("event name"), valibot.regex(EXTERNAL_EVENT_NAME_REGEX, "Event name must contain only letters, digits, underscores, hyphens, and dots (e.g., \"external_event\", \"webhook.received\", \"my-event_123\")"), valibot.maxLength(MAX_EVENT_NAME_LENGTH, `The event name must not be longer than ${MAX_EVENT_NAME_LENGTH} characters`));
}
/**
* Schema for field names.
* Validates that the field name matches the XSD fieldName pattern:
* can either contain only [a-zA-Z0-9_\-\.\[\]] or be set to *.
*/
function fieldNameSchema() {
	return valibot.pipe(require_schemas.nonEmptyStringValueSchema("field name"), valibot.regex(FIELD_NAME_REGEX, "Field name must contain only letters (a-z, A-Z), numbers (0-9), underscores (_), dashes (-), dots (.), and square brackets ([, ]), or be exactly \"*\""));
}
/** Validates that a text field contains only characters accepted by the Adobe I/O Events API. */
function ioEventsTextSchema(name) {
	return valibot.regex(IO_EVENTS_TEXT_REGEX, `${name} can only contain letters, numbers, spaces, underscores, hyphens, dots, colons, parentheses, commas, @, and /`);
}
/**
* Schema for field objects in Commerce events.
* Each field has a required name and an optional source.
*/
function commerceEventFieldSchema() {
	return valibot.object({
		name: fieldNameSchema(),
		source: valibot.optional(require_schemas.stringValueSchema("field source"))
	});
}
/** Schema for event provider configuration */
const ProviderSchema = valibot.object({
	label: valibot.pipe(require_schemas.nonEmptyStringValueSchema("provider label"), ioEventsTextSchema("Provider label"), valibot.maxLength(MAX_LABEL_LENGTH, `The provider label must not be longer than ${MAX_LABEL_LENGTH} characters`)),
	description: valibot.pipe(require_schemas.nonEmptyStringValueSchema("provider description"), ioEventsTextSchema("Provider description"), valibot.maxLength(MAX_DESCRIPTION_LENGTH$1, `The provider description must not be longer than ${MAX_DESCRIPTION_LENGTH$1} characters`)),
	key: valibot.optional(valibot.pipe(require_schemas.alphaNumericOrHyphenSchema("provider key"), valibot.maxLength(MAX_KEY_LENGTH, `The provider key must not be longer than ${MAX_KEY_LENGTH} characters`)))
});
/** Schema for base shared properties between event types. */
const BaseEventSchema = valibot.object({
	label: valibot.pipe(require_schemas.nonEmptyStringValueSchema("event label"), ioEventsTextSchema("Event label"), valibot.maxLength(MAX_LABEL_LENGTH, `The event label must not be longer than ${MAX_LABEL_LENGTH} characters`)),
	description: valibot.pipe(require_schemas.nonEmptyStringValueSchema("event description"), ioEventsTextSchema("Event description"), valibot.maxLength(MAX_DESCRIPTION_LENGTH$1, `The event description must not be longer than ${MAX_DESCRIPTION_LENGTH$1} characters`)),
	runtimeActions: valibot.array(valibot.pipe(require_schemas.nonEmptyStringValueSchema("runtime action"), valibot.regex(/^[a-z0-9-]+\/[a-z0-9-]+$/i, "Runtime action must be in the format \"<package>/<action>\" (e.g., \"my-package/my-action\")")), "Expected an array of runtime actions in the format <package>/<action>")
});
/**
* Schema for rule operator values.
* Valid operators for Commerce event filtering rules.
*/
const OPERATORS = [
	"greaterThan",
	"lessThan",
	"equal",
	"regex",
	"in",
	"onChange"
];
const ruleOperatorSchema = valibot.union(OPERATORS.map((op) => valibot.literal(op)), `Operator must be one of: ${OPERATORS.join(", ")}`);
/** Schema for Commerce event rule configuration */
const CommerceEventRuleSchema = valibot.object({
	field: require_schemas.nonEmptyStringValueSchema("rule field"),
	operator: ruleOperatorSchema,
	value: require_schemas.nonEmptyStringValueSchema("rule value")
});
/** Schema for Commerce event configuration */
const CommerceEventSchema = valibot.object({
	...BaseEventSchema.entries,
	name: commerceEventNameSchema(),
	fields: valibot.pipe(valibot.array(commerceEventFieldSchema(), "Expected an array of event field objects with a 'name' property"), valibot.minLength(1, "The Commerce event configuration must define at least one field")),
	rules: valibot.optional(valibot.array(CommerceEventRuleSchema, "Expected an array of event rules with field, operator, and value")),
	destination: valibot.optional(require_schemas.nonEmptyStringValueSchema("destination")),
	hipaa_audit_required: valibot.optional(require_schemas.booleanValueSchema("hipaa_audit_required")),
	priority: valibot.optional(require_schemas.booleanValueSchema("priority")),
	force: valibot.optional(require_schemas.booleanValueSchema("force"))
});
/** Schema for external event configuration */
const ExternalEventSchema = valibot.object({
	...BaseEventSchema.entries,
	name: externalEventNameSchema()
});
/** Schema for Commerce event source configuration */
const CommerceEventSourceSchema = valibot.object({
	provider: ProviderSchema,
	events: valibot.array(CommerceEventSchema, "Expected an array of Commerce events")
});
/** Schema for external event source configuration */
const ExternalEventSourceSchema = valibot.object({
	provider: ProviderSchema,
	events: valibot.array(ExternalEventSchema, "Expected an array of external events")
});
/** Schema for eventing configuration with separate commerce and external arrays */
const EventingSchema = valibot.object({
	commerce: valibot.optional(valibot.pipe(valibot.array(CommerceEventSourceSchema, "Expected an array of Commerce event sources"), valibot.check((sources) => new Set(sources.map((s) => s.provider.label)).size === sources.length, "Commerce provider labels must be unique"))),
	external: valibot.optional(valibot.pipe(valibot.array(ExternalEventSourceSchema, "Expected an array of external event sources"), valibot.check((sources) => new Set(sources.map((s) => s.provider.label)).size === sources.length, "External provider labels must be unique")))
});
/**
* Check if config has commerce event sources.
* @param config - The configuration to check.
*/
function hasCommerceEvents(config) {
	return Array.isArray(config?.eventing?.commerce) && config.eventing.commerce.length > 0;
}
/**
* Check if config has external event sources.
* @param config - The configuration to check.
*/
function hasExternalEvents(config) {
	return Array.isArray(config?.eventing?.external) && config.eventing.external.length > 0;
}
/**
* Check if config has any eventing configuration.
* @param config - The configuration to check.
*/
function hasEventing(config) {
	return config.eventing !== void 0;
}

//#endregion
//#region source/config/schema/installation.ts
const MAX_DESCRIPTION_LENGTH = 255;
const MAX_NAME_LENGTH = 255;
const MAX_MESSAGE_LENGTH = 1e3;
/**
* Regex for script paths that can be relative or absolute.
* Must end with .js extension
* Examples:
*   - "./scripts/setup.js"
*   - "./setup.js"
*   - "../../scripts/setup.js"
*/
const SCRIPT_PATH_REGEX = /^(?:\.{0,2}\/)*[\w-/]*[\w-]+\.js$/;
/**
* Schema for custom installation step configuration
*/
const CustomInstallationStepSchema = valibot.object({
	script: valibot.pipe(require_schemas.nonEmptyStringValueSchema("script path"), valibot.regex(SCRIPT_PATH_REGEX, "Script path must end with .js (e.g., \"./setup.js\", \"./scripts/setup.js\", or \"../../scripts/setup.js\")")),
	name: valibot.pipe(require_schemas.nonEmptyStringValueSchema("step name"), valibot.maxLength(MAX_NAME_LENGTH, `The step name must not be longer than ${MAX_NAME_LENGTH} characters`)),
	description: valibot.pipe(require_schemas.nonEmptyStringValueSchema("step description"), valibot.maxLength(MAX_DESCRIPTION_LENGTH, `The step description must not be longer than ${MAX_DESCRIPTION_LENGTH} characters`))
});
/**
* Schema for installation messages configuration
*/
const MessagesSchema = valibot.object({
	preInstallation: valibot.optional(valibot.pipe(require_schemas.nonEmptyStringValueSchema("preInstallation message"), valibot.maxLength(MAX_MESSAGE_LENGTH, `The preInstallation message must not be longer than ${MAX_MESSAGE_LENGTH} characters`))),
	postInstallation: valibot.optional(valibot.pipe(require_schemas.nonEmptyStringValueSchema("postInstallation message"), valibot.maxLength(MAX_MESSAGE_LENGTH, `The postInstallation message must not be longer than ${MAX_MESSAGE_LENGTH} characters`)))
});
/**
* Schema for installation configuration
*/
const InstallationSchema = valibot.object({
	messages: valibot.optional(MessagesSchema),
	customInstallationSteps: valibot.pipe(valibot.optional(valibot.array(CustomInstallationStepSchema, "Expected an array of custom installation steps")), valibot.check((input) => {
		const steps = input || [];
		return new Set(steps.map((step) => step.name)).size === steps.length;
	}, "Duplicate step names detected in custom installation steps. Each step must have a unique name."))
});
/**
* Check if config has custom installation settings.
* @param config - The configuration to check.
*/
function hasCustomInstallation(config) {
	return config.installation !== void 0;
}
/**
* Check if config has custom installation steps.
* @param config - The configuration to check.
*/
function hasCustomInstallationSteps(config) {
	return Array.isArray(config?.installation?.customInstallationSteps) && config.installation.customInstallationSteps.length > 0;
}

//#endregion
//#region source/config/schema/webhooks.ts
/** Schema for webhook field configuration (name and optional source). */
const WebhookFieldSchema = valibot.object({
	name: require_schemas.nonEmptyStringValueSchema("field name"),
	source: valibot.optional(require_schemas.stringValueSchema("field source"))
});
/** Schema for webhook rule configuration (field, operator, value). */
const WebhookRuleSchema = valibot.object({
	field: require_schemas.nonEmptyStringValueSchema("rule field"),
	operator: require_schemas.nonEmptyStringValueSchema("rule operator"),
	value: require_schemas.nonEmptyStringValueSchema("rule value")
});
/** Schema for webhook header configuration (name, value). */
const WebhookHeaderSchema = valibot.object({
	name: require_schemas.nonEmptyStringValueSchema("header name"),
	value: require_schemas.nonEmptyStringValueSchema("header value")
});
/** batch_name and hook_name must contain only letters, numbers, and underscores. */
const WEBHOOK_IDENTIFIER_REGEX = /^[a-zA-Z0-9_]+$/;
/** Category for conflict detection: validation, append, or modification. */
const CATEGORIES = [
	"validation",
	"append",
	"modification"
];
const CategorySchema = valibot.picklist(CATEGORIES, `Webhook category must be one of: ${CATEGORIES.join(", ")}`);
/** Schema for the nested webhook payload without url — used when runtimeAction resolves the URL at runtime. */
const WebhookDefinitionBaseSchema = valibot.object({
	webhook_method: require_schemas.nonEmptyStringValueSchema("webhook_method"),
	webhook_type: require_schemas.nonEmptyStringValueSchema("webhook_type"),
	batch_name: valibot.pipe(require_schemas.nonEmptyStringValueSchema("batch_name"), valibot.regex(WEBHOOK_IDENTIFIER_REGEX, "batch_name must contain only letters, numbers, and underscores")),
	batch_order: valibot.optional(require_schemas.positiveNumberValueSchema("batch_order")),
	hook_name: valibot.pipe(require_schemas.nonEmptyStringValueSchema("hook_name"), valibot.regex(WEBHOOK_IDENTIFIER_REGEX, "hook_name must contain only letters, numbers, and underscores")),
	priority: valibot.optional(require_schemas.positiveNumberValueSchema("priority")),
	required: valibot.optional(require_schemas.booleanValueSchema("required")),
	soft_timeout: valibot.optional(require_schemas.positiveNumberValueSchema("soft_timeout")),
	timeout: valibot.optional(require_schemas.positiveNumberValueSchema("timeout")),
	method: require_schemas.nonEmptyStringValueSchema("HTTP method"),
	fallback_error_message: valibot.optional(require_schemas.stringValueSchema("fallback_error_message")),
	ttl: valibot.optional(require_schemas.positiveNumberValueSchema("ttl")),
	fields: valibot.optional(valibot.array(WebhookFieldSchema, "Expected an array of webhook field objects")),
	rules: valibot.optional(valibot.array(WebhookRuleSchema, "Expected an array of webhook rule objects")),
	headers: valibot.optional(valibot.array(WebhookHeaderSchema, "Expected an array of webhook header objects"))
});
/** Schema for the nested webhook payload with a required url. */
const WebhookDefinitionWithUrlSchema = valibot.object({
	...WebhookDefinitionBaseSchema.entries,
	url: valibot.pipe(require_schemas.stringValueSchema("webhook URL"), valibot.url("The 'url' field must be a valid absolute URL (e.g., 'https://example.com/webhook')"))
});
/** Schema for a webhook entry that resolves its URL from a runtime action. */
const WebhookEntryWithRuntimeActionSchema = valibot.object({
	label: require_schemas.nonEmptyStringValueSchema("webhook label"),
	description: require_schemas.nonEmptyStringValueSchema("webhook description"),
	category: valibot.optional(CategorySchema),
	runtimeAction: require_schemas.nonEmptyStringValueSchema("runtimeAction"),
	requireAdobeAuth: valibot.optional(require_schemas.booleanValueSchema("requireAdobeAuth")),
	webhook: WebhookDefinitionBaseSchema
});
/** Schema for a webhook entry that provides an explicit URL. */
const WebhookEntryWithUrlSchema = valibot.object({
	label: require_schemas.nonEmptyStringValueSchema("webhook label"),
	description: require_schemas.nonEmptyStringValueSchema("webhook description"),
	category: valibot.optional(CategorySchema),
	webhook: WebhookDefinitionWithUrlSchema
});
/** Schema for a single webhook entry — either runtimeAction (no url) or explicit url (no runtimeAction). */
const WebhookEntrySchema = valibot.union([WebhookEntryWithRuntimeActionSchema, WebhookEntryWithUrlSchema], "Each webhook entry must define either a 'runtimeAction' (to resolve the URL from a runtime action) or an explicit 'url' inside the 'webhook' object, but not both");
/** Schema for the optional webhooks array (when present, must have at least one item). */
const WebhooksSchema = valibot.optional(valibot.pipe(valibot.array(WebhookEntrySchema, "Expected an array of webhook entries"), valibot.minLength(1, "webhooks array must contain at least one webhook when present")));
/**
* Check if config has webhooks (non-empty array).
* @param config - The configuration to check.
*/
function hasWebhooks(config) {
	return Array.isArray(config?.webhooks) && config.webhooks.length > 0;
}

//#endregion
Object.defineProperty(exports, 'AdminUiSdkSchema', {
  enumerable: true,
  get: function () {
    return AdminUiSdkSchema;
  }
});
Object.defineProperty(exports, 'EventingSchema', {
  enumerable: true,
  get: function () {
    return EventingSchema;
  }
});
Object.defineProperty(exports, 'InstallationSchema', {
  enumerable: true,
  get: function () {
    return InstallationSchema;
  }
});
Object.defineProperty(exports, 'WebhooksSchema', {
  enumerable: true,
  get: function () {
    return WebhooksSchema;
  }
});
Object.defineProperty(exports, 'hasAdminUiSdk', {
  enumerable: true,
  get: function () {
    return hasAdminUiSdk;
  }
});
Object.defineProperty(exports, 'hasCommerceEvents', {
  enumerable: true,
  get: function () {
    return hasCommerceEvents;
  }
});
Object.defineProperty(exports, 'hasCustomInstallation', {
  enumerable: true,
  get: function () {
    return hasCustomInstallation;
  }
});
Object.defineProperty(exports, 'hasCustomInstallationSteps', {
  enumerable: true,
  get: function () {
    return hasCustomInstallationSteps;
  }
});
Object.defineProperty(exports, 'hasEventing', {
  enumerable: true,
  get: function () {
    return hasEventing;
  }
});
Object.defineProperty(exports, 'hasExternalEvents', {
  enumerable: true,
  get: function () {
    return hasExternalEvents;
  }
});
Object.defineProperty(exports, 'hasWebhooks', {
  enumerable: true,
  get: function () {
    return hasWebhooks;
  }
});
