// @adobe/aio-commerce-lib-core/dist/es/headers/index.mjs
/**
 * @license
 *
 * Copyright 2026 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */

import { t as CommerceSdkValidationError } from "../error-WnQqveay.mjs";
import { literal, nonEmpty, object, pipe, rawCheck, rawTransform, record, safeParse, string, trim, undefined as undefined$1, union } from "valibot";
import * as camelcaseModule from "camelcase";

//#region source/headers/camelcase.ts
const camelcase = camelcaseModule.default.default ?? camelcaseModule.default;

//#endregion
//#region source/headers/helpers.ts
/**
* Extracts a header value from an object with case-insensitive lookup.
* This is useful for handling HTTP headers which can be case-insensitive per RFC 7230.
*
* When enabled via options, comma-separated header values (per RFC 9110) can be
* automatically split into an array. This is useful when multiple header values
* are combined into a single comma-separated string.
*
* @param headers The headers object to search in.
* @param name The header name to look for (case-insensitive).
* @param options Optional settings to control header value processing.
*
* @example
* ```typescript
* const headers = { authorization: "Bearer token123" };
* const auth = getHeader(headers, "Authorization"); // "Bearer token123"
* ```
*
* @example
* ```typescript
* // Comma-separated string can be split when enabled (per RFC 9110)
* const headers = { "Example-Field": "Foo, Bar" };
* const value = getHeader(headers, "Example-Field", { splitCommaSeparated: true }); // ["Foo", "Bar"]
* ```
*
* @see https://datatracker.ietf.org/doc/html/rfc7230 - HTTP/1.1 Message Syntax and Routing
* @see https://datatracker.ietf.org/doc/html/rfc9110#name-field-lines-and-combined-fi - HTTP Semantics: Field Lines and Combined Field Values
*/
function getHeader(headers, name, options) {
	const { splitCommaSeparated = false } = options ?? {};
	let value;
	if (Object.hasOwn(headers, name)) value = headers[name];
	else {
		const lowerName = name.toLowerCase();
		for (const key in headers) if (Object.hasOwn(headers, key) && key.toLowerCase() === lowerName) {
			value = headers[key];
			break;
		}
	}
	if (splitCommaSeparated && typeof value === "string" && value.includes(",")) value = value.split(",").map((v) => v.trim());
	return value;
}
/**
* Extracts the `__ow_headers` object from App Builder runtime action parameters.
*
* @param params The action parameters from App Builder runtime.
* @throws {CommerceSdkValidationError} If `__ow_headers` is missing or not an object.
*
* @example
* ```typescript
* export async function main(params) {
*   const headers = getHeadersFromParams(params);
*   const apiKey = headers["x-api-key"];
* }
* ```
*/
function getHeadersFromParams(params) {
	const parsed = safeParse(object({ __ow_headers: record(string(), union([string(), undefined$1()])) }), params);
	if (!parsed.success) throw new CommerceSdkValidationError("Missing __ow_headers in action params", { issues: parsed.issues });
	return parsed.output.__ow_headers;
}

//#endregion
//#region source/headers/validation.ts
/**
* Returns an array of missing or empty headers using Valibot validation.
* Case-insensitive by default.
*
* Unlike {@link assertRequiredHeaders}, this function does not throw an error.
* Instead, it returns an array of header names that are missing or empty.
* This is useful when you need to check headers without immediately failing.
*
* @param headers The headers object to validate.
* @param requiredHeaders Array of required header names.
* @returns Array of header names that are missing or empty. Empty array if all headers are present and non-empty.
*
* @example
* ```typescript
* const headers = { "x-api-key": "key123" };
* const missing = getMissingOrEmptyHeaders(headers, ["x-api-key", "Authorization"]);
* // Returns: ["Authorization"]
* ```
*
* @example
* ```typescript
* const headers = { "x-api-key": "", "Authorization": "Bearer token" };
* const missing = getMissingOrEmptyHeaders(headers, ["x-api-key", "Authorization"]);
* // Returns: ["x-api-key"] (empty string is considered missing)
* ```
*/
function getMissingOrEmptyHeaders(headers, requiredHeaders) {
	const normalized = {};
	if (headers && typeof headers === "object" && !Array.isArray(headers)) for (const key of Object.keys(headers)) {
		const value = headers[key];
		if (value !== void 0 && value !== null) normalized[key.toLowerCase()] = value;
	}
	const parsed = safeParse(buildHeadersSchema(requiredHeaders.map((h) => literal(h))), normalized);
	if (parsed.success) return [];
	const lookup = new Map(requiredHeaders.map((h) => [h.toLowerCase(), h]));
	return parsed.issues.flatMap((issue) => {
		const key = issue.path?.[0]?.key;
		const original = typeof key === "string" ? lookup.get(key.toLowerCase()) : null;
		return original ? [original] : [];
	});
}
/**
* Builds a Valibot schema that checks:
* - Required headers exist (case-insensitive)
* - Each required header has a non-empty string value
*/
function buildHeadersSchema(requiredHeaders) {
	const buildSchema = () => pipe(string("Expected a string value for the header"), rawCheck(({ dataset, addIssue }) => {
		if (!dataset.typed) return;
		const value = dataset.value;
		if (value.trim() === "") {
			addIssue({
				expected: "a non-empty string value",
				received: `"${value}"`,
				message: "Header value cannot be empty or contain only whitespace"
			});
			return;
		}
		if (value.includes(",")) {
			const parts = value.split(",").map((part) => part.trim());
			if (!parts.some((part) => part !== "")) addIssue({
				expected: "at least one non-empty value in comma-separated list",
				received: `[${parts.join(", ")}]`,
				message: "Header value contains only empty or whitespace values when split by comma"
			});
		}
	}));
	const shape = {};
	for (const schema of requiredHeaders) shape[schema.literal.toLowerCase()] = buildSchema();
	return object(shape);
}
/**
* Validates that required headers are present in the headers object.
* Case-insensitive by default (HTTP standard).
*
* This is an advanced validation function that accepts Valibot `LiteralSchema` objects
* instead of plain strings. Use {@link assertRequiredHeaders} for the simpler string-based API.
*
* @param headers The headers object to validate.
* @param requiredHeaders Array of required header names as Valibot `LiteralSchema` objects.
* @throws {CommerceSdkValidationError} If validation fails.
*
* @example
* ```typescript
* import { literal } from "valibot";
* import { assertRequiredHeadersSchema } from "@adobe/aio-commerce-lib-core/headers";
*
* const headers = getHeadersFromParams(params);
* assertRequiredHeadersSchema(headers, [
*   literal("x-api-key"),
*   literal("Authorization"),
* ]);
* ```
*/
function assertRequiredHeadersSchema(headers, requiredHeaders) {
	const normalized = {};
	if (headers && typeof headers === "object" && !Array.isArray(headers)) for (const key of Object.keys(headers)) {
		const value = headers[key];
		if (value !== void 0 && value !== null) normalized[key.toLowerCase()] = value;
	}
	const parsed = safeParse(buildHeadersSchema(requiredHeaders), normalized);
	if (!parsed.success) throw new CommerceSdkValidationError("The headers object does not match the required headers (case-insensitive)", { issues: parsed.issues });
}
/**
* Validates that required headers are present and non-empty in the headers object.
* Performs case-insensitive validation with {@link getHeader}.
*
* @param headers The headers object to validate.
* @param requiredHeaders Array of required header names.
* @throws {CommerceSdkValidationError} If any required headers are missing or empty.
*
* @example
* ```typescript
* const headers = { "x-api-key": "key123", "Authorization": "Bearer token" };
* assertRequiredHeaders(headers, ["x-api-key", "Authorization"]);
* // Headers are validated, use getHeader() for safe access
* const apiKey = getHeader(headers, "x-api-key")!;
* ```
*
* @example
* ```typescript
* try {
*   assertRequiredHeaders({}, ["x-api-key", "Authorization"]);
* } catch (error) {
*   console.error(error.display());
*   // "The headers object does not match the required headers (case-insensitive)"
*  // ├── Schema validation error at x-api-key → Expected a non-empty string value
*  // └── Schema validation error at Authorization → Expected a non-empty string value
* }
* ```
*/
function assertRequiredHeaders(headers, requiredHeaders) {
	assertRequiredHeadersSchema(headers, requiredHeaders.map((header) => literal(header)));
}

//#endregion
//#region source/headers/accessor.ts
const LEADING_UNDERSCORES = /^_+/;
/**
* Converts a header name to a camelCase key aligned with `type-fest`'s `CamelCase` type.
*
* `camelcase` v9+ preserves leading underscores (e.g. `_foo_bar` → `_fooBar`), but
* `type-fest`'s `CamelCase` — used by `HttpHeaderAccessorMap` — drops them
* (e.g. `_foo_bar` → `fooBar`). Without stripping them first, runtime keys would diverge
* from the inferred TypeScript type.
*
* @remarks This function must be removed and replaced with a bare `camelcase(header)` call
* once the two helpers agree on leading-underscore handling. That cleanup is a breaking
* change and must land in a major release.
*/
function toHeaderCamelKey(header) {
	return camelcase(header.replace(LEADING_UNDERSCORES, ""));
}
/**
* Creates a type-safe header accessor object with validated required headers.
* Header names are normalized to camelCase for consistent access.
*
* @param headers The headers object from runtime params.
* @param requiredHeaders Array of required header names.
* @param options Optional settings to control header value processing (e.g., `splitCommaSeparated`).
* @throws {CommerceSdkValidationError} If any required headers are missing or empty.
*
* @example
* ```typescript
* const headers = getHeadersFromParams(params);
* const accessor = createHeaderAccessor(headers, ["x-api-key", "Authorization"]);
*
* // TypeScript knows these are non-empty strings (normalized to camelCase)
* const apiKey = accessor.xApiKey;       // string
* const auth = accessor.authorization;   // string
* ```
*
* @example
* ```typescript
* // With options to enable comma-separated value splitting
* const accessor = createHeaderAccessor(
*   headers,
*   ["cache-control"],
*   { splitCommaSeparated: true }
* );
* // accessor.cacheControl may be string | string[] depending on comma presence
* ```
*/
function createHeaderAccessor(headers, requiredHeaders, options) {
	assertRequiredHeaders(headers, requiredHeaders);
	const accessor = {};
	for (const header of requiredHeaders) {
		const camelKey = toHeaderCamelKey(header);
		accessor[camelKey] = getHeader(headers, header, options);
	}
	return accessor;
}

//#endregion
//#region source/headers/auth.ts
/**
* Type guard to check if authorization is Bearer type.
*/
function isBearerAuth(auth) {
	return auth.scheme === "Bearer";
}
/**
* Type guard to check if authorization is Basic type.
*/
function isBasicAuth(auth) {
	return auth.scheme === "Basic";
}
/**
* Type guard to check if authorization is OAuth type.
*/
function isOAuth(auth) {
	return auth.scheme === "OAuth";
}
/**
* Parses key="value" or key=value parameters from a string.
* Used for OAuth and other schemes with comma-separated key-value pairs.
*
* @param parametersString The parameters string.
*/
function parseKeyValueParameters(parametersString) {
	const params = {};
	const paramPattern = /(\w+)=(?:"([^"]*)"|([^,\s]+))/g;
	let match = paramPattern.exec(parametersString);
	while (match !== null) {
		const [, key, quotedValue, unquotedValue] = match;
		params[key] = quotedValue === void 0 ? unquotedValue : quotedValue;
		match = paramPattern.exec(parametersString);
	}
	return params;
}
const authorizationSchema = pipe(string("Expected a string value"), trim(), nonEmpty("Expected a non-empty string value"), rawTransform(({ dataset: { value: authorization }, addIssue, NEVER }) => {
	const spaceIndex = authorization.indexOf(" ");
	if (spaceIndex === -1) {
		addIssue({ message: "Invalid Authorization header format" });
		return NEVER;
	}
	return {
		scheme: authorization.slice(0, spaceIndex),
		parameters: authorization.slice(spaceIndex + 1).trim()
	};
}));
/**
* Parses an Authorization header value into its scheme and parameters.
*
* @param authorization The Authorization header value.
* @throws {CommerceSdkValidationError} If the authorization header is malformed.
*
* @example
* ```typescript
* const auth = parseAuthorization("Bearer token123");
* // { scheme: "Bearer", token: "token123" }
* ```
*
* @example
* ```typescript
* const auth = parseAuthorization("Basic dXNlcjpwYXNz");
* // { scheme: "Basic", credentials: "dXNlcjpwYXNz" }
* ```
*
* @example
* ```typescript
* const auth = parseAuthorization('OAuth oauth_consumer_key="key", oauth_token="token"');
* // { scheme: "OAuth", parameters: { oauth_consumer_key: "key", oauth_token: "token" } }
* ```
*
* @example
* ```typescript
* const auth = parseAuthorization('Digest realm="Example", nonce="abc"');
* // { scheme: "Digest", rawParameters: 'realm="Example", nonce="abc"' }
*
* const parameters = auth.parseParameters();
* // { realm: "Example", nonce: "abc" }
* ```
*/
function parseAuthorization(authorization) {
	const parsed = safeParse(authorizationSchema, authorization);
	if (!parsed.success) throw new CommerceSdkValidationError("Invalid Authorization header format", { issues: parsed.issues });
	const { scheme, parameters } = parsed.output;
	if (scheme === "Bearer") return {
		scheme: "Bearer",
		token: parameters
	};
	if (scheme === "Basic") return {
		scheme: "Basic",
		credentials: parameters
	};
	if (scheme === "OAuth") return {
		scheme: "OAuth",
		parameters: parseKeyValueParameters(parameters)
	};
	return {
		scheme,
		rawParameters: parameters,
		parseParameters: () => parseKeyValueParameters(parameters)
	};
}
const bearerSchema = object({ scheme: literal("Bearer") });
/**
* Parses and validates a Bearer authorization header.
*
* @param authorization The Authorization header value.
* @throws {CommerceSdkValidationError} If the authorization header is not a Bearer token or if the token is empty.
*
* @example
* ```typescript
* const auth = parseBearerToken("Bearer eyJhbGci...");
* console.log(auth.token); // "eyJhbGci..."
* ```
*
* @example
* ```typescript
* try {
*   parseBearerToken("Basic dXNlcjpwYXNz");
* } catch (error) {
*   console.error(error.message); // "Authorization header must be a Bearer token"
* }
* ```
*/
function parseBearerToken(authorization) {
	const auth = parseAuthorization(authorization);
	const parsed = safeParse(bearerSchema, auth);
	if (!parsed.success) throw new CommerceSdkValidationError("Authorization header must be a Bearer token", { issues: parsed.issues });
	return auth;
}
const basicSchema = object({ scheme: literal("Basic") });
/**
* Parses and validates a Basic authorization header.
*
* @param authorization The Authorization header value.
* @throws {CommerceSdkValidationError} If the authorization header is not Basic auth.
*
* @example
* ```typescript
* const auth = parseBasicToken("Basic dXNlcjpwYXNz");
* console.log(auth.credentials); // "dXNlcjpwYXNz"
* ```
*/
function parseBasicToken(authorization) {
	const auth = parseAuthorization(authorization);
	const parsed = safeParse(basicSchema, auth);
	if (!parsed.success) throw new CommerceSdkValidationError("Authorization header must be Basic authentication", { issues: parsed.issues });
	return auth;
}
const oauthSchema = object({ scheme: literal("OAuth") });
/**
* Parses and validates an OAuth 1.0 authorization header.
*
* @param authorization The Authorization header value.
* @throws {CommerceSdkValidationError} If the authorization header is not OAuth.
*
* @example
* ```typescript
* const auth = parseOAuthToken('OAuth oauth_consumer_key="key", oauth_token="token"');
* console.log(auth.parameters.oauth_consumer_key); // "key"
* console.log(auth.parameters.oauth_token); // "token"
* ```
*/
function parseOAuthToken(authorization) {
	const auth = parseAuthorization(authorization);
	const parsed = safeParse(oauthSchema, auth);
	if (!parsed.success) throw new CommerceSdkValidationError("Authorization header must be OAuth", { issues: parsed.issues });
	return auth;
}

//#endregion
export { assertRequiredHeaders, createHeaderAccessor, getHeader, getHeadersFromParams, getMissingOrEmptyHeaders, isBasicAuth, isBearerAuth, isOAuth, parseAuthorization, parseBasicToken, parseBearerToken, parseOAuthToken };
