// @adobe/aio-commerce-lib-core/dist/es/responses/index.mjs
/**
 * @license
 *
 * Copyright 2026 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */

//#region source/responses/helpers.ts
/**
* Creates a standardized error response for runtime actions
* @see https://developer.adobe.com/app-builder/docs/guides/runtime_guides/creating-actions#unsuccessful-response
*
* @template TBody - Additional error body properties beyond the required message field
* @template THeaders - Custom response headers
*
* @param statusCode - HTTP status code (e.g., 400, 404, 500)
* @param payload - Error response configuration
* @param payload.message - Human-readable error message (required)
* @param payload.body - Optional additional error details to include in the response body
* @param payload.headers - Optional custom response headers
*
* @returns Standardized error response object with type discriminator
*
* @example
* ```typescript
* // Simple error with just a message
* const error = buildErrorResponse(404, {
*   message: 'Resource not found'
* });
*
* // Error with additional body data
* const error = buildErrorResponse(400, {
*   message: 'Invalid request',
*   body: { field: 'email', code: 'INVALID_FORMAT' }
* });
*
* // Error with custom headers
* const error = buildErrorResponse(429, {
*   message: 'Rate limit exceeded',
*   headers: { 'Retry-After': '60' }
* });
* ```
*/
function buildErrorResponse(statusCode, payload) {
	return {
		type: "error",
		error: {
			...payload?.headers && { headers: payload.headers },
			statusCode,
			body: payload.body
		}
	};
}
/**
* Creates a standardized success response for runtime actions
* @see https://developer.adobe.com/app-builder/docs/guides/runtime_guides/creating-actions#successful-response
*
* @template TBody - Response body properties
* @template THeaders - Custom response headers
*
* @param statusCode - HTTP status code (typically 200, 201, 204, etc.)
* @param payload - Success response configuration
* @param payload.message - Human-readable success message (required)
* @param payload.body - Optional additional response data to include in the response body
* @param payload.headers - Optional custom response headers
*
* @returns Standardized success response object with type discriminator
*
* @example
* ```typescript
* // Simple success response
* const response = buildSuccessResponse(200, {
*   message: 'Operation successful'
* });
*
* // Success with additional body data
* const response = buildSuccessResponse(201, {
*   message: 'Resource created',
*   body: { id: '456', created: true },
*   headers: { 'Location': '/api/resources/456' }
* });
* ```
*/
function buildSuccessResponse(statusCode, payload) {
	return {
		type: "success",
		statusCode,
		...payload?.headers && { headers: payload.headers },
		...payload?.body && { body: payload.body }
	};
}

//#endregion
//#region source/responses/presets.ts
const HTTP_OK = 200;
const HTTP_CREATED = 201;
const HTTP_ACCEPTED = 202;
const HTTP_NON_AUTHORITATIVE_INFORMATION = 203;
const HTTP_NO_CONTENT = 204;
const HTTP_BAD_REQUEST = 400;
const HTTP_UNAUTHORIZED = 401;
const HTTP_FORBIDDEN = 403;
const HTTP_NOT_FOUND = 404;
const HTTP_METHOD_NOT_ALLOWED = 405;
const HTTP_CONFLICT = 409;
const HTTP_INTERNAL_SERVER_ERROR = 500;
function curryBuildSuccessResponse(code) {
	return (payload) => buildSuccessResponse(code, typeof payload === "string" ? { body: { message: payload } } : payload);
}
function curryBuildErrorResponse(code) {
	return (payload) => buildErrorResponse(code, typeof payload === "string" ? { body: { message: payload } } : payload);
}
/**
* Creates a success response with the HTTP status code 200.
* See {@link buildSuccessResponse} for details on the response payload.
*/
const ok = curryBuildSuccessResponse(200);
/**
* Creates a success response with the HTTP status code 201.
* See {@link buildSuccessResponse} for details on the response payload.
*/
const created = curryBuildSuccessResponse(201);
/**
* Creates a success response with the HTTP status code 202.
* See {@link buildSuccessResponse} for details on the response payload.
*/
const accepted = curryBuildSuccessResponse(202);
/**
* Creates a success response with the HTTP status code 203.
* See {@link buildSuccessResponse} for details on the response payload.
*/
const nonAuthoritativeInformation = curryBuildSuccessResponse(203);
/**
* Creates a success response with the HTTP status code 204 (No Content).
* This response has no body.
*/
const noContent = () => buildSuccessResponse(204);
/**
* Creates an error response with the HTTP status code 400.
* See {@link buildErrorResponse} for details on the response payload.
*/
const badRequest = curryBuildErrorResponse(400);
/**
* Creates an error response with the HTTP status code 401.
* See {@link buildErrorResponse} for details on the response payload.
*/
const unauthorized = curryBuildErrorResponse(401);
/**
* Creates an error response with the HTTP status code 403.
* See {@link buildErrorResponse} for details on the response payload.
*/
const forbidden = curryBuildErrorResponse(403);
/**
* Creates an error response with the HTTP status code 404.
* See {@link buildErrorResponse} for details on the response payload.
*/
const notFound = curryBuildErrorResponse(404);
/**
* Creates an error response with the HTTP status code 405.
* See {@link buildErrorResponse} for details on the response payload.
*/
const methodNotAllowed = curryBuildErrorResponse(405);
/**
* Creates an error response with the HTTP status code 409.
* See {@link buildErrorResponse} for details on the response payload.
*/
const conflict = curryBuildErrorResponse(409);
/**
* Creates an error response with the HTTP status code 500.
* See {@link buildErrorResponse} for details on the response payload.
*/
const internalServerError = curryBuildErrorResponse(500);

//#endregion
export { HTTP_ACCEPTED, HTTP_BAD_REQUEST, HTTP_CONFLICT, HTTP_CREATED, HTTP_FORBIDDEN, HTTP_INTERNAL_SERVER_ERROR, HTTP_METHOD_NOT_ALLOWED, HTTP_NON_AUTHORITATIVE_INFORMATION, HTTP_NOT_FOUND, HTTP_NO_CONTENT, HTTP_OK, HTTP_UNAUTHORIZED, accepted, badRequest, buildErrorResponse, buildSuccessResponse, conflict, created, forbidden, internalServerError, methodNotAllowed, noContent, nonAuthoritativeInformation, notFound, ok, unauthorized };
