// @adobe/aio-commerce-lib-api/dist/es/hooks-Dc1jLa88.mjs
/**
 * @license
 *
 * Copyright 2026 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */

import { getImsAuthProvider, getIntegrationAuthProvider } from "@adobe/aio-commerce-lib-auth";

//#region source/utils/auth/hooks.ts
/**
* Type guard to check if the given auth object is an auth provider.
* @param auth The auth object to check.
*/
function isAuthProvider(auth) {
	return "getHeaders" in auth;
}
/**
* Builds a before request hook for integration authentication.
* @param integrationAuth The integration authentication parameters or integration auth provider.
*/
function buildIntegrationAuthBeforeRequestHook(integrationAuth) {
	const integrationAuthProvider = isAuthProvider(integrationAuth) ? integrationAuth : getIntegrationAuthProvider(integrationAuth);
	return (request) => {
		const headers = integrationAuthProvider.getHeaders(request.method.toUpperCase(), request.url);
		request.headers.set("Authorization", headers.Authorization);
	};
}
/**
* Builds a before request hook for IMS authentication.
* @param imsAuth The IMS authentication parameters or IMS auth provider.
*/
function buildImsAuthBeforeRequestHook(imsAuth) {
	const imsAuthProvider = isAuthProvider(imsAuth) ? imsAuth : getImsAuthProvider(imsAuth);
	return async (request) => {
		const headers = await imsAuthProvider.getHeaders();
		request.headers.set("Authorization", headers.Authorization);
		const xApiKey = headers["x-api-key"];
		if (xApiKey) request.headers.set("x-api-key", xApiKey);
	};
}

//#endregion
export { buildIntegrationAuthBeforeRequestHook as n, isAuthProvider as r, buildImsAuthBeforeRequestHook as t };
