// @adobe/aio-commerce-lib-api/dist/es/index.mjs
/**
 * @license
 *
 * Copyright 2026 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */

import { n as buildIntegrationAuthBeforeRequestHook, r as isAuthProvider, t as buildImsAuthBeforeRequestHook } from "./hooks-Dc1jLa88.mjs";
import { forwardImsAuthProvider, isImsAuthProvider, resolveAuthParams } from "@adobe/aio-commerce-lib-auth";
import { allNonEmpty } from "@adobe/aio-commerce-lib-core/params";
import * as kyModule from "ky";

//#region source/lib/api-client.ts
/** A client that binds a set of {@link ApiFunction} to a given HTTP client. */
var ApiClient = class {
	/**
	* Creates a new API client that binds a set of {@link ApiFunction} to a given HTTP client.
	* @param client - The HTTP client to bind the API functions to.
	* @param functions - The API functions to bind to the HTTP client.
	*/
	static create(client, functions) {
		const wrapped = {};
		for (const key in functions) if (Object.hasOwn(functions, key)) {
			const fn = functions[key];
			wrapped[key] = ((...args) => fn(client, ...args));
		}
		return wrapped;
	}
};

//#endregion
//#region source/utils/auth/ims-scopes.ts
/**
* Ensures the correct scopes are set for the given IMS authentication parameters.
* @param imsAuth - The IMS authentication parameters.
* @param requiredScopes - The required scopes.
*/
function ensureImsScopes(imsAuth, requiredScopes) {
	const scopes = new Set([...imsAuth.scopes ?? [], ...requiredScopes]);
	return {
		...imsAuth,
		scopes: Array.from(scopes)
	};
}

//#endregion
//#region source/utils/http/ky.ts
const ky = kyModule.default.default ?? kyModule.default;
/**
* Creates a Ky instance with the provided options.
* @param options - The options for the Ky instance.
*/
function createKy(options) {
	return ky.create(options);
}
/**
* Extends the given Ky instance with the provided options if they are provided.
* @param ky - The Ky instance to extend.
* @param options - The options to extend the Ky instance with.
*/
function optionallyExtendKy(kyInstance, options) {
	return options ? kyInstance.extend(options) : kyInstance;
}

//#endregion
//#region source/lib/commerce/helpers.ts
/** A regex matching a regular SaaS API URL, with a tenant ID and optional trailing slash. */
const COMMERCE_API_URL_REGEX = /^([a-zA-Z0-9-]+\.)?api\.commerce\.adobe\.com\/[a-zA-Z0-9-]+\/?$/;
const COMMERCE_IMS_REQUIRED_SCOPES = ["openid", "additional_info.projectedProductContext"];
const COMMERCE_SAAS_IMS_REQUIRED_SCOPES = [...COMMERCE_IMS_REQUIRED_SCOPES, "commerce.accs"];
/**
* Gets the Commerce URL for the given configuration and flavor.
* @param config The Commerce HTTP client configuration.
*/
function getCommerceUrl(config) {
	const { baseUrl, storeViewCode, version } = config;
	const commerceUrl = new URL(baseUrl.endsWith("/") ? baseUrl : `${baseUrl}/`);
	const uppercasedVersion = version?.toUpperCase();
	if (config.flavor === "paas") commerceUrl.pathname += `rest/${storeViewCode}/${uppercasedVersion}`;
	else if (config.flavor === "saas") commerceUrl.pathname += `${uppercasedVersion}`;
	return commerceUrl.toString();
}
/**
* Builds the appropriate auth hook based on the auth provider type.
* Supports both IMS and Integration authentication.
* @param auth The authentication provider (IMS or Integration).
*/
function buildAuthBeforeRequestHook(flavor, auth) {
	if (isImsAuthProvider(auth) || "clientId" in auth) return isAuthProvider(auth) ? buildImsAuthBeforeRequestHook(auth) : buildImsAuthBeforeRequestHook(ensureImsScopes(auth, flavor === "saas" ? COMMERCE_SAAS_IMS_REQUIRED_SCOPES : COMMERCE_IMS_REQUIRED_SCOPES));
	return buildIntegrationAuthBeforeRequestHook(auth);
}
/**
* Builds a Commerce HTTP client for PaaS.
* @param params The parameters for building the Commerce PaaS HTTP client.
*/
function buildCommerceHttpClientPaaS(params) {
	const { auth, config, fetchOptions } = params;
	return optionallyExtendKy(createKy({
		prefixUrl: getCommerceUrl(config),
		hooks: { beforeRequest: [buildAuthBeforeRequestHook("paas", auth)] }
	}), fetchOptions);
}
/**
* Builds a Commerce HTTP client for SaaS.
* @param params The parameters for building the Commerce SaaS HTTP client.
*/
function buildCommerceHttpClientSaaS(params) {
	const { auth, config, fetchOptions } = params;
	return optionallyExtendKy(createKy({
		prefixUrl: getCommerceUrl(config),
		hooks: { beforeRequest: [buildAuthBeforeRequestHook("saas", auth), (request) => {
			request.headers.set("Store", config.storeViewCode);
		}] }
	}), fetchOptions);
}
/**
* Type guard to check if params are for PaaS
*/
function isPaaSParams(params) {
	return params.config.flavor === "paas";
}
/**
* Type guard to check if params are for SaaS
*/
function isSaaSParams(params) {
	return params.config.flavor === "saas";
}
/**
* Builds the Commerce HTTP client for the given parameters.
* @param params The configuration, authentication and fetch options parameters.
*/
function buildCommerceHttpClient(params) {
	const flavor = params.config.flavor;
	if (isPaaSParams(params)) return buildCommerceHttpClientPaaS(params);
	if (isSaaSParams(params)) return buildCommerceHttpClientSaaS(params);
	throw new Error(`Invalid Commerce configuration. Unknown flavor: ${flavor}`);
}
/**
* Type guard to check if a value is a valid {@link CommerceFlavor}.
* @param input The value to check.
* @returns True if the value is a valid CommerceFlavor, false otherwise.
*/
function isFlavor(input) {
	return input === "saas" || input === "paas";
}
/**
* Resolves the {@link CommerceFlavor} from the given API URL.
* @param apiUrl The API URL to resolve the {@link CommerceFlavor} from.
*/
function resolveCommerceFlavorFromApiUrl(apiUrl) {
	const { hostname, pathname } = new URL(apiUrl);
	const hostAndPath = `${hostname}${pathname}`;
	return COMMERCE_API_URL_REGEX.test(hostAndPath) ? "saas" : "paas";
}
/**
* Resolves the {@link CommerceHttpClientParams} from the given App Builder action inputs.
* @param params The App Builder action inputs to resolve the {@link CommerceHttpClientParams} from.
* @throws If the base API URL or the authentication parameters cannot be resolved.
* @example
* ```typescript
* import { resolveCommerceHttpClientParams, AdobeCommerceHttpClient } from "@adobe/aio-commerce-lib-api/commerce";
*
* // Some App Builder runtime action that needs the Commerce HTTP client
* export function main(params) {
*   const commerceHttpClientParams = resolveCommerceHttpClientParams(params);
*   const commerceClient = new AdobeCommerceHttpClient(commerceHttpClientParams);
* }
* ```
*/
function resolveCommerceHttpClientParams(params, options = {}) {
	if (allNonEmpty(params, ["AIO_COMMERCE_API_BASE_URL"])) {
		const baseUrl = String(params.AIO_COMMERCE_API_BASE_URL);
		const flavor = isFlavor(params.AIO_COMMERCE_API_FLAVOR) ? params.AIO_COMMERCE_API_FLAVOR : resolveCommerceFlavorFromApiUrl(baseUrl);
		const { tryForwardAuthProvider = false } = options;
		let clientAuth;
		if (tryForwardAuthProvider) clientAuth = forwardImsAuthProvider(params);
		else {
			const authParams = resolveAuthParams(params);
			if (flavor === "saas" && authParams.strategy !== "ims") throw new Error("Resolved incorrect auth parameters for SaaS. Only IMS auth is supported");
			clientAuth = authParams;
		}
		return {
			auth: clientAuth,
			config: {
				baseUrl,
				flavor
			}
		};
	}
	throw new Error("Can not resolve CommerceHttpClientParams without an AIO_COMMERCE_API_BASE_URL. Please provide one.");
}

//#endregion
//#region source/lib/http-client-base.ts
/**
* Base class for HTTP clients.
* @template T The type of the configuration object.
*/
var HttpClientBase = class HttpClientBase {
	/**
	* Creates a new HTTP client instance.
	* @param config The configuration used by the HTTP client.
	* @param httpClient The actual HTTP client instance.
	*/
	constructor(config, httpClient) {
		this.config = Object.freeze(config);
		this.setHttpClient(httpClient);
	}
	/**
	* Sets the HTTP client instance.
	* @param httpClient The HTTP client instance to set.
	*/
	setHttpClient(httpClient) {
		this.httpClient = Object.freeze(httpClient);
		this.get = this.httpClient.get.bind(this.httpClient);
		this.post = this.httpClient.post.bind(this.httpClient);
		this.put = this.httpClient.put.bind(this.httpClient);
		this.delete = this.httpClient.delete.bind(this.httpClient);
		this.patch = this.httpClient.patch.bind(this.httpClient);
		this.head = this.httpClient.head.bind(this.httpClient);
		this.retry = this.httpClient.retry.bind(this.httpClient);
		this.stop = this.httpClient.stop;
	}
	/**
	* Extends the current HTTP client instance with the given options.
	* @param options The options to extend the HTTP client with.
	*/
	extend(options) {
		const client = Object.freeze(this.httpClient.extend(options));
		return new HttpClientBase(this.config, client);
	}
};

//#endregion
//#region source/lib/commerce/http-client.ts
const DEFAULT_STORE_VIEW_CODE = "all";
const DEFAULT_API_VERSION = "V1";
/**
* A Ky-based HTTP client used to make requests to the Commerce API.
* @see https://github.com/sindresorhus/ky
*/
var AdobeCommerceHttpClient = class extends HttpClientBase {
	/**
	* Creates a new Commerce HTTP client instance.
	* @param params The parameters for building the Commerce HTTP client.
	*/
	constructor(params) {
		const config = {
			...params.config,
			storeViewCode: params.config.storeViewCode ?? DEFAULT_STORE_VIEW_CODE,
			version: params.config.version ?? DEFAULT_API_VERSION
		};
		const httpClient = buildCommerceHttpClient({
			...params,
			config
		});
		super(config, httpClient);
	}
};

//#endregion
//#region source/lib/io-events/helpers.ts
const IO_EVENTS_IMS_REQUIRED_SCOPES = ["adobeio_api"];
/**
* Builds the Adobe I/O Events HTTP client for the given parameters.
* @param params The configuration, authentication and fetch options parameters.
*/
function buildIoEventsHttpClient(params) {
	const { auth, config, fetchOptions } = params;
	const beforeRequestAuthHook = isAuthProvider(auth) ? buildImsAuthBeforeRequestHook(auth) : buildImsAuthBeforeRequestHook(ensureImsScopes(auth, IO_EVENTS_IMS_REQUIRED_SCOPES));
	const adobeIoBaseUrl = config.baseUrl;
	return optionallyExtendKy(createKy({
		prefixUrl: adobeIoBaseUrl,
		headers: { Accept: "application/hal+json" },
		hooks: { beforeRequest: [beforeRequestAuthHook] }
	}), fetchOptions);
}
/**
* Resolves the {@link IoEventsHttpClientParams} from the given App Builder action inputs.
* @param params The App Builder action inputs to resolve the {@link IoEventsHttpClientParams} from.
* @throws {Error} If the authentication parameters cannot be resolved or if non-IMS auth is detected.
* @example
* ```typescript
* import { resolveIoEventsHttpClientParams, AdobeIoEventsHttpClient } from "@adobe/aio-commerce-lib-api/io-events";
*
* export const main = async function (params: Record<string, unknown>) {
*   // Automatically resolves IMS auth params from environment variables
*   const clientParams = resolveIoEventsHttpClientParams(params);
*   const eventsClient = new AdobeIoEventsHttpClient(clientParams);
*
*   // Use the client
*   const response = await eventsClient.get("{orgId}/providers").json();
*   return { statusCode: 200, body: response };
* };
* ```
*/
function resolveIoEventsHttpClientParams(params, options = {}) {
	const { tryForwardAuthProvider = false } = options;
	let clientAuth;
	if (tryForwardAuthProvider) clientAuth = forwardImsAuthProvider(params);
	else {
		const authParams = resolveAuthParams(params);
		if (authParams.strategy !== "ims") throw new Error("Resolved incorrect auth parameters for I/O Events. Only IMS auth is supported");
		clientAuth = authParams;
	}
	return {
		auth: clientAuth,
		config: { baseUrl: params.AIO_EVENTS_API_BASE_URL ? String(params.AIO_EVENTS_API_BASE_URL) : void 0 }
	};
}

//#endregion
//#region source/lib/io-events/http-client.ts
const DEFAULT_IO_EVENTS_BASE_URL = "https://api.adobe.io/events";
/**
* A Ky-based HTTP client used to make requests to the Adobe I/O Events API.
* @see https://github.com/sindresorhus/ky
*/
var AdobeIoEventsHttpClient = class extends HttpClientBase {
	/**
	* Creates a new Adobe I/O Events HTTP client instance.
	* @param params The parameters for building the Adobe I/O Events HTTP client.
	*/
	constructor(params) {
		const config = {
			...params.config,
			baseUrl: params.config?.baseUrl ?? DEFAULT_IO_EVENTS_BASE_URL
		};
		const httpClient = buildIoEventsHttpClient({
			...params,
			config
		});
		super(config, httpClient);
	}
};

//#endregion
export { AdobeCommerceHttpClient, AdobeIoEventsHttpClient, ApiClient, resolveCommerceHttpClientParams, resolveIoEventsHttpClientParams };
