// @adobe/aio-commerce-lib-app/dist/cjs/config/index.cjs
/**
 * @license
 *
 * Copyright 2026 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
const require_config = require('../config-D-s0GSep.cjs');
const require_webhooks = require('../webhooks-tbZy9G3p.cjs');
const require_validate = require('../validate-oAOpyFyb.cjs');

exports.CommerceAppConfigSchemas = require_validate.CommerceAppConfigSchemas;
exports.defineConfig = require_config.defineConfig;
exports.getConfigDomains = require_validate.getConfigDomains;
exports.hasAdminUiSdk = require_webhooks.hasAdminUiSdk;
exports.hasBusinessConfig = require_validate.hasBusinessConfig;
exports.hasBusinessConfigSchema = require_validate.hasBusinessConfigSchema;
exports.hasCommerceEvents = require_webhooks.hasCommerceEvents;
exports.hasConfigDomain = require_validate.hasConfigDomain;
exports.hasCustomInstallation = require_webhooks.hasCustomInstallation;
exports.hasCustomInstallationSteps = require_webhooks.hasCustomInstallationSteps;
exports.hasEventing = require_webhooks.hasEventing;
exports.hasExternalEvents = require_webhooks.hasExternalEvents;
exports.hasMetadata = require_validate.hasMetadata;
exports.hasWebhooks = require_webhooks.hasWebhooks;
exports.parseCommerceAppConfig = require_config.parseCommerceAppConfig;
exports.readCommerceAppConfig = require_config.readCommerceAppConfig;
exports.resolveCommerceAppConfig = require_config.resolveCommerceAppConfig;
exports.validateCommerceAppConfig = require_validate.validateCommerceAppConfig;
exports.validateCommerceAppConfigDomain = require_validate.validateCommerceAppConfigDomain;
