// @adobe/aio-commerce-lib-core/dist/cjs/error-sJoC2Xru.cjs
/**
 * @license
 *
 * Copyright 2026 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */

//#region \0rolldown/runtime.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __copyProps = (to, from, except, desc) => {
	if (from && typeof from === "object" || typeof from === "function") {
		for (var keys = __getOwnPropNames(from), i = 0, n = keys.length, key; i < n; i++) {
			key = keys[i];
			if (!__hasOwnProp.call(to, key) && key !== except) {
				__defProp(to, key, {
					get: ((k) => from[k]).bind(null, key),
					enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
				});
			}
		}
	}
	return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
	value: mod,
	enumerable: true
}) : target, mod));

//#endregion
let util = require("util");
util = __toESM(util, 1);
let ansis = require("ansis");
let valibot = require("valibot");

//#region source/error/base-error.ts
/**
* Base class for all the errors in the AIO Commerce SDK.
* @example
* ```ts
* class ValidationError extends CommerceSdkErrorBase {
*   constructor(message: string, options: ValidationErrorOptions) {
*     super(message, options);
*   }
* }
*
* const err = new ValidationError("Invalid value", {
*   tag: "ValidationError",
*   field: "name",
*   value: "John Doe",
* });
*
* console.log(err.toJSON());
* ```
*/
var CommerceSdkErrorBase = class CommerceSdkErrorBase extends Error {
	/**
	* Constructs a new CommerceSdkErrorBase instance. This is an abstract class so you
	* should not instantiate it directly. Only invoke this constructor from a subclass.
	*
	* @param message - A human-friendly description of the error.
	* @param options - Optional error options (additional information).
	*/
	constructor(message, options) {
		const { traceId, ...baseOptions } = options ?? {};
		super(message, baseOptions);
		Object.setPrototypeOf(this, new.target.prototype);
		if (Error.captureStackTrace) Error.captureStackTrace(this, new.target);
		this.name = new.target.name || "CommerceSdkError";
		this.traceId = traceId;
	}
	/**
	* Checks if the error is any CommerceSdkErrorBase instance.
	* @example
	* ```ts
	* class ValidationError extends CommerceSdkErrorBase {}
	* const err = new ValidationError("Invalid", {});
	*
	* CommerceSdkErrorBase.isSdkError(err); // true
	* ValidationError.isSdkError(err); // true
	* CommerceSdkErrorBase.isSdkError(new Error("Regular")); // false
	* ```
	*/
	static isSdkError(error) {
		return error instanceof CommerceSdkErrorBase;
	}
	/** Returns the full stack trace of the error and its causes. */
	get fullStack() {
		let out = this.stack ?? "";
		let cause = this.cause;
		while (cause instanceof Error) {
			out += `\nCaused by: ${cause.stack ?? cause.message}`;
			cause = cause.cause;
		}
		return out;
	}
	/** Returns the root cause of the error. */
	get rootCause() {
		let cause = this.cause;
		while (cause) if (typeof cause === "object" && cause !== null && "cause" in cause) cause = cause.cause;
		else break;
		return cause;
	}
	/** Converts the error to a JSON-like representation. */
	toJSON() {
		return {
			name: this.name,
			message: this.message,
			stack: this.fullStack,
			cause: this.cause,
			traceId: this.traceId
		};
	}
	/**
	* Returns a pretty string representation of the error.
	* @param inspect - Whether to inspect the error (returns a more detailed string, useful for debugging).
	*/
	toString(inspect = true) {
		if (inspect) return util.default.inspect(this, {
			depth: null,
			colors: true,
			sorted: true,
			maxStringLength: Number.POSITIVE_INFINITY
		});
		return super.toString();
	}
};

//#endregion
//#region source/error/validation-error.ts
const LAST_RETURN_CHAR = "└── ";
const RETURN_CHAR = "├── ";
/** Maps issue kinds to their corresponding error titles for display purposes. */
const ISSUE_KIND_TO_ERROR_TITLE = {
	schema: "Schema validation error",
	transformation: "Transformation error",
	validation: "Input error"
};
function getReturnChar(isLastItem, withColor) {
	const char = isLastItem ? LAST_RETURN_CHAR : RETURN_CHAR;
	return withColor ? (0, ansis.cyanBright)(char) : char;
}
function getKindText(kind, withColor) {
	const text = ISSUE_KIND_TO_ERROR_TITLE[kind] ?? "Unknown issue kind";
	return withColor ? (0, ansis.yellowBright)(text) : text;
}
function getPathText(dotPath, withColor) {
	if (!dotPath) return "";
	return withColor ? `${(0, ansis.cyanBright)(dotPath)}${(0, ansis.whiteBright)((0, ansis.dim)(" →"))}` : `${dotPath} →`;
}
function issueToDisplay(issues, withColor = true) {
	return (issues?.map((issue, index) => {
		const returnChar = getReturnChar(index === issues.length - 1, withColor);
		const kindText = getKindText(issue.kind, withColor);
		const path = getPathText((0, valibot.getDotPath)(issue), withColor);
		const message = withColor ? (0, ansis.whiteBright)(issue.message) : issue.message;
		return `${returnChar} ${`${kindText} ${withColor ? (0, ansis.whiteBright)("at") : "at"} ${path} ${message}`}`;
	}))?.join("\n") ?? "";
}
function displayValidationError(error, withColor = true) {
	const display = issueToDisplay(error.issues, withColor);
	return `${withColor ? (0, ansis.whiteBright)(error.message) : error.message}\n${display}`;
}
/**
* Represents a validation error in the Commerce SDK.
* This error should be thrown when an input does not conform to the expected schema.
* It contains a list of issues that describe the validation errors.
*
* @example
* ```ts
* const error = new CommerceSdkValidationError("Invalid input", {
*   // `someIssues` is in scope, returned by some `valibot` operation.
*   issues: someIssues
* });
*
* console.log(error.display());
* ```
*/
var CommerceSdkValidationError = class extends CommerceSdkErrorBase {
	/**
	* Constructs a new {@link CommerceSdkValidationError} instance.
	*
	* @param message - A human-friendly description of the validation error.
	* @param options - Options for the validation error, including the issues that caused the error.
	*/
	constructor(message, { issues, ...options }) {
		super(message, options);
		this.issues = issues;
	}
	/**
	* Returns a pretty string (and colored) representation of the validation error.
	* @param withColor Whether to use color in the output.
	*/
	display(withColor = true) {
		return displayValidationError(this, withColor);
	}
};

//#endregion
Object.defineProperty(exports, 'CommerceSdkErrorBase', {
  enumerable: true,
  get: function () {
    return CommerceSdkErrorBase;
  }
});
Object.defineProperty(exports, 'CommerceSdkValidationError', {
  enumerable: true,
  get: function () {
    return CommerceSdkValidationError;
  }
});
Object.defineProperty(exports, '__toESM', {
  enumerable: true,
  get: function () {
    return __toESM;
  }
});
