// @adobe/aio-commerce-lib-app/dist/es/commands/index.mjs
#!/usr/bin/env node
/**
 * @license
 *
 * Copyright 2026 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */

import { t as stringifyError } from "../error-_RMwtrUA.mjs";
import { a as getExecCommand, c as isESM, i as detectPackageManager, l as makeOutputDirFor, n as readCommerceAppConfig, o as getInstallCommand, r as resolveCommerceAppConfig, s as getProjectRootDirectory, t as parseCommerceAppConfig, u as readPackageJson } from "../config-3I4eLyyR.mjs";
import { a as hasCustomInstallationSteps, d as hasAdminUiSdk } from "../webhooks-B5oq9ULL.mjs";
import { o as getConfigDomains, r as requiresInstallation, t as validateCommerceAppConfig, u as hasBusinessConfigSchema } from "../validate-Dfajrp99.mjs";
import { t as inspect } from "../logging-DbJHXSKh.mjs";
import { CommerceSdkValidationError } from "@adobe/aio-commerce-lib-core/error";
import consola, { consola as consola$1 } from "consola";
import path, { basename, dirname, extname, join, relative, resolve, sep } from "path";
import { hasDynamicSchema } from "@adobe/aio-commerce-lib-config";
import * as prettier from "prettier";
import { existsSync, readFileSync, writeFileSync } from "fs";
import { access, readFile, rm, writeFile } from "fs/promises";
import { fileURLToPath } from "url";
import { Document, YAMLMap, YAMLSeq, isMap, isPair, isSeq, parseDocument } from "yaml";
import NpmPackageJson from "@npmcli/package-json";
import { colorize, formatTree } from "consola/utils";
import { build } from "esbuild";
import stringify, { stringify as stringify$1 } from "safe-stable-stringify";
import { execSync, spawnSync } from "child_process";
import config from "@adobe/aio-lib-core-config";
import aioIms from "@adobe/aio-lib-ims";
import dotenv from "dotenv";

//#region source/commands/constants.ts
/** The name of the package containing all the generated runtime actions */
const PACKAGE_NAME = "app-management";
/** The path to the directory containing the generated files */
const GENERATED_PATH = ".generated";
/** The folder name containing the runtime actions */
const ACTIONS_FOLDER = `actions/${PACKAGE_NAME}`;
/** The path to the directory containing the generated actions */
const GENERATED_ACTIONS_PATH = `${GENERATED_PATH}/${ACTIONS_FOLDER}`;
/** The name of the extension point for extensibility. */
const EXTENSIBILITY_EXTENSION_POINT_ID = "commerce/extensibility/1";
/** The name of the extension point for configuration. */
const CONFIGURATION_EXTENSION_POINT_ID = "commerce/configuration/1";
/** The name of the extension point for Admin UI SDK. */
const BACKEND_UI_EXTENSION_POINT_ID = "commerce/backend-ui/1";
/** The name of the package containing the admin UI SDK generated actions */
const ADMIN_UI_SDK_PACKAGE_NAME = "admin-ui-sdk";
/** The path to the directory containing the generated Admin UI SDK actions. */
const ADMIN_UI_SDK_ACTIONS_PATH = `${GENERATED_PATH}/actions/registration`;
/** The name of the configuration schema file */
const APP_MANIFEST_FILE = "app.commerce.manifest.json";
/** The name of the generated runtime-safe app config module (dynamic schemas only). */
const RUNTIME_APP_CONFIG_FILE = "app.commerce.config.js";
/** Package import alias for the generated runtime-safe app config module. */
const APP_CONFIG_IMPORT_ALIAS = "#app.commerce.config";
/** The name of the commerce app configuration file */
const COMMERCE_APP_CONFIG_FILE = "app.commerce.config";
/** The name of the configuration schema file */
const CONFIG_SCHEMA_FILE_NAME = "configuration-schema.json";
/** The name of the project package file */
const PACKAGE_JSON_FILE = "package.json";
/** The name of the app configuration file */
const APP_CONFIG_FILE = "app.config.yaml";
/** The name of the install configuration file */
const INSTALL_YAML_FILE = "install.yaml";
/** Get the path to the extension point folder */
function getExtensionPointFolderPath(extensionPointId) {
	return `src/${extensionPointId.replaceAll("/", "-")}`;
}

//#endregion
//#region source/commands/utils.ts
/** Format file content using prettier, inferring the parser from the file path. */
function prettierFormat(content, filepath) {
	return prettier.format(content, {
		semi: true,
		quoteStyle: "double",
		arrowParens: "always",
		bracketSameLine: true,
		bracketSpacing: true,
		trailingComma: "all",
		tabWidth: 2,
		useTabs: false,
		printWidth: 80,
		filepath
	});
}
/** Load the app commerce config */
async function loadAppManifest() {
	const appConfig = await parseCommerceAppConfig();
	consola.debug("Loaded app commerce config");
	return appConfig;
}
/**
* Whether the app config contains anything that can't be JSON-serialized as-is
* and therefore requires generating a runtime-safe ESM config module.
*/
function hasDynamicAppConfig(appConfig) {
	const schema = appConfig.businessConfig?.schema;
	return Array.isArray(schema) && hasDynamicSchema(schema);
}
/**
* Path to an extension point's `.generated` directory, relative to the project root.
* @param extensionPointId - The extension point ID, e.g. "commerce/extensibility/1"
*/
function getGeneratedDir(extensionPointId) {
	return join(getExtensionPointFolderPath(extensionPointId), GENERATED_PATH);
}
/**
* Path to an extension point's generated actions directory, relative to the project root.
* @param extensionPointId - The extension point ID, e.g. "commerce/extensibility/1"
*/
function getActionsDir(extensionPointId) {
	return join(getExtensionPointFolderPath(extensionPointId), GENERATED_ACTIONS_PATH);
}
/**
* Path to a specific generated action file, relative to the project root.
* @param extensionPointId - The extension point ID, e.g. "commerce/extensibility/1"
* @param actionName - The name of the action, e.g. "my-action"
*/
function getActionPath(extensionPointId, actionName) {
	return join(getActionsDir(extensionPointId), `${actionName}.js`);
}
/**
* Path to an extension point's `ext.config.yaml`, relative to the project root.
* @param extensionPointId - The extension point ID, e.g. "commerce/extensibility/1"
*/
function getExtConfigPath(extensionPointId) {
	return join(getExtensionPointFolderPath(extensionPointId), "ext.config.yaml");
}
/**
* Path to an Admin UI SDK generated actions directory, relative to the project root.
* @param extensionPointId - The extension point ID, e.g. "commerce/backend-ui/1"
*/
function getAdminUiSdkActionsDir(extensionPointId) {
	return join(getExtensionPointFolderPath(extensionPointId), ADMIN_UI_SDK_ACTIONS_PATH);
}
/**
* Path to the generated Admin UI SDK registration action file, relative to the project root.
* @param extensionPointId - The extension point ID, e.g. "commerce/backend-ui/1"
*/
function getAdminUiSdkRegistrationActionPath(extensionPointId) {
	return join(getAdminUiSdkActionsDir(extensionPointId), "index.js");
}
/** Path to the generated app manifest JSON file, relative to the project root. */
function getManifestPath() {
	return join(getGeneratedDir(EXTENSIBILITY_EXTENSION_POINT_ID), APP_MANIFEST_FILE);
}
/**
* Path to the generated runtime-safe app config ESM module, relative to the
* project root. Used when the business config schema contains dynamic options.
*/
function getRuntimeAppConfigPath() {
	return join(getGeneratedDir(EXTENSIBILITY_EXTENSION_POINT_ID), RUNTIME_APP_CONFIG_FILE);
}
/** Path to the generated configuration schema file, relative to the project root. */
function getSchemaPath() {
	return join(getGeneratedDir(CONFIGURATION_EXTENSION_POINT_ID), CONFIG_SCHEMA_FILE_NAME);
}

//#endregion
//#region source/commands/generate/actions/config.ts
/** The list of Commerce variables that are required for the runtime actions */
const COMMERCE_VARIABLES = [
	"AIO_COMMERCE_AUTH_IMS_CLIENT_ID",
	"AIO_COMMERCE_AUTH_IMS_CLIENT_SECRETS",
	"AIO_COMMERCE_AUTH_IMS_TECHNICAL_ACCOUNT_ID",
	"AIO_COMMERCE_AUTH_IMS_TECHNICAL_ACCOUNT_EMAIL",
	"AIO_COMMERCE_AUTH_IMS_ORG_ID",
	"AIO_COMMERCE_AUTH_IMS_SCOPES"
];
/** The inputs for the generated runtime actions */
const COMMERCE_ACTION_INPUTS = Object.fromEntries(COMMERCE_VARIABLES.map((variable) => [variable, `$${variable}`]));
const CUSTOM_IMPORTS_PLACEHOLDER = "// {{CUSTOM_SCRIPTS_IMPORTS}}";
const CUSTOM_SCRIPTS_MAP_PLACEHOLDER = "// {{CUSTOM_SCRIPTS_MAP}}";
const CUSTOM_SCRIPTS_LOADER_PLACEHOLDER = "// {{CUSTOM_SCRIPTS_LOADER}}";
const REGISTRATION_JSON_PLACEHOLDER = "// {{REGISTRATION_JSON}}";
/**
* Creates a runtime action configuration.
* @param actionName - The name of the action.
* @param options - Optional configuration options.
*/
function createActionDefinition(actionName, config = {}, options = {}) {
	const def = {
		...options,
		function: `${GENERATED_ACTIONS_PATH}/${actionName}.js`,
		web: options.web ?? "yes",
		runtime: "nodejs:24",
		annotations: {
			"require-adobe-auth": true,
			final: true
		}
	};
	if (config.requiresEncryptionKey) def.inputs = {
		...def.inputs,
		AIO_COMMERCE_CONFIG_ENCRYPTION_KEY: "$AIO_COMMERCE_CONFIG_ENCRYPTION_KEY"
	};
	return def;
}
/**
* Gets the runtime actions to be generated from the ext.config.yaml configuration.
* @param extConfig - The ext.config.yaml configuration.
*/
function getRuntimeActions(extConfig, dir) {
	return Object.entries(extConfig.runtimeManifest?.packages?.["app-management"]?.actions ?? {}).map(([name, _]) => ({
		name,
		templateFile: join(dir, `${name}.js.template`)
	}));
}
/**
* Builds the ext.config.yaml configuration for the extensibility extension.
* @param features - The features that are enabled for the app.
*/
function buildAppManagementExtConfig(appConfig) {
	const extConfig = {
		hooks: { "pre-app-build": "EXTENSION=extensibility/1 $packageExec aio-commerce-lib-app hooks pre-app-build" },
		operations: { workerProcess: [{
			type: "action",
			impl: `${PACKAGE_NAME}/app-config`
		}] },
		runtimeManifest: { packages: { [PACKAGE_NAME]: {
			license: "Apache-2.0",
			actions: { "app-config": createActionDefinition("app-config") }
		} } }
	};
	const needsInstallAction = requiresInstallation(appConfig);
	const hasPasswordFieldsInSchema = hasBusinessConfigSchema(appConfig) && appConfig.businessConfig.schema.some((field) => field.type === "password");
	if (needsInstallAction) {
		extConfig.operations.workerProcess.push({
			type: "action",
			impl: `${PACKAGE_NAME}/installation`
		});
		extConfig.runtimeManifest.packages[PACKAGE_NAME].actions.installation = createActionDefinition("installation", { requiresEncryptionKey: hasPasswordFieldsInSchema }, {
			inputs: {
				...COMMERCE_ACTION_INPUTS,
				LOG_LEVEL: "$LOG_LEVEL"
			},
			limits: { timeout: 6e5 }
		});
	}
	return extConfig;
}
/** Builds the ext.config.yaml configuration for the business configuration extension. */
function buildBusinessConfigurationExtConfig() {
	const actions = [{
		name: "config",
		templateFile: "config.js.template",
		requiresEncryptionKey: true
	}, {
		name: "scope-tree",
		templateFile: "scope-tree.js.template"
	}];
	return {
		hooks: { "pre-app-build": "EXTENSION=configuration/1 $packageExec aio-commerce-lib-app hooks pre-app-build" },
		operations: { workerProcess: actions.map((action) => ({
			type: "action",
			impl: `${PACKAGE_NAME}/${action.name}`
		})) },
		runtimeManifest: { packages: { [PACKAGE_NAME]: {
			license: "Apache-2.0",
			actions: Object.fromEntries(actions.map((action) => [action.name, createActionDefinition(action.name, action)]))
		} } }
	};
}
/**
* Builds the ext.config.yaml configuration for the Admin UI SDK backend-ui extension.
*/
function buildAdminUiSdkExtConfig() {
	return {
		hooks: { "pre-app-build": "EXTENSION=backend-ui/1 $packageExec aio-commerce-lib-app hooks pre-app-build" },
		operations: { view: [{
			type: "web",
			impl: "index.html"
		}] },
		web: "web-src",
		runtimeManifest: { packages: { [ADMIN_UI_SDK_PACKAGE_NAME]: {
			license: "Apache-2.0",
			actions: { registration: {
				function: `${ADMIN_UI_SDK_ACTIONS_PATH}/index.js`,
				web: "yes",
				runtime: "nodejs:24",
				annotations: {
					"require-adobe-auth": true,
					final: true
				}
			} }
		} } }
	};
}

//#endregion
//#region ../../packages-private/scripting-utils/source/yaml/helpers.ts
/**
* Read a YAML file and return a {@link Document}
* @param path - The path to the YAML file
*/
async function readYamlFile(path) {
	let doc = new Document();
	if (existsSync(path)) try {
		doc = parseDocument(await readFile(path, "utf-8"), { keepSourceTokens: true });
	} catch {
		const file = basename(path);
		throw new Error(`Failed to parse ${file}`);
	}
	if (doc.contents === null) doc.contents = new YAMLMap();
	return doc;
}
/**
* Generic helper to get or create a YAML node at the given path
* @param doc - The YAML document
* @param path - The path to the node
* @param options - The options for the node
* @param typeConfig - Configuration for the node type
* @returns The existing or newly created node
*/
function getOrCreateNode(doc, path, options, typeConfig) {
	const node = doc.getIn(path);
	if (node) {
		if (!typeConfig.typeGuard(node)) throw new Error(`Expected ${typeConfig.typeName} at path "${path.join(".")}".`);
		return node;
	}
	if (doc.hasIn(path)) doc.deleteIn(path);
	const pair = doc.createPair(path.at(-1), typeConfig.createNode());
	options.onBeforeCreate?.(pair);
	if (path.length === 1) doc.add(pair);
	else doc.addIn(path.slice(0, -1), pair);
	return doc.getIn(path);
}
/**
* Get or create a sequence at the given path
* @param doc - The YAML document
* @param path - The path to the sequence
* @param options - The options for the sequence
*/
function getOrCreateSeq(doc, path, options) {
	return getOrCreateNode(doc, path, options ?? {}, {
		typeGuard: isSeq,
		createNode: () => new YAMLSeq(),
		typeName: "sequence"
	});
}
/**
* Get or create a map at the given path
* @param doc - The YAML document
* @param path - The path to the map
* @param options - The options for the map
*/
function getOrCreateMap(doc, path, options) {
	return getOrCreateNode(doc, path, options ?? {}, {
		typeGuard: isMap,
		createNode: () => new YAMLMap(),
		typeName: "map"
	});
}
/**
* Read the `inputs` map from an existing YAML action definition into a plain
* object. Used by codegen to merge developer-added inputs with the ones the
* generator manages, so regeneration does not strip user-added entries.
*
* @param existingAction - The YAML map representing the existing action.
*/
function getExistingInputs(existingAction) {
	const inputs = existingAction?.get("inputs");
	if (!isMap(inputs)) return {};
	return Object.fromEntries(inputs.items.map((item) => {
		if (!isPair(item)) return [];
		return [String(item.key), item.value];
	}));
}
/**
* Read a string field from an existing YAML action definition. Used by codegen
* to preserve developer-set values (e.g. `runtime`) across regeneration
* instead of overwriting them.
*
* @param key - The field to read.
* @param existingAction - The YAML map representing the existing action.
*/
function getExistingString(key, existingAction) {
	const value = existingAction?.get(key);
	return typeof value === "string" ? value : void 0;
}

//#endregion
//#region ../../packages-private/scripting-utils/source/yaml/codegen.ts
/**
* Create an ext.config.yaml file
* @param path - The path to the ext.config.yaml file
* @param config - The config to build
* @param doc - The document to modify (a new one is created if not provided)
*/
async function createOrUpdateExtConfig(path, config, doc) {
	const extConfigDoc = doc ?? new Document({});
	if (config.web !== void 0) buildWeb(extConfigDoc, config.web);
	config.hooks ??= {};
	config.runtimeManifest ??= { packages: {} };
	await buildHooks(extConfigDoc, config.hooks);
	if (config.operations !== void 0) buildOperations(extConfigDoc, config.operations);
	buildRuntimeManifest(extConfigDoc, config.runtimeManifest);
	await writeExtConfig(path, extConfigDoc);
	return extConfigDoc;
}
/**
* Set the top-level `web` scalar only if it's missing.
* Preserves any user-customized value already present.
*/
function buildWeb(extConfig, web) {
	if (!extConfig.hasIn(["web"])) extConfig.add(extConfig.createPair("web", web));
}
/**
* Build the definition for a runtime action.
*
* Developer-added `inputs` keys (anything not in the generator-managed set) on
* `existingAction` are preserved so that hand-written entries — for example
* factory credentials for `dynamicList` fields — survive regeneration.
*
* A `runtime` already set on `existingAction` is likewise preserved, so a
* developer can pin a different Node runtime (e.g. `nodejs:24`) in
* `ext.config.yaml` once without codegen reverting it on the next run.
*
* @param action - The action definition to build.
* @param existingAction - The action's previous YAML definition, if any.
*/
function buildActionDefinition(action, existingAction) {
	const actionDef = new YAMLMap();
	const existingInputs = getExistingInputs(existingAction);
	const existingRuntime = getExistingString("runtime", existingAction);
	const managedInputs = { LOG_LEVEL: "$LOG_LEVEL" };
	actionDef.set("function", action.function);
	actionDef.set("web", action.web ?? "yes");
	actionDef.set("runtime", existingRuntime ?? action.runtime ?? "nodejs:24");
	actionDef.set("inputs", {
		...existingInputs,
		...managedInputs,
		...action.inputs ?? {}
	});
	actionDef.set("annotations", { ...action.annotations ?? {
		"require-adobe-auth": true,
		final: true
	} });
	const includes = action.include ?? [];
	if (includes.length > 0) {
		const itemSeq = new YAMLSeq();
		itemSeq.flow = true;
		for (const [source, target] of includes) itemSeq.items.push(source, target);
		const seq = new YAMLSeq();
		seq.items.push(itemSeq);
		actionDef.set("include", seq);
	}
	if (action.limits !== void 0) actionDef.set("limits", action.limits);
	return actionDef;
}
/**
* Build the `operations` section of the `ext.config.yaml` file
* @param extConfig - The ext.config.yaml file
* @param operations - The operations to build
*/
function buildOperations(extConfig, operations) {
	const ourOps = operations.workerProcess ?? [];
	if (ourOps.length > 0) {
		getOrCreateMap(extConfig, ["operations"], { onBeforeCreate: (pair) => {
			pair.key.spaceBefore = true;
		} });
		const workerProcess = getOrCreateSeq(extConfig, ["operations", "workerProcess"], { onBeforeCreate: (pair) => {
			pair.key.commentBefore = " These worker processes definitions are auto-generated. Do not remove or manually edit.";
		} });
		workerProcess.items = [];
		workerProcess.items.push(...ourOps.map((op) => {
			const map = new YAMLMap();
			map.set("type", op.type);
			map.set("impl", op.impl);
			return map;
		}));
	} else if (extConfig.hasIn(["operations", "workerProcess"])) extConfig.deleteIn(["operations", "workerProcess"]);
	if (operations.view !== void 0) {
		getOrCreateMap(extConfig, ["operations"], { onBeforeCreate: (pair) => {
			pair.key.spaceBefore = true;
		} });
		const view = getOrCreateSeq(extConfig, ["operations", "view"]);
		if (view.items.length === 0) view.items.push(...operations.view.map((v) => {
			const map = new YAMLMap();
			map.set("type", v.type);
			map.set("impl", v.impl);
			return map;
		}));
	}
	const operationsMap = extConfig.getIn(["operations"]);
	if (operationsMap instanceof YAMLMap && operationsMap.items.length === 0) extConfig.deleteIn(["operations"]);
}
/**
* Build the `runtimeManifest` section of the `ext.config.yaml` file
* @param extConfig - The ext.config.yaml file
* @param packages - The packages to build
*/
function buildRuntimeManifest(extConfig, manifest) {
	getOrCreateMap(extConfig, ["runtimeManifest"], { onBeforeCreate: (pair) => {
		pair.key.spaceBefore = true;
	} });
	const packages = manifest.packages ?? {};
	getOrCreateMap(extConfig, ["runtimeManifest", "packages"]);
	for (const [name, pkg] of Object.entries(packages)) {
		const packageDef = getOrCreateMap(extConfig, [
			"runtimeManifest",
			"packages",
			name
		], { onBeforeCreate: (pair) => {
			pair.key.commentBefore = " This package definition is auto-generated. Do not remove or manually edit.";
		} });
		const existingActions = packageDef.get("actions");
		const actions = new YAMLMap();
		packageDef.set("license", pkg.license ?? "Apache-2.0");
		packageDef.set("actions", actions);
		for (const [actionName, action] of Object.entries(pkg.actions ?? {})) {
			const existingAction = isMap(existingActions) ? existingActions.get(actionName) : void 0;
			const actionDef = buildActionDefinition(action, isMap(existingAction) ? existingAction : void 0);
			actions.set(actionName, actionDef);
		}
	}
}
/**
* Build the `hooks` section of the `ext.config.yaml` file
* @param extConfig - The ext.config.yaml file
* @param hooks - The hooks to build
*/
async function buildHooks(extConfig, hooks) {
	const generatedHooks = `[${Object.keys(hooks).join(", ")}]`;
	const hooksMap = getOrCreateMap(extConfig, ["hooks"], { onBeforeCreate: (pair) => {
		pair.key.spaceBefore = true;
		pair.key.commentBefore = ` The ${generatedHooks} hooks are auto-generated. Do not remove or manually edit.`;
	} });
	const execCommand = getExecCommand(await detectPackageManager());
	for (const [name, command] of Object.entries(hooks)) {
		const fullCommand = `${command.replaceAll("$packageExec", execCommand)}`;
		const prevValue = (hooksMap.get(name) ?? "").trim();
		if (prevValue.endsWith("js") || prevValue.endsWith("ts")) throw new Error(`Conflicting hook definition found. The "${name}" hook needs to be a command, not a script.`);
		if (prevValue !== "" && !prevValue.includes(fullCommand)) hooksMap.set(name, `${prevValue} && ${fullCommand}`);
		else if (prevValue === "") hooksMap.set(name, fullCommand);
	}
}
/**
* Write the ext.config.yaml file
* @param configPath - The path to the ext.config.yaml file
* @param config - The config to write
*/
async function writeExtConfig(configPath, doc) {
	await writeFile(configPath, doc.toString({
		indent: 2,
		lineWidth: 0,
		defaultStringType: "PLAIN"
	}), "utf-8");
}

//#endregion
//#region source/commands/generate/actions/lib.ts
const __dirname = dirname(fileURLToPath(import.meta.url));
/** The path to the action templates directory, sibling to this file. */
const TEMPLATES_DIR = join(__dirname, "templates");
const TYPESCRIPT_CONFIG_EXTENSIONS = new Set([
	".ts",
	".mts",
	".cts"
]);
const LEADING_DOT_SLASH_PATTERN = /^\.\//u;
const DYNAMIC_APP_CONFIG_IMPORT = `import appConfig from "${APP_CONFIG_IMPORT_ALIAS}";`;
const STATIC_APP_MANIFEST_IMPORT = "import appConfig from \"../../app.commerce.manifest.json\" with { type: \"json\" };";
const STATIC_CONFIG_SCHEMA_IMPORT = "import configSchema from \"../../configuration-schema.json\" with { type: \"json\" };";
/** Normalize a path for use as an ESM import specifier. */
function normalizeImportPath(path) {
	const normalizedPath = path.split(sep).join("/");
	return normalizedPath.startsWith(".") ? normalizedPath : `./${normalizedPath}`;
}
/** Normalize a generated file path for package.json imports. */
function normalizePackageJsonPath(path) {
	return normalizeImportPath(path).replace(LEADING_DOT_SLASH_PATTERN, "./");
}
/** Add or update the package import alias used by dynamic generated actions. */
async function updateAppConfigImportAlias(projectRoot, runtimeConfigPath) {
	const pkg = await NpmPackageJson.load(projectRoot);
	const existingImports = typeof pkg.content.imports === "object" && pkg.content.imports !== null ? pkg.content.imports : {};
	pkg.update({ imports: {
		...existingImports,
		[APP_CONFIG_IMPORT_ALIAS]: normalizePackageJsonPath(runtimeConfigPath)
	} });
	await pkg.save();
}
/** Remove the package import alias used only by dynamic generated actions. */
async function deleteAppConfigImportAlias(projectRoot) {
	const pkg = await NpmPackageJson.load(projectRoot);
	const existingImports = typeof pkg.content.imports === "object" && pkg.content.imports !== null ? pkg.content.imports : {};
	if (!("#app.commerce.config" in existingImports)) return;
	const imports = { ...existingImports };
	delete imports[APP_CONFIG_IMPORT_ALIAS];
	if (Object.keys(imports).length > 0) pkg.update({ imports });
	else pkg.content.imports = void 0;
	await pkg.save();
}
/** Write an ESM passthrough module for JavaScript app config files. */
async function writeJavaScriptAppConfigModule(configFilePath, outputPath) {
	await writeFile(outputPath, [
		"// This file has been auto-generated by `@adobe/aio-commerce-lib-app`",
		"// Do not modify this file directly",
		"",
		`import appConfig from "${normalizeImportPath(relative(dirname(outputPath), configFilePath))}";`,
		"",
		"export default appConfig;",
		""
	].join("\n"), "utf-8");
}
/** Bundle a TypeScript app config file into runtime-safe ESM. */
async function bundleTypeScriptAppConfigModule(configFilePath, outputPath) {
	await build({
		entryPoints: [configFilePath],
		outfile: outputPath,
		bundle: true,
		format: "esm",
		platform: "node",
		target: "node22",
		packages: "external",
		logLevel: "silent"
	});
}
/** Generate the runtime-safe app config module used by generated runtime actions. */
async function generateRuntimeAppConfigModule() {
	const projectRoot = await getProjectRootDirectory();
	const runtimeConfigPath = getRuntimeAppConfigPath();
	await makeOutputDirFor(dirname(runtimeConfigPath));
	const outputPath = join(projectRoot, runtimeConfigPath);
	const configFilePath = await resolveCommerceAppConfig(projectRoot);
	if (!configFilePath) throw new Error("Dynamic list options require an app.commerce.config.* file so generated actions can import the option factories at runtime.");
	consola$1.info("Generating JS config module...");
	if (TYPESCRIPT_CONFIG_EXTENSIONS.has(extname(configFilePath))) await bundleTypeScriptAppConfigModule(configFilePath, outputPath);
	else await writeJavaScriptAppConfigModule(configFilePath, outputPath);
	await updateAppConfigImportAlias(projectRoot, runtimeConfigPath);
	consola$1.success(`Generated JS config module: ${relative(process.cwd(), outputPath)}`);
}
/** Remove the generated app config module when generated actions use static JSON. */
async function cleanupRuntimeAppConfigModule() {
	const projectRoot = await getProjectRootDirectory();
	await rm(join(projectRoot, getRuntimeAppConfigPath()), { force: true });
	await deleteAppConfigImportAlias(projectRoot);
}
/**
* Prepare runtime-side config artifacts consumed by generated actions: the
* bundled ESM module and its `#app.commerce.config` package import alias.
*
* Generates them for dynamic schemas, cleans them up for static ones.
*/
async function prepareRuntimeAppConfigModule(appManifest) {
	if (hasDynamicAppConfig(appManifest)) {
		await generateRuntimeAppConfigModule();
		return;
	}
	await cleanupRuntimeAppConfigModule();
}
/** Switch static JSON imports to the generated runtime config module for dynamic schemas. */
function applyRuntimeAppConfigImport(template, actionName) {
	if (actionName === "app-config" || actionName === "installation") return template.replace(STATIC_APP_MANIFEST_IMPORT, DYNAMIC_APP_CONFIG_IMPORT);
	if (actionName === "config") return template.replace(STATIC_CONFIG_SCHEMA_IMPORT, DYNAMIC_APP_CONFIG_IMPORT).replace("const args = { configSchema };", "const args = { configSchema: appConfig.businessConfig.schema };");
	return template;
}
/**
* Read an extension point's ext.config.yaml file and return it as a JS object.
* @param extensionPointId - The extension point ID to read the config for.
*/
async function readExtConfig(extensionPointId) {
	const extConfigPath = join(await getProjectRootDirectory(), getExtConfigPath(extensionPointId));
	try {
		return {
			path: extConfigPath,
			doc: await readYamlFile(extConfigPath)
		};
	} catch (error) {
		const reason = error instanceof Error ? error.message : "Unexpected error while reading the file.";
		throw new Error([
			`Could not read ext.config.yaml for ${extensionPointId} at ${extConfigPath}.`,
			`Make sure the file exists and is valid YAML. You can regenerate it with "aio-commerce-lib-app generate actions".`,
			`Reason: ${reason}`
		].join("\n"));
	}
}
/** Update the ext.config.yaml file */
async function updateExtConfig(appConfig, extensionPointId) {
	consola$1.info(`Updating ext.config.yaml for ${extensionPointId}...`);
	await makeOutputDirFor(getExtensionPointFolderPath(extensionPointId));
	const { path: extConfigPath, doc: extConfigDoc } = await readExtConfig(extensionPointId);
	let extConfig;
	switch (extensionPointId) {
		case EXTENSIBILITY_EXTENSION_POINT_ID:
			extConfig = buildAppManagementExtConfig(appConfig);
			break;
		case CONFIGURATION_EXTENSION_POINT_ID:
			extConfig = buildBusinessConfigurationExtConfig();
			break;
		case BACKEND_UI_EXTENSION_POINT_ID:
			extConfig = buildAdminUiSdkExtConfig();
			break;
	}
	await createOrUpdateExtConfig(extConfigPath, extConfig, extConfigDoc);
	return extConfig;
}
/** Generate the action files */
async function generateActionFiles(appManifest, actions, extensionPointId, templatesDir) {
	consola$1.start("Generating runtime actions...");
	await makeOutputDirFor(getActionsDir(extensionPointId));
	const projectRoot = await getProjectRootDirectory();
	const outputFiles = [];
	const useDynamicImports = hasDynamicAppConfig(appManifest);
	for (const action of actions) {
		let template = await readFile(join(templatesDir, action.templateFile), "utf-8");
		if (useDynamicImports) template = applyRuntimeAppConfigImport(template, action.name);
		if (action.name === "installation") {
			const scriptsTemplate = await generateCustomScriptsTemplate(await readFile(join(templatesDir, "app-management", "custom-scripts.js.template"), "utf-8"), appManifest);
			template = applyCustomScripts(template, scriptsTemplate);
		}
		const actionPath = join(projectRoot, getActionPath(extensionPointId, action.name));
		await writeFile(actionPath, template, "utf-8");
		outputFiles.push(` ${relative(process.cwd(), actionPath)}`);
	}
	consola$1.success(`Generated ${actions.length} action(s)`);
	consola$1.log.raw(formatTree(outputFiles));
}
/**
* Applies the given custom scripts template code to the given installation template.
* @param installationTemplate - The installation code runtime action template
* @param customScriptsTemplate - The custom scripts dynamically generated template.
*/
function applyCustomScripts(installationTemplate, customScriptsTemplate) {
	if (customScriptsTemplate !== null) return installationTemplate.replace(CUSTOM_SCRIPTS_LOADER_PLACEHOLDER, customScriptsTemplate).replace("const args = { appConfig };", "const args = { appConfig, customScriptsLoader };");
	consola$1.debug("No custom installation steps found, skipping custom-scripts.js generation...");
	return installationTemplate.replace(CUSTOM_SCRIPTS_LOADER_PLACEHOLDER, "// No custom installation scripts configured");
}
/**
* Generate the installation template with dynamic custom script imports
*/
async function generateCustomScriptsTemplate(template, appManifest) {
	if (!hasCustomInstallationSteps(appManifest)) return null;
	const projectRoot = await getProjectRootDirectory();
	const installationActionDir = join(projectRoot, getActionsDir(EXTENSIBILITY_EXTENSION_POINT_ID));
	const customSteps = appManifest.installation.customInstallationSteps;
	const importStatements = customSteps.map((step, index) => {
		let relativeImportPath = relative(installationActionDir, join(projectRoot, step.script));
		if (!relativeImportPath.startsWith(".")) relativeImportPath = `./${relativeImportPath}`;
		relativeImportPath = relativeImportPath.replace(/\\/g, "/");
		return `import * as ${`customScript${index}`} from "${relativeImportPath}";`;
	}).join("\n");
	const scriptMap = customSteps.map((step, index) => {
		const entry = `"${step.script}": ${`customScript${index}`},`;
		return entry.padStart(entry.length + 6);
	}).join("\n");
	return template.replace(CUSTOM_IMPORTS_PLACEHOLDER, importStatements).replace(CUSTOM_SCRIPTS_MAP_PLACEHOLDER, scriptMap);
}
/**
* Generates `registration/index.js` with the Admin UI SDK registration config
* inlined as a JS object literal.
* @param appManifest - The validated app config; must satisfy `hasAdminUiSdk`.
* @param extensionPointId - The extension point ID that owns the registration action.
* @param templatesDir - The root directory containing the action templates.
*/
async function generateRegistrationActionFile(appManifest, extensionPointId, templatesDir = TEMPLATES_DIR) {
	consola$1.start("Generating Admin UI SDK registration action...");
	await makeOutputDirFor(getAdminUiSdkActionsDir(extensionPointId));
	const projectRoot = await getProjectRootDirectory();
	const template = await readFile(join(templatesDir, "admin-ui-sdk", "registration.js.template"), "utf-8");
	const { registration } = appManifest.adminUiSdk;
	const actionPath = join(projectRoot, getAdminUiSdkRegistrationActionPath(extensionPointId));
	await writeFile(actionPath, await prettierFormat(template.replace(REGISTRATION_JSON_PLACEHOLDER, `const registration = ${stringify(registration)};`), actionPath), "utf-8");
	consola$1.success(`Generated registration action at ${relative(process.cwd(), actionPath)}`);
}

//#endregion
//#region source/commands/generate/actions/main.ts
/**
* Runs the generate actions command for the given extension point.
* @param appManifest - The app manifest, used to determine which actions to generate and their content.
* @param templatesDir - The directory to load templates from, for testing purposes. Defaults to the generated actions template root.
*/
async function run$5(appManifest, templatesDir = TEMPLATES_DIR) {
	await prepareRuntimeAppConfigModule(appManifest);
	await generateActionFiles(appManifest, getRuntimeActions(await updateExtConfig(appManifest, EXTENSIBILITY_EXTENSION_POINT_ID), "app-management"), EXTENSIBILITY_EXTENSION_POINT_ID, templatesDir);
	if (hasBusinessConfigSchema(appManifest)) await generateActionFiles(appManifest, getRuntimeActions(await updateExtConfig(appManifest, CONFIGURATION_EXTENSION_POINT_ID), "business-configuration"), CONFIGURATION_EXTENSION_POINT_ID, templatesDir);
	if (hasAdminUiSdk(appManifest)) {
		await updateExtConfig(appManifest, BACKEND_UI_EXTENSION_POINT_ID);
		await generateRegistrationActionFile(appManifest, BACKEND_UI_EXTENSION_POINT_ID, templatesDir);
	}
}
/** Run the generate actions command */
async function exec$5() {
	try {
		await run$5(await loadAppManifest());
	} catch (error) {
		if (error instanceof CommerceSdkValidationError) consola$1.error(error.display());
		else consola$1.error(error);
		process.exit(1);
	}
}

//#endregion
//#region source/commands/generate/manifest/main.ts
async function run$4(appConfig) {
	if (hasDynamicAppConfig(appConfig)) {
		const stalePath = join(await getProjectRootDirectory(), getManifestPath());
		if (await access(stalePath).then(() => true, () => false)) {
			await rm(stalePath, { force: true });
			consola$1.success(`Removed stale ${APP_MANIFEST_FILE}`);
		}
		return;
	}
	consola$1.info("Generating app manifest...");
	const contents = stringify$1(appConfig, null, 2);
	await writeFile(join(await makeOutputDirFor(getGeneratedDir(EXTENSIBILITY_EXTENSION_POINT_ID)), APP_MANIFEST_FILE), contents, "utf-8");
	consola$1.success(`Generated ${APP_MANIFEST_FILE}\n`);
}
/** Run the generate manifest command */
async function exec$4() {
	try {
		await run$4(await loadAppManifest());
	} catch (error) {
		if (error instanceof CommerceSdkValidationError) consola$1.error(error.display());
		else consola$1.error(error);
		process.exit(1);
	}
}

//#endregion
//#region ../../packages-private/scripting-utils/source/filesystem/helpers.ts
/**
* Touch a file, creating it if it doesn't exist.
* @param filePath The path of the file to touch
*/
async function touch(filePath) {
	await writeFile(filePath, "", { flag: "wx" }).catch((e) => {
		if (e.code !== "EEXIST") throw e;
	});
}

//#endregion
//#region source/commands/generate/schema/main.ts
async function run$3(appConfig) {
	if (!hasBusinessConfigSchema(appConfig)) {
		consola$1.debug("Business configuration schema not found in application configuration. Nothing to do.");
		return;
	}
	const projectDir = await getProjectRootDirectory();
	const envPath = join(projectDir, ".env");
	await touch(envPath);
	process.loadEnvFile(envPath);
	if (appConfig.businessConfig.schema.some((field) => field.type === "password")) {
		const packageExec = getExecCommand(await detectPackageManager(projectDir));
		if ("AIO_COMMERCE_CONFIG_ENCRYPTION_KEY" in process.env && String(process.env.AIO_COMMERCE_CONFIG_ENCRYPTION_KEY).trim().length > 0) execSync(`${packageExec} aio-commerce-lib-config encryption validate`);
		else execSync(`${packageExec} aio-commerce-lib-config encryption setup`);
	}
	if (hasDynamicAppConfig(appConfig)) {
		const stalePath = join(projectDir, getSchemaPath());
		if (await access(stalePath).then(() => true, () => false)) {
			await rm(stalePath, { force: true });
			consola$1.success(`Removed stale ${CONFIG_SCHEMA_FILE_NAME}`);
		}
		return;
	}
	consola$1.info("Generating configuration schema...");
	const outputDir = await makeOutputDirFor(getGeneratedDir(CONFIGURATION_EXTENSION_POINT_ID));
	const contents = stringify$1(appConfig.businessConfig.schema, null, 2);
	await writeFile(join(outputDir, CONFIG_SCHEMA_FILE_NAME), contents, "utf-8");
	consola$1.success(`Generated ${CONFIG_SCHEMA_FILE_NAME}\n`);
}
/** Run the generate schema command */
async function exec$3() {
	try {
		await run$3(await loadAppManifest());
	} catch (error) {
		if (error instanceof CommerceSdkValidationError) consola$1.error(error.display());
		else consola$1.error(error);
		process.exit(1);
	}
}

//#endregion
//#region source/commands/hooks/postinstall.ts
async function run$2(appManifest, templatesDir) {
	await run$5(appManifest, templatesDir);
	await run$4(appManifest);
	await run$3(appManifest);
}
/** Runs the postinstall hook */
async function exec$2() {
	consola.debug("Running lib-app postinstall hook");
	try {
		await run$2(await loadAppManifest());
	} catch (error) {
		if (error instanceof CommerceSdkValidationError) consola.error(error.display());
		else consola.error(error);
		process.exit(1);
	}
}

//#endregion
//#region ../../packages-private/scripting-utils/source/env.ts
/**
* This module exports shared environment utilities for the AIO Commerce SDK.
* @packageDocumentation
*/
const { context } = aioIms;
const IMS_KEYS = {
	client_id: "AIO_COMMERCE_AUTH_IMS_CLIENT_ID",
	client_secrets: "AIO_COMMERCE_AUTH_IMS_CLIENT_SECRETS",
	technical_account_email: "AIO_COMMERCE_AUTH_IMS_TECHNICAL_ACCOUNT_EMAIL",
	technical_account_id: "AIO_COMMERCE_AUTH_IMS_TECHNICAL_ACCOUNT_ID",
	scopes: "AIO_COMMERCE_AUTH_IMS_SCOPES",
	ims_org_id: "AIO_COMMERCE_AUTH_IMS_ORG_ID"
};
/**
* Replaces or creates an environment variable in a .env file
* @param filePath - The path to the .env file
* @param key - The environment variable key to replace or create
* @param value - The new value for the environment variable
*/
function replaceEnvVar(filePath, key, value) {
	const envPath = path.resolve(filePath);
	const envLines = readFileSync(envPath, "utf8").split("\n");
	const formattedValue = value.includes(" ") ? `"${value}"` : value;
	let keyExists = false;
	const updatedLines = envLines.map((line) => {
		if (line.trim().startsWith("#") || !line.includes("=")) return line;
		const [currentKey] = line.split("=");
		if (currentKey === key) {
			keyExists = true;
			return `${key}=${formattedValue}`;
		}
		return line;
	});
	if (!keyExists) updatedLines.push(`${key}=${formattedValue}`);
	writeFileSync(envPath, updatedLines.join("\n"), "utf8");
}
/** Returns the path to the .env file. */
function resolveEnvPath() {
	const envPath = process.env.INIT_CWD ? `${process.env.INIT_CWD}/.env` : ".env";
	return path.resolve(envPath);
}
/** Resolves the IMS server to server context from the project workspace credentials. */
function resolveImsS2SContext() {
	const [credential] = (config.get("project.workspace.details.credentials") ?? []).filter(({ integration_type }) => integration_type === "oauth_server_to_server").map(({ name }) => name) ?? [];
	if (!credential) return Promise.resolve(null);
	return context.get(credential);
}
/**
* Syncs the IMS credentials environment variables from the configured IMS context in
* the .env file, in a way that is compatible with `@adobe/aio-commerce-lib-auth`.
*/
async function syncImsCredentials() {
	const envPath = resolveEnvPath();
	if (!existsSync(envPath)) return {
		ok: false,
		reason: "missing-env"
	};
	const envVars = dotenv.parse(readFileSync(envPath, "utf8"));
	const imsContext = await resolveImsS2SContext();
	if (!imsContext) return {
		ok: false,
		reason: "no-ims-context"
	};
	const { data } = imsContext;
	for (const [key, value] of Object.entries(data)) {
		const oauthKey = IMS_KEYS[key];
		if (!oauthKey) continue;
		if (!envVars[oauthKey] || envVars[oauthKey] !== value) replaceEnvVar(envPath, oauthKey, value);
	}
	return { ok: true };
}

//#endregion
//#region source/commands/hooks/pre-app-build.ts
/**
* Runs the pre-app-build hook for the given extension.
* @param extension - The extension to run the hook for.
* @param templatesDir - The directory to load templates from, for testing purposes. Defaults to the generated actions template root.
*/
async function run$1(extension, templatesDir = TEMPLATES_DIR) {
	const appManifest = await loadAppManifest();
	await prepareRuntimeAppConfigModule(appManifest);
	if (extension === "extensibility/1") {
		const { doc: extensibilityExtConfig } = await readExtConfig(EXTENSIBILITY_EXTENSION_POINT_ID);
		await run$4(appManifest);
		await generateActionFiles(appManifest, getRuntimeActions(extensibilityExtConfig.toJS(), "app-management"), EXTENSIBILITY_EXTENSION_POINT_ID, templatesDir);
		consola.info("Syncing IMS credentials...");
		await syncImsCredentials();
		return;
	}
	if (extension === "configuration/1") {
		const { doc: businessConfigExtConfig } = await readExtConfig(CONFIGURATION_EXTENSION_POINT_ID);
		await run$3(appManifest);
		await generateActionFiles(appManifest, getRuntimeActions(businessConfigExtConfig.toJS(), "business-configuration"), CONFIGURATION_EXTENSION_POINT_ID, templatesDir);
		return;
	}
	if (extension === "backend-ui/1") {
		if (hasAdminUiSdk(appManifest)) await generateRegistrationActionFile(appManifest, BACKEND_UI_EXTENSION_POINT_ID);
		return;
	}
	throw new Error(`Unsupported extension: ${extension}`);
}
/** Runs the pre-app-build hook */
async function exec$1() {
	consola.debug("Running lib-app pre-app-build hook");
	const extension = process.env.EXTENSION;
	try {
		if (!extension) throw new Error("EXTENSION environment variable is not set");
		await run$1(extension);
	} catch (error) {
		if (error instanceof CommerceSdkValidationError) consola.error(error.display());
		else consola.error(error);
		process.exit(1);
	}
}

//#endregion
//#region source/commands/init/constants.ts
/** The default values for the features of the commerce app config. */
const DOMAIN_DEFAULTS = {
	businessConfig: { schema: [{
		type: "list",
		name: "sampleList",
		label: "Sample List",
		selectionMode: "multiple",
		default: ["a"],
		options: [{
			label: "Option A",
			value: "a"
		}, {
			label: "Option B",
			value: "b"
		}]
	}, {
		type: "text",
		name: "sampleText",
		label: "Sample Text",
		default: "Hello, world!"
	}] },
	"eventing.commerce": [{
		provider: {
			label: "Commerce Events Provider",
			description: "A description for your Commerce Events provider."
		},
		events: [{
			name: "plugin.sample_event",
			fields: [{ name: "*" }],
			label: "Sample Event",
			description: "Use case description for the event.",
			runtimeActions: ["my-package/handle-sample-event"]
		}]
	}],
	"eventing.external": [{
		provider: {
			label: "External Events Provider",
			description: "A description for your External Events provider."
		},
		events: [{
			name: "be-observer.sample_event",
			label: "Sample Event",
			description: "Use case description for the event.",
			runtimeActions: ["my-package/handle-sample-event"]
		}]
	}],
	"installation.customInstallationSteps": [{
		name: "sample-step",
		description: "Use case description for the step.",
		script: "./path/to/script.js"
	}],
	webhooks: [{
		label: "Sample Webhook",
		description: "A sample Commerce Webhook handler.",
		runtimeAction: "my-package/handle-webhook",
		category: "modification",
		webhook: {
			webhook_method: "plugin.sample.event",
			webhook_type: "after",
			batch_name: "my_app",
			hook_name: "sample_hook",
			method: "POST"
		}
	}, {
		label: "Sample Webhook with URL",
		description: "A sample Commerce Webhook handler using an explicit URL.",
		category: "modification",
		webhook: {
			webhook_method: "plugin.sample.event",
			webhook_type: "after",
			batch_name: "my_app",
			hook_name: "sample_hook_url",
			method: "POST",
			url: "https://example.com/webhook"
		}
	}]
};

//#endregion
//#region source/commands/init/utils.ts
/**
* Convert the given init options to answers for scaffolding the app config. This is useful for non-interactive use cases, such as running the command with a config file or from another script.
* @param flags - The initialization options
*/
function initFlagsToScaffoldAppAnswers(flags) {
	const configFile = `${COMMERCE_APP_CONFIG_FILE}.${flags.configFormat}`;
	return {
		appName: flags.appName,
		configFile,
		configFormat: flags.configFormat,
		domains: new Set(flags.domains)
	};
}
/** Prompt the user to scaffold a new Commerce App configuration. When options are provided, prompts are skipped. */
async function promptForCommerceAppConfig() {
	const configFormat = await consola.prompt("What format do you want to use for the config file?", {
		type: "select",
		default: "ts",
		initial: "ts",
		cancel: "reject",
		options: [{
			label: "TypeScript",
			value: "ts",
			hint: "Recommended"
		}, {
			label: "JavaScript",
			value: "js"
		}]
	});
	const appName = await consola.prompt("What is the name of your app?", {
		type: "text",
		placeholder: "Application Display Name",
		default: "My Application",
		cancel: "reject"
	});
	const featuresToAdd = await consola.prompt(`What features do you want to add to your app? ${colorize("gray", "Space to select/deselect, Enter to submit")}`, {
		type: "multiselect",
		initial: [],
		required: true,
		cancel: "reject",
		options: [
			{
				label: "Business Configuration",
				value: "businessConfig.schema"
			},
			{
				label: "Commerce Events",
				value: "eventing.commerce"
			},
			{
				label: "Commerce Webhooks",
				value: "webhooks"
			},
			{
				label: "External Events",
				value: "eventing.external"
			},
			{
				label: "Custom Installation Steps",
				value: "installation.customInstallationSteps"
			}
		]
	});
	const configFile = `${COMMERCE_APP_CONFIG_FILE}.${configFormat}`;
	const domains = new Set(featuresToAdd);
	consola.log("");
	return {
		appName,
		configFile,
		domains,
		configFormat
	};
}
/** Create the default commerce app config file content */
async function getDefaultCommerceAppConfig(cwd, { appName, configFormat, domains }) {
	const needsESM = await isESM(cwd) || configFormat === "ts";
	const exportKeyword = needsESM ? "export default" : "module.exports =";
	const importStatement = needsESM ? "import { defineConfig } from \"@adobe/aio-commerce-lib-app/config\";" : "const { defineConfig } = require(\"@adobe/aio-commerce-lib-app/config\");";
	const defaultConfig = { metadata: {
		id: appName.toLowerCase().replace(/ /g, "-"),
		displayName: appName,
		version: "1.0.0",
		description: "A custom Adobe Commerce application. Fill description for your app."
	} };
	if (domains.has("businessConfig.schema")) defaultConfig.businessConfig = { schema: DOMAIN_DEFAULTS.businessConfig.schema };
	if (domains.has("eventing.commerce")) defaultConfig.eventing = {
		...defaultConfig.eventing,
		commerce: DOMAIN_DEFAULTS["eventing.commerce"]
	};
	if (domains.has("eventing.external")) defaultConfig.eventing = {
		...defaultConfig.eventing,
		external: DOMAIN_DEFAULTS["eventing.external"]
	};
	if (domains.has("installation.customInstallationSteps")) defaultConfig.installation = { customInstallationSteps: DOMAIN_DEFAULTS["installation.customInstallationSteps"] };
	if (domains.has("webhooks")) defaultConfig.webhooks = DOMAIN_DEFAULTS.webhooks;
	return [
		importStatement,
		"",
		`${exportKeyword} defineConfig(${inspect(defaultConfig, { colors: false })})`
	].join("\n");
}
/**
* Add an extension point to the app config file.
* @param extensionPointId - The id of the extension point.
* @param rootDirectory - The root directory of the project.
* @param commentBefore - The comment to add before the extension point include pair.
*/
async function addExtensionPointToAppConfig(extensionPointId, rootDirectory, commentBefore) {
	const appConfigPath = join(rootDirectory, APP_CONFIG_FILE);
	let doc;
	const includePath = relative(rootDirectory, join(rootDirectory, getExtConfigPath(extensionPointId)));
	try {
		doc = await readYamlFile(appConfigPath);
	} catch (error) {
		const fallbackContent = `extensions:\n  ${extensionPointId}:\n    $include: "${includePath}"`;
		throw new Error(`Failed to parse ${APP_CONFIG_FILE}. \nPlease add manually: \n\n${fallbackContent}`, { cause: error });
	}
	if (doc.getIn([
		"extensions",
		extensionPointId,
		"$include"
	]) === includePath) {
		consola.success(`Extension "${extensionPointId}" already configured in ${APP_CONFIG_FILE}`);
		return;
	}
	consola.info(`Adding extension "${extensionPointId}" to ${APP_CONFIG_FILE}...`);
	const extensions = getOrCreateMap(doc, ["extensions"], { onBeforeCreate: (pair) => {
		pair.key.spaceBefore = true;
	} });
	getOrCreateMap(doc, ["extensions", extensionPointId], { onBeforeCreate: (pair) => {
		pair.key.spaceBefore = extensions.items.length > 0;
		pair.key.commentBefore = commentBefore;
	} }).set("$include", includePath);
	await writeFile(appConfigPath, doc.toString(), "utf-8");
}
/**
* Add an extension point to the install.yaml file.
* @param extensionPointId - The id of the extension point.
* @param rootDirectory - The root directory of the project.
* @param commentBefore - The comment to add before the extension point include pair.
*/
async function addExtensionPointToInstallYaml(extensionPointId, rootDirectory, commentBefore) {
	const installYamlPath = join(rootDirectory, INSTALL_YAML_FILE);
	let doc;
	let extensions;
	try {
		doc = await readYamlFile(installYamlPath);
		extensions = getOrCreateSeq(doc, ["extensions"], { onBeforeCreate: (pair) => {
			pair.key.spaceBefore = true;
		} });
	} catch (error) {
		const fallbackContent = `\nextensions:\n  - extensionPointId: ${extensionPointId}`;
		throw new Error(`Failed to parse ${INSTALL_YAML_FILE}. \nPlease add manually: \n\n${fallbackContent}`, { cause: error });
	}
	if (extensions.items.some((item) => isMap(item) && item.get("extensionPointId") === extensionPointId)) {
		consola.success(`Extension "${extensionPointId}" already configured in ${INSTALL_YAML_FILE}`);
		return;
	}
	consola.info(`Adding extension "${extensionPointId}" to ${INSTALL_YAML_FILE}...`);
	const extension = doc.createPair("extensionPointId", extensionPointId);
	extension.key.spaceBefore = extensions.items.length > 0;
	extension.key.commentBefore = commentBefore;
	extensions.items.push(extension);
	await writeFile(installYamlPath, doc.toString(), "utf-8");
}

//#endregion
//#region source/commands/init/lib.ts
/** Ensure app.commerce.config file exists, allow creating if it doesn't. When options are provided, prompts are skipped. */
async function ensureCommerceAppConfig(cwd = process.cwd(), formatConfig = true, flags) {
	let config = null;
	try {
		config = await readCommerceAppConfig(cwd);
	} catch {}
	if (config) try {
		const validatedConfig = validateCommerceAppConfig(config);
		consola$1.success(`${COMMERCE_APP_CONFIG_FILE} found and is valid. Continuing...`);
		return {
			config: validatedConfig,
			domains: getConfigDomains(validatedConfig)
		};
	} catch (error) {
		throw new Error(`${COMMERCE_APP_CONFIG_FILE} is invalid`, { cause: error });
	}
	if (!flags) {
		consola$1.warn(`${COMMERCE_APP_CONFIG_FILE} not found.`);
		if (!await consola$1.prompt(`Do you want to create a ${"app.commerce.config"} file? (y/n)`, {
			type: "confirm",
			initial: true,
			default: false
		})) throw new Error("Initialization cancelled.");
	}
	const answers = flags ? initFlagsToScaffoldAppAnswers(flags) : await promptForCommerceAppConfig();
	try {
		const configContent = await getDefaultCommerceAppConfig(cwd, answers);
		const path = join(await getProjectRootDirectory(cwd), answers.configFile);
		consola$1.info(`Creating ${answers.configFile}...`);
		if (formatConfig) await writeFile(path, await prettierFormat(configContent, path), "utf-8");
		else await writeFile(path, configContent, "utf-8");
		const config = await parseCommerceAppConfig(cwd);
		consola$1.success(`Created ${answers.configFile}`);
		return {
			config,
			domains: answers.domains
		};
	} catch (error) {
		throw new Error(`Failed to create ${answers.configFile}`, { cause: error });
	}
}
/**
* Ensure a package.json exists and detect the package manager.
* @param cwd - Directory to check; defaults to `process.cwd()`
*/
async function ensurePackageJson(cwd = process.cwd()) {
	const existing = await readPackageJson(cwd);
	let packageJson;
	if (existing) packageJson = existing;
	else {
		consola$1.warn("package.json not found. Creating one...");
		packageJson = {
			name: "my-commerce-app",
			version: "1.0.0",
			private: true
		};
		await writeFile(join(resolve(cwd), PACKAGE_JSON_FILE), JSON.stringify(packageJson, null, 2), "utf-8");
		consola$1.success("Wrote package.json");
	}
	const packageManager = await detectPackageManager(cwd);
	const execCommand = getExecCommand(packageManager);
	return {
		packageJson,
		packageManager,
		execCommand
	};
}
/**
* Register the `hooks postinstall` script in package.json.
* @param execCommand - Prefix for running local binaries (e.g. `pnpm exec`, `npx`)
* @param cwd - Directory containing the package.json to update
*/
async function writePostinstallHook(execCommand, cwd = process.cwd()) {
	const postinstallScript = `${execCommand} aio-commerce-lib-app hooks postinstall`;
	const pkg = await NpmPackageJson.load(cwd);
	const existing = pkg.content.scripts?.postinstall;
	if (existing === postinstallScript || existing?.includes(postinstallScript)) {
		consola$1.success(`postinstall script already configured in ${PACKAGE_JSON_FILE}`);
		return;
	}
	const nextPostinstall = existing ? `${existing} && ${postinstallScript}` : postinstallScript;
	if (existing) consola$1.warn(`${PACKAGE_JSON_FILE} already has a postinstall script. Adding a new one...`);
	else consola$1.info(`Adding postinstall script to ${PACKAGE_JSON_FILE}...`);
	pkg.update({ scripts: {
		...pkg.content.scripts,
		postinstall: nextPostinstall
	} });
	await pkg.save();
	consola$1.success(`Added postinstall script to ${PACKAGE_JSON_FILE}`);
}
/** Ensure app.config.yaml has the extension reference */
async function ensureAppConfig(domains, cwd = process.cwd()) {
	const rootDirectory = await getProjectRootDirectory(cwd);
	if (domains.has("businessConfig.schema")) await addExtensionPointToAppConfig(CONFIGURATION_EXTENSION_POINT_ID, rootDirectory, " This extension is required for business configuration. Do not remove.");
	if (domains.has("adminUiSdk")) await addExtensionPointToAppConfig(BACKEND_UI_EXTENSION_POINT_ID, rootDirectory, " This extension is required for Admin UI SDK. Do not remove.");
	await addExtensionPointToAppConfig(EXTENSIBILITY_EXTENSION_POINT_ID, rootDirectory, " This extension is required for app management. Do not remove.");
}
/**
* Install the given dependencies.
* @param packageManager - The detected package manager
* @param dependencies - Package specifiers to install; no-op when empty
* @param cwd - Working directory for the install command
*/
function runInstall(packageManager, dependencies, cwd = process.cwd()) {
	if (dependencies.length === 0) return;
	const dependencyListString = dependencies.map((dependency) => `  - ${dependency}`).join("\n");
	consola$1.start([`Installing the following dependencies with ${packageManager}:\n${dependencyListString}\n`, "This may take a few seconds...\n"].join("\n"));
	const { command, args } = getInstallCommand(packageManager, dependencies);
	const displayCommand = [command, ...args].join(" ");
	const result = spawnSync(command, args, {
		cwd,
		stdio: "inherit"
	});
	if (result.error || result.status !== 0) throw new Error(`Failed to install dependencies automatically. Please install manually: ${displayCommand}`, { cause: result.error ?? /* @__PURE__ */ new Error(`Install exited with code ${result.status}`) });
	consola$1.log("");
	consola$1.success("Dependencies installed successfully");
}
/**
* Install the domain-specific dependencies derived from the selected domains.
* @param packageManager - The detected package manager
* @param domains - Domains enabled in the commerce app config
* @param cwd - Working directory for the install command
*/
function installDependencies(packageManager, domains, cwd = process.cwd()) {
	const packages = [];
	if (domains.has("businessConfig.schema")) packages.push(`@adobe/aio-commerce-lib-config@^1.5.1`);
	if (packages.length === 0) {
		consola$1.info("No additional domain dependencies to install.");
		return;
	}
	runInstall(packageManager, packages, cwd);
}
/** Run the generation command */
async function runGeneration(appConfig, execCommand) {
	try {
		await run$5(appConfig);
		await run$4(appConfig);
		await run$3(appConfig);
	} catch (error) {
		throw new Error(`Failed to run generation command. Please run manually: ${execCommand} aio-commerce-lib-app generate all`, { cause: error });
	}
}
/** Ensure install.yaml has the extension reference */
async function ensureInstallYaml(domains, cwd = process.cwd()) {
	const rootDirectory = await getProjectRootDirectory(cwd);
	if (domains.has("businessConfig.schema")) await addExtensionPointToInstallYaml(CONFIGURATION_EXTENSION_POINT_ID, rootDirectory, " This extension is required for business configuration. Do not remove.");
	if (domains.has("adminUiSdk")) await addExtensionPointToInstallYaml(BACKEND_UI_EXTENSION_POINT_ID, rootDirectory, " This extension is required for Admin UI SDK. Do not remove.");
	await addExtensionPointToInstallYaml(EXTENSIBILITY_EXTENSION_POINT_ID, rootDirectory, " This extension is required for app management. Do not remove.");
}

//#endregion
//#region source/commands/init/main.ts
const REQUIRED_DEPENDENCIES = ["@adobe/aio-commerce-sdk", `@adobe/aio-commerce-lib-app@1.7.0`];
/** Initialize the project */
async function run(flags, extraOptions) {
	consola$1.start("Initializing app...");
	const { execCommand, packageManager } = await ensurePackageJson();
	runInstall(packageManager, REQUIRED_DEPENDENCIES);
	const { config, domains } = await ensureCommerceAppConfig(process.cwd(), extraOptions?.formatConfig ?? true, flags);
	installDependencies(packageManager, domains);
	const pkg = await NpmPackageJson.load(process.cwd());
	pkg.update({
		name: config.metadata.id,
		version: config.metadata.version,
		description: config.metadata.description
	});
	await pkg.save();
	await runGeneration(config, execCommand);
	await ensureAppConfig(domains);
	await ensureInstallYaml(domains);
	await writePostinstallHook(execCommand);
	consola$1.success("Initialization complete!");
	consola$1.box([
		"Next steps:",
		"  - Review and customize app.commerce.config.*",
		"  - Build and deploy your app"
	].join("\n"));
}
/** Run the init command */
async function exec() {
	try {
		await run();
	} catch (error) {
		if (error instanceof CommerceSdkValidationError) consola$1.error(error.display());
		else consola$1.error(error);
		process.exit(1);
	}
}

//#endregion
//#region source/commands/index.ts
const NAMESPACE = "@adobe/aio-commerce-lib-app";
const [command, subcommand] = process.argv.slice(2);
const USAGE = `
Usage: ${NAMESPACE} <command> [target]

Commands:
  init                 Initialize the project (recommended for first-time setup)
  
  generate <target>    Generate artifacts
    all                Generate app manifest and runtime actions
    actions            Generate runtime actions only
    manifest           Generate app manifest only
    schema             Generate configuration schema only

  help                 Show this help message

Examples:
  ${NAMESPACE} init
  ${NAMESPACE} generate all
  ${NAMESPACE} generate manifest
  ${NAMESPACE} generate actions
  ${NAMESPACE} generate schema
`;
/** Run all generate targets in sequence */
async function generateAll() {
	await exec$5();
	await exec$4();
	await exec$3();
}
/** Command handlers registry mapping command names to their subcommand handlers */
const COMMANDS = {
	init: exec,
	hooks: {
		"pre-app-build": exec$1,
		postinstall: exec$2
	},
	generate: {
		actions: exec$5,
		manifest: exec$4,
		schema: exec$3,
		all: generateAll
	}
};
/**
* Generic command handler that dispatches to specific command implementations
* @param commandName - The name of the command (for error messages)
* @param target - The target subcommand to execute
* @param handlers - Map of target names to handler functions
*/
async function handleCommand(commandName, target, handlers) {
	function invalidTargetError(message) {
		consola.error(message);
		consola.info(`Available targets: ${Object.keys(handlers).join(" | ")}`);
		process.exit(1);
	}
	if (!target) invalidTargetError(`No ${commandName} target specified`);
	const handler = handlers[target];
	if (!handler) invalidTargetError(`Unknown ${commandName} target: ${target}`);
	await handler();
}
async function main() {
	if (!command || command === "help" || command === "--help" || command === "-h") {
		consola.log.raw(USAGE);
		process.exit(0);
	}
	try {
		const handlers = COMMANDS[command];
		if (!handlers) {
			consola.error(`Unknown command: ${command}`);
			consola.log.raw(USAGE);
			process.exit(1);
		}
		if (typeof handlers === "function") {
			await handlers();
			return;
		}
		await handleCommand(command, subcommand, handlers);
	} catch (error) {
		consola.error(stringifyError(error));
	}
}
main();

//#endregion
export {  };
