// @adobe/aio-commerce-lib-auth/dist/cjs/commands/index.cjs
#!/usr/bin/env node
/**
 * @license
 *
 * Copyright 2026 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */

const require_chunk = require('../chunk-Cozy-1ZF.cjs');
let _adobe_aio_commerce_lib_core_error = require("@adobe/aio-commerce-lib-core/error");
let _adobe_aio_lib_ims = require("@adobe/aio-lib-ims");
_adobe_aio_lib_ims = require_chunk.__toESM(_adobe_aio_lib_ims, 1);
let consola = require("consola");
consola = require_chunk.__toESM(consola, 1);
let fs = require("fs");
let path = require("path");
path = require_chunk.__toESM(path, 1);
let _adobe_aio_lib_core_config = require("@adobe/aio-lib-core-config");
_adobe_aio_lib_core_config = require_chunk.__toESM(_adobe_aio_lib_core_config, 1);
let dotenv = require("dotenv");
dotenv = require_chunk.__toESM(dotenv, 1);

//#region ../../packages-private/scripting-utils/source/error.ts
/**
* This module exports shared error utilities for the AIO Commerce SDK.
* @packageDocumentation
*/
/**
* Stringify an error to a human-friendly string.
* @param error - The error to stringify.
*/
function stringifyError(error) {
	if (error instanceof _adobe_aio_commerce_lib_core_error.CommerceSdkValidationError) return error.display();
	return error instanceof Error ? error.message : String(error);
}

//#endregion
//#region ../../packages-private/scripting-utils/source/env.ts
/**
* This module exports shared environment utilities for the AIO Commerce SDK.
* @packageDocumentation
*/
const { context } = _adobe_aio_lib_ims.default;
const IMS_KEYS = {
	client_id: "AIO_COMMERCE_AUTH_IMS_CLIENT_ID",
	client_secrets: "AIO_COMMERCE_AUTH_IMS_CLIENT_SECRETS",
	technical_account_email: "AIO_COMMERCE_AUTH_IMS_TECHNICAL_ACCOUNT_EMAIL",
	technical_account_id: "AIO_COMMERCE_AUTH_IMS_TECHNICAL_ACCOUNT_ID",
	scopes: "AIO_COMMERCE_AUTH_IMS_SCOPES",
	ims_org_id: "AIO_COMMERCE_AUTH_IMS_ORG_ID"
};
/**
* Replaces or creates an environment variable in a .env file
* @param filePath - The path to the .env file
* @param key - The environment variable key to replace or create
* @param value - The new value for the environment variable
*/
function replaceEnvVar(filePath, key, value) {
	const envPath = path.default.resolve(filePath);
	const envLines = (0, fs.readFileSync)(envPath, "utf8").split("\n");
	const formattedValue = value.includes(" ") ? `"${value}"` : value;
	let keyExists = false;
	const updatedLines = envLines.map((line) => {
		if (line.trim().startsWith("#") || !line.includes("=")) return line;
		const [currentKey] = line.split("=");
		if (currentKey === key) {
			keyExists = true;
			return `${key}=${formattedValue}`;
		}
		return line;
	});
	if (!keyExists) updatedLines.push(`${key}=${formattedValue}`);
	(0, fs.writeFileSync)(envPath, updatedLines.join("\n"), "utf8");
}
/** Returns the path to the .env file. */
function resolveEnvPath() {
	const envPath = process.env.INIT_CWD ? `${process.env.INIT_CWD}/.env` : ".env";
	return path.default.resolve(envPath);
}
/** Resolves the IMS server to server context from the project workspace credentials. */
function resolveImsS2SContext() {
	const [credential] = (_adobe_aio_lib_core_config.default.get("project.workspace.details.credentials") ?? []).filter(({ integration_type }) => integration_type === "oauth_server_to_server").map(({ name }) => name) ?? [];
	if (!credential) return Promise.resolve(null);
	return context.get(credential);
}
/**
* Syncs the IMS credentials environment variables from the configured IMS context in
* the .env file, in a way that is compatible with `@adobe/aio-commerce-lib-auth`.
*/
async function syncImsCredentials() {
	const envPath = resolveEnvPath();
	if (!(0, fs.existsSync)(envPath)) return {
		ok: false,
		reason: "missing-env"
	};
	const envVars = dotenv.default.parse((0, fs.readFileSync)(envPath, "utf8"));
	const imsContext = await resolveImsS2SContext();
	if (!imsContext) return {
		ok: false,
		reason: "no-ims-context"
	};
	const { data } = imsContext;
	for (const [key, value] of Object.entries(data)) {
		const oauthKey = IMS_KEYS[key];
		if (!oauthKey) continue;
		if (!envVars[oauthKey] || envVars[oauthKey] !== value) replaceEnvVar(envPath, oauthKey, value);
	}
	return { ok: true };
}

//#endregion
//#region source/commands/sync-ims-credentials.ts
/** Synchronizes the current context IMS credentials to their AIO_COMMERCE_IMS_AUTH counterparts. */
async function run() {
	consola.default.start("Syncing IMS credentials...");
	try {
		const result = await syncImsCredentials();
		if (!result.ok) {
			switch (result.reason) {
				case "missing-env":
					consola.default.warn(".env not found — run `aio app use` to configure your workspace.");
					break;
				case "no-ims-context":
					consola.default.warn("No IMS context configured — run `aio login` to authenticate.");
					break;
				default: result.reason;
			}
			return;
		}
		consola.default.success("IMS credentials successfully synced to their AIO_COMMERCE_IMS_AUTH counterparts!");
	} catch (error) {
		consola.default.error(stringifyError(error));
		process.exit(1);
	}
}

//#endregion
//#region source/commands/index.ts
const NAMESPACE = "@adobe/aio-commerce-lib-auth";
const [command, subcommand] = process.argv.slice(2);
const USAGE = `
Usage: ${NAMESPACE} <command> [target]

Commands:
  sync-ims-credentials                 Syncs the IMS credentials from the current context to the .env file.

  help                                 Show this help message

Examples:
  ${NAMESPACE} sync-ims-credentials
`;
/** Command handlers registry mapping command names to their subcommand handlers */
const COMMANDS = { "sync-ims-credentials": run };
/**
* Generic command handler that dispatches to specific command implementations
* @param commandName - The name of the command (for error messages)
* @param target - The target subcommand to execute
* @param handlers - Map of target names to handler functions
*/
async function handleCommand(commandName, target, handlers) {
	if (!target) {
		const availableTargets = Object.keys(handlers).join(", ");
		consola.default.error(`No ${commandName} target specified`);
		consola.default.info(`Available targets: ${availableTargets}`);
		process.exit(1);
	}
	const handler = handlers[target];
	if (!handler) {
		const availableTargets = Object.keys(handlers).join(", ");
		consola.default.error(`Unknown ${commandName} target: ${target}`);
		consola.default.info(`Available targets: ${availableTargets}`);
		process.exit(1);
	}
	await handler();
}
async function main() {
	if (!command || command === "help" || command === "--help" || command === "-h") {
		consola.default.log.raw(USAGE);
		process.exit(0);
	}
	try {
		const handlers = COMMANDS[command];
		if (!handlers) {
			consola.default.error(`Unknown command: ${command}`);
			consola.default.log.raw(USAGE);
			process.exit(1);
		}
		if (typeof handlers === "function") {
			await handlers();
			return;
		}
		await handleCommand(command, subcommand, handlers);
	} catch (error) {
		consola.default.error(stringifyError(error));
		process.exit(1);
	}
}
main();

//#endregion
