// @adobe/aio-commerce-lib-app/dist/cjs/actions/installation/index.cjs
/**
 * @license
 *
 * Copyright 2026 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
const require_chunk = require('../../chunk-Cozy-1ZF.cjs');
const require_schemas = require('../../schemas-FngDuIeo.cjs');
const require_management = require('../../management-Dehz9qSJ.cjs');
const require_validate = require('../../validate-oAOpyFyb.cjs');
const require_router = require('../../router-BOR7T6qR.cjs');
let valibot = require("valibot");
valibot = require_chunk.__toESM(valibot, 1);
let _adobe_aio_commerce_lib_core_responses = require("@adobe/aio-commerce-lib-core/responses");
let _adobe_aio_lib_files = require("@adobe/aio-lib-files");
let _adobe_aio_lib_state = require("@adobe/aio-lib-state");
let openwhisk = require("openwhisk");
openwhisk = require_chunk.__toESM(openwhisk, 1);

//#region ../../packages-private/common-utils/source/storage/files-store.ts
/** Default directory prefix. */
const DEFAULT_DIR_PREFIX = "store";
/**
* Creates a generic key-value store backed by @adobe/aio-lib-files.
* Provides persistent storage that survives beyond TTL.
*
* @typeParam T - The type of data to store.
* @param options - Configuration options for the store.
* @returns A KeyValueStore implementation.
*
* @example
* ```typescript
* interface UserProfile {
*   id: string;
*   name: string;
*   email: string;
* }
*
* const store = await createFilesStore<UserProfile>({
*   dirPrefix: "profiles",
* });
*
* await store.put("user-123", { id: "123", name: "John", email: "john@example.com" });
* const profile = await store.get("user-123");
* ```
*/
async function createFilesStore(options = {}) {
	const { dirPrefix = DEFAULT_DIR_PREFIX } = options;
	return new FilesStore(await (0, _adobe_aio_lib_files.init)(), dirPrefix);
}
/**
* Key-value store implementation using @adobe/aio-lib-files.
*/
var FilesStore = class {
	constructor(files, dirPrefix) {
		this.files = files;
		this.dirPrefix = dirPrefix;
	}
	async get(key) {
		const filePath = this.buildFilePath(key);
		try {
			const content = await this.files.read(filePath);
			if (!content) return null;
			return JSON.parse(content.toString("utf8"));
		} catch {
			return null;
		}
	}
	async put(key, data) {
		const filePath = this.buildFilePath(key);
		await this.files.write(filePath, JSON.stringify(data));
	}
	async delete(key) {
		const filePath = this.buildFilePath(key);
		try {
			const result = await this.files.delete(filePath);
			return Array.isArray(result) && result.length > 0;
		} catch {
			return false;
		}
	}
	buildFilePath(key) {
		return `${this.dirPrefix}/${key}.json`;
	}
};

//#endregion
//#region ../../packages-private/common-utils/source/storage/state-store.ts
/** Default TTL for state entries (3 hours in seconds). */
const DEFAULT_TTL_SECONDS = 10800;
/** Default key prefix. */
const DEFAULT_KEY_PREFIX = "store";
/**
* Creates a generic key-value store backed by @adobe/aio-lib-state.
* Provides fast, TTL-based caching for temporary data.
*
* @typeParam T - The type of data to store.
* @param options - Configuration options for the store.
* @returns A KeyValueStore implementation.
*
* @example
* ```typescript
* interface UserSession {
*   userId: string;
*   token: string;
* }
*
* const store = await createStateStore<UserSession>({
*   keyPrefix: "session",
*   ttlSeconds: 3600, // 1 hour
* });
*
* await store.put("user-123", { userId: "123", token: "abc" });
* const session = await store.get("user-123");
* ```
*/
async function createStateStore(options = {}) {
	const { keyPrefix = DEFAULT_KEY_PREFIX, ttlSeconds = DEFAULT_TTL_SECONDS } = options;
	return new StateStore(await (0, _adobe_aio_lib_state.init)(), keyPrefix, ttlSeconds);
}
/**
* Key-value store implementation using @adobe/aio-lib-state.
*/
var StateStore = class {
	constructor(state, keyPrefix, ttlSeconds) {
		this.state = state;
		this.keyPrefix = keyPrefix;
		this.ttlSeconds = ttlSeconds;
	}
	async get(key) {
		const fullKey = this.buildKey(key);
		const result = await this.state.get(fullKey);
		if (!result?.value) return null;
		try {
			return JSON.parse(result.value);
		} catch {
			return null;
		}
	}
	async put(key, data) {
		const fullKey = this.buildKey(key);
		const value = JSON.stringify(data);
		await this.state.put(fullKey, value, { ttl: this.ttlSeconds });
	}
	async delete(key) {
		const fullKey = this.buildKey(key);
		try {
			await this.state.delete(fullKey);
			return true;
		} catch {
			return false;
		}
	}
	buildKey(key) {
		return `${this.keyPrefix}-${key}`;
	}
};

//#endregion
//#region ../../packages-private/common-utils/source/storage/combined-store.ts
/** Default TTL for cache (10 minutes). */
const DEFAULT_CACHE_TTL_SECONDS = 600;
/**
* Creates a combined key-value store that uses:
* - lib-state for fast cache during active operations
* - lib-files for persistent storage
*
* Read strategy: cache first, then persistent storage
* Write strategy:
* - Always write to cache for fast reads
* - Write to persistent storage based on shouldPersist predicate
*
* @typeParam T - The type of data to store.
* @param options - Configuration options for the stores.
*
* @example
* ```typescript
* interface Task {
*   id: string;
*   status: "pending" | "completed";
*   result?: unknown;
* }
*
* const store = await createCombinedStore<Task>({
*   cache: { keyPrefix: "task", ttlSeconds: 600 },
*   persistent: {
*     dirPrefix: "tasks",
*     shouldPersist: (task) => task.status === "completed",
*   },
* });
*
* // During processing: writes to cache only
* await store.put("task-1", { id: "1", status: "pending" });
*
* // On completion (as shouldPersist indicates): writes to both cache and persistent
* await store.put("task-1", { id: "1", status: "completed", result: {} });
* ```
*/
async function createCombinedStore(options = {}) {
	const { cache = {}, persistent = {} } = options;
	const { shouldPersist, ...filesOptions } = persistent;
	const cacheOptions = {
		ttlSeconds: DEFAULT_CACHE_TTL_SECONDS,
		...cache
	};
	const [cacheStore, persistentStore] = await Promise.all([createStateStore(cacheOptions), createFilesStore(filesOptions)]);
	return new CombinedStore(cacheStore, persistentStore, shouldPersist);
}
/**
* Combined key-value store implementation.
* Uses lib-state for cache and lib-files for persistence.
*/
var CombinedStore = class {
	constructor(cache, persistent, shouldPersist) {
		this.cache = cache;
		this.persistent = persistent;
		this.shouldPersist = shouldPersist;
	}
	async get(key) {
		const cached = await this.cache.get(key);
		if (cached !== null && cached !== void 0) return cached;
		const persisted = await this.persistent.get(key);
		if (persisted !== null && persisted !== void 0) {
			await this.cache.put(key, persisted).catch(() => {});
			return persisted;
		}
		return null;
	}
	async put(key, data) {
		await this.cache.put(key, data);
		if (!this.shouldPersist || this.shouldPersist(data)) await this.persistent.put(key, data);
	}
	async delete(key) {
		const [cacheDeleted, persistentDeleted] = await Promise.all([this.cache.delete?.(key) ?? Promise.resolve(false), this.persistent.delete?.(key) ?? Promise.resolve(false)]);
		return cacheDeleted || persistentDeleted;
	}
};

//#endregion
//#region source/management/installation/schema.ts
/** Schema for validating Adobe I/O app credentials required for installation. */
const AppDataSchema = valibot.object({
	consumerOrgId: require_schemas.nonEmptyStringValueSchema("consumerOrgId"),
	orgName: require_schemas.nonEmptyStringValueSchema("orgName"),
	projectId: require_schemas.nonEmptyStringValueSchema("projectId"),
	projectName: require_schemas.nonEmptyStringValueSchema("projectName"),
	projectTitle: require_schemas.nonEmptyStringValueSchema("projectTitle"),
	workspaceId: require_schemas.nonEmptyStringValueSchema("workspaceId"),
	workspaceName: require_schemas.nonEmptyStringValueSchema("workspaceName"),
	workspaceTitle: require_schemas.nonEmptyStringValueSchema("workspaceTitle")
});

//#endregion
//#region source/actions/installation/schema.ts
/** Request body for POST / and POST /validation (shared shape) */
const InstallationRequestBodySchema = valibot.object({
	appData: AppDataSchema,
	commerceBaseUrl: valibot.string(),
	commerceEnv: valibot.string(),
	ioEventsUrl: valibot.string(),
	ioEventsEnv: valibot.string()
});

//#endregion
//#region source/actions/installation/router.ts
const DEFAULT_ACTION_NAME = "app-management/installation";
/** Creates a workflow state store with the given prefix. */
function createWorkflowStore(prefix) {
	return createCombinedStore({
		cache: { keyPrefix: prefix },
		persistent: {
			dirPrefix: prefix,
			shouldPersist: require_management.isCompletedState
		}
	});
}
/** Creates the installation state store. */
function createInstallationStore() {
	return createWorkflowStore("installation");
}
/** Creates the uninstallation state store. */
function createUninstallationStore() {
	return createWorkflowStore("uninstallation");
}
/** Returns the storage key used to store the current installation ID. */
function getStorageKey() {
	return "current";
}
/**
* Merges rawParams with body fields, overriding API URLs.
* Shared by POST /, POST /execution, POST /uninstallation, POST /uninstallation/execution.
*/
function buildWorkflowParams(body, rawParams) {
	return {
		...rawParams,
		appData: body.appData,
		AIO_EVENTS_API_BASE_URL: body.ioEventsUrl,
		AIO_COMMERCE_AUTH_IMS_ENVIRONMENT: body.ioEventsEnv,
		AIO_COMMERCE_API_BASE_URL: body.commerceBaseUrl,
		AIO_COMMERCE_API_FLAVOR: body.commerceEnv
	};
}
/**
* Builds an InstallationContext from merged workflow params.
* Shared by POST /execution and POST /uninstallation/execution.
*/
function buildInstallationContext(params, appConfig, logFn) {
	return {
		appData: params.appData,
		params,
		logger: logFn,
		customScripts: params.customScriptsLoader?.(appConfig, logFn) ?? {}
	};
}
/**
* Reads state from a store and returns 200 with body or 204.
* Shared by GET / and GET /uninstallation.
*/
async function readStateFromStore(store, logFn) {
	const state = await store.get(getStorageKey());
	if (state) {
		logFn(`Found state: ${state.status}`);
		return (0, _adobe_aio_commerce_lib_core_responses.ok)({ body: state });
	}
	logFn("No state found");
	return (0, _adobe_aio_commerce_lib_core_responses.noContent)();
}
/**
* Creates hooks to sync installation state to storage.
*/
function createInstallationHooks(store, logFn) {
	const logAndSave = async (message, data) => {
		logFn(message);
		await store.put(getStorageKey(), data);
	};
	return {
		onInstallationStart: (state) => logAndSave("Installation started", state),
		onInstallationFailure: (state) => logAndSave("Installation failed", state),
		onInstallationSuccess: (state) => logAndSave(state.status === "succeeded" && state.metadata?.isRetry ? "Installation succeeded on retry" : "Installation succeeded", state),
		onStepStart: (event, state) => logAndSave(`Step started: ${event.stepName}`, state),
		onStepSuccess: (event, state) => logAndSave(`Step succeeded: ${event.stepName}`, state),
		onStepFailure: (event, state) => logAndSave(`Step failed: ${event.stepName} — ${event.error.message ?? `(key: ${event.error.key})`}`, state)
	};
}
/**
* Installation action router.
*
* Routes:
* - GET /                            Get current installation status
* - POST /                           Start installation (creates plan, invokes execution async)
* - POST /execution                  Execute installation (internal, called async)
* - POST /validation                 Pre-installation validation
* - POST /uninstallation             Start uninstallation (creates plan, invokes execution async)
* - GET /uninstallation              Get current uninstallation status
* - POST /uninstallation/execution   Execute uninstallation (internal, called async)
* - DELETE /uninstallation           Clear uninstallation state only (no offboarding)
*/
const router = new require_router.HttpActionRouter().use(require_router.logger({ name: () => "installation" }));
/**
* GET /installation/execution - Get current execution status
*
* Flow:
* 1. Find execution in state store
* 2. If found: return execution plan with step statuses
* 3. If not found: return empty status
*/
router.get("/", { handler: async (_req, { logger }) => {
	logger.debug("Getting installation execution status...");
	return readStateFromStore(await createInstallationStore(), (msg) => logger.debug(msg));
} });
/**
* POST / - Start installation
*
* Flow:
* 1. Find execution in state store
* 2. If found and (pending/in-progress or succeeded): return 409 Conflict
* 3. If not found or failed: create plan, invoke execution async, return 202 Accepted
*/
router.post("/", {
	body: InstallationRequestBodySchema,
	handler: async (req, { logger, rawParams }) => {
		const { appData, commerceBaseUrl } = req.body;
		logger.debug(`Starting installation for app "${appData.projectName}" (workspace: "${appData.workspaceName}", commerce: "${commerceBaseUrl}")`);
		const store = await createInstallationStore();
		const existingState = await store.get(getStorageKey());
		if (existingState) {
			if (require_management.isInProgressState(existingState)) {
				logger.debug(`Installation already in progress: ${existingState.status}`);
				return (0, _adobe_aio_commerce_lib_core_responses.conflict)(`Installation is already ${existingState.status}. Wait for it to complete.`);
			}
			if (require_management.isSucceededState(existingState)) {
				logger.debug("Installation already succeeded");
				return (0, _adobe_aio_commerce_lib_core_responses.conflict)("Installation has already completed successfully.");
			}
			logger.debug("Previous installation failed, allowing retry");
		}
		const rawAppConfig = rawParams.appConfig;
		if (!rawAppConfig) return (0, _adobe_aio_commerce_lib_core_responses.internalServerError)("The app config is missing. Does the action receive it as a parameter?");
		const appConfig = require_validate.validateCommerceAppConfig(rawAppConfig);
		const initialState = require_management.createInitialInstallationState({ config: appConfig });
		logger.debug(`Created initial state: ${initialState.id}`);
		await store.put(getStorageKey(), initialState);
		const ow = (0, openwhisk.default)();
		const mergedParams = buildWorkflowParams(req.body, rawParams);
		const activation = await ow.actions.invoke({
			name: DEFAULT_ACTION_NAME,
			blocking: false,
			result: false,
			params: {
				...mergedParams,
				initialState,
				appConfig,
				__ow_path: "/execution",
				__ow_method: "post"
			}
		});
		logger.debug(`Async execution started: ${activation.activationId}`);
		return (0, _adobe_aio_commerce_lib_core_responses.accepted)({ body: {
			message: "Installation started",
			activationId: activation.activationId,
			...initialState
		} });
	}
});
/**
* POST /installation/execution - Execute installation
* @internal - Do not add to OpenAPI Spec.
*
* This endpoint is called asynchronously by POST /installation.
* It runs the actual installation workflow and saves state.
*/
router.post("/execution", { handler: async (_req, { logger, rawParams }) => {
	const params = rawParams;
	const { initialState, appConfig: rawAppConfig, appData, AIO_COMMERCE_API_BASE_URL } = params;
	if (!initialState) return (0, _adobe_aio_commerce_lib_core_responses.badRequest)("initialState is required for execution");
	if (!rawAppConfig) return (0, _adobe_aio_commerce_lib_core_responses.badRequest)("appConfig is required for execution");
	const appConfig = require_validate.validateCommerceAppConfig(rawAppConfig);
	const store = await createInstallationStore();
	const hooks = createInstallationHooks(store, (msg) => logger.debug(msg));
	const installationContext = buildInstallationContext(params, appConfig, logger);
	logger.debug(`Executing installation ${initialState.id} for app "${appData.projectName}" (workspace: "${appData.workspaceName}", commerce: "${AIO_COMMERCE_API_BASE_URL}")`);
	const result = await require_management.runInstallation({
		installationContext,
		config: appConfig,
		initialState,
		hooks
	});
	await store.put(getStorageKey(), result);
	logger.debug(`Installation completed: ${result.status}`);
	if (require_management.isFailedState(result)) return (0, _adobe_aio_commerce_lib_core_responses.internalServerError)({ body: {
		message: "Installation failed",
		error: result.error,
		state: result
	} });
	return (0, _adobe_aio_commerce_lib_core_responses.ok)({ body: result });
} });
/**
* POST /installation/validation - Pre-installation validation
*
* Synchronously validates the step tree before installation begins.
* Accepts the same request body as POST / (installation start) so the
* frontend can reuse the same parameters without any extra mapping.
*
* Flow:
* 1. Build a ValidationContext from the request parameters
* 2. Call runValidation() — traverses the step tree and collects issues
* 3. Return the structured ValidationResult immediately (no async invoke)
*/
router.post("/validation", {
	body: InstallationRequestBodySchema,
	handler: async (req, { logger, rawParams }) => {
		logger.debug("Running pre-installation validation...");
		const rawAppConfig = rawParams.appConfig;
		if (!rawAppConfig) return (0, _adobe_aio_commerce_lib_core_responses.internalServerError)("The app config is missing. Does the action receive it as a parameter?");
		const appConfig = require_validate.validateCommerceAppConfig(rawAppConfig);
		const { appData, ...params } = buildWorkflowParams(req.body, rawParams);
		const result = await require_management.runValidation({
			validationContext: {
				appData,
				params,
				logger
			},
			config: appConfig
		});
		logger.debug(`Validation complete — valid: ${result.valid}, errors: ${result.summary.errors}, warnings: ${result.summary.warnings}`);
		return (0, _adobe_aio_commerce_lib_core_responses.ok)({ body: result });
	}
});
/**
* GET /uninstallation - Get current uninstallation status
*
* Returns 200 with state if an uninstallation has been started, 204 otherwise.
*/
router.get("/uninstallation", { handler: async (_req, { logger }) => {
	logger.debug("Getting uninstallation execution status...");
	return readStateFromStore(await createUninstallationStore(), (msg) => logger.debug(msg));
} });
/**
* POST /uninstallation - Start uninstallation (async)
*
* Flow:
* 1. Check uninstallation store for existing state
* 2. If in-progress: return 409 Conflict
* 3. Create initial uninstall state, save to store
* 4. Invoke POST /uninstallation/execution async via openwhisk
* 5. Return 202 Accepted with initial state
*/
router.post("/uninstallation", {
	body: InstallationRequestBodySchema,
	handler: async (req, { logger, rawParams }) => {
		const { appData, commerceBaseUrl } = req.body;
		logger.debug(`Starting uninstallation for app "${appData.projectName}" (workspace: "${appData.workspaceName}", commerce: "${commerceBaseUrl}")`);
		const rawAppConfig = rawParams.appConfig;
		if (!rawAppConfig) return (0, _adobe_aio_commerce_lib_core_responses.internalServerError)("The app config is missing. Does the action receive it as a parameter?");
		const appConfig = require_validate.validateCommerceAppConfig(rawAppConfig);
		const store = await createUninstallationStore();
		const existingState = await store.get(getStorageKey());
		if (existingState && require_management.isInProgressState(existingState)) {
			logger.debug(`Uninstallation already in progress: ${existingState.status}`);
			return (0, _adobe_aio_commerce_lib_core_responses.conflict)("Uninstallation is already in progress. Wait for it to complete.");
		}
		const initialState = require_management.createInitialUninstallationState({ config: appConfig });
		logger.debug(`Created initial uninstall state: ${initialState.id}`);
		await store.put(getStorageKey(), initialState);
		const workflowParams = buildWorkflowParams(req.body, rawParams);
		const activation = await (0, openwhisk.default)().actions.invoke({
			name: DEFAULT_ACTION_NAME,
			blocking: false,
			result: false,
			params: {
				...workflowParams,
				initialState,
				appConfig,
				__ow_path: "/uninstallation/execution",
				__ow_method: "post"
			}
		});
		logger.debug(`Async uninstallation started: ${activation.activationId}`);
		return (0, _adobe_aio_commerce_lib_core_responses.accepted)({ body: {
			message: "Uninstallation started",
			activationId: activation.activationId,
			...initialState
		} });
	}
});
/**
* POST /uninstallation/execution - Execute uninstallation
* @internal - Do not add to OpenAPI Spec.
*
* Flow:
* 1. Build InstallationContext from params
* 2. Run uninstallation workflow with hooks (hooks persist state per step)
* 3. Save final state to uninstallation store
* 4. On success, clear installation store
* 5. Return 200 on success, 500 on failure
*/
router.post("/uninstallation/execution", { handler: async (_req, { logger, rawParams }) => {
	const params = rawParams;
	const { initialState, appConfig: rawAppConfig, appData, AIO_COMMERCE_API_BASE_URL } = params;
	if (!initialState) return (0, _adobe_aio_commerce_lib_core_responses.badRequest)("initialState is required for execution");
	if (!rawAppConfig) return (0, _adobe_aio_commerce_lib_core_responses.badRequest)("appConfig is required for execution");
	const appConfig = require_validate.validateCommerceAppConfig(rawAppConfig);
	const store = await createUninstallationStore();
	const hooks = createInstallationHooks(store, (msg) => logger.debug(msg));
	const installationContext = buildInstallationContext(params, appConfig, logger);
	logger.debug(`Executing uninstallation ${initialState.id} for app "${appData.projectName}" (workspace: "${appData.workspaceName}", commerce: "${AIO_COMMERCE_API_BASE_URL}")`);
	const result = await require_management.runUninstallation({
		installationContext,
		config: appConfig,
		initialState,
		hooks
	});
	await store.put(getStorageKey(), result);
	logger.debug(`Uninstallation completed: ${result.status}`);
	if (require_management.isSucceededState(result)) {
		await (await createInstallationStore()).delete(getStorageKey());
		logger.debug("Cleared installation state after successful uninstallation");
	}
	if (require_management.isFailedState(result)) return (0, _adobe_aio_commerce_lib_core_responses.internalServerError)({ body: {
		message: "Uninstallation failed",
		error: result.error,
		state: result
	} });
	return (0, _adobe_aio_commerce_lib_core_responses.ok)({ body: result });
} });
/**
* DELETE /uninstallation - Clear uninstallation state
*
* Removes the stored uninstallation state without triggering any offboarding.
*/
router.delete("/uninstallation", { handler: async (_req, { logger }) => {
	logger.debug("Clearing uninstallation state...");
	await (await createUninstallationStore()).delete(getStorageKey());
	logger.debug("Uninstallation state cleared");
	return (0, _adobe_aio_commerce_lib_core_responses.noContent)();
} });

//#endregion
//#region source/actions/installation/index.ts
/**
* Factory to create the route handler for the `installation` action.
* @param args - The arguments required to create the runtime action.
*/
const installationRuntimeAction = (args) => async (params) => {
	const handler = router.handler();
	const { appConfig, customScriptsLoader } = args;
	return await handler({
		...params,
		appConfig,
		customScriptsLoader
	});
};

//#endregion
exports.installationRuntimeAction = installationRuntimeAction;
