// @adobe/aio-commerce-lib-events/dist/cjs/utils-Da725lXV.cjs
/**
 * @license
 *
 * Copyright 2026 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */

//#region \0rolldown/runtime.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __exportAll = (all, no_symbols) => {
	let target = {};
	for (var name in all) {
		__defProp(target, name, {
			get: all[name],
			enumerable: true
		});
	}
	if (!no_symbols) {
		__defProp(target, Symbol.toStringTag, { value: "Module" });
	}
	return target;
};
var __copyProps = (to, from, except, desc) => {
	if (from && typeof from === "object" || typeof from === "function") {
		for (var keys = __getOwnPropNames(from), i = 0, n = keys.length, key; i < n; i++) {
			key = keys[i];
			if (!__hasOwnProp.call(to, key) && key !== except) {
				__defProp(to, key, {
					get: ((k) => from[k]).bind(null, key),
					enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
				});
			}
		}
	}
	return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
	value: mod,
	enumerable: true
}) : target, mod));

//#endregion
let _adobe_aio_commerce_lib_core_error = require("@adobe/aio-commerce-lib-core/error");
let valibot = require("valibot");
valibot = __toESM(valibot, 1);

//#region ../../packages-private/common-utils/source/valibot/schemas.ts
const ALPHANUMERIC_OR_UNDERSCORE_REGEX = {
	any: /^[a-zA-Z0-9_]+$/,
	lowercase: /^[a-z0-9_]+$/,
	uppercase: /^[A-Z0-9_]+$/
};
const ALPHANUMERIC_OR_UNDERSCORE_OR_HYPHEN_REGEX = {
	any: /^[a-zA-Z0-9_-]+$/,
	lowercase: /^[a-z0-9_-]+$/,
	uppercase: /^[A-Z0-9_-]+$/
};
/**
* A schema for a string value.
* @param name The name of the field this schema refers to.
*/
function stringValueSchema(name) {
	return valibot.string(`Expected a string value for '${name}'`);
}
/**
* A schema for a non-empty string value.
* @param name The name of the field this schema refers to.
*/
function nonEmptyStringValueSchema(name) {
	return valibot.pipe(stringValueSchema(name), valibot.nonEmpty(`The value of "${name}" must not be empty`));
}
/**
* A schema for a boolean value.
* @param name The name of the field this schema refers to.
*/
function booleanValueSchema(name) {
	return valibot.boolean(`Expected a boolean value for '${name}'`);
}
/**
* A schema for a string that only contains alphanumeric characters and underscores.
* @param name The name of the field this schema refers to.
* @param casing The allowed casing for the string (default: "any").
*/
function alphaNumericOrUnderscoreSchema(name, casing = "any") {
	const casingLabel = casing === "any" ? "" : ` (${casing} only)`;
	return valibot.pipe(stringValueSchema(name), valibot.regex(ALPHANUMERIC_OR_UNDERSCORE_REGEX[casing], `Only alphanumeric characters and underscores are allowed in string value of "${name}"${casingLabel}`));
}
/**
* A schema for a string that only contains alphanumeric characters, underscores, and hyphens.
* @param name The name of the field this schema refers to.
* @param casing The allowed casing for the string (default: "any").
*/
function alphaNumericOrUnderscoreOrHyphenSchema(name, casing = "any") {
	const casingLabel = casing === "any" ? "" : ` (${casing} only)`;
	return valibot.pipe(stringValueSchema(name), valibot.regex(ALPHANUMERIC_OR_UNDERSCORE_OR_HYPHEN_REGEX[casing], `Only alphanumeric characters, underscores, and hyphens are allowed in string value of "${name}"${casingLabel}`));
}

//#endregion
//#region ../../packages-private/common-utils/source/valibot/utils.ts
/**
* Parses the input using the provided schema and throws a {@link CommerceSdkValidationError} error if the input is invalid.
* @param schema - The schema to use for parsing.
* @param input - The input to parse.
* @param message - Optional custom error message for the validation error.
*/
function parseOrThrow(schema, input, message) {
	const result = valibot.safeParse(schema, input);
	if (!result.success) throw new _adobe_aio_commerce_lib_core_error.CommerceSdkValidationError(message ?? "Invalid input", { issues: result.issues });
	return result.output;
}

//#endregion
Object.defineProperty(exports, '__exportAll', {
  enumerable: true,
  get: function () {
    return __exportAll;
  }
});
Object.defineProperty(exports, '__toESM', {
  enumerable: true,
  get: function () {
    return __toESM;
  }
});
Object.defineProperty(exports, 'alphaNumericOrUnderscoreOrHyphenSchema', {
  enumerable: true,
  get: function () {
    return alphaNumericOrUnderscoreOrHyphenSchema;
  }
});
Object.defineProperty(exports, 'alphaNumericOrUnderscoreSchema', {
  enumerable: true,
  get: function () {
    return alphaNumericOrUnderscoreSchema;
  }
});
Object.defineProperty(exports, 'booleanValueSchema', {
  enumerable: true,
  get: function () {
    return booleanValueSchema;
  }
});
Object.defineProperty(exports, 'nonEmptyStringValueSchema', {
  enumerable: true,
  get: function () {
    return nonEmptyStringValueSchema;
  }
});
Object.defineProperty(exports, 'parseOrThrow', {
  enumerable: true,
  get: function () {
    return parseOrThrow;
  }
});
Object.defineProperty(exports, 'stringValueSchema', {
  enumerable: true,
  get: function () {
    return stringValueSchema;
  }
});
