// @adobe/aio-commerce-lib-app/dist/es/config-3I4eLyyR.mjs
/**
 * @license
 *
 * Copyright 2026 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */

import { t as stringifyError } from "./error-_RMwtrUA.mjs";
import { t as validateCommerceAppConfig } from "./validate-Dfajrp99.mjs";
import { dirname, join, parse } from "path";
import { existsSync } from "fs";
import { access, mkdir, readFile } from "fs/promises";
import { resolveCommand } from "package-manager-detector";
import { detect, getUserAgent } from "package-manager-detector/detect";

//#region source/config/lib/define.ts
/**
* Helper to type-safely define the app config.
* @param config - The app config to define.
*
* @example
* ```typescript
* import { defineConfig } from "@adobe/aio-commerce-lib-app";
*
* // In app.commerce.config.js
* export default defineConfig({
*   // You get autocompletion and type-safety for the config object.
*   businessConfig: { ... }
* });
*/
function defineConfig(config) {
	return config;
}

//#endregion
//#region ../../packages-private/scripting-utils/source/project.ts
/**
* This module exports shared project utilities for the AIO Commerce SDK.
* @packageDocumentation
*/
/**
* Find a file by walking up parent directories
* @param name - The file name to search for (or array of file names)
* @param options - Search options
* @returns The path to the file, or undefined if not found
*/
async function findUp(name, options = {}) {
	const names = Array.isArray(name) ? name : [name];
	const cwd = options.cwd || process.cwd();
	const { root } = parse(cwd);
	const stopAt = options.stopAt || root;
	let currentDir = cwd;
	while (true) {
		for (const fileName of names) {
			const filePath = join(currentDir, fileName);
			try {
				await access(filePath);
				return filePath;
			} catch {}
		}
		if (currentDir === stopAt || currentDir === root) return;
		currentDir = dirname(currentDir);
	}
}
/**
* Find the nearest package.json file in the current working directory or its parents
* @param cwd The current working directory
*/
async function findNearestPackageJson(cwd = process.cwd()) {
	const packageJsonPath = await findUp("package.json", { cwd });
	if (!packageJsonPath) return null;
	return packageJsonPath;
}
/**
* Read the package.json file
* @param cwd The current working directory
*/
async function readPackageJson(cwd = process.cwd()) {
	const packageJsonPath = await findNearestPackageJson(cwd);
	if (!packageJsonPath) return null;
	return JSON.parse(await readFile(packageJsonPath, "utf-8"));
}
/**
* Check if the current working directory is an ESM project.
* @param cwd The current working directory
*/
async function isESM(cwd = process.cwd()) {
	const packageJson = await readPackageJson(cwd);
	if (!packageJson) throw new Error("Could not find a `package.json` file in the current working directory or its parents.");
	return packageJson.type === "module";
}
/**
* Get the root directory of the project
* @param cwd The current working directory
*/
async function getProjectRootDirectory(cwd = process.cwd()) {
	const packageJsonPath = await findNearestPackageJson(cwd);
	if (!packageJsonPath) throw new Error("Could not find a the root directory of the project. `package.json` file not found.");
	return dirname(packageJsonPath);
}
/**
* Create the output directory for the given file or folder (relative to the project root)
* @param fileOrFolder - The file or folder to create
*/
async function makeOutputDirFor(fileOrFolder) {
	const outputDir = join(await getProjectRootDirectory(), fileOrFolder);
	if (!existsSync(outputDir)) await mkdir(outputDir, { recursive: true });
	return outputDir;
}
const VALID_PACKAGE_MANAGERS = [
	"npm",
	"pnpm",
	"yarn",
	"bun"
];
/**
* Type guard asserting that a value is a supported `PackageManager`.
* @param value - The value to check
*/
function isValidPackageManager(value) {
	return value !== void 0 && VALID_PACKAGE_MANAGERS.includes(value);
}
/**
* Detect the package manager for a project.
* @param cwd - Directory to start detection from; defaults to `process.cwd()`
*/
async function detectPackageManager(cwd = process.cwd()) {
	const result = await detect({ cwd: await getProjectRootDirectory(cwd) });
	if (isValidPackageManager(result?.name)) return result.name;
	const fromUserAgent = getUserAgent() ?? void 0;
	if (isValidPackageManager(fromUserAgent)) return fromUserAgent;
	return "npm";
}
/**
* Get the exec command that runs a **locally installed** binary from
* `node_modules/.bin` for the given package manager.
* @param packageManager - The detected package manager
*/
function getExecCommand(packageManager) {
	const resolved = resolveCommand(packageManager, "execute-local", []);
	if (!resolved) return "npx";
	return [resolved.command, ...resolved.args].filter(Boolean).join(" ");
}
/**
* Get the command to install the given dependencies with the given package
* manager (e.g. `pnpm add foo bar`, `npm i foo bar`).
* @param packageManager - The detected package manager
* @param packages - Package specifiers to install (e.g. `["foo@^1.0"]`)
*/
function getInstallCommand(packageManager, packages) {
	const resolved = resolveCommand(packageManager, "add", packages);
	if (!resolved) return {
		command: "npm",
		args: ["install", ...packages]
	};
	return {
		args: resolved.args.filter((a) => Boolean(a)),
		command: resolved.command
	};
}

//#endregion
//#region source/config/lib/parser.ts
const configPaths = Object.freeze([
	"app.commerce.config.js",
	"app.commerce.config.ts",
	"app.commerce.config.cjs",
	"app.commerce.config.mjs",
	"app.commerce.config.mts",
	"app.commerce.config.cts"
]);
/**
* Try to find (up to the nearest package.json file) the app config file.
*
* Searches for config files in the following order of priority:
* 1. `app.commerce.config.js` - JavaScript (CommonJS or ESM)
* 2. `app.commerce.config.ts` - TypeScript
* 3. `app.commerce.config.cjs` - CommonJS
* 4. `app.commerce.config.mjs` - ES Module
* 5. `app.commerce.config.mts` - ES Module TypeScript
* 6. `app.commerce.config.cts` - CommonJS TypeScript
*
* @param cwd The current working directory
* @returns The path to the config file, or null if not found
*/
async function resolveCommerceAppConfig(cwd = process.cwd()) {
	const packageJsonPath = await findNearestPackageJson(cwd);
	if (!packageJsonPath) return null;
	const rootDirectory = dirname(packageJsonPath);
	for (const configPath of configPaths) {
		const configFilePath = await findUp(configPath, {
			cwd,
			stopAt: rootDirectory
		});
		if (configFilePath) return configFilePath;
	}
	return null;
}
/**
* Read the commerce app config file as-is, without validating it.
*
* Supports multiple config file formats (see {@link resolveCommerceAppConfig} for the list).
* The config file must export a default export with the configuration object.
*
* @param cwd The current working directory
* @returns The raw config object from the file
* @throws {Error} If no config file is found or if the file doesn't export a default export
*/
async function readCommerceAppConfig(cwd = process.cwd()) {
	const configFilePath = await resolveCommerceAppConfig(cwd);
	if (!configFilePath) throw new Error("Could not find a commerce app config file in the current working directory or its parents.");
	let config = null;
	try {
		const { createJiti } = await import("jiti");
		config = await createJiti(import.meta.url).import(configFilePath);
	} catch (error) {
		const message = stringifyError(error);
		throw new Error(`Failed to read commerce app config file at ${configFilePath}: ${message}`, { cause: error });
	}
	if (!(config && "default" in config) || config.default === void 0 || config.default === null || typeof config.default === "object" && Object.keys(config.default).length === 0) throw new Error("Commerce app config file does not export a default export or exports an empty object. Make sure you use `export default` or `module.exports = { /* your config */ }`");
	return config.default;
}
/**
* Read the commerce app config file and parse its contents into its schema.
*
* Supports multiple config file formats (see {@link resolveCommerceAppConfig} for the list).
* The config file must export a default export with the configuration object.
*
* @param cwd The current working directory
* @returns The validated and parsed config object
* @throws {Error} If no config file is found, if the file doesn't export a default export, or if validation fails
*/
async function parseCommerceAppConfig(cwd = process.cwd()) {
	return validateCommerceAppConfig(await readCommerceAppConfig(cwd));
}

//#endregion
export { getExecCommand as a, isESM as c, defineConfig as d, detectPackageManager as i, makeOutputDirFor as l, readCommerceAppConfig as n, getInstallCommand as o, resolveCommerceAppConfig as r, getProjectRootDirectory as s, parseCommerceAppConfig as t, readPackageJson as u };
