// @adobe/aio-commerce-lib-app/dist/es/management-BUX-w2Ji.mjs
/**
 * @license
 *
 * Copyright 2026 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */

import { t as stringifyError } from "./error-_RMwtrUA.mjs";
import { a as hasCustomInstallationSteps, c as hasEventing, d as hasAdminUiSdk, l as hasExternalEvents, n as hasWebhooks, s as hasCommerceEvents } from "./webhooks-B5oq9ULL.mjs";
import { t as inspect } from "./logging-DbJHXSKh.mjs";
import { CommerceSdkValidationError } from "@adobe/aio-commerce-lib-core/error";
import * as v from "valibot";
import { unwrapHttpError } from "@adobe/aio-commerce-lib-api/utils";
import * as camelcaseModule from "camelcase";
import { createAdminUiApiClient } from "@adobe/aio-commerce-lib-admin-ui/api";
import { resolveCommerceHttpClientParams, resolveIoEventsHttpClientParams } from "@adobe/aio-commerce-lib-api";
import { resolveAuthParams, resolveImsAuthParams } from "@adobe/aio-commerce-lib-auth";
import { createCustomCommerceEventsApiClient, createEventProvider, createEventSubscription, deleteEventProvider, deleteEventSubscription, getAllEventProviders, getAllEventSubscriptions, updateEventingConfiguration } from "@adobe/aio-commerce-lib-events/commerce";
import { createCustomAdobeIoEventsApiClient, createEventMetadataForProvider, createEventProvider as createEventProvider$1, createRegistration, deleteEventMetadataForProvider, deleteEventProvider as deleteEventProvider$1, deleteRegistration, getAllEventProviders as getAllEventProviders$1, getAllRegistrations } from "@adobe/aio-commerce-lib-events/io-events";
import { createCustomCommerceWebhooksApiClient, getWebhookList, subscribeWebhook, unsubscribeWebhook } from "@adobe/aio-commerce-lib-webhooks/api";

//#region ../../packages-private/common-utils/source/valibot/utils.ts
/**
* Parses the input using the provided schema and throws a {@link CommerceSdkValidationError} error if the input is invalid.
* @param schema - The schema to use for parsing.
* @param input - The input to parse.
* @param message - Optional custom error message for the validation error.
*/
function parseOrThrow(schema, input, message) {
	const result = v.safeParse(schema, input);
	if (!result.success) throw new CommerceSdkValidationError(message ?? "Invalid input", { issues: result.issues });
	return result.output;
}

//#endregion
//#region source/management/installation/workflow/hooks.ts
/** Helper to call a hook if it exists. */
async function callHook(hooks, hookName, ...args) {
	const hook = hooks?.[hookName];
	if (hook) await hook(...args);
}

//#endregion
//#region source/management/installation/workflow/step.ts
/** Check if a step is a leaf step. */
function isLeafStep(step) {
	return step.type === "leaf";
}
/** Check if a step is a branch step. */
function isBranchStep(step) {
	return step.type === "branch";
}
/**
* Define a leaf step (executable, no children).
*
* @example
* ```typescript
* const createProviders = defineLeafStep({
*   name: "providers",
*   meta: { install: { label: "Create Providers", description: "Creates I/O Events providers" } },
*   install: async ({ config, stepContext }) => {
*     const { eventsClient } = stepContext;
*     return eventsClient.createProvider(config.eventing);
*   },
* });
* ```
*/
function defineLeafStep(options) {
	return {
		type: "leaf",
		name: options.name,
		meta: options.meta,
		when: options.when,
		install: options.install,
		uninstall: options.uninstall,
		validate: options.validate
	};
}
/**
* Define a branch step (container with children, no runner).
*
* @example
* ```typescript
* const eventing = defineBranchStep({
*   name: "eventing",
*   meta: { install: { label: "Eventing", description: "Sets up I/O Events" } },
*   when: hasEventing,
*   context: async (ctx) => ({ eventsClient: await createEventsClient(ctx) }),
*   children: [commerceEventsStep, externalEventsStep],
* });
* ```
*/
function defineBranchStep(options) {
	return {
		type: "branch",
		name: options.name,
		meta: options.meta,
		when: options.when,
		context: options.context,
		children: options.children,
		validate: options.validate
	};
}

//#endregion
//#region source/management/installation/workflow/utils.ts
/** Returns the current time as an ISO string. */
function nowIsoString() {
	return (/* @__PURE__ */ new Date()).toISOString();
}
/** Sets a value at a nested path in the data object. */
function setAtPath(data, path, value) {
	const lastKey = path.at(-1);
	if (!lastKey) return;
	let current = data;
	for (let i = 0; i < path.length - 1; i++) {
		const key = path[i];
		current[key] ??= {};
		current = current[key];
	}
	current[lastKey] = value;
}
/** Gets a value at a nested path in the data object. */
function getAtPath(data, path) {
	let current = data;
	for (const key of path) {
		if (current == null || typeof current !== "object") return;
		current = current[key];
	}
	return current;
}
/** Creates an installation error from an exception. */
async function createInstallationError(err, path, key = "STEP_EXECUTION_FAILED") {
	return {
		path,
		key,
		message: await unwrapHttpError(err)
	};
}
/** Creates a succeeded installation state. */
function createSucceededState(base) {
	return {
		...base,
		status: "succeeded",
		completedAt: nowIsoString()
	};
}
/** Creates a failed installation state. */
function createFailedState(base, error) {
	return {
		...base,
		status: "failed",
		completedAt: nowIsoString(),
		error
	};
}

//#endregion
//#region source/management/installation/workflow/runner.ts
/**
* Creates an initial installation state from a root step and config.
*
* Filters steps based on their `when` conditions and builds a
* tree structure with all steps set to "pending".
*/
function createInitialState(options) {
	const { rootStep, config, mode } = options;
	return {
		id: crypto.randomUUID(),
		startedAt: nowIsoString(),
		status: "in-progress",
		step: buildInitialStepStatus(rootStep, config, [], mode),
		data: null
	};
}
/**
* Creates a retry state from a failed state.
* Preserves succeeded steps and their data so the workflow resumes from
* the failed step rather than restarting from scratch.
*/
function createRetryState(failedState) {
	return {
		id: failedState.id,
		startedAt: failedState.startedAt,
		status: "in-progress",
		step: resetFailedSteps(failedState.step),
		data: failedState.data
	};
}
/** Recursively resets non-succeeded steps back to "pending". */
function resetFailedSteps(step) {
	return {
		...step,
		status: step.status === "succeeded" ? "succeeded" : "pending",
		children: step.children.map(resetFailedSteps)
	};
}
/**
* Executes a workflow from an initial state. Returns the final state (never throws).
*/
async function executeWorkflow(options) {
	return executeWorkflowWithMode(options, "install");
}
/**
* Executes an uninstall workflow from an initial state. Returns the final state (never throws).
* Steps with an `uninstall` handler get it called; steps without are silently skipped.
*/
async function executeUninstallWorkflow(options) {
	return executeWorkflowWithMode(options, "uninstall");
}
/**
* Internal implementation shared by executeWorkflow and executeUninstallWorkflow.
*/
async function executeWorkflowWithMode(options, mode) {
	const { rootStep, installationContext, config, initialState, hooks } = options;
	const step = structuredClone(initialState.step);
	const context = {
		installationContext,
		config,
		id: initialState.id,
		startedAt: initialState.startedAt,
		step,
		data: initialState.data,
		error: null,
		hooks,
		mode
	};
	await callHook(hooks, "onInstallationStart", snapshot(context));
	try {
		await executeStep(rootStep, context.step, {}, context);
		const succeeded = createSucceededState({
			id: context.id,
			startedAt: context.startedAt,
			step: context.step,
			data: context.data
		});
		await callHook(hooks, "onInstallationSuccess", succeeded);
		return succeeded;
	} catch (err) {
		const error = context.error ?? await createInstallationError(err, [], "INSTALLATION_FAILED");
		const failed = createFailedState({
			step: context.step,
			id: context.id,
			startedAt: context.startedAt,
			data: context.data
		}, error);
		await callHook(hooks, "onInstallationFailure", failed);
		return failed;
	}
}
/**
* Builds initial step status from a step definition.
* Filters steps based on their `when` conditions.
*/
function buildInitialStepStatus(step, config, parentPath, mode) {
	const path = [...parentPath, step.name];
	const children = [];
	if (isBranchStep(step) && step.children.length > 0) for (const child of step.children) {
		if (child.when && !child.when(config)) continue;
		children.push(buildInitialStepStatus(child, config, path, mode));
	}
	return {
		id: crypto.randomUUID(),
		name: step.name,
		path,
		meta: mode === "uninstall" && step.meta.uninstall ? step.meta.uninstall : step.meta.install,
		status: "pending",
		children
	};
}
/** Snapshot current execution as InProgressInstallationState. */
function snapshot(context) {
	return {
		id: context.id,
		startedAt: context.startedAt,
		status: "in-progress",
		step: context.step,
		data: context.data
	};
}
/** Executes a single step (branch or leaf) recursively. */
async function executeStep(step, stepStatus, inherited, context) {
	if (stepStatus.status === "succeeded") return;
	const { path } = stepStatus;
	const isLeaf = isLeafStep(step);
	stepStatus.status = "in-progress";
	await callHook(context.hooks, "onStepStart", {
		path,
		stepName: step.name,
		isLeaf
	}, snapshot(context));
	try {
		if (isBranchStep(step)) await executeBranchStep(step, stepStatus, inherited, context);
		else if (isLeafStep(step)) await executeLeafStep(step, stepStatus, inherited, context);
		stepStatus.status = "succeeded";
		context.data ??= {};
		await callHook(context.hooks, "onStepSuccess", {
			path,
			stepName: step.name,
			isLeaf,
			result: getAtPath(context.data, path)
		}, snapshot(context));
	} catch (err) {
		stepStatus.status = "failed";
		context.error ??= await createInstallationError(err, path);
		await callHook(context.hooks, "onStepFailure", {
			path,
			stepName: step.name,
			isLeaf,
			error: context.error
		}, snapshot(context));
		throw err;
	}
}
/** Executes a branch step by processing its children. */
async function executeBranchStep(step, stepStatus, inherited, context) {
	let childContext = inherited;
	if (step.context) {
		const stepContext = await step.context(context.installationContext);
		childContext = {
			...inherited,
			...stepContext
		};
	}
	for (const child of stepStatus.children) {
		const childStep = step.children.find((c) => c.name === child.name);
		if (!childStep) throw new Error(`Step "${child.name}" not found`);
		await executeStep(childStep, child, childContext, context);
	}
}
/** Executes a leaf step and stores its result, or runs uninstall if in uninstall mode. */
async function executeLeafStep(step, stepStatus, inherited, context) {
	const executionContext = {
		...context.installationContext,
		...inherited
	};
	if (context.mode === "uninstall") {
		if (step.uninstall) await step.uninstall(context.config, executionContext);
		return;
	}
	const result = await step.install(context.config, executionContext);
	context.data ??= {};
	setAtPath(context.data, stepStatus.path, result);
}

//#endregion
//#region source/management/installation/workflow/types.ts
/** Type guard for pending installation state. */
/** Type guard for in-progress installation state. */
function isInProgressState(state) {
	return state.status === "in-progress";
}
/** Type guard for succeeded installation state. */
function isSucceededState(state) {
	return state.status === "succeeded";
}
/** Type guard for failed installation state. */
function isFailedState(state) {
	return state.status === "failed";
}
/** Type guard for completed installation state (succeeded or failed). */
function isCompletedState(state) {
	return state.status === "succeeded" || state.status === "failed";
}

//#endregion
//#region source/management/installation/workflow/validation.ts
/**
* Runs validation over the full step tree, returning a structured result.
*
* - Respects `when` conditions (skips steps that don't apply to the config)
* - Calls each step's optional `validate` handler
* - Sets up branch context factories before validating children
* - Never throws; all errors from validate handlers are caught and reported as issues
*/
async function validateStepTree(options) {
	const { rootStep, validationContext, config } = options;
	const result = await validateStep(rootStep, config, validationContext, []);
	const summary = aggregateSummary(result);
	return {
		valid: summary.errors === 0 && summary.warnings === 0,
		result,
		summary
	};
}
/** Recursively validates a single step and its children. */
async function validateStep(step, config, context, parentPath) {
	const path = [...parentPath, step.name];
	const issues = await runStepValidation(step, config, context);
	const children = [];
	if (isBranchStep(step) && step.children.length > 0) {
		const resolved = await resolveBranchContext(step, context);
		issues.push(...resolved.issues);
		for (const child of step.children) {
			if (child.when && !child.when(config)) continue;
			children.push(await validateStep(child, config, resolved.childContext, path));
		}
	}
	return {
		name: step.name,
		path,
		meta: step.meta.install,
		issues,
		children
	};
}
/** Resolves the child context for a branch step, reporting errors as issues. */
async function resolveBranchContext(step, context) {
	if (!step.context) return {
		childContext: context,
		issues: []
	};
	try {
		const stepContext = await step.context(context);
		return {
			childContext: {
				...context,
				...stepContext
			},
			issues: []
		};
	} catch (err) {
		return {
			childContext: context,
			issues: [{
				code: "VALIDATION_CONTEXT_ERROR",
				message: err instanceof Error ? err.message : String(err),
				severity: "error"
			}]
		};
	}
}
/** Runs a step's validate handler, catching any thrown errors as issues. */
async function runStepValidation(step, config, context) {
	if (!step.validate) return [];
	try {
		return await step.validate(config, context);
	} catch (err) {
		return [{
			code: "VALIDATION_HANDLER_ERROR",
			message: err instanceof Error ? err.message : String(err),
			severity: "error"
		}];
	}
}
/** Recursively aggregates issue counts across the full validation tree. */
function aggregateSummary(result) {
	let errors = 0;
	let warnings = 0;
	for (const issue of result.issues) if (issue.severity === "error") errors++;
	else if (issue.severity === "warning") warnings++;
	for (const child of result.children) {
		const childSummary = aggregateSummary(child);
		errors += childSummary.errors;
		warnings += childSummary.warnings;
	}
	return {
		totalIssues: errors + warnings,
		errors,
		warnings
	};
}

//#endregion
//#region source/management/installation/custom-installation/custom-scripts.ts
const camelcase = camelcaseModule.default.default ?? camelcaseModule.default;
const ScriptModuleSchema = v.union([v.function(), v.object({
	install: v.function(),
	uninstall: v.optional(v.function())
})]);
/**
* Validates that a loaded script export can be executed as a custom installation step.
*
* @param module - The loaded script export to validate.
* @throws If the export is not a function or an object with an `install` function.
*/
function assertScriptModule(module) {
	parseOrThrow(ScriptModuleSchema, module, "Invalid script module format. Expected a function or an object with an `install` method.");
}
/**
* Parses and validates a custom script module from the provided customScripts context.
* It supports both direct function exports and default exports.
*
* @param customScripts - The customScripts context containing the loaded modules
* @param script - The script path to resolve the module for
*/
function getScriptModule(customScripts, script) {
	let scriptModule = customScripts[script];
	if (!scriptModule) throw new Error(`Script ${script} not found in customScripts context. Make sure the script is defined in the configuration and the action was generated with custom scripts support.`);
	if (typeof scriptModule === "object" && "default" in scriptModule) scriptModule = scriptModule.default;
	assertScriptModule(scriptModule);
	return scriptModule;
}
function resolveCustomScriptHandler(scriptModule, handler) {
	if (typeof scriptModule === "function") return handler === "install" ? scriptModule : null;
	if (handler in scriptModule && typeof scriptModule[handler] === "function") return scriptModule[handler];
	return null;
}
/**
* Creates a leaf step for executing a single custom installation script.
*/
function createCustomScriptStep(scriptConfig) {
	const { script, name, description } = scriptConfig;
	return defineLeafStep({
		name: camelcase(name),
		meta: { install: {
			label: name,
			description
		} },
		install: async (config, context) => {
			const { logger } = context;
			const customScripts = context.customScripts || {};
			logger.info(`Executing custom installation script: ${name}`);
			logger.debug(`Script path: ${script}`);
			const scriptResult = await resolveCustomScriptHandler(getScriptModule(customScripts, script), "install")(config, context);
			logger.info(`Successfully executed script: ${name}`);
			return {
				script,
				data: scriptResult
			};
		},
		uninstall: async (config, context) => {
			const { logger } = context;
			const customScripts = context.customScripts || {};
			logger.debug(`Uninstalling custom script: ${name}`);
			const uninstall = resolveCustomScriptHandler(getScriptModule(customScripts, script), "uninstall");
			if (!uninstall) {
				logger.debug(`Script ${script} does not export an uninstall function, skipping uninstall.`);
				return;
			}
			await uninstall(config, context);
			logger.info(`Successfully uninstalled script: ${name}`);
		}
	});
}
/**
* Creates child steps dynamically based on the custom installation scripts
* defined in the configuration. Each script becomes a separate leaf step.
*/
function createCustomScriptSteps(config) {
	if (!hasCustomInstallationSteps(config)) return [];
	const steps = config.installation.customInstallationSteps;
	if (new Set(steps.map((step) => step.name)).size !== steps.length) throw new Error("Duplicate step names detected in custom installation steps. Each step must have a unique name.");
	return steps.map((scriptConfig) => createCustomScriptStep(scriptConfig));
}

//#endregion
//#region source/management/installation/custom-installation/branch.ts
/** Root custom installation step that executes custom installation scripts. */
const customInstallationStepBase = defineBranchStep({
	name: "customInstallationSteps",
	meta: {
		install: {
			label: "Custom Installation Steps",
			description: "Executes custom installation scripts defined in the application configuration"
		},
		uninstall: {
			label: "Custom Uninstallation Steps",
			description: "Executes custom uninstallation scripts defined in the application configuration"
		}
	},
	when: hasCustomInstallationSteps,
	children: []
});
/**
* Creates the custom installation step with dynamic children based on config.
* Each custom script becomes a direct child step.
*/
function createCustomInstallationStep(config) {
	return {
		...customInstallationStepBase,
		children: createCustomScriptSteps(config)
	};
}

//#endregion
//#region source/management/installation/custom-installation/define.ts
/**
* Define a custom installation step with type-safe parameters.
*
* This helper provides type safety and IDE autocompletion for custom installation scripts.
* Accepts either a plain function (install only) or an object with `install` and optional
* `uninstall` handlers.
*
* @example Plain function (install only):
* ```typescript
* import { defineCustomInstallationStep } from "@adobe/aio-commerce-lib-app/management";
*
* export default defineCustomInstallationStep(async (config, context) => {
*   const { logger, params } = context;
*   logger.info(`Setting up ${config.metadata.displayName}...`);
*   return { status: "success" };
* });
* ```
*
* @example Object form with install and uninstall:
* ```typescript
* import { defineCustomInstallationStep } from "@adobe/aio-commerce-lib-app/management";
*
* export default defineCustomInstallationStep({
*   install: async (config, context) => {
*     context.logger.info(`Registering ${config.metadata.displayName}...`);
*     return { status: "success" };
*   },
*   uninstall: async (config, context) => {
*     context.logger.info(`Removing ${config.metadata.displayName}...`);
*   },
* });
* ```
*/
function defineCustomInstallationStep(handlerOrDefinition) {
	return handlerOrDefinition;
}

//#endregion
//#region source/management/installation/utils/http-error.ts
/**
* Unwraps an error via `unwrapHttpError`, prefixes it, logs it at error level,
* and throws a new `Error` with the enriched message. Returns `Promise<never>`
* so it composes cleanly inside `.catch()` without widening the surrounding type.
*/
async function throwHttpError(logger, error, prefix) {
	const message = `${prefix}: ${await unwrapHttpError(error)}`;
	logger.error(message);
	throw new Error(message);
}

//#endregion
//#region source/management/installation/admin-ui-sdk/helpers.ts
/**
* Registers the extension with Commerce via POST /V1/adminuisdk/extension.
*
* @param context - The execution context providing the Admin UI SDK client and logger.
*/
async function registerExtension(context) {
	const { adminUiClient, appData, logger } = context;
	const extensionName = process.env.__OW_NAMESPACE;
	if (!extensionName) throw new Error("__OW_NAMESPACE environment variable is not set");
	logger.info(`Registering Admin UI SDK extension: ${appData.projectName}`);
	const { extensionId } = await adminUiClient.registerExtension({
		extensionName,
		extensionTitle: appData.projectTitle,
		extensionUrl: `https://${extensionName}.adobeio-static.net/index.html`,
		extensionWorkspace: appData.workspaceName
	}).catch((error) => throwHttpError(logger, error, "Failed to register Admin UI SDK extension"));
	logger.info(`Admin UI SDK extension registered successfully: ${extensionId}`);
}
/**
* Unregisters the extension from Commerce via DELETE /V1/adminuisdk/extension/:workspace_name/:extension_name.
* Best-effort: errors are logged as warnings and do not stop the uninstall workflow.
*
* @param context - The execution context providing the Admin UI SDK client and logger.
*/
async function unregisterExtension(context) {
	const { adminUiClient, appData, logger } = context;
	const extensionName = process.env.__OW_NAMESPACE;
	if (!extensionName) {
		logger.warn("__OW_NAMESPACE environment variable is not set; skipping Admin UI SDK extension unregistration. Continuing uninstall.");
		return;
	}
	logger.info(`Unregistering Admin UI SDK extension "${extensionName}" from workspace "${appData.workspaceName}"...`);
	try {
		await adminUiClient.unregisterExtension({
			workspaceName: appData.workspaceName,
			extensionName
		});
		logger.info(`Admin UI SDK extension "${extensionName}" unregistered successfully.`);
	} catch (error) {
		const msg = await unwrapHttpError(error);
		logger.warn(`Failed to unregister Admin UI SDK extension "${extensionName}": ${msg}. Continuing uninstall.`);
	}
}

//#endregion
//#region source/management/installation/admin-ui-sdk/utils.ts
function createAdminUiClient(params) {
	return createAdminUiApiClient(resolveCommerceHttpClientParams(params, { tryForwardAuthProvider: true }));
}
/** Creates the Admin UI SDK step context with a lazy-initialized API client. */
const createAdminUiSdkStepContext = (installation) => {
	const { params } = installation;
	let adminUiClient = null;
	return { get adminUiClient() {
		if (adminUiClient === null) adminUiClient = createAdminUiClient(params);
		return adminUiClient;
	} };
};

//#endregion
//#region source/management/installation/admin-ui-sdk/branch.ts
/** Leaf step that registers the extension (POST) on install and unregisters it (DELETE) on uninstall. */
const registerExtensionStep = defineLeafStep({
	name: "register-extension",
	meta: {
		install: {
			label: "Register Extension",
			description: "Registers the Admin UI SDK extension in Adobe Commerce"
		},
		uninstall: {
			label: "Unregister Extension",
			description: "Removes the Admin UI SDK extension from Adobe Commerce"
		}
	},
	install: (_, context) => registerExtension(context),
	uninstall: (_, context) => unregisterExtension(context)
});
/** Branch step for setting up the Admin UI SDK extension registration. */
const adminUiSdkStep = defineBranchStep({
	name: "admin-ui-sdk",
	meta: {
		install: {
			label: "Admin UI SDK",
			description: "Registers the extension with Adobe Commerce Admin UI SDK"
		},
		uninstall: {
			label: "Admin UI SDK",
			description: "Removes the extension from Adobe Commerce Admin UI SDK"
		}
	},
	when: hasAdminUiSdk,
	context: createAdminUiSdkStepContext,
	children: [registerExtensionStep]
});

//#endregion
//#region source/management/installation/events/utils.ts
const COMMERCE_PROVIDER_TYPE = "dx_commerce_events";
const EXTERNAL_PROVIDER_TYPE = "3rd_party_custom_events";
const PROVIDER_TYPE_TO_LABEL = {
	[COMMERCE_PROVIDER_TYPE]: "Commerce",
	[EXTERNAL_PROVIDER_TYPE]: "External"
};
/** Max characters taken from `metadata.id` in the I/O Events provider `instance_id`. */
const METADATA_ID_MAX_LENGTH_FOR_INSTANCE_ID = 100;
/**
* Generates a unique instance ID for I/O Events for this app deployment.
* Uses `{metadata.id (first 100 chars)}-{providerKeyOrSlug}-{workspaceId}` (lowercased).
*
* @param metadata - The metadata of the application
* @param provider - The event provider (optional `key`, else label is slugified)
* @param workspaceId - Adobe I/O Developer Console workspace ID for this deployment
*/
function generateInstanceId(metadata, provider, workspaceId) {
	const appId = metadata.id.slice(0, METADATA_ID_MAX_LENGTH_FOR_INSTANCE_ID);
	const slugLabel = provider.label.toLowerCase().replace(/\s+/g, "-");
	return `${appId}-${provider.key ?? slugLabel}-${workspaceId}`.toLowerCase();
}
/**
* Old version of instanceId generator which can be not unique within the same ORG.
*
* @param metadata - The metadata of the application
* @param provider - The event provider for which to generate the instance ID
* @deprecated use {@link generateInstanceId} instead
*/
function generateInstanceIdDeprecated(metadata, provider) {
	const slugLabel = provider.label.toLowerCase().replace(/\s+/g, "-");
	return `${metadata.id}-${provider.key ?? slugLabel}`.toLowerCase();
}
/**
* Find an existing event provider by its instance ID.
* @param allProviders - The list of all existing event providers.
* @param instanceId - The instance ID to search for.
*/
function findExistingProvider(allProviders, instanceId) {
	return allProviders.find((provider) => provider.instance_id === instanceId) ?? null;
}
/**
* Find existing event metadata by its event name.
* @param allMetadata - The list of all existing event metadata.
* @param eventName - The event name to search for.
*/
function findExistingProviderMetadata(allMetadata, eventName) {
	return allMetadata.find((meta) => meta.event_code === eventName) ?? null;
}
/**
* Find existing event registrations by client ID and name.
* @param allRegistrations - The list of all existing event registrations.
* @param clientId - The client ID of the workspace where the registration was created.
* @param name - The name of the registration to search for.
*/
function findExistingRegistrations(allRegistrations, clientId, name) {
	return allRegistrations.find((reg) => reg.client_id === clientId && reg.name === name);
}
/**
* Generates a namespaced event name by combining the application ID with the event name.
* @param metadata
* @param name
*/
function getNamespacedEvent(metadata, name) {
	return `${metadata.id}.${name}`.toLowerCase();
}
/**
* Get the fully qualified name of an event for I/O Events based on the provider type.
* @param name - The name of the event.
* @param providerType - The type of the event provider.
*/
function getIoEventCode(name, providerType) {
	return providerType === "dx_commerce_events" ? `com.adobe.commerce.${name}` : name;
}
/** Maps a provider's metadata type to its human-readable label ("Commerce" or "External"). */
function getProviderTypeLabel(provider) {
	return PROVIDER_TYPE_TO_LABEL[provider.provider_metadata];
}
/**
* Generates a registration name based on the provider type, provider label, and runtime action.
*
* Prefixes with the provider type label ("Commerce" or "External") for readability, then the
* provider's own label to ensure uniqueness when multiple providers of the same type route
* events to the same runtime action.
*
* @param provider - The provider this registration is associated to.
* @param runtimeAction - The runtime action this registration points to.
*/
function getRegistrationName(provider, runtimeAction) {
	const [packageName, actionName] = runtimeAction.split("/").map(kebabToTitleCase);
	return `${getProviderTypeLabel(provider)} Event Registration: ${provider.label} - ${actionName} (${packageName})`;
}
/**
* Returns the registration name in the legacy format used by SDK versions that built
* the name from the generic provider type label ("Commerce" / "External") rather than
* the provider's own label. Used during uninstall to match registrations that were
* created before the naming change.
*
* @param provider - The provider this registration is associated to.
* @param runtimeAction - The runtime action this registration points to.
* @deprecated Use {@link getRegistrationName} for all new registrations.
*/
function getLegacyRegistrationName(provider, runtimeAction) {
	const [packageName, actionName] = runtimeAction.split("/").map(kebabToTitleCase);
	return `${getProviderTypeLabel(provider)} Event Registration: ${actionName} (${packageName})`;
}
/**
* Generates a registration name and description based on the provider, events, and runtime action.
* @param provider - The provider this registration is associated to.
* @param events - The events routed by this registration.
* @param runtimeAction - The runtime action this registration points to.
*/
function getRegistrationDescription(provider, events, runtimeAction) {
	return [
		"This registration was automatically created by @adobe/aio-commerce-lib-app. ",
		`It belongs to the provider "${provider.label}" (instance ID: ${provider.instance_id}). `,
		`It routes ${events.length} event(s) to the runtime action "${runtimeAction}".`
	].join("\n");
}
/**
* Converts a kebab-case string to Title Case.
* @param str - The kebab-case string to convert.
*/
function kebabToTitleCase(str) {
	return str.split("-").map((word) => word.charAt(0).toUpperCase() + word.slice(1)).join(" ");
}
/**
* Groups events by their runtime actions. Since each event can have multiple
* runtime actions, this function creates a mapping where each unique runtime
* action points to all events that target it.
*
* @param events - The events to group by runtime actions.
*/
function groupEventsByRuntimeActions(events) {
	const actionEventsMap = /* @__PURE__ */ new Map();
	for (const event of events) for (const runtimeAction of event.runtimeActions) {
		const existingEvents = actionEventsMap.get(runtimeAction) ?? [];
		actionEventsMap.set(runtimeAction, [...existingEvents, event]);
	}
	return actionEventsMap;
}
function findExistingSubscription(allSubscriptions, eventName) {
	return allSubscriptions.get(eventName) ?? null;
}
/**
* Builds the payload to send to Commerce when configuring Eventing.
* Returns `null` when no update call is needed.
*
* @param initialParams - Initial Commerce Eventing configuration parameters.
* @param existingData - Existing Commerce Eventing state from the API.
*/
function getCommerceEventingConfigurationUpdateParams(initialParams, existingData) {
	const { isDefaultProviderConfigured, isDefaultWorkspaceConfigurationEmpty } = existingData;
	if (isDefaultProviderConfigured && !isDefaultWorkspaceConfigurationEmpty) return null;
	const { workspace_configuration, ...configWithoutWorkspace } = initialParams;
	let updateParams = { enabled: true };
	if (isDefaultWorkspaceConfigurationEmpty) {
		if (!workspace_configuration) throw new Error("Workspace configuration is required to enable Commerce Eventing when there is not an existing one.");
		updateParams.workspace_configuration = workspace_configuration;
	}
	if (!isDefaultProviderConfigured) updateParams = {
		...updateParams,
		...configWithoutWorkspace
	};
	return updateParams;
}
/**
* Sanitizes a Commerce Eventing identifier.
* Preserves underscores, converts spaces to underscores, lowercases, and strips the rest.
*
* @param value - The raw identifier value to normalize.
*/
function sanitizeEventingIdentifier(value) {
	return value.toLowerCase().replace(/\s+/g, "_").replace(/[^a-z0-9_]/g, "");
}
/**
* Creates a partially filled workspace configuration object based on the app credentials and parameters.
* This configuration is used when creating an event provider in Commerce.
*
* @param context - The execution context containing app credentials and parameters.
*/
function makeWorkspaceConfig(context) {
	const { appData, params } = context;
	const { consumerOrgId, orgName, projectId, projectName, projectTitle, workspaceId, workspaceName, workspaceTitle } = appData;
	const authParams = resolveAuthParams(params);
	if (authParams.strategy !== "ims") throw new Error("Failed to resolve IMS authentication parameters from the runtime action inputs.");
	const { clientId, clientSecrets, technicalAccountEmail, technicalAccountId, imsOrgId, scopes } = authParams;
	return { project: {
		id: projectId,
		name: projectName,
		title: projectTitle,
		org: {
			id: consumerOrgId,
			name: orgName,
			ims_org_id: imsOrgId
		},
		workspace: {
			id: workspaceId,
			name: workspaceName,
			title: workspaceTitle,
			action_url: `https://${process.env.__OW_NAMESPACE}.adobeioruntime.net`,
			app_url: `https://${process.env.__OW_NAMESPACE}.adobeio-static.net`,
			details: { credentials: [{
				id: "000000",
				name: `aio-${workspaceId}`,
				integration_type: "oauth_server_to_server",
				oauth_server_to_server: {
					client_id: clientId,
					client_secrets: clientSecrets,
					technical_account_email: technicalAccountEmail,
					technical_account_id: technicalAccountId,
					scopes: scopes.map((scope) => scope.trim())
				}
			}] }
		}
	} };
}
/**
* Retrieves the current existing data and returns it in a normalized way.
* @param context The execution context.
*/
async function getIoEventsExistingData(context) {
	const { ioEventsClient, appData } = context;
	const appCredentials = {
		consumerOrgId: appData.consumerOrgId,
		projectId: appData.projectId,
		workspaceId: appData.workspaceId
	};
	const { _embedded: { providers: existingProviders } } = await ioEventsClient.getAllEventProviders({
		consumerOrgId: appData.consumerOrgId,
		withEventMetadata: true
	});
	const providersWithMetadata = existingProviders.map((providerHal) => {
		const { _embedded, _links, ...providerData } = providerHal;
		const actualMetadata = (_embedded?.eventmetadata ?? []).map(({ _embedded, _links, ...meta }) => ({
			...meta,
			sample: _embedded?.sample_event ?? null
		}));
		return {
			...providerData,
			metadata: actualMetadata
		};
	});
	const { _embedded: { registrations: registrationsHal } } = await ioEventsClient.getAllRegistrations(appCredentials);
	return {
		providersWithMetadata,
		registrations: registrationsHal.map(({ _links, ...reg }) => reg)
	};
}
/**
* Retrieves the current existing Commerce eventing data and returns it in a normalized way.
* @param context - The execution context.
*/
async function getCommerceEventingExistingData(context) {
	const { commerceEventsClient } = context;
	const existingProviders = await commerceEventsClient.getAllEventProviders();
	const existingSubscriptions = await commerceEventsClient.getAllEventSubscriptions();
	const defaultProvider = existingProviders.find((provider) => !("id" in provider)) ?? null;
	const isDefaultProviderConfigured = defaultProvider !== null;
	return {
		isDefaultProviderConfigured,
		isDefaultWorkspaceConfigurationEmpty: isDefaultProviderConfigured ? !defaultProvider.workspace_configuration?.trim() : true,
		providers: existingProviders,
		subscriptions: new Map(existingSubscriptions.map((subscription) => [subscription.name, subscription]))
	};
}

//#endregion
//#region source/management/installation/events/helpers.ts
/**
* Creates an event provider if it does not already exist.
* @param params - The parameters necessary to create the provider.
*/
async function createIoEventProvider(params) {
	const { context, provider } = params;
	const { appData, ioEventsClient, logger } = context;
	const appCredentials = {
		consumerOrgId: appData.consumerOrgId,
		projectId: appData.projectId,
		workspaceId: appData.workspaceId
	};
	logger.info(`Creating provider "${provider.label}" with instance ID "${provider.instanceId}"`);
	return ioEventsClient.createEventProvider({
		...appCredentials,
		label: provider.label,
		providerType: provider.type,
		instanceId: provider.instanceId,
		description: provider.description
	}).then((res) => {
		logger.info(`Provider "${provider.label}" created with ID '${res.id}'`);
		return res;
	}).catch((error) => throwHttpError(logger, error, `Failed to create I/O Events provider '${provider.label}'`));
}
/**
* Creates or retrieves an existing I/O Events provider.
* @param params - Parameters needed to create or get the provider.
* @param existingData - Existing I/O Events data.
*/
async function createOrGetIoEventProvider(params, existingData) {
	const { context, provider } = params;
	const { logger } = context;
	const { instanceId, ...providerData } = provider;
	const existing = findExistingProvider(existingData, instanceId);
	if (existing) {
		logger.info(`Provider "${provider.label}" already exists with ID "${existing.id}", skipping creation.`);
		return existing;
	}
	return createIoEventProvider({
		context,
		provider: {
			...providerData,
			instanceId
		}
	});
}
/**
* Creates an event metadata for a given provider and event.
* @param params - The parameters necessary to create the event metadata.
*/
async function createIoEventProviderEventMetadata(params) {
	const { context, event, provider, type, metadata } = params;
	const { appData, ioEventsClient, logger } = context;
	const appCredentials = {
		consumerOrgId: appData.consumerOrgId,
		projectId: appData.projectId,
		workspaceId: appData.workspaceId
	};
	const eventCode = getIoEventCode(getNamespacedEvent(metadata, event.name), type);
	logger.info(`Creating event metadata (${eventCode}) for provider "${provider.label}" (instance ID: ${provider.instance_id}))`);
	return ioEventsClient.createEventMetadataForProvider({
		...appCredentials,
		providerId: provider.id,
		label: event.label,
		description: event.description,
		eventCode
	}).then((res) => {
		logger.info(`Event metadata "${event.label}" created for provider "${provider.label}"`);
		return res;
	}).catch((error) => throwHttpError(logger, error, `Failed to register I/O Events metadata for '${eventCode}'`));
}
/**
* Creates or retrieves existing I/O Events metadata for a given provider and event.
* @param params - The parameters necessary to create or get the event metadata.
* @param existingData - Existing I/O Events metadata for the provider.
*/
async function createOrGetIoProviderEventMetadata(params, existingData) {
	const { context, event, provider, type, metadata } = params;
	const { logger } = context;
	const existing = findExistingProviderMetadata(existingData, getIoEventCode(getNamespacedEvent(metadata, event.name), type));
	if (existing) {
		logger.info(`Event metadata "${event.label}" already exists for provider "${provider.label}" (instance ID: ${provider.instance_id}), skipping creation.`);
		return existing;
	}
	return createIoEventProviderEventMetadata(params);
}
/**
* Creates an event registration bound to the given provider ID for a set of events targeting the given runtime action.
* @param params - The parameters necessary to create the registration.
*/
async function createIoEventRegistration(params) {
	const { context, events, provider, runtimeAction, metadata } = params;
	const { appData, ioEventsClient, logger, params: runtimeParams } = context;
	const appCredentials = {
		consumerOrgId: appData.consumerOrgId,
		projectId: appData.projectId,
		workspaceId: appData.workspaceId
	};
	logger.info(`Creating registration(s) to runtime action "${runtimeAction}" for ${events.length} event(s) with provider "${provider.label}" (instance ID: ${provider.instance_id}))`);
	const name = getRegistrationName(provider, runtimeAction);
	const description = getRegistrationDescription(provider, events, runtimeAction);
	const payload = {
		...appCredentials,
		name,
		clientId: runtimeParams.AIO_COMMERCE_AUTH_IMS_CLIENT_ID,
		description,
		deliveryType: "webhook",
		runtimeAction,
		enabled: true,
		eventsOfInterest: events.map((event) => ({
			providerId: provider.id,
			eventCode: getIoEventCode(getNamespacedEvent(metadata, event.name), provider.provider_metadata)
		}))
	};
	return ioEventsClient.createRegistration(payload).then((res) => {
		logger.info(`Registration "${name}" created for provider "${provider.label}" with ID '${res.id}'`);
		return res;
	}).catch((error) => throwHttpError(logger, error, `Failed to create I/O Events registration '${name}'`));
}
/**
* Creates or retrieves an existing I/O Events registration.
* @param params - Parameters needed to create or get the registration.
* @param existingData - Existing I/O Events data.
*/
async function createOrGetIoEventRegistration(params, registrations) {
	const { context, provider, runtimeAction } = params;
	const { logger, params: runtimeParams } = context;
	const name = getRegistrationName(provider, runtimeAction);
	const existing = findExistingRegistrations(registrations, runtimeParams.AIO_COMMERCE_AUTH_IMS_CLIENT_ID, name);
	if (existing) {
		logger.info(`Registration "${name}" already exists for provider "${provider.label}" with ID '${existing.id}', skipping creation.`);
		return existing;
	}
	return createIoEventRegistration(params);
}
/**
* Ensures Commerce Eventing is configured with the given configuration, updating it if it already exists.
* @param params - The parameters necessary to configure Commerce Eventing.
* @param existingData - Existing Commerce Eventing data.
*/
async function configureCommerceEventing(params, existingData) {
	const { context, config } = params;
	const { commerceEventsClient, logger } = context;
	logger.info("Starting configuration of the Commerce Eventing Module");
	const updateParams = getCommerceEventingConfigurationUpdateParams(config, existingData);
	if (updateParams === null) {
		logger.info("Commerce Eventing Module is already configured, skipping configuration step.");
		return;
	}
	logger.info(`Updating Commerce Eventing Module configuration with the following data: [${Object.keys(updateParams).join(", ")}]`);
	return commerceEventsClient.updateEventingConfiguration(updateParams).then((success) => {
		if (success) {
			logger.info("Commerce Eventing Module configured successfully.");
			return;
		}
		throw new Error("Something went wrong while configuring Commerce Eventing Module. Response was not successful but no error was thrown.");
	}).catch((err) => throwHttpError(logger, err, "Failed to configure Adobe Commerce eventing"));
}
/**
* Creates an event provider in Commerce for a given {@link IoEventProvider}.
* @param params - The parameters necessary to create the Commerce provider.
*/
async function createCommerceProvider(params) {
	const { context, provider } = params;
	const { commerceEventsClient, logger } = context;
	logger.info(`Creating Commerce provider "${provider.label}" with instance ID "${provider.instance_id}"`);
	return commerceEventsClient.createEventProvider({
		...provider,
		provider_id: provider.id
	}).then((res) => {
		logger.info(`Commerce provider "${provider.label}" created with ID '${res.provider_id}'`);
		return res;
	}).catch((err) => throwHttpError(logger, err, `Failed to create Adobe Commerce event provider '${provider.label}'`));
}
/**
* Creates or retrieves an existing Commerce event provider.
*
* Checks by `provider_id` first, since Commerce validates subscriptions against that field.
* A previous partial run may have left multiple records sharing the same `instance_id` but
* with different `provider_id` values; checking by `provider_id` first avoids re-creating a
* provider that is already correctly registered.
*
* @param params - Parameters needed to create or get the Commerce provider.
* @param existingProviders - Existing Commerce providers.
*/
async function createOrGetCommerceProvider(params, existingProviders) {
	const { context, provider } = params;
	const { logger } = context;
	const existingByProviderId = existingProviders.find((p) => p.provider_id === provider.id);
	if (existingByProviderId) {
		logger.info(`Provider "${provider.label}" already exists with provider ID "${provider.id}", skipping creation.`);
		return existingByProviderId;
	}
	const existingByInstanceId = findExistingProvider(existingProviders, provider.instance_id);
	if (existingByInstanceId) logger.warn(`Provider "${provider.label}" exists (instance ID: ${provider.instance_id}) but with a different provider ID ("${existingByInstanceId.provider_id}" → "${provider.id}"). Registering updated provider in Commerce.`);
	return createCommerceProvider(params);
}
/**
* Creates an event subscription in Commerce for the given event and provider.
* @param params - The parameters necessary to create the Commerce event subscription.
*/
async function createCommerceEventSubscription(params) {
	const { context, metadata, provider, event } = params;
	const { logger, commerceEventsClient } = context;
	const eventName = getNamespacedEvent(metadata, event.config.name);
	logger.info(`Creating event subscription for event "${event.config.name}" to provider "${provider.label}" (instance ID: ${provider.instance_id})`);
	const eventSpec = {
		name: eventName,
		parent: event.config.name,
		fields: event.config.fields,
		rules: event.config.rules,
		provider_id: provider.id,
		destination: event.config.destination,
		hipaa_audit_required: event.config.hipaa_audit_required,
		priority: event.config.priority,
		force: event.config.force
	};
	logger.debug(`Event subscription specification for event "${event.config.name}": ${inspect(eventSpec)}`);
	return commerceEventsClient.createEventSubscription(eventSpec).then((_res) => {
		logger.info(`Created event subscription for event "${event.config.name}" to provider "${provider.label} (instance ID: ${provider.instance_id})"`);
		return eventSpec;
	}).catch((err) => throwHttpError(logger, err, `Failed to create Adobe Commerce event subscription for '${event.config.name}'`));
}
/**
* Creates or retrieves an existing Commerce event subscription.
* @param params - Parameters needed to create or get the subscription.
* @param existingData - Map of existing Commerce subscriptions keyed by event name.
*/
async function createOrGetCommerceEventSubscription(params, existingData) {
	const { context, metadata, event } = params;
	const { logger } = context;
	const existing = findExistingSubscription(existingData, getNamespacedEvent(metadata, event.config.name));
	if (existing) {
		logger.info(`Subscription for event "${event.config.name}" already exists, skipping creation.`);
		return existing;
	}
	return createCommerceEventSubscription(params);
}
/**
* Onboards a single event source to I/O Events by creating/retrieving the provider and its event metadata.
* This is the shared logic inside the loop for both commerce and external event installation steps.
*
* @param params - Configuration for onboarding a single event source
* @param existingData - Existing I/O Events data
*/
async function onboardIoEvents(params, existingData) {
	const { providersWithMetadata, registrations } = existingData;
	const { context, metadata, provider, providerType, events } = params;
	const instanceId = generateInstanceId(metadata, provider, context.appData.workspaceId);
	const providerData = await createOrGetIoEventProvider({
		context,
		provider: {
			...provider,
			instanceId,
			type: providerType
		}
	}, providersWithMetadata);
	const metadataPromises = events.map((event) => createOrGetIoProviderEventMetadata({
		metadata,
		context,
		type: providerType,
		provider: providerData,
		event
	}, providersWithMetadata.find((p) => p.id === providerData.id)?.metadata ?? []));
	const actionEventsMap = groupEventsByRuntimeActions(events);
	const registrationPromises = Array.from(actionEventsMap.entries()).map(([runtimeAction, groupedEvents]) => createOrGetIoEventRegistration({
		metadata,
		context,
		events: groupedEvents,
		provider: providerData,
		runtimeAction
	}, registrations));
	const [metadataData, registrationsData] = await Promise.all([Promise.all(metadataPromises), Promise.all(registrationPromises)]);
	return {
		providerData,
		eventsData: events.map((event, index) => {
			const eventRegistrations = event.runtimeActions.map((runtimeAction) => {
				return registrationsData[Array.from(actionEventsMap.keys()).indexOf(runtimeAction)];
			});
			return {
				config: {
					...event,
					providerType
				},
				data: {
					metadata: metadataData[index],
					registrations: eventRegistrations
				}
			};
		})
	};
}
/**
* Onboards Commerce Eventing by creating event subscriptions and configuring the Eventing Module.
* @param params - The parameters necessary to onboard Commerce Eventing.
*/
async function onboardCommerceEventing(params, existingData) {
	const { context, metadata, ioData } = params;
	const { events, provider, workspaceConfiguration } = ioData;
	const instanceId = provider.instance_id;
	const subscriptions = [];
	const { workspace_configuration: _, ...commerceProviderData } = await createOrGetCommerceProvider({
		context,
		provider: {
			id: provider.id,
			instance_id: instanceId,
			label: provider.label,
			description: provider.description,
			workspace_configuration: workspaceConfiguration
		}
	}, existingData.providers);
	for (const event of events) subscriptions.push(await createOrGetCommerceEventSubscription({
		context,
		metadata,
		provider,
		event
	}, existingData.subscriptions));
	return {
		commerceProvider: commerceProviderData,
		subscriptions
	};
}
/**
* Deletes all I/O Events registrations for the given provider.
* Registration names are reconstructed deterministically using the same logic as during installation.
* Errors are caught and logged so that uninstall remains best-effort.
*/
async function deleteIoEventRegistrations(providerData, provider, events, registrations, context) {
	const { ioEventsClient, appData, logger, params: runtimeParams } = context;
	const appCredentials = {
		consumerOrgId: appData.consumerOrgId,
		projectId: appData.projectId,
		workspaceId: appData.workspaceId
	};
	const actionEventsMap = groupEventsByRuntimeActions(events);
	const runtimeActions = Array.from(actionEventsMap.keys());
	const registrationNames = new Set([...runtimeActions.map((runtimeAction) => getRegistrationName(providerData, runtimeAction)), ...runtimeActions.map((runtimeAction) => getLegacyRegistrationName(providerData, runtimeAction))]);
	const providerRegistrations = registrations.filter((reg) => reg.client_id === runtimeParams.AIO_COMMERCE_AUTH_IMS_CLIENT_ID && registrationNames.has(reg.name));
	if (providerRegistrations.length === 0) {
		logger.info(`No I/O Events registrations found for provider "${provider.label}" (instance ID: "${providerData.instance_id}").`);
		return;
	}
	logger.info(`Deleting ${providerRegistrations.length} I/O Events registration(s) for provider "${provider.label}" (instance ID: "${providerData.instance_id}")...`);
	for (const registration of providerRegistrations) {
		logger.info(`Deleting registration "${registration.name}" (ID: ${registration.id})...`);
		try {
			await ioEventsClient.deleteRegistration({
				...appCredentials,
				registrationId: registration.registration_id
			});
			logger.info(`Deleted registration "${registration.name}" (ID: ${registration.id}).`);
		} catch (error) {
			const msg = await unwrapHttpError(error);
			logger.warn(`Failed to delete I/O Events registration "${registration.name}" (ID: ${registration.id}): ${msg}. Continuing uninstall.`);
		}
	}
}
/**
* Deletes all event metadata entries from the given I/O Events provider.
* Errors are caught and logged so that uninstall remains best-effort.
*/
async function deleteIoEventMetadata(providerData, provider, context) {
	const { ioEventsClient, appData, logger } = context;
	const appCredentials = {
		consumerOrgId: appData.consumerOrgId,
		projectId: appData.projectId,
		workspaceId: appData.workspaceId
	};
	const eventMetadataList = providerData.metadata ?? [];
	if (eventMetadataList.length === 0) {
		logger.info(`No event metadata found for provider "${provider.label}" (ID: ${providerData.id}).`);
		return;
	}
	logger.info(`Deleting ${eventMetadataList.length} event metadata entry(s) for provider "${provider.label}" (ID: ${providerData.id})...`);
	for (const eventMetadata of eventMetadataList) {
		logger.info(`Deleting event metadata "${eventMetadata.event_code}" from provider "${providerData.id}"...`);
		try {
			await ioEventsClient.deleteEventMetadataForProvider({
				...appCredentials,
				providerId: providerData.id,
				eventCode: eventMetadata.event_code
			});
			logger.info(`Deleted event metadata "${eventMetadata.event_code}" from provider "${providerData.id}".`);
		} catch (error) {
			const msg = await unwrapHttpError(error);
			logger.warn(`Failed to delete I/O Events metadata "${eventMetadata.event_code}" from provider "${providerData.id}": ${msg}. Continuing uninstall.`);
		}
	}
}
/**
* Deletes a single I/O Events provider.
* Errors are caught and logged so that uninstall remains best-effort.
*/
async function deleteIoEventProvider(providerData, provider, context) {
	const { ioEventsClient, appData, logger } = context;
	const appCredentials = {
		consumerOrgId: appData.consumerOrgId,
		projectId: appData.projectId,
		workspaceId: appData.workspaceId
	};
	logger.info(`Deleting I/O Events provider "${provider.label}" (ID: ${providerData.id})...`);
	try {
		await ioEventsClient.deleteEventProvider({
			...appCredentials,
			providerId: providerData.id
		});
		logger.info(`Deleted I/O Events provider "${provider.label}" (ID: ${providerData.id}).`);
	} catch (error) {
		const msg = await unwrapHttpError(error);
		logger.warn(`Failed to delete I/O Events provider "${provider.label}" (ID: ${providerData.id}): ${msg}. Continuing uninstall.`);
	}
}
/**
* Offboards a single event source from I/O Events by deleting, in order:
* 1. All registrations that reference events from this provider.
* 2. All event metadata entries on the provider.
* 3. The provider itself.
*
* This is the reverse of {@link onboardIoEvents} and is called during uninstall.
* All deletion errors are caught and logged so that uninstall remains best-effort.
*
* @param params - Configuration identifying the provider to offboard.
* @param existingData - Current I/O Events data (providers and registrations).
*/
async function offboardIoEvents(params, existingData) {
	const { context, metadata, provider, events } = params;
	const { appData, logger } = context;
	const instanceId = generateInstanceId(metadata, provider, appData.workspaceId);
	const instanceIdOldVersion = generateInstanceIdDeprecated(metadata, provider);
	const providerData = existingData.providersWithMetadata.find((p) => p.instance_id === instanceId || p.instance_id === instanceIdOldVersion);
	if (!providerData) {
		logger.info(`No I/O Events provider found with instance ID "${instanceId}", skipping offboarding.`);
		return;
	}
	await deleteIoEventRegistrations(providerData, provider, events, existingData.registrations, context);
	await deleteIoEventMetadata(providerData, provider, context);
	await deleteIoEventProvider(providerData, provider, context);
}
/**
* Deletes all Commerce event subscriptions for the given events.
* Subscriptions are matched by their namespaced name, built the same way as during installation.
* Errors are caught and logged so that uninstall remains best-effort.
*/
async function deleteCommerceEventSubscriptions(events, metadata, provider, existingSubscriptions, context) {
	const { commerceEventsClient, logger } = context;
	logger.info(`Unsubscribing Commerce event subscriptions for provider "${provider.label}"...`);
	for (const event of events) {
		const eventName = getNamespacedEvent(metadata, event.name);
		if (!existingSubscriptions.has(eventName)) {
			logger.info(`No Commerce subscription found for event "${event.name}" (namespaced: "${eventName}"), skipping.`);
			continue;
		}
		logger.info(`Unsubscribing Commerce event subscription for "${event.name}" (namespaced: "${eventName}")...`);
		try {
			await commerceEventsClient.deleteEventSubscription({ name: eventName });
			logger.info(`Unsubscribed Commerce event subscription for "${eventName}".`);
		} catch (error) {
			const msg = await unwrapHttpError(error);
			logger.warn(`Failed to unsubscribe Adobe Commerce event subscription for "${eventName}": ${msg}. Continuing uninstall.`);
		}
	}
}
/**
* Deletes a single Commerce-side event provider.
* The provider is matched by its deterministic `instance_id`. If not found, deletion is skipped.
* Errors are caught and logged so that uninstall remains best-effort.
*/
async function deleteCommerceEventProvider(metadata, provider, existingProviders, context) {
	const { commerceEventsClient, appData, logger } = context;
	const instanceId = generateInstanceId(metadata, provider, appData.workspaceId);
	const instanceIdOldVersion = generateInstanceIdDeprecated(metadata, provider);
	const commerceProvider = existingProviders.find((p) => p.instance_id === instanceId || p.instance_id === instanceIdOldVersion);
	if (!commerceProvider) {
		logger.info(`No Commerce event provider found with instance ID "${instanceId}", skipping provider deletion.`);
		return;
	}
	logger.info(`Deleting Commerce event provider "${provider.label}" (provider_id: ${commerceProvider.provider_id})...`);
	try {
		await commerceEventsClient.deleteEventProvider({ provider_id: commerceProvider.provider_id });
		logger.info(`Deleted Commerce event provider "${provider.label}" (provider_id: ${commerceProvider.provider_id}).`);
	} catch (error) {
		const msg = await unwrapHttpError(error);
		logger.warn(`Failed to delete Adobe Commerce event provider "${provider.label}" (provider_id: ${commerceProvider.provider_id}): ${msg}. Continuing uninstall.`);
	}
}
/**
* Offboards Commerce eventing for a single provider. Performs the following steps in order:
* 1. Unsubscribes all event subscriptions that were created for the given provider.
* 2. Deletes the Commerce-side event provider itself.
*
* Subscriptions are matched by their namespaced name, which is deterministic and built the
* same way as during {@link onboardCommerceEventing}. The provider is matched by its
* `instance_id`. Missing subscriptions or providers are silently skipped. All errors are
* caught and logged so that uninstall remains best-effort.
*
* @param params - Configuration identifying the provider and its events to offboard.
* @param existingData - Current Commerce eventing data (providers and subscriptions).
*/
async function offboardCommerceEventing(params, existingData) {
	const { context, metadata, provider, events } = params;
	await deleteCommerceEventSubscriptions(events, metadata, provider, existingData.subscriptions, context);
	await deleteCommerceEventProvider(metadata, provider, existingData.providers, context);
}

//#endregion
//#region source/management/installation/events/commerce.ts
/** Leaf step for installing commerce event sources. */
const commerceEventsStep = defineLeafStep({
	name: "commerce",
	meta: {
		install: {
			label: "Configure Commerce Events",
			description: "Sets up I/O Events for Adobe Commerce event sources"
		},
		uninstall: {
			label: "Remove Commerce Events",
			description: "Removes I/O Events for Adobe Commerce event sources"
		}
	},
	when: hasCommerceEvents,
	install: createCommerceEvents,
	uninstall: removeCommerceEvents
});
/**
* Creates all needed entities for Eventing to work with Commerce and Adobe I/O Events.
* @param config - The configuration of the app, with commerce events.
* @param context - The execution context for the events installation.
*/
async function createCommerceEvents(config, context) {
	const { logger } = context;
	logger.debug("Starting installation of Commerce Events with config:", config);
	const stepData = [];
	const workspaceConfiguration = JSON.stringify(makeWorkspaceConfig(context));
	const existingIoEventsData = await getIoEventsExistingData(context);
	const commerceEventingExistingData = await getCommerceEventingExistingData(context);
	for (let i = 0; i < config.eventing.commerce.length; i++) {
		const { provider, events } = config.eventing.commerce[i];
		const { providerData, eventsData } = await onboardIoEvents({
			context,
			metadata: config.metadata,
			provider,
			events,
			providerType: COMMERCE_PROVIDER_TYPE
		}, existingIoEventsData);
		if (i === 0) await configureCommerceEventing({
			context,
			config: {
				enabled: true,
				merchant_id: sanitizeEventingIdentifier(context.appData.orgName),
				environment_id: sanitizeEventingIdentifier(context.appData.projectName),
				instance_id: providerData.instance_id,
				workspace_configuration: workspaceConfiguration
			}
		}, commerceEventingExistingData);
		const { commerceProvider, subscriptions } = await onboardCommerceEventing({
			context,
			metadata: config.metadata,
			provider,
			ioData: {
				provider: providerData,
				events: eventsData,
				workspaceConfiguration
			}
		}, commerceEventingExistingData);
		stepData.push({ provider: {
			config: provider,
			data: {
				ioEvents: providerData,
				commerce: commerceProvider,
				events: eventsData.map(({ config, data }, index) => ({
					config,
					data: {
						...data,
						subscription: subscriptions[index]
					}
				}))
			}
		} });
	}
	logger.debug("Completed Commerce Events installation step.");
	return stepData;
}
/**
* Remove all created for Commerce eventing created durint installation
* @param config - The configuration of the app, with commerce events.
* @param context - The execution context for the events installation.
*/
async function removeCommerceEvents(config, context) {
	const { logger } = context;
	logger.debug("Starting uninstall of Commerce Events with config:", config);
	const [existingIoEventsData, commerceEventingExistingData] = await Promise.all([getIoEventsExistingData(context), getCommerceEventingExistingData(context)]);
	for (const { provider, events } of config.eventing.commerce) {
		await offboardCommerceEventing({
			context,
			metadata: config.metadata,
			provider,
			events
		}, commerceEventingExistingData);
		await offboardIoEvents({
			context,
			metadata: config.metadata,
			provider,
			events
		}, existingIoEventsData);
	}
	logger.debug("Completed Commerce Events uninstall step.");
}

//#endregion
//#region source/management/installation/events/context.ts
/**
* Create a custom Commerce API Client with only the operations we need for optimal package size.
* @param params - The runtime action params to resolve the client params from.
*/
function createCommerceEventsApiClient(params) {
	const commerceClientParams = resolveCommerceHttpClientParams(params, { tryForwardAuthProvider: true });
	commerceClientParams.fetchOptions ??= {};
	commerceClientParams.fetchOptions.timeout = 1e3 * 60 * 2;
	return createCustomCommerceEventsApiClient(commerceClientParams, {
		createEventProvider,
		getAllEventProviders,
		createEventSubscription,
		deleteEventProvider,
		deleteEventSubscription,
		getAllEventSubscriptions,
		updateEventingConfiguration
	});
}
/**
* Create a custom Adobe I/O Events API Client with only the operations we need for optimal package size.
* @param params - The runtime action params to resolve the client params from.
*/
function createIoEventsApiClient(params) {
	const ioEventsClientParams = resolveIoEventsHttpClientParams(params);
	ioEventsClientParams.fetchOptions ??= {};
	ioEventsClientParams.fetchOptions.timeout = 1e3 * 60 * 2;
	return createCustomAdobeIoEventsApiClient(ioEventsClientParams, {
		createEventProvider: createEventProvider$1,
		createEventMetadataForProvider,
		createRegistration,
		deleteEventMetadataForProvider,
		deleteEventProvider: deleteEventProvider$1,
		deleteRegistration,
		getAllEventProviders: getAllEventProviders$1,
		getAllRegistrations
	});
}
/** Creates the events step context with lazy-initialized API clients. */
function createEventsStepContext(installation) {
	const { params } = installation;
	let commerceEventsClient = null;
	let ioEventsClient = null;
	return {
		get commerceEventsClient() {
			if (commerceEventsClient === null) commerceEventsClient = createCommerceEventsApiClient(params);
			return commerceEventsClient;
		},
		get ioEventsClient() {
			if (ioEventsClient === null) ioEventsClient = createIoEventsApiClient(params);
			return ioEventsClient;
		}
	};
}

//#endregion
//#region source/management/installation/events/external.ts
/** Leaf step for installing external event sources. */
const externalEventsStep = defineLeafStep({
	name: "external",
	meta: {
		install: {
			label: "Configure External Events",
			description: "Sets up I/O Events for external event sources"
		},
		uninstall: {
			label: "Remove External Events",
			description: "Removes I/O Events for external event sources"
		}
	},
	when: hasExternalEvents,
	install: createExternalEvents,
	uninstall: removeExternalEvents
});
/**
* Creates all needed entities for External Events to work with Adobe I/O Events.
* @param config - The configuration of the app, with external events.
* @param context - The execution context for the events installation.
*/
async function createExternalEvents(config, context) {
	const { logger } = context;
	logger.debug("Starting installation of External Events with config:", config);
	const stepData = [];
	const existingIoEventsData = await getIoEventsExistingData(context);
	for (const { provider, events } of config.eventing.external) {
		const { providerData, eventsData } = await onboardIoEvents({
			context,
			metadata: config.metadata,
			provider,
			events,
			providerType: EXTERNAL_PROVIDER_TYPE
		}, existingIoEventsData);
		stepData.push({ provider: {
			config: provider,
			data: {
				ioEvents: providerData,
				events: {
					config: events,
					data: eventsData
				}
			}
		} });
	}
	logger.debug("Completed External Events installation step.");
	return stepData;
}
/**
* Removed all created entities for External Events during the installation
* @param config - The configuration of the app, with external events.
* @param context - The execution context for the events installation.
*/
async function removeExternalEvents(config, context) {
	const { logger } = context;
	logger.debug("Starting uninstall of External Events with config:", config);
	const existingIoEventsData = await getIoEventsExistingData(context);
	for (const { provider, events } of config.eventing.external) await offboardIoEvents({
		context,
		metadata: config.metadata,
		provider,
		events
	}, existingIoEventsData);
	logger.debug("Completed External Events uninstall step.");
}

//#endregion
//#region source/management/installation/events/branch.ts
/** Root eventing step that contains commerce and external event sub-steps. */
const eventingStep = defineBranchStep({
	name: "eventing",
	meta: {
		install: {
			label: "Eventing",
			description: "Sets up the I/O Events and the Commerce events required by the application"
		},
		uninstall: {
			label: "Eventing",
			description: "Removes the I/O Events and Commerce events configured by the application"
		}
	},
	when: hasEventing,
	context: createEventsStepContext,
	children: [commerceEventsStep, externalEventsStep]
});

//#endregion
//#region source/management/installation/webhooks/context.ts
/**
* Create a custom Commerce Webhooks API Client with only the operations needed for installation.
* @param params - The runtime action params to resolve the client params from.
*/
function createCommerceWebhooksApiClient(params) {
	const commerceClientParams = resolveCommerceHttpClientParams(params, { tryForwardAuthProvider: true });
	commerceClientParams.fetchOptions ??= {};
	commerceClientParams.fetchOptions.timeout = 1e3 * 60 * 2;
	return createCustomCommerceWebhooksApiClient(commerceClientParams, {
		getWebhookList,
		subscribeWebhook,
		unsubscribeWebhook
	});
}
/** Creates the webhooks step context with a lazy-initialized API client. */
function createWebhooksStepContext(installation) {
	const { params } = installation;
	let commerceWebhooksClient = null;
	return { get commerceWebhooksClient() {
		if (commerceWebhooksClient === null) commerceWebhooksClient = createCommerceWebhooksApiClient(params);
		return commerceWebhooksClient;
	} };
}

//#endregion
//#region source/management/installation/webhooks/helpers.ts
/** Matches any character that is not a valid identifier character (letter, digit, or underscore). */
const NON_IDENTIFIER_CHAR_REGEX = /[^a-zA-Z0-9_]/g;
/** Matches two or more consecutive underscores. */
const MULTIPLE_UNDERSCORES_REGEX = /_+/g;
/** Matches the `.magento` segment in plugin webhook method names (e.g. `plugin.magento.foo`). */
const PLUGIN_MAGENTO_REGEX = /^plugin\.magento\./;
const ENVIRONMENT_PRODUCTION = "production";
const ENVIRONMENT_STAGING = "staging";
/**
* Validates that no modification webhooks in the app config conflict with webhooks
* already registered in Commerce by another app.
*
* A conflict is: Commerce has a webhook with the same `webhook_method` and `webhook_type`
* that does NOT belong to this app (i.e. different `batch_name` or `hook_name` after prefix).
*
* Returns a `ValidationIssue` with code `WEBHOOK_CONFLICTS` and `details.conflictedWebhooks` listing
* every conflicting Commerce webhook when conflicts are found, or an empty array otherwise.
*
* @param config - The app config (must have a non-empty `webhooks` array).
* @param context - The webhooks execution context (provides the Commerce API client and logger).
*/
async function validateWebhookConflicts(config, context) {
	const { logger, commerceWebhooksClient } = context;
	const modificationWebhooks = config.webhooks.filter((entry) => entry.category === "modification");
	if (modificationWebhooks.length === 0) {
		logger.debug("No modification webhooks to validate, skipping conflict check.");
		return [];
	}
	logger.debug(`Validating ${modificationWebhooks.length} modification webhook(s) for conflicts...`);
	const existingWebhooks = await commerceWebhooksClient.getWebhookList();
	const idPrefix = buildWebhookIdPrefix(config.metadata.id);
	const conflictedWebhooks = [];
	for (const entry of modificationWebhooks) {
		const { webhook } = entry;
		const resolvedBatch = `${idPrefix}${webhook.batch_name}`;
		const resolvedHook = `${idPrefix}${webhook.hook_name}`;
		for (const existing of existingWebhooks) if (existing.webhook_method === webhook.webhook_method && existing.webhook_type === webhook.webhook_type && !(existing.batch_name === resolvedBatch && existing.hook_name === resolvedHook)) {
			conflictedWebhooks.push({
				label: entry.label,
				...existing
			});
			break;
		}
	}
	if (conflictedWebhooks.length > 0) return [{
		code: "WEBHOOK_CONFLICTS",
		message: `Webhook conflicts detected: ${conflictedWebhooks.length} webhook(s) already registered for the same method and type by another app`,
		severity: "warning",
		details: { conflictedWebhooks }
	}];
	logger.info("No webhook conflicts found.");
	return [];
}
/**
* Subscribes each webhook from the app config to Adobe Commerce.
* Throws on the first failure, aborting any remaining subscriptions.
*
* @param config - The app config (must have a non-empty `webhooks` array).
* @param context - The webhooks execution context (provides the Commerce API client and logger).
*/
async function createWebhookSubscriptions(config, context) {
	const { logger, commerceWebhooksClient, params } = context;
	logger.info(`Subscribing ${config.webhooks.length} webhook(s) to Commerce...`);
	const idPrefix = buildWebhookIdPrefix(config.metadata.id);
	const subscribedWebhooks = [];
	const existingWebhooks = await commerceWebhooksClient.getWebhookList();
	for (const entry of config.webhooks) {
		const { webhook } = entry;
		const resolvedUrl = "runtimeAction" in entry ? generateUrlForRuntimeAction(entry.runtimeAction) : entry.webhook.url;
		logger.debug(`Subscribing webhook "${getWebhookName(webhook)}" (runtimeAction: ${"runtimeAction" in entry ? entry.runtimeAction : "none"})`);
		const resolvedWebhook = {
			...webhook,
			url: resolvedUrl,
			batch_name: `${idPrefix}${webhook.batch_name}`,
			hook_name: `${idPrefix}${webhook.hook_name}`,
			..."runtimeAction" in entry && entry.requireAdobeAuth !== false && { developer_console_oauth: resolveDeveloperConsoleOAuthCredentials(params) }
		};
		subscribedWebhooks.push(await createOrGetWebhookSubscription(existingWebhooks, commerceWebhooksClient, resolvedWebhook, logger));
	}
	logger.info(`Webhook subscriptions complete: ${subscribedWebhooks.length} subscribed.`);
	return { subscribedWebhooks };
}
/**
* Unsubscribes each webhook from the app config in Adobe Commerce.
* If a webhook is not found in the existing list, it is silently skipped (idempotent).
*
* @param config - The app config (must have a non-empty `webhooks` array).
* @param context - The webhooks execution context (provides the Commerce API client and logger).
*/
async function deleteWebhookSubscriptions(config, context) {
	const { logger, commerceWebhooksClient } = context;
	logger.info(`Unsubscribing ${config.webhooks.length} webhook(s) from Commerce...`);
	const idPrefix = buildWebhookIdPrefix(config.metadata.id);
	const unsubscribedWebhooks = [];
	const existingWebhooks = await commerceWebhooksClient.getWebhookList();
	for (const entry of config.webhooks) {
		const { webhook } = entry;
		const resolvedBatch = `${idPrefix}${webhook.batch_name}`;
		const resolvedHook = `${idPrefix}${webhook.hook_name}`;
		const params = {
			webhook_method: webhook.webhook_method,
			webhook_type: webhook.webhook_type,
			batch_name: resolvedBatch,
			hook_name: resolvedHook
		};
		if (!isWebhookInList(existingWebhooks, params)) {
			logger.debug(`Webhook not found, skipping unsubscribe: ${getWebhookName(webhook)}`);
			continue;
		}
		try {
			await deleteWebhookSubscription(commerceWebhooksClient, webhook, params);
			logger.info(`Unsubscribed webhook: ${getWebhookName(webhook)}`);
			unsubscribedWebhooks.push(params);
		} catch (error) {
			logger.warn(`${stringifyError(error)}. Continuing uninstall.`);
		}
	}
	logger.info(`Webhook unsubscriptions complete: ${unsubscribedWebhooks.length} unsubscribed.`);
	return { unsubscribedWebhooks };
}
/**
* Subscribes a single webhook to Commerce, skipping the API call if the webhook
* is already subscribed (matched by webhook_method, webhook_type, batch_name, hook_name).
*/
async function createOrGetWebhookSubscription(existingWebhooks, client, resolvedWebhook, logger) {
	if (isWebhookInList(existingWebhooks, resolvedWebhook)) {
		logger.info(`Webhook already subscribed, skipping: ${getWebhookName(resolvedWebhook)}`);
		return resolvedWebhook;
	}
	const subscribed = await createWebhookSubscription(client, resolvedWebhook);
	logger.info(`Subscribed webhook: ${getWebhookName(resolvedWebhook)}`);
	return subscribed;
}
/**
* Re-throws `err` as a new Error with an enriched message that includes the webhook name
* and the unwrapped HTTP response body (if available).
*/
async function rethrowWithWebhookName(err, webhookName, operation) {
	const msg = await unwrapHttpError(err);
	throw new Error(`Failed to ${operation} webhook subscription for "${webhookName}": ${msg}`);
}
/**
* Subscribes a single webhook to Commerce, enriching the error with the webhook name
* if the API responds with a string `message`.
*/
async function createWebhookSubscription(client, resolvedWebhook) {
	try {
		await client.subscribeWebhook(resolvedWebhook);
		return resolvedWebhook;
	} catch (err) {
		return await rethrowWithWebhookName(err, getWebhookName(resolvedWebhook), "create");
	}
}
/**
* Unsubscribes a single webhook from Commerce, enriching the error with the webhook name
* if the API responds with a string `message`.
*/
async function deleteWebhookSubscription(client, resolvedWebhook, params) {
	try {
		await client.unsubscribeWebhook(params);
	} catch (err) {
		return rethrowWithWebhookName(err, getWebhookName(resolvedWebhook), "delete");
	}
}
/**
* Resolves and validates the IMS credentials required for `developer_console_oauth`.
*
* Delegates parsing and validation to `resolveImsAuthParams` from `aio-commerce-lib-auth`,
* which correctly handles `AIO_COMMERCE_AUTH_IMS_CLIENT_SECRETS` whether it arrives as a
* real array or as a JSON-stringified array string.
*/
function resolveDeveloperConsoleOAuthCredentials(params) {
	const { AIO_COMMERCE_AUTH_IMS_ENVIRONMENT: imsEnvironment, ...imsParams } = params;
	const { clientId, clientSecrets, imsOrgId } = resolveImsAuthParams(imsParams);
	return {
		client_id: clientId,
		client_secret: clientSecrets[0],
		org_id: imsOrgId,
		environment: !imsEnvironment || String(imsEnvironment).startsWith("prod") ? ENVIRONMENT_PRODUCTION : ENVIRONMENT_STAGING
	};
}
/**
* Returns true when a webhook with the given four-part identity exists in the list.
*
* The identity check uses: webhook_method, webhook_type, batch_name, hook_name.
* `webhook_method` is normalised before comparison to handle the case where Commerce strips the
* `.magento` segment from plugin webhook methods on storage
* (e.g. `plugin.magento.foo` and `plugin.foo` are treated as the same method).
*/
function isWebhookInList(existing, candidate) {
	const normalizedCandidate = normalizeWebhookMethod(candidate.webhook_method);
	return existing.some((w) => normalizeWebhookMethod(w.webhook_method) === normalizedCandidate && w.webhook_type === candidate.webhook_type && w.batch_name === candidate.batch_name && w.hook_name === candidate.hook_name);
}
/**
* Normalises a webhook method name by removing the `.magento` segment that Commerce
* may drop when persisting plugin webhook methods.
*
* @example
* normalizeWebhookMethod("plugin.magento.foo.bar") // → "plugin.foo.bar"
* normalizeWebhookMethod("plugin.foo.bar")         // → "plugin.foo.bar" (unchanged)
*/
function normalizeWebhookMethod(method) {
	return method.replace(PLUGIN_MAGENTO_REGEX, "plugin.");
}
/**
* Generates a URL for a given runtime action using the AIO Runtime API host and namespace.
* @param runtimeAction
* @return The generated URL for the runtime action.
*/
function generateUrlForRuntimeAction(runtimeAction) {
	const namespace = process.env.__OW_NAMESPACE;
	if (!namespace) throw new Error(`Cannot generate URL for runtime action "${runtimeAction}": namespace environment variable is not set.`);
	return `https://${namespace}.adobeioruntime.net/api/v1/web/${runtimeAction}`;
}
/**
* Builds a prefix string from the app ID to namespace webhook batch/hook names.
* Non-identifier characters are replaced with underscores; consecutive underscores
* are collapsed to one; a trailing underscore is appended. The result is lowercased
* to ensure consistent matching regardless of input casing.
*
* @example
* ```typescript
* buildWebhookIdPrefix("my--app.v2") // => "my_app_v2_"
* buildWebhookIdPrefix("MyApp") // => "myapp_"
* ```
* @param appId - The app ID to build the prefix from.
* @return The built prefix string.
*/
function buildWebhookIdPrefix(appId) {
	const prefix = appId.toLowerCase().replace(NON_IDENTIFIER_CHAR_REGEX, "_").replace(MULTIPLE_UNDERSCORES_REGEX, "_");
	return prefix.endsWith("_") ? prefix : `${prefix}_`;
}
/**
* Generates a name for a webhook based on its method and type.
*
* @param webhook
* @return A string in the format "webhook_method:webhook_type" to identify the webhook.
*/
function getWebhookName(webhook) {
	return `${webhook.webhook_method}:${webhook.webhook_type}`;
}

//#endregion
//#region source/management/installation/webhooks/branch.ts
const subscriptionsStep = defineLeafStep({
	name: "subscriptions",
	meta: {
		install: {
			label: "Create Subscriptions",
			description: "Creates webhook subscriptions in Adobe Commerce"
		},
		uninstall: {
			label: "Delete Subscriptions",
			description: "Deletes webhook subscriptions from Adobe Commerce"
		}
	},
	validate: (config, context) => validateWebhookConflicts(config, context),
	install: (config, context) => createWebhookSubscriptions(config, context),
	uninstall: async (config, context) => {
		await deleteWebhookSubscriptions(config, context);
	}
});
/** Branch step for setting up Commerce webhooks. */
const webhooksStep = defineBranchStep({
	name: "webhooks",
	meta: {
		install: {
			label: "Webhooks",
			description: "Sets up Commerce webhooks"
		},
		uninstall: {
			label: "Webhooks",
			description: "Removes Commerce webhooks"
		}
	},
	when: hasWebhooks,
	context: createWebhooksStepContext,
	children: [subscriptionsStep]
});

//#endregion
//#region source/management/installation/root.ts
/**
* Creates the default child steps built-in in the library with dynamic children based on the config.
*/
function createDefaultChildSteps(config) {
	return [
		eventingStep,
		webhooksStep,
		adminUiSdkStep,
		createCustomInstallationStep(config)
	];
}
/**
* Creates a root installation step with dynamic children based on the config.
*/
function createRootInstallationStep(config) {
	return defineBranchStep({
		name: "installation",
		meta: { install: {
			label: "Installation",
			description: "App installation workflow"
		} },
		children: createDefaultChildSteps(config)
	});
}
/**
* Creates a root uninstallation step with dynamic children based on the config.
*/
function createRootUninstallationStep(config) {
	return defineBranchStep({
		name: "uninstallation",
		meta: { install: {
			label: "Uninstallation",
			description: "App uninstallation workflow"
		} },
		children: createDefaultChildSteps(config)
	});
}

//#endregion
//#region source/management/installation/runner.ts
/**
* Creates an initial installation state from the config and step definitions.
* Filters steps based on their `when` conditions and builds a tree structure
* with all steps set to "pending".
*/
function createInitialInstallationState(options) {
	const { config } = options;
	return createInitialState({
		rootStep: createRootInstallationStep(config),
		config
	});
}
/**
* Runs the full installation workflow. Returns the final state (never throws).
*
* Retries once on failure. `onInstallationFailure` only fires if both attempts fail;
* `isRetry: true` is set on the result when the retry succeeds.
*/
async function runInstallation(options) {
	const { installationContext, config, initialState, hooks } = options;
	const rootStep = createRootInstallationStep(config);
	const firstResult = await executeWorkflow({
		rootStep,
		installationContext,
		config,
		initialState,
		hooks: {
			...hooks,
			onInstallationFailure: void 0,
			onStepFailure: void 0
		}
	});
	if (firstResult.status === "succeeded") return firstResult;
	const { error } = firstResult;
	installationContext.logger.warn(`Installation attempt 1 failed: step "${error.path.join(".")}", key "${error.key}"${error.message ? `, message "${error.message}"` : ""}. Retrying once.`);
	return {
		...await executeWorkflow({
			rootStep,
			installationContext,
			config,
			initialState: createRetryState(firstResult),
			hooks: hooks && {
				...hooks,
				onInstallationSuccess: (state) => hooks.onInstallationSuccess?.({
					...state,
					metadata: { isRetry: true }
				}),
				onInstallationFailure: (state) => hooks.onInstallationFailure?.({
					...state,
					metadata: { isRetry: true }
				})
			}
		}),
		metadata: { isRetry: true }
	};
}
/**
* Creates an initial uninstallation state from the config and step definitions.
*/
function createInitialUninstallationState(options) {
	const { config } = options;
	return createInitialState({
		rootStep: createRootUninstallationStep(config),
		config,
		mode: "uninstall"
	});
}
/**
* Runs the full uninstallation workflow. Returns the final state (never throws).
*/
function runUninstallation(options) {
	const { installationContext, config, initialState, hooks } = options;
	return executeUninstallWorkflow({
		rootStep: createRootUninstallationStep(config),
		installationContext,
		config,
		initialState,
		hooks
	});
}
/**
* Runs pre-installation validation over the full step tree.
*
* Traverses the same step hierarchy used during installation but only calls
* each step's optional `validate` handler rather than executing side effects.
* Always resolves (never throws). Returns a structured result with per-step
* issues and an aggregated summary.
*/
function runValidation(options) {
	const { validationContext, config } = options;
	return validateStepTree({
		rootStep: createRootInstallationStep(config),
		validationContext,
		config
	});
}

//#endregion
export { runValidation as a, isFailedState as c, runUninstallation as i, isInProgressState as l, createInitialUninstallationState as n, defineCustomInstallationStep as o, runInstallation as r, isCompletedState as s, createInitialInstallationState as t, isSucceededState as u };
