// @adobe/aio-commerce-lib-app/dist/es/router-D58JJHHA.mjs
/**
 * @license
 *
 * Copyright 2026 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */

import { badRequest, internalServerError, methodNotAllowed, notFound } from "@adobe/aio-commerce-lib-core/responses";
import AioLogger from "@adobe/aio-lib-core-logging";
import { parse } from "regexparam";

//#region ../../packages-private/common-utils/source/actions/http/middleware/logger.ts
/**
* Creates a logger middleware that adds logging capabilities to the context.
*
* @example
* ```typescript
* router.use(logger({ level: "debug", name: () => "my-logger-name" }));
*
* router.get("/test", {
*   handler: (req, ctx) => {
*     ctx.logger.info("Hello world");
*     return ok({ body: {} });
*   },
* });
* ```
*/
function logger({ name, ...restOptions } = {}) {
	return (ctx) => {
		const params = ctx.rawParams;
		return { logger: AioLogger(`${params.__ow_method}-${name?.(ctx) ?? process.env.__OW_ACTION_NAME}`, {
			level: `${params.LOG_LEVEL ?? "info"}`,
			...restOptions
		}) };
	};
}

//#endregion
//#region ../../packages-private/common-utils/source/actions/http/utils.ts
/**
* Validates input against a Standard Schema and returns a result.
*
* @template TInput - The input type expected by the schema
* @template TOutput - The output type produced by the schema
* @param schema - A Standard Schema v1 compliant schema
* @param input - The input data to validate
* @returns A promise resolving to either success with validated data or failure with issues
*
* @example
* ```typescript
* const result = await validateSchema(mySchema, userInput);
* if (result.success) {
*   console.log(result.data); // Typed as TOutput
* } else {
*   console.error(result.issues); // Validation errors
* }
* ```
*/
async function validateSchema(schema, input) {
	const result = await schema["~standard"].validate(input);
	if (result.issues) return {
		success: false,
		issues: result.issues.map((issue) => ({
			message: issue.message,
			path: issue.path?.map((segment) => typeof segment === "object" && segment !== null && "key" in segment ? segment.key : segment)
		}))
	};
	return {
		success: true,
		data: result.value
	};
}
/**
* Parses a request body from OpenWhisk/Runtime.
* Handles multiple formats:
* - Base64-encoded strings (__ow_body)
* - Already-parsed objects
* - Body properties mixed into args (web actions with JSON content-type)
*
* @param owBody - Body from __ow_body (base64 string, JSON string, or object)
* @param args - Full args object to extract body from if __ow_body is not present
*
* @example
* ```typescript
* const body = parseRequestBody(params.__ow_body, params);
* ```
*/
function parseRequestBody(owBody, args) {
	if (owBody) {
		if (typeof owBody === "object") return owBody;
		if (typeof owBody === "string") {
			try {
				return JSON.parse(owBody);
			} catch {}
			try {
				const decoded = Buffer.from(owBody, "base64").toString();
				return JSON.parse(decoded);
			} catch {}
		}
	}
	if (args && typeof args === "object") {
		const body = {};
		for (const [key, value] of Object.entries(args)) if (!key.startsWith("__ow_")) body[key] = value;
		return body;
	}
	return {};
}
/**
* Parses query parameters from OpenWhisk/Runtime format.
*
* @param queryString - Query string from __ow_query
* @param fallbackParams - Fallback params object (used when __ow_query is not present)
* @returns Parsed query parameters as a record
*
* @example
* ```typescript
* const query = parseQueryParams(params.__ow_query, params);
* ```
*/
function parseQueryParams(queryString, fallbackParams) {
	if (queryString) return Object.fromEntries(new URLSearchParams(queryString));
	if (fallbackParams) {
		const { __ow_method, __ow_path, __ow_headers, __ow_body, __ow_query, ...rest } = fallbackParams;
		return rest;
	}
	return {};
}

//#endregion
//#region ../../packages-private/common-utils/source/actions/http/router.ts
/**
* HTTP router for Adobe I/O Runtime actions.
* Provides type-safe routing with schema validation and OpenWhisk integration.
*
* @example
* ```typescript
* const router = new HttpActionRouter();
*
* router.get("/users/:id", {
*   handler: (req) => ok({ id: req.params.id, context: req.context })
* });
*
* // Add context builders
* router.use(async (base) => ({
*   user: await getUser(base.rawParams.__ow_headers?.authorization),
* }));
*
* export const main = router.handler();
* ```
*/
var HttpActionRouter = class {
	constructor() {
		this.routes = [];
		this.contextBuilders = [];
	}
	/**
	* Internal method to add a route to the router.
	*/
	addRoute(method, path, config) {
		const { pattern, keys } = parse(path);
		this.routes.push({
			method,
			pattern,
			keys,
			params: config.params,
			body: config.body,
			query: config.query,
			handler: config.handler
		});
		return this;
	}
	/**
	* Register a GET route.
	*
	* @example
	* ```typescript
	* router.get("/users/:id", {
	*   handler: (req) => ok({ id: req.params.id })
	* });
	* ```
	*/
	get(path, config) {
		return this.addRoute("GET", path, config);
	}
	/**
	* Register a POST route.
	*
	* @example
	* ```typescript
	* router.post("/users", {
	*   body: userSchema,
	*   handler: (req) => created(req.body)
	* });
	* ```
	*/
	post(path, config) {
		return this.addRoute("POST", path, config);
	}
	/**
	* Register a PUT route.
	*
	* @example
	* ```typescript
	* router.put("/users/:id", {
	*   body: userSchema,
	*   handler: (req) => ok(req.body)
	* });
	* ```
	*/
	put(path, config) {
		return this.addRoute("PUT", path, config);
	}
	/**
	* Register a PATCH route.
	*
	* @example
	* ```typescript
	* router.patch("/users/:id", {
	*   body: partialUserSchema,
	*   handler: (req) => ok(req.body)
	* });
	* ```
	*/
	patch(path, config) {
		return this.addRoute("PATCH", path, config);
	}
	/**
	* Register a DELETE route.
	*
	* @example
	* ```typescript
	* router.delete("/users/:id", {
	*   handler: (req) => noContent()
	* });
	* ```
	*/
	delete(path, config) {
		return this.addRoute("DELETE", path, config);
	}
	/**
	* Register a context builder that runs before route handlers.
	* Context builders can add properties to the request context.
	* Multiple builders are executed in order and their results are merged.
	*
	* The returned router has an updated context type that includes the new properties,
	* enabling type-safe access in route handlers.
	*
	* @param builder - Function that receives base context and returns additional context
	* @returns The router instance with updated context type for chaining
	*
	* @example
	* ```typescript
	* const router = new HttpActionRouter()
	*   .use(logger()) // HttpActionRouter<BaseContext & { logger: Logger }>
	*   .use(auth());  // HttpActionRouter<BaseContext & { logger: Logger } & { user: User }>
	*
	* router.get("/me", {
	*   handler: (req, ctx) => {
	*     ctx.logger.info("Hello"); // ✅ typed
	*     return ok({ body: ctx.user }); // ✅ typed
	*   },
	* });
	* ```
	*/
	use(builder) {
		this.contextBuilders.push(builder);
		return this;
	}
	/**
	* Builds the full context by running all context builders.
	*/
	async buildContext(args) {
		let context = { rawParams: args };
		for (const builder of this.contextBuilders) {
			const result = await builder(context);
			if (result) context = {
				...context,
				...result
			};
		}
		return context;
	}
	/**
	* Validates and extracts route parameters.
	*/
	async validateParams(route, params) {
		if (!route.params) return {
			success: true,
			data: params
		};
		const result = await validateSchema(route.params, params);
		if (!result.success) return {
			success: false,
			issues: result.issues
		};
		return {
			success: true,
			data: result.data
		};
	}
	/**
	* Validates request body.
	*/
	async validateBody(route, body) {
		if (!route.body) return {
			success: true,
			data: body
		};
		const result = await validateSchema(route.body, body);
		if (!result.success) return {
			success: false,
			issues: result.issues
		};
		return {
			success: true,
			data: result.data
		};
	}
	/**
	* Validates query parameters.
	*/
	async validateQuery(route, query) {
		if (!route.query) return {
			success: true,
			data: query
		};
		const result = await validateSchema(route.query, query);
		if (!result.success) return {
			success: false,
			issues: result.issues
		};
		return {
			success: true,
			data: result.data
		};
	}
	/** Handles a matched route by validating inputs and calling the handler. */
	async handleRoute(route, match, body, query, headers, method, path, context) {
		const params = {};
		route.keys.forEach((key, i) => {
			params[key] = decodeURIComponent(match[i + 1] || "");
		});
		const paramsResult = await this.validateParams(route, params);
		if (!paramsResult.success) return badRequest({ body: {
			message: "Invalid route parameters",
			issues: paramsResult.issues
		} });
		const bodyResult = await this.validateBody(route, body);
		if (!bodyResult.success) return badRequest({ body: {
			message: "Invalid request body",
			issues: bodyResult.issues
		} });
		const queryResult = await this.validateQuery(route, query);
		if (!queryResult.success) return badRequest({ body: {
			message: "Invalid query parameters",
			issues: queryResult.issues
		} });
		try {
			return await route.handler({
				params: paramsResult.data,
				body: bodyResult.data,
				query: queryResult.data,
				headers,
				method,
				path
			}, context);
		} catch (err) {
			console.error("Handler error:", err);
			return internalServerError({ body: {
				message: "Internal server error",
				error: err instanceof Error ? err.message : "Unknown error"
			} });
		}
	}
	/**
	* Creates an OpenWhisk/Runtime action handler from the registered routes.
	*
	* @example
	* ```typescript
	* const router = new HttpActionRouter();
	* router.get("/hello", { handler: () => ok({ message: "Hello!" }) });
	*
	* export const main = router.handler();
	* ```
	*/
	handler() {
		return async (args) => {
			const method = (args.__ow_method ?? "get").toUpperCase();
			const rawPath = args.__ow_path ?? "/";
			const path = rawPath.startsWith("/") ? rawPath : `/${rawPath}`;
			const headers = args.__ow_headers ?? {};
			const body = parseRequestBody(args.__ow_body, args);
			const query = parseQueryParams(args.__ow_query, args);
			const context = await this.buildContext(args);
			const matchedMethods = [];
			for (const route of this.routes) {
				const match = route.pattern.exec(path);
				if (!match) continue;
				matchedMethods.push(route.method);
				if (route.method !== method) continue;
				const response = await this.handleRoute(route, match, body, query, headers, method, path, context);
				if (response) return response;
			}
			if (matchedMethods.length > 0) return methodNotAllowed(`Method ${method} not allowed`);
			return notFound(`No route matches ${path}`);
		};
	}
};

//#endregion
export { logger as n, HttpActionRouter as t };
