// @adobe/aio-commerce-lib-app/dist/es/schemas-hRovglBF.mjs
/**
 * @license
 *
 * Copyright 2026 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */

import * as v from "valibot";

//#region ../../packages-private/common-utils/source/valibot/schemas.ts
const ALPHANUMERIC_OR_HYPHEN_REGEX = {
	any: /^[a-zA-Z0-9-]+$/,
	lowercase: /^[a-z0-9-]+$/,
	uppercase: /^[A-Z0-9-]+$/
};
/**
* A schema for a number value.
* @param name The name of the field this schema refers to.
*/
function numberValueSchema(name) {
	return v.number(`Expected a number value for '${name}'`);
}
/**
* A schema for a positive number value (including zero).
* @param name The name of the field this schema refers to.
*/
function positiveNumberValueSchema(name) {
	return v.pipe(numberValueSchema(name), v.minValue(0, `The value of ${name} must be a non-negative number`));
}
/**
* A schema for a string value.
* @param name The name of the field this schema refers to.
*/
function stringValueSchema(name) {
	return v.string(`Expected a string value for '${name}'`);
}
/**
* A schema for a non-empty string value.
* @param name The name of the field this schema refers to.
*/
function nonEmptyStringValueSchema(name) {
	return v.pipe(stringValueSchema(name), v.nonEmpty(`The value of "${name}" must not be empty`));
}
/**
* A schema for a boolean value.
* @param name The name of the field this schema refers to.
*/
function booleanValueSchema(name) {
	return v.boolean(`Expected a boolean value for '${name}'`);
}
/**
* A schema for a string that only contains alphanumeric characters and hyphens.
* @param name The name of the field this schema refers to.
* @param casing The allowed casing for the string (default: "any").
*/
function alphaNumericOrHyphenSchema(name, casing = "any") {
	const casingLabel = casing === "any" ? "" : ` (${casing} only)`;
	return v.pipe(stringValueSchema(name), v.regex(ALPHANUMERIC_OR_HYPHEN_REGEX[casing], `Only alphanumeric characters and hyphens are allowed in string value of "${name}"${casingLabel}`));
}

//#endregion
export { stringValueSchema as a, positiveNumberValueSchema as i, booleanValueSchema as n, nonEmptyStringValueSchema as r, alphaNumericOrHyphenSchema as t };
