// @adobe/aio-commerce-lib-app/dist/es/validate-Dfajrp99.mjs
/**
 * @license
 *
 * Copyright 2026 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */

import { t as alphaNumericOrHyphenSchema } from "./schemas-hRovglBF.mjs";
import { a as hasCustomInstallationSteps, d as hasAdminUiSdk, i as hasCustomInstallation, l as hasExternalEvents, n as hasWebhooks, o as EventingSchema, r as InstallationSchema, s as hasCommerceEvents, t as WebhooksSchema, u as AdminUiSdkSchema } from "./webhooks-B5oq9ULL.mjs";
import { CommerceSdkValidationError } from "@adobe/aio-commerce-lib-core/error";
import { SchemaBusinessConfig } from "@adobe/aio-commerce-lib-config";
import * as v from "valibot";

//#region source/config/schema/business-configuration.ts
/**
* Check if config has business config.
* @param config - The configuration to check.
*/
function hasBusinessConfig(config) {
	return config.businessConfig !== void 0;
}
/**
* Check if config has business config schema.
* @param config - The configuration to check.
*/
function hasBusinessConfigSchema(config) {
	return config.businessConfig?.schema !== void 0 && config.businessConfig.schema.length > 0;
}

//#endregion
//#region source/config/schema/metadata.ts
const MAX_ID_LENGTH = 100;
const MAX_DESCRIPTION_LENGTH = 255;
const MAX_DISPLAY_NAME_LENGTH = 50;
const NUMERIC_IDENTIFIER = "(0|[1-9]\\d*)";
const SEMVER_REGEX = new RegExp(`^${NUMERIC_IDENTIFIER}\\.${NUMERIC_IDENTIFIER}\\.${NUMERIC_IDENTIFIER}$`);
function nonEmptyString(fieldName) {
	return v.pipe(v.string(`Expected a string for the ${fieldName}`), v.nonEmpty(`The ${fieldName} must not be empty`));
}
/** The schema for the metadata of the application. */
const MetadataSchema = v.object({
	id: v.pipe(alphaNumericOrHyphenSchema("application id (metadata.id)"), v.maxLength(MAX_ID_LENGTH, `The application id must not be longer than ${MAX_ID_LENGTH} characters`)),
	displayName: v.pipe(nonEmptyString("application display name"), v.maxLength(MAX_DISPLAY_NAME_LENGTH, `The application display name must not be longer than ${MAX_DISPLAY_NAME_LENGTH} characters`)),
	description: v.pipe(nonEmptyString("metadata description"), v.maxLength(MAX_DESCRIPTION_LENGTH, `The metadata description must not be longer than ${MAX_DESCRIPTION_LENGTH} characters`)),
	version: v.pipe(nonEmptyString("version"), v.regex(SEMVER_REGEX, "The version must follow semantic versioning (semver) format: Major.Minor.Patch (e.g., '1.0.0', '2.3.1')"))
});
/**
* Check if config has metadata.
* @param config - The configuration to check.
*/
function hasMetadata(config) {
	return config.metadata !== void 0;
}

//#endregion
//#region source/config/schema/domains.ts
/** The individual validatable domains of the app config. */
const CommerceAppConfigSchemas = {
	metadata: MetadataSchema,
	businessConfig: SchemaBusinessConfig,
	eventing: EventingSchema,
	installation: InstallationSchema,
	webhooks: WebhooksSchema,
	"businessConfig.schema": v.unwrap(SchemaBusinessConfig.entries.schema),
	"eventing.commerce": v.unwrap(EventingSchema.entries.commerce),
	"eventing.external": v.unwrap(EventingSchema.entries.external),
	"installation.customInstallationSteps": v.unwrap(InstallationSchema.entries.customInstallationSteps),
	adminUiSdk: AdminUiSdkSchema
};
/**
* Get the config domains that are present in the config.
* @param config - The configuration to check.
*/
function getConfigDomains(config) {
	const withCommerceEvents = hasCommerceEvents(config);
	const withExternalEvents = hasExternalEvents(config);
	const domains = {
		metadata: hasMetadata(config),
		businessConfig: hasBusinessConfig(config),
		eventing: withCommerceEvents || withExternalEvents,
		installation: hasCustomInstallation(config),
		webhooks: hasWebhooks(config),
		adminUiSdk: hasAdminUiSdk(config),
		"businessConfig.schema": hasBusinessConfigSchema(config),
		"eventing.commerce": withCommerceEvents,
		"eventing.external": withExternalEvents,
		"installation.customInstallationSteps": hasCustomInstallationSteps(config)
	};
	const domainsList = Object.entries(domains).filter(([_, value]) => value).map(([key]) => key);
	return new Set(domainsList);
}
/**
* Check if the config has a specific domain.
* @param config - The configuration to check.
* @param domain - The domain to check.
*/
function hasConfigDomain(config, domain) {
	return getConfigDomains(config).has(domain);
}

//#endregion
//#region source/config/schema/app.ts
/** biome-ignore-all lint/performance/noBarrelFile: We want to have all the schema exports in one file. */
/** The schema used to validate the commerce app config file. */
const CommerceAppConfigSchema = v.looseObject({
	metadata: MetadataSchema,
	businessConfig: v.optional(SchemaBusinessConfig),
	eventing: v.optional(EventingSchema),
	adminUiSdk: v.optional(AdminUiSdkSchema),
	installation: v.optional(InstallationSchema),
	webhooks: v.optional(WebhooksSchema)
});
/**
* Returns true if the given config domains require installation, false otherwise.
* @param domains - The config domains to check.
*/
function requiresInstallationFromDomains(domains) {
	return [
		"installation.customInstallationSteps",
		"eventing.commerce",
		"eventing.external",
		"webhooks",
		"adminUiSdk"
	].some((feature) => domains.has(feature));
}
/**
* Returns true if the given config requires installation, false otherwise.
* @param config - The commerce app config to check.
*/
function requiresInstallation(config) {
	return requiresInstallationFromDomains(getConfigDomains(config));
}

//#endregion
//#region source/config/lib/validate.ts
const commerceAppConfigDomainsSchema = v.picklist(Object.keys(CommerceAppConfigSchemas));
/**
* Validates a complete commerce app configuration object against the schema.
*
* @param config - The configuration object to validate.
* @returns The validated and typed configuration output model.
*
* @throws {CommerceSdkValidationError} If the configuration is invalid, with
* detailed validation issues included.
*
* @example
* ```typescript
* const config = {
*   businessConfiguration: {
*     // ... configuration data
*   }
* };
*
* try {
*   const validatedConfig = validateCommerceAppConfig(config);
*   // Use validatedConfig safely
* } catch (error) {
*   if (error instanceof CommerceSdkValidationError) {
*     console.error('Validation failed:', error.display());
*   }
* }
* ```
*/
function validateCommerceAppConfig(config) {
	const validatedConfig = v.safeParse(CommerceAppConfigSchema, config);
	if (!validatedConfig.success) throw new CommerceSdkValidationError("Invalid commerce app config", { issues: validatedConfig.issues });
	return validatedConfig.output;
}
/**
* Validates a specific domain configuration within the commerce app config.
*
* This function validates only a specific domain's configuration rather than
* the entire commerce app configuration object. It first validates that the
* domain name is valid, then validates the configuration data against the
* schema for that specific domain.
*
* @template T - The type of the domain, constrained to valid domain names.
*
* @param config - The domain configuration object to validate.
* @param domain - The name of the domain to validate (e.g., 'businessConfiguration').
* @returns The validated and typed configuration for the specified domain.
*
* @throws {CommerceSdkValidationError} If the domain name is invalid or if the
* configuration doesn't match the domain's schema.
*
* @example
* ```typescript
* const businessConfig = {
*   fields: [
*     {
*       name: 'category',
*       type: 'dropdown',
*       // ... field configuration
*     }
*   ]
* };
*
* try {
*   const validatedConfig = validateCommerceAppConfigDomain(
*     businessConfig,
*     'businessConfig'
*   );
*   // Use validatedConfig safely
* } catch (error) {
*   if (error instanceof CommerceSdkValidationError) {
*     console.error('Domain validation failed:', error.issues);
*   }
* }
* ```
*/
function validateCommerceAppConfigDomain(config, domain) {
	const domainSchema = v.safeParse(commerceAppConfigDomainsSchema, domain);
	if (!domainSchema.success) throw new CommerceSdkValidationError("Invalid commerce app config domain", { issues: domainSchema.issues });
	const domainConfigSchema = CommerceAppConfigSchemas[domain];
	const validatedConfig = v.safeParse(domainConfigSchema, config);
	if (!validatedConfig.success) throw new CommerceSdkValidationError(`Invalid commerce app config: ${domain}`, { issues: validatedConfig.issues });
	return validatedConfig.output;
}

//#endregion
export { CommerceAppConfigSchemas as a, hasMetadata as c, requiresInstallationFromDomains as i, hasBusinessConfig as l, validateCommerceAppConfigDomain as n, getConfigDomains as o, requiresInstallation as r, hasConfigDomain as s, validateCommerceAppConfig as t, hasBusinessConfigSchema as u };
