// @adobe/aio-commerce-lib-app/dist/es/webhooks-B5oq9ULL.mjs
/**
 * @license
 *
 * Copyright 2026 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */

import { a as stringValueSchema, i as positiveNumberValueSchema, n as booleanValueSchema, r as nonEmptyStringValueSchema, t as alphaNumericOrHyphenSchema } from "./schemas-hRovglBF.mjs";
import * as v from "valibot";

//#region source/config/schema/admin-ui-sdk.ts
const SANDBOX_VALUES = [
	"allow-downloads",
	"allow-modals",
	"allow-popups"
];
function isSandboxValue(value) {
	return SANDBOX_VALUES.includes(value);
}
const SandboxSchema = v.pipe(v.string("Expected a string value for \"sandbox\""), v.check((val) => val.split(" ").every(isSandboxValue), `sandbox must contain only single-space-separated values from: ${SANDBOX_VALUES.map((t) => `"${t}"`).join(", ")}`));
const ColumnTypeSchema = v.picklist([
	"boolean",
	"date",
	"float",
	"integer",
	"string"
]);
const ColumnAlignSchema = v.picklist([
	"left",
	"right",
	"center"
]);
const ViewButtonLevelSchema = v.picklist([
	-1,
	0,
	1
]);
const MassActionConfirmSchema = v.object({
	title: v.optional(nonEmptyStringValueSchema("confirm title")),
	message: v.optional(nonEmptyStringValueSchema("confirm message"))
});
const ViewButtonConfirmSchema = v.object({ message: v.optional(nonEmptyStringValueSchema("confirm message")) });
const iframeActionEntries = {
	displayIframe: v.optional(booleanValueSchema("displayIframe")),
	timeout: v.optional(positiveNumberValueSchema("timeout")),
	sandbox: v.optional(SandboxSchema)
};
const GridColumnPropertySchema = v.object({
	label: nonEmptyStringValueSchema("column label"),
	columnId: nonEmptyStringValueSchema("column ID"),
	type: ColumnTypeSchema,
	align: ColumnAlignSchema
});
const GridColumnsSchema = v.object({
	data: v.object({ meshId: nonEmptyStringValueSchema("mesh ID") }),
	properties: v.pipe(v.array(GridColumnPropertySchema), v.minLength(1, "At least one grid column property is required"))
});
const massActionBaseEntries = {
	actionId: nonEmptyStringValueSchema("mass action ID"),
	label: nonEmptyStringValueSchema("mass action label"),
	title: v.optional(nonEmptyStringValueSchema("mass action page title")),
	confirm: v.optional(MassActionConfirmSchema),
	path: nonEmptyStringValueSchema("mass action path"),
	...iframeActionEntries
};
const sandboxDisplayIframeCheck = v.forward(v.partialCheck([["sandbox"], ["displayIframe"]], (input) => input.sandbox === void 0 || input.displayIframe === true, "sandbox is only relevant when displayIframe is set to true"), ["sandbox"]);
function withSandboxDisplayIframeCheck(schema) {
	return v.pipe(schema, sandboxDisplayIframeCheck);
}
function createMassActionSchema(variantEntries) {
	return withSandboxDisplayIframeCheck(v.strictObject({
		...massActionBaseEntries,
		...variantEntries
	}));
}
const OrderMassActionSchema = createMassActionSchema({ selectionLimit: v.optional(positiveNumberValueSchema("selectionLimit")) });
const ProductMassActionSchema = createMassActionSchema({ productSelectLimit: v.optional(positiveNumberValueSchema("productSelectLimit")) });
const CustomerMassActionSchema = createMassActionSchema({ customerSelectLimit: v.optional(positiveNumberValueSchema("customerSelectLimit")) });
const OrderViewButtonSchema = withSandboxDisplayIframeCheck(v.object({
	buttonId: nonEmptyStringValueSchema("view button ID"),
	label: nonEmptyStringValueSchema("view button label"),
	confirm: v.optional(ViewButtonConfirmSchema),
	path: nonEmptyStringValueSchema("view button path"),
	level: v.optional(ViewButtonLevelSchema),
	sortOrder: v.optional(positiveNumberValueSchema("sortOrder")),
	...iframeActionEntries
}));
const CustomFeeSchema = v.object({
	id: nonEmptyStringValueSchema("custom fee ID"),
	label: nonEmptyStringValueSchema("custom fee label"),
	value: v.number("Custom fee value must be a number"),
	orderMinimumAmount: v.optional(v.number("orderMinimumAmount must be a number")),
	applyFeeOnLastInvoice: v.optional(booleanValueSchema("applyFeeOnLastInvoice")),
	applyFeeOnLastCreditMemo: v.optional(booleanValueSchema("applyFeeOnLastCreditMemo"))
});
const OrderExtensionPointsSchema = v.object({
	massActions: v.optional(v.array(OrderMassActionSchema)),
	gridColumns: v.optional(GridColumnsSchema),
	viewButtons: v.optional(v.array(OrderViewButtonSchema)),
	customFees: v.optional(v.array(CustomFeeSchema))
});
const ProductExtensionPointsSchema = v.object({
	massActions: v.optional(v.array(ProductMassActionSchema)),
	gridColumns: v.optional(GridColumnsSchema)
});
const CustomerExtensionPointsSchema = v.object({
	massActions: v.optional(v.array(CustomerMassActionSchema)),
	gridColumns: v.optional(GridColumnsSchema)
});
const MassActionBannerSchema = v.object({
	actionId: nonEmptyStringValueSchema("mass action ID"),
	successMessage: v.optional(nonEmptyStringValueSchema("success message")),
	errorMessage: v.optional(nonEmptyStringValueSchema("error message"))
});
const OrderViewButtonBannerSchema = v.object({
	buttonId: nonEmptyStringValueSchema("view button ID"),
	successMessage: v.optional(nonEmptyStringValueSchema("success message")),
	errorMessage: v.optional(nonEmptyStringValueSchema("error message"))
});
const BannerNotificationSchema = v.object({
	massActions: v.optional(v.object({
		order: v.optional(v.array(MassActionBannerSchema)),
		product: v.optional(v.array(MassActionBannerSchema)),
		customer: v.optional(v.array(MassActionBannerSchema))
	})),
	orderViewButtons: v.optional(v.array(OrderViewButtonBannerSchema))
});
const MenuItemSchema = v.object({
	id: nonEmptyStringValueSchema("menu item ID"),
	title: v.optional(nonEmptyStringValueSchema("menu item title")),
	parent: v.optional(nonEmptyStringValueSchema("menu item parent")),
	sortOrder: v.optional(v.number()),
	isSection: v.optional(booleanValueSchema("isSection")),
	sandbox: v.optional(SandboxSchema)
});
/**
* Schema for the Admin UI SDK registration parameters (for the `adminUiSdk.registration` config section).
* @see https://developer.adobe.com/commerce/extensibility/admin-ui-sdk/extension-points/ for more details.
*/
const AdminUiSdkRegistrationSchema = v.object({
	menuItems: v.optional(v.array(MenuItemSchema)),
	order: v.optional(OrderExtensionPointsSchema),
	product: v.optional(ProductExtensionPointsSchema),
	customer: v.optional(CustomerExtensionPointsSchema),
	bannerNotification: v.optional(BannerNotificationSchema)
});
/**
* Schema for Admin UI SDK configuration.
* @experimental
*/
const AdminUiSdkSchema = v.object({ registration: AdminUiSdkRegistrationSchema });
/**
* Check if config has Admin UI SDK registration configuration.
* @experimental
*/
function hasAdminUiSdk(config) {
	return config.adminUiSdk !== void 0 && config.adminUiSdk.registration !== void 0;
}

//#endregion
//#region source/config/schema/eventing.ts
const MAX_DESCRIPTION_LENGTH$1 = 255;
const MAX_LABEL_LENGTH = 100;
const MAX_KEY_LENGTH = 50;
const MAX_EVENT_NAME_LENGTH = 180;
/**
* Regex for Commerce event names that must start with "plugin." or "observer."
* followed by one or more dot-separated lowercase segments containing letters
* and underscores only.
* Examples: "observer.order_placed", "plugin.sales.api.order_management.place"
*/
const COMMERCE_EVENT_NAME_REGEX = /^(?:plugin|observer)\.[a-z_]+(?:\.[a-z_]+)*$/;
/**
* Regex for external event names.
* Allows word characters (letters, digits, underscore), hyphens, underscores, and dots.
* Examples: "external_event", "webhook.received", "my-event_123"
*/
const EXTERNAL_EVENT_NAME_REGEX = /^[\w\-_.]+$/;
/**
* Regex for field names according to XSD fieldName pattern.
* Field name can either contain only [a-zA-Z0-9_\-\.\[\]] or be set to *.
*/
const FIELD_NAME_REGEX = /^([a-zA-Z0-9_\-.[\]]+|\*)$/;
/**
* Regex for Adobe I/O Events API text fields (label, description).
* Valid characters per API: letters, numbers, spaces, underscores, hyphens,
* dots, colons, parentheses, commas, @, and /.
*/
const IO_EVENTS_TEXT_REGEX = /^[a-zA-Z0-9 _.:()\-,@/]+$/;
/**
* Schema for Commerce event names.
* Validates that the event name starts with "plugin." or "observer."
* followed by one or more dot-separated lowercase segments containing letters
* and underscores only.
*/
function commerceEventNameSchema() {
	return v.pipe(nonEmptyStringValueSchema("event name"), v.regex(COMMERCE_EVENT_NAME_REGEX, "Event name must start with \"plugin.\" or \"observer.\" followed by one or more dot-separated lowercase segments containing letters and underscores only (e.g., \"observer.order_placed\", \"plugin.sales.api.order_management.place\")"), v.maxLength(MAX_EVENT_NAME_LENGTH, `The event name must not be longer than ${MAX_EVENT_NAME_LENGTH} characters`));
}
/**
* Schema for external event names.
* Validates that the event name contains only word characters (letters, digits, underscore),
* hyphens, underscores, and dots.
*/
function externalEventNameSchema() {
	return v.pipe(nonEmptyStringValueSchema("event name"), v.regex(EXTERNAL_EVENT_NAME_REGEX, "Event name must contain only letters, digits, underscores, hyphens, and dots (e.g., \"external_event\", \"webhook.received\", \"my-event_123\")"), v.maxLength(MAX_EVENT_NAME_LENGTH, `The event name must not be longer than ${MAX_EVENT_NAME_LENGTH} characters`));
}
/**
* Schema for field names.
* Validates that the field name matches the XSD fieldName pattern:
* can either contain only [a-zA-Z0-9_\-\.\[\]] or be set to *.
*/
function fieldNameSchema() {
	return v.pipe(nonEmptyStringValueSchema("field name"), v.regex(FIELD_NAME_REGEX, "Field name must contain only letters (a-z, A-Z), numbers (0-9), underscores (_), dashes (-), dots (.), and square brackets ([, ]), or be exactly \"*\""));
}
/** Validates that a text field contains only characters accepted by the Adobe I/O Events API. */
function ioEventsTextSchema(name) {
	return v.regex(IO_EVENTS_TEXT_REGEX, `${name} can only contain letters, numbers, spaces, underscores, hyphens, dots, colons, parentheses, commas, @, and /`);
}
/**
* Schema for field objects in Commerce events.
* Each field has a required name and an optional source.
*/
function commerceEventFieldSchema() {
	return v.object({
		name: fieldNameSchema(),
		source: v.optional(stringValueSchema("field source"))
	});
}
/** Schema for event provider configuration */
const ProviderSchema = v.object({
	label: v.pipe(nonEmptyStringValueSchema("provider label"), ioEventsTextSchema("Provider label"), v.maxLength(MAX_LABEL_LENGTH, `The provider label must not be longer than ${MAX_LABEL_LENGTH} characters`)),
	description: v.pipe(nonEmptyStringValueSchema("provider description"), ioEventsTextSchema("Provider description"), v.maxLength(MAX_DESCRIPTION_LENGTH$1, `The provider description must not be longer than ${MAX_DESCRIPTION_LENGTH$1} characters`)),
	key: v.optional(v.pipe(alphaNumericOrHyphenSchema("provider key"), v.maxLength(MAX_KEY_LENGTH, `The provider key must not be longer than ${MAX_KEY_LENGTH} characters`)))
});
/** Schema for base shared properties between event types. */
const BaseEventSchema = v.object({
	label: v.pipe(nonEmptyStringValueSchema("event label"), ioEventsTextSchema("Event label"), v.maxLength(MAX_LABEL_LENGTH, `The event label must not be longer than ${MAX_LABEL_LENGTH} characters`)),
	description: v.pipe(nonEmptyStringValueSchema("event description"), ioEventsTextSchema("Event description"), v.maxLength(MAX_DESCRIPTION_LENGTH$1, `The event description must not be longer than ${MAX_DESCRIPTION_LENGTH$1} characters`)),
	runtimeActions: v.array(v.pipe(nonEmptyStringValueSchema("runtime action"), v.regex(/^[a-z0-9-]+\/[a-z0-9-]+$/i, "Runtime action must be in the format \"<package>/<action>\" (e.g., \"my-package/my-action\")")), "Expected an array of runtime actions in the format <package>/<action>")
});
/**
* Schema for rule operator values.
* Valid operators for Commerce event filtering rules.
*/
const OPERATORS = [
	"greaterThan",
	"lessThan",
	"equal",
	"regex",
	"in",
	"onChange"
];
const ruleOperatorSchema = v.union(OPERATORS.map((op) => v.literal(op)), `Operator must be one of: ${OPERATORS.join(", ")}`);
/** Schema for Commerce event rule configuration */
const CommerceEventRuleSchema = v.object({
	field: nonEmptyStringValueSchema("rule field"),
	operator: ruleOperatorSchema,
	value: nonEmptyStringValueSchema("rule value")
});
/** Schema for Commerce event configuration */
const CommerceEventSchema = v.object({
	...BaseEventSchema.entries,
	name: commerceEventNameSchema(),
	fields: v.pipe(v.array(commerceEventFieldSchema(), "Expected an array of event field objects with a 'name' property"), v.minLength(1, "The Commerce event configuration must define at least one field")),
	rules: v.optional(v.array(CommerceEventRuleSchema, "Expected an array of event rules with field, operator, and value")),
	destination: v.optional(nonEmptyStringValueSchema("destination")),
	hipaa_audit_required: v.optional(booleanValueSchema("hipaa_audit_required")),
	priority: v.optional(booleanValueSchema("priority")),
	force: v.optional(booleanValueSchema("force"))
});
/** Schema for external event configuration */
const ExternalEventSchema = v.object({
	...BaseEventSchema.entries,
	name: externalEventNameSchema()
});
/** Schema for Commerce event source configuration */
const CommerceEventSourceSchema = v.object({
	provider: ProviderSchema,
	events: v.array(CommerceEventSchema, "Expected an array of Commerce events")
});
/** Schema for external event source configuration */
const ExternalEventSourceSchema = v.object({
	provider: ProviderSchema,
	events: v.array(ExternalEventSchema, "Expected an array of external events")
});
/** Schema for eventing configuration with separate commerce and external arrays */
const EventingSchema = v.object({
	commerce: v.optional(v.pipe(v.array(CommerceEventSourceSchema, "Expected an array of Commerce event sources"), v.check((sources) => new Set(sources.map((s) => s.provider.label)).size === sources.length, "Commerce provider labels must be unique"))),
	external: v.optional(v.pipe(v.array(ExternalEventSourceSchema, "Expected an array of external event sources"), v.check((sources) => new Set(sources.map((s) => s.provider.label)).size === sources.length, "External provider labels must be unique")))
});
/**
* Check if config has commerce event sources.
* @param config - The configuration to check.
*/
function hasCommerceEvents(config) {
	return Array.isArray(config?.eventing?.commerce) && config.eventing.commerce.length > 0;
}
/**
* Check if config has external event sources.
* @param config - The configuration to check.
*/
function hasExternalEvents(config) {
	return Array.isArray(config?.eventing?.external) && config.eventing.external.length > 0;
}
/**
* Check if config has any eventing configuration.
* @param config - The configuration to check.
*/
function hasEventing(config) {
	return config.eventing !== void 0;
}

//#endregion
//#region source/config/schema/installation.ts
const MAX_DESCRIPTION_LENGTH = 255;
const MAX_NAME_LENGTH = 255;
const MAX_MESSAGE_LENGTH = 1e3;
/**
* Regex for script paths that can be relative or absolute.
* Must end with .js extension
* Examples:
*   - "./scripts/setup.js"
*   - "./setup.js"
*   - "../../scripts/setup.js"
*/
const SCRIPT_PATH_REGEX = /^(?:\.{0,2}\/)*[\w-/]*[\w-]+\.js$/;
/**
* Schema for custom installation step configuration
*/
const CustomInstallationStepSchema = v.object({
	script: v.pipe(nonEmptyStringValueSchema("script path"), v.regex(SCRIPT_PATH_REGEX, "Script path must end with .js (e.g., \"./setup.js\", \"./scripts/setup.js\", or \"../../scripts/setup.js\")")),
	name: v.pipe(nonEmptyStringValueSchema("step name"), v.maxLength(MAX_NAME_LENGTH, `The step name must not be longer than ${MAX_NAME_LENGTH} characters`)),
	description: v.pipe(nonEmptyStringValueSchema("step description"), v.maxLength(MAX_DESCRIPTION_LENGTH, `The step description must not be longer than ${MAX_DESCRIPTION_LENGTH} characters`))
});
/**
* Schema for installation messages configuration
*/
const MessagesSchema = v.object({
	preInstallation: v.optional(v.pipe(nonEmptyStringValueSchema("preInstallation message"), v.maxLength(MAX_MESSAGE_LENGTH, `The preInstallation message must not be longer than ${MAX_MESSAGE_LENGTH} characters`))),
	postInstallation: v.optional(v.pipe(nonEmptyStringValueSchema("postInstallation message"), v.maxLength(MAX_MESSAGE_LENGTH, `The postInstallation message must not be longer than ${MAX_MESSAGE_LENGTH} characters`)))
});
/**
* Schema for installation configuration
*/
const InstallationSchema = v.object({
	messages: v.optional(MessagesSchema),
	customInstallationSteps: v.pipe(v.optional(v.array(CustomInstallationStepSchema, "Expected an array of custom installation steps")), v.check((input) => {
		const steps = input || [];
		return new Set(steps.map((step) => step.name)).size === steps.length;
	}, "Duplicate step names detected in custom installation steps. Each step must have a unique name."))
});
/**
* Check if config has custom installation settings.
* @param config - The configuration to check.
*/
function hasCustomInstallation(config) {
	return config.installation !== void 0;
}
/**
* Check if config has custom installation steps.
* @param config - The configuration to check.
*/
function hasCustomInstallationSteps(config) {
	return Array.isArray(config?.installation?.customInstallationSteps) && config.installation.customInstallationSteps.length > 0;
}

//#endregion
//#region source/config/schema/webhooks.ts
/** Schema for webhook field configuration (name and optional source). */
const WebhookFieldSchema = v.object({
	name: nonEmptyStringValueSchema("field name"),
	source: v.optional(stringValueSchema("field source"))
});
/** Schema for webhook rule configuration (field, operator, value). */
const WebhookRuleSchema = v.object({
	field: nonEmptyStringValueSchema("rule field"),
	operator: nonEmptyStringValueSchema("rule operator"),
	value: nonEmptyStringValueSchema("rule value")
});
/** Schema for webhook header configuration (name, value). */
const WebhookHeaderSchema = v.object({
	name: nonEmptyStringValueSchema("header name"),
	value: nonEmptyStringValueSchema("header value")
});
/** batch_name and hook_name must contain only letters, numbers, and underscores. */
const WEBHOOK_IDENTIFIER_REGEX = /^[a-zA-Z0-9_]+$/;
/** Category for conflict detection: validation, append, or modification. */
const CATEGORIES = [
	"validation",
	"append",
	"modification"
];
const CategorySchema = v.picklist(CATEGORIES, `Webhook category must be one of: ${CATEGORIES.join(", ")}`);
/** Schema for the nested webhook payload without url — used when runtimeAction resolves the URL at runtime. */
const WebhookDefinitionBaseSchema = v.object({
	webhook_method: nonEmptyStringValueSchema("webhook_method"),
	webhook_type: nonEmptyStringValueSchema("webhook_type"),
	batch_name: v.pipe(nonEmptyStringValueSchema("batch_name"), v.regex(WEBHOOK_IDENTIFIER_REGEX, "batch_name must contain only letters, numbers, and underscores")),
	batch_order: v.optional(positiveNumberValueSchema("batch_order")),
	hook_name: v.pipe(nonEmptyStringValueSchema("hook_name"), v.regex(WEBHOOK_IDENTIFIER_REGEX, "hook_name must contain only letters, numbers, and underscores")),
	priority: v.optional(positiveNumberValueSchema("priority")),
	required: v.optional(booleanValueSchema("required")),
	soft_timeout: v.optional(positiveNumberValueSchema("soft_timeout")),
	timeout: v.optional(positiveNumberValueSchema("timeout")),
	method: nonEmptyStringValueSchema("HTTP method"),
	fallback_error_message: v.optional(stringValueSchema("fallback_error_message")),
	ttl: v.optional(positiveNumberValueSchema("ttl")),
	fields: v.optional(v.array(WebhookFieldSchema, "Expected an array of webhook field objects")),
	rules: v.optional(v.array(WebhookRuleSchema, "Expected an array of webhook rule objects")),
	headers: v.optional(v.array(WebhookHeaderSchema, "Expected an array of webhook header objects"))
});
/** Schema for the nested webhook payload with a required url. */
const WebhookDefinitionWithUrlSchema = v.object({
	...WebhookDefinitionBaseSchema.entries,
	url: v.pipe(stringValueSchema("webhook URL"), v.url("The 'url' field must be a valid absolute URL (e.g., 'https://example.com/webhook')"))
});
/** Schema for a webhook entry that resolves its URL from a runtime action. */
const WebhookEntryWithRuntimeActionSchema = v.object({
	label: nonEmptyStringValueSchema("webhook label"),
	description: nonEmptyStringValueSchema("webhook description"),
	category: v.optional(CategorySchema),
	runtimeAction: nonEmptyStringValueSchema("runtimeAction"),
	requireAdobeAuth: v.optional(booleanValueSchema("requireAdobeAuth")),
	webhook: WebhookDefinitionBaseSchema
});
/** Schema for a webhook entry that provides an explicit URL. */
const WebhookEntryWithUrlSchema = v.object({
	label: nonEmptyStringValueSchema("webhook label"),
	description: nonEmptyStringValueSchema("webhook description"),
	category: v.optional(CategorySchema),
	webhook: WebhookDefinitionWithUrlSchema
});
/** Schema for a single webhook entry — either runtimeAction (no url) or explicit url (no runtimeAction). */
const WebhookEntrySchema = v.union([WebhookEntryWithRuntimeActionSchema, WebhookEntryWithUrlSchema], "Each webhook entry must define either a 'runtimeAction' (to resolve the URL from a runtime action) or an explicit 'url' inside the 'webhook' object, but not both");
/** Schema for the optional webhooks array (when present, must have at least one item). */
const WebhooksSchema = v.optional(v.pipe(v.array(WebhookEntrySchema, "Expected an array of webhook entries"), v.minLength(1, "webhooks array must contain at least one webhook when present")));
/**
* Check if config has webhooks (non-empty array).
* @param config - The configuration to check.
*/
function hasWebhooks(config) {
	return Array.isArray(config?.webhooks) && config.webhooks.length > 0;
}

//#endregion
export { hasCustomInstallationSteps as a, hasEventing as c, hasAdminUiSdk as d, hasCustomInstallation as i, hasExternalEvents as l, hasWebhooks as n, EventingSchema as o, InstallationSchema as r, hasCommerceEvents as s, WebhooksSchema as t, AdminUiSdkSchema as u };
