// @adobe/aio-commerce-lib-events/dist/cjs/io-events/index.cjs
/**
 * @license
 *
 * Copyright 2026 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
const require_utils = require('../utils-Da725lXV.cjs');
let valibot = require("valibot");
valibot = require_utils.__toESM(valibot, 1);
let _adobe_aio_commerce_lib_api = require("@adobe/aio-commerce-lib-api");

//#region source/io-events/api/event-metadata/schema.ts
const BASE64_ENCODED_STRING_REGEX = /^[A-Za-z0-9+/]*={0,2}$/;
const BASE64_ENCODED_STRING_LENGTH_REMAINDER = 4;
function isBase64EncodedString(value) {
	return BASE64_ENCODED_STRING_REGEX.test(value) && value.length % BASE64_ENCODED_STRING_LENGTH_REMAINDER === 0;
}
function encodeToBase64(value) {
	return Buffer.from(value).toString("base64");
}
function tryDecodeFromBase64(value) {
	if (isBase64EncodedString(value)) return Buffer.from(value, "base64").toString("utf-8");
	return value;
}
function sampleEventTemplateSchema(fieldName) {
	return valibot.pipe(valibot.union([
		valibot.pipe(require_utils.stringValueSchema(fieldName), valibot.transform(tryDecodeFromBase64), valibot.parseJson(void 0, `Expected valid JSON string for property '${fieldName}'`), valibot.record(valibot.string(), valibot.unknown())),
		valibot.record(valibot.string(), valibot.unknown()),
		valibot.array(valibot.unknown())
	]), valibot.stringifyJson(void 0, `Expected valid JSON data for property '${fieldName}'`), valibot.transform(encodeToBase64));
}
/** The schema of the parameters received by the GET `providers/:id/eventmetadata` Adobe I/O Events API endpoint. */
const GetAllEventMetadataForProviderSchema = valibot.object({ providerId: require_utils.stringValueSchema("providerId") });
/** The schema of the parameters received by the GET `providers/:id/eventmetadata/:code` Adobe I/O Events API endpoint. */
const GetEventMetadataForEventAndProviderSchema = valibot.object({
	eventCode: require_utils.stringValueSchema("eventCode"),
	providerId: require_utils.stringValueSchema("providerId")
});
/** The schema of the parameters received by the POST `providers/:id/eventmetadata` Adobe I/O Events API endpoint. */
const CreateEventMetadataForProviderSchema = valibot.object({
	consumerOrgId: require_utils.stringValueSchema("consumerOrgId"),
	projectId: require_utils.stringValueSchema("projectId"),
	workspaceId: require_utils.stringValueSchema("workspaceId"),
	providerId: require_utils.stringValueSchema("providerId"),
	label: require_utils.stringValueSchema("label"),
	description: require_utils.stringValueSchema("description"),
	eventCode: require_utils.stringValueSchema("eventCode"),
	sampleEventTemplate: valibot.optional(sampleEventTemplateSchema("sampleEventTemplate"))
});
/** The schema of the parameters received by the DELETE `providers/:id/eventmetadata/:code` Adobe I/O Events API endpoint. */
const DeleteEventMetadataForProviderSchema = valibot.object({
	consumerOrgId: require_utils.stringValueSchema("consumerOrgId"),
	projectId: require_utils.stringValueSchema("projectId"),
	workspaceId: require_utils.stringValueSchema("workspaceId"),
	providerId: require_utils.stringValueSchema("providerId"),
	eventCode: require_utils.stringValueSchema("eventCode")
});

//#endregion
//#region source/io-events/api/event-metadata/endpoints.ts
var endpoints_exports$2 = /* @__PURE__ */ require_utils.__exportAll({
	createEventMetadataForProvider: () => createEventMetadataForProvider,
	deleteEventMetadataForProvider: () => deleteEventMetadataForProvider,
	getAllEventMetadataForProvider: () => getAllEventMetadataForProvider,
	getEventMetadataForEventAndProvider: () => getEventMetadataForEventAndProvider
});
/**
* Gets all event metadata for a specific provider.
* @see https://developer.adobe.com/events/docs/api#operation/getByProviderId
*
* @param httpClient - The {@link AdobeIoEventsHttpClient} to use to make the request.
* @param params - The parameters to get the event metadata with.
* @param fetchOptions - The {@link Options} to use to make the request.
*
* @throws A {@link CommerceSdkValidationError} If the parameters are in the wrong format.
* @throws An {@link HTTPError} If the status code is not 2XX.
*/
async function getAllEventMetadataForProvider(httpClient, params, fetchOptions) {
	const validatedParams = require_utils.parseOrThrow(GetAllEventMetadataForProviderSchema, params);
	return httpClient.get(`providers/${validatedParams.providerId}/eventmetadata`, fetchOptions).json();
}
/**
* Gets event metadata for a specific event code and provider.
* @see https://developer.adobe.com/events/docs/api#operation/getByProviderIdAndEventCode
*
* @param httpClient - The {@link AdobeIoEventsHttpClient} to use to make the request.
* @param params - The parameters to get the event metadata with.
* @param fetchOptions - The {@link Options} to use to make the request.
*
* @throws A {@link CommerceSdkValidationError} If the parameters are in the wrong format.
* @throws An {@link HTTPError} If the status code is not 2XX.
*/
async function getEventMetadataForEventAndProvider(httpClient, params, fetchOptions) {
	const validatedParams = require_utils.parseOrThrow(GetEventMetadataForEventAndProviderSchema, params);
	return httpClient.get(`providers/${validatedParams.providerId}/eventmetadata/${validatedParams.eventCode}`, fetchOptions).json();
}
/**
* Creates event metadata for a provider.
* @see https://developer.adobe.com/events/docs/api#operation/postEventMetadata
*
* @param httpClient - The {@link AdobeIoEventsHttpClient} to use to make the request.
* @param params - The parameters to create the event metadata with.
* @param fetchOptions - The {@link Options} to use to make the request.
*
* @throws A {@link CommerceSdkValidationError} If the parameters are in the wrong format.
* @throws An {@link HTTPError} If the status code is not 2XX.
*/
async function createEventMetadataForProvider(httpClient, params, fetchOptions) {
	const validatedParams = require_utils.parseOrThrow(CreateEventMetadataForProviderSchema, params);
	return httpClient.post(`${validatedParams.consumerOrgId}/${validatedParams.projectId}/${validatedParams.workspaceId}/providers/${validatedParams.providerId}/eventmetadata`, {
		...fetchOptions,
		json: {
			label: validatedParams.label,
			description: validatedParams.description,
			event_code: validatedParams.eventCode,
			sample_event_template: validatedParams.sampleEventTemplate
		}
	}).json();
}
/**
* Deletes event metadata for a specific event code and provider.
* @see https://developer.adobe.com/events/docs/api#operation/deleteEventMetadata
*
* @param httpClient - The {@link AdobeIoEventsHttpClient} to use to make the request.
* @param params - The parameters to delete the event metadata with.
* @param fetchOptions - The {@link Options} to use to make the request.
*
* @throws A {@link CommerceSdkValidationError} If the parameters are in the wrong format.
* @throws An {@link HTTPError} If the status code is not 2XX.
*/
async function deleteEventMetadataForProvider(httpClient, params, fetchOptions) {
	const validatedParams = require_utils.parseOrThrow(DeleteEventMetadataForProviderSchema, params);
	return httpClient.delete(`${validatedParams.consumerOrgId}/${validatedParams.projectId}/${validatedParams.workspaceId}/providers/${validatedParams.providerId}/eventmetadata/${validatedParams.eventCode}`, fetchOptions).then((_res) => {});
}

//#endregion
//#region source/utils/query-params.ts
/**
* Sets a query parameter if the value is truthy.
* @param queryParams The URLSearchParams instance to set the query parameter on.
* @param key The key of the query parameter to set.
* @param value The value of the query parameter to set.
*/
function setQueryParamIfTruthy(queryParams, key, value) {
	if (value) queryParams.set(key, `${value}`);
}
/**
* Sets an array query parameter if the values are truthy.
* @param queryParams The URLSearchParams instance to set the query parameter on.
* @param key The key of the query parameter to set.
* @param values The values of the query parameter to set.
*/
function setArrayQueryParam(queryParams, key, values) {
	if (values !== void 0 && values.length > 0) for (const value of values) queryParams.set(key, `${value}`);
}

//#endregion
//#region source/io-events/lib/schema.ts
/**
* The provider types that this library is able to handle.
* We deliberately don't support any other provider types.
*/
const VALID_EVENT_PROVIDER_TYPES = ["dx_commerce_events", "3rd_party_custom_events"];
/** The data residency regions that this I/O Events API supports. */
const VALID_DATA_RESIDENCY_REGIONS = ["va6", "irl1"];
const EventProviderTypeSchema = valibot.picklist(VALID_EVENT_PROVIDER_TYPES);
const DataResidencyRegionSchema = valibot.picklist(VALID_DATA_RESIDENCY_REGIONS);

//#endregion
//#region source/io-events/api/event-providers/schema.ts
const EventProviderListAllParamsSchema = valibot.object({
	consumerOrgId: require_utils.stringValueSchema("consumerOrgId"),
	withEventMetadata: valibot.optional(require_utils.booleanValueSchema("withEventMetadata")),
	filterBy: valibot.optional(valibot.object({
		instanceId: valibot.optional(require_utils.stringValueSchema("instanceId")),
		providerTypes: valibot.optional(valibot.union([valibot.array(EventProviderTypeSchema, "Expected an array of event provider types")]))
	}))
});
const EventProviderGetByIdParamsSchema = valibot.object({
	providerId: require_utils.stringValueSchema("providerId"),
	withEventMetadata: valibot.optional(require_utils.booleanValueSchema("withEventMetadata"))
});
const EventProviderCreateParamsSchema = valibot.object({
	consumerOrgId: require_utils.stringValueSchema("consumerOrgId"),
	projectId: require_utils.stringValueSchema("projectId"),
	workspaceId: require_utils.stringValueSchema("workspaceId"),
	label: require_utils.stringValueSchema("label"),
	description: valibot.optional(require_utils.stringValueSchema("description")),
	docsUrl: valibot.optional(require_utils.stringValueSchema("docsUrl")),
	instanceId: valibot.optional(require_utils.stringValueSchema("instanceId")),
	providerType: valibot.optional(EventProviderTypeSchema),
	dataResidencyRegion: valibot.optional(DataResidencyRegionSchema)
});
const EventProviderDeleteParamsSchema = valibot.object({
	consumerOrgId: require_utils.stringValueSchema("consumerOrgId"),
	projectId: require_utils.stringValueSchema("projectId"),
	workspaceId: require_utils.stringValueSchema("workspaceId"),
	providerId: require_utils.stringValueSchema("providerId")
});

//#endregion
//#region source/io-events/api/event-providers/endpoints.ts
var endpoints_exports$1 = /* @__PURE__ */ require_utils.__exportAll({
	createEventProvider: () => createEventProvider,
	deleteEventProvider: () => deleteEventProvider,
	getAllEventProviders: () => getAllEventProviders,
	getEventProviderById: () => getEventProviderById
});
/**
* Lists all event providers for the given consumer organization ID.
* @see https://developer.adobe.com/events/docs/api#operation/getProvidersByConsumerOrgId
*
* @param httpClient - The {@link AdobeIoEventsHttpClient} to use to make the request.
* @param params - The parameters to list the event providers with.
* @param fetchOptions - The {@link Options} to use to make the request.
*/
async function getAllEventProviders(httpClient, params, fetchOptions) {
	const validatedParams = require_utils.parseOrThrow(EventProviderListAllParamsSchema, params);
	const queryParams = new URLSearchParams();
	const { providerTypes = [], instanceId } = validatedParams.filterBy ?? {};
	setArrayQueryParam(queryParams, "providerMetadataIds", providerTypes);
	setQueryParamIfTruthy(queryParams, "instanceId", instanceId);
	setQueryParamIfTruthy(queryParams, "eventmetadata", validatedParams.withEventMetadata);
	return httpClient.get(`${validatedParams.consumerOrgId}/providers`, {
		...fetchOptions,
		searchParams: queryParams
	}).json();
}
/**
* Gets an event provider by ID.
* @see https://developer.adobe.com/events/docs/api#operation/getProvidersById
*
* @param httpClient - The {@link AdobeIoEventsHttpClient} to use to make the request.
* @param params - The parameters to get the event provider by.
* @param fetchOptions - The {@link Options} to use to make the request.
*
* @throws A {@link CommerceSdkValidationError} If the parameters are in the wrong format.
* @throws An {@link HTTPError} If the status code is not 2XX.
*/
async function getEventProviderById(httpClient, params, fetchOptions) {
	const validatedParams = require_utils.parseOrThrow(EventProviderGetByIdParamsSchema, params);
	const searchParams = new URLSearchParams();
	setQueryParamIfTruthy(searchParams, "eventmetadata", validatedParams.withEventMetadata);
	return httpClient.get(`providers/${params.providerId}`, {
		...fetchOptions,
		searchParams
	}).json();
}
/**
* Creates an event provider.
* @see https://developer.adobe.com/events/docs/api#operation/createProvider
*
* @param httpClient - The {@link AdobeIoEventsHttpClient} to use to make the request.
* @param params - The parameters to create the event provider with.
* @param fetchOptions - The {@link Options} to use to make the request.
*
* @throws A {@link CommerceSdkValidationError} If the parameters are in the wrong format.
* @throws An {@link HTTPError} If the status code is not 2XX.
*/
async function createEventProvider(httpClient, params, fetchOptions) {
	const validatedParams = require_utils.parseOrThrow(EventProviderCreateParamsSchema, params);
	return httpClient.post(`${validatedParams.consumerOrgId}/${validatedParams.projectId}/${validatedParams.workspaceId}/providers`, {
		...fetchOptions,
		json: {
			...validatedParams,
			docs_url: validatedParams.docsUrl,
			provider_metadata: validatedParams.providerType,
			instance_id: validatedParams.instanceId,
			data_residency_region: validatedParams.dataResidencyRegion
		}
	}).json();
}
/**
* Deletes an event provider.
* @see https://developer.adobe.com/events/docs/api#operation/deleteProvider
*
* @param httpClient - The {@link AdobeIoEventsHttpClient} to use to make the request.
* @param params - The parameters to delete the event provider with.
* @param fetchOptions - The {@link Options} to use to make the request.
*
* @throws A {@link CommerceSdkValidationError} If the parameters are in the wrong format.
* @throws An {@link HTTPError} If the status code is not 2XX.
*/
async function deleteEventProvider(httpClient, params, fetchOptions) {
	const validatedParams = require_utils.parseOrThrow(EventProviderDeleteParamsSchema, params);
	return httpClient.delete(`${validatedParams.consumerOrgId}/${validatedParams.projectId}/${validatedParams.workspaceId}/providers/${validatedParams.providerId}`, fetchOptions).then((_res) => {});
}

//#endregion
//#region source/io-events/api/event-providers/shorthands.ts
var shorthands_exports = /* @__PURE__ */ require_utils.__exportAll({
	create3rdPartyCustomEventProvider: () => create3rdPartyCustomEventProvider,
	createCommerceEventProvider: () => createCommerceEventProvider,
	getAll3rdPartyCustomEventProviders: () => getAll3rdPartyCustomEventProviders,
	getAllCommerceEventProviders: () => getAllCommerceEventProviders
});
/**
* Lists all Commerce (of type `dx_commerce_events`) event providers for the given consumer organization ID.
* @see https://developer.adobe.com/events/docs/api#operation/getProvidersByConsumerOrgId
*
* @param httpClient - The {@link AdobeIoEventsHttpClient} to use to make the request.
* @param params - The parameters to list the event providers with.
* @param fetchOptions - The {@link Options} to use to make the request.
*
* @throws A {@link CommerceSdkValidationError} If the parameters are in the wrong format.
* @throws An {@link HTTPError} If the status code is not 2XX.
*/
async function getAllCommerceEventProviders(httpClient, params, fetchOptions) {
	return getAllEventProviders(httpClient, {
		...params,
		filterBy: {
			...params.filterBy,
			providerTypes: ["dx_commerce_events"]
		}
	}, fetchOptions);
}
/**
* Lists all 3rd Party Custom Events (of type `3rd_party_custom_events`) event providers for the given consumer organization ID.
* @see https://developer.adobe.com/events/docs/api#operation/getProvidersByConsumerOrgId
*
* @param httpClient - The {@link AdobeIoEventsHttpClient} to use to make the request.
* @param params - The parameters to list the event providers with.
* @param fetchOptions - The {@link Options} to use to make the request.
*
* @throws A {@link CommerceSdkValidationError} If the parameters are in the wrong format.
* @throws An {@link HTTPError} If the status code is not 2XX.
*/
async function getAll3rdPartyCustomEventProviders(httpClient, params, fetchOptions) {
	return getAllEventProviders(httpClient, {
		...params,
		filterBy: {
			...params.filterBy,
			providerTypes: ["3rd_party_custom_events"]
		}
	}, fetchOptions);
}
/**
* Creates a Commerce (of type `dx_commerce_events`) event provider.
* @see https://developer.adobe.com/events/docs/api#operation/createProvider
*
* @param httpClient - The {@link AdobeIoEventsHttpClient} to use to make the request.
* @param params - The parameters to create the event provider with.
* @param fetchOptions - The {@link Options} to use to make the request.
*
* @throws A {@link CommerceSdkValidationError} If the parameters are in the wrong format.
* @throws An {@link HTTPError} If the status code is not 2XX.
*/
async function createCommerceEventProvider(httpClient, params, fetchOptions) {
	return createEventProvider(httpClient, {
		...params,
		providerType: "dx_commerce_events"
	}, fetchOptions);
}
/**
* Creates a 3rd Party Custom Events (of type `3rd_party_custom_events`) event provider.
* @see https://developer.adobe.com/events/docs/api#operation/createProvider
*
* @param httpClient - The {@link AdobeIoEventsHttpClient} to use to make the request.
* @param params - The parameters to create the event provider with.
* @param fetchOptions - The {@link Options} to use to make the request.
*
* @throws A {@link CommerceSdkValidationError} If the parameters are in the wrong format.
* @throws An {@link HTTPError} If the status code is not 2XX.
*/
async function create3rdPartyCustomEventProvider(httpClient, params, fetchOptions) {
	return createEventProvider(httpClient, {
		...params,
		providerType: "3rd_party_custom_events"
	}, fetchOptions);
}

//#endregion
//#region source/io-events/api/event-registrations/schema.ts
/** Valid delivery types for event registrations. */
const VALID_DELIVERY_TYPES = [
	"webhook",
	"webhook_batch",
	"journal",
	"aws_eventbridge"
];
/** Regex pattern for AWS region validation. */
const AWS_REGION_REGEX = /^(us|ca|eu|af|ap|cn|me|sa|il)-[a-z]+-\d$/;
/** Regex pattern for AWS account ID validation (12 digits). */
const AWS_ACCOUNT_ID_REGEX = /^\d{12}$/;
/** Schema for delivery type validation. */
const DeliveryTypeSchema = valibot.picklist(VALID_DELIVERY_TYPES, `Expected delivery type to be one of: ${VALID_DELIVERY_TYPES.join(", ")}`);
/** Schema for events of interest. */
const EventsOfInterestSchema = valibot.object({
	providerId: require_utils.stringValueSchema("providerId"),
	eventCode: require_utils.stringValueSchema("eventCode")
});
/** Schema for AWS EventBridge destination metadata. */
const DestinationMetadataSchema = valibot.object({
	awsRegion: valibot.optional(valibot.pipe(require_utils.stringValueSchema("awsRegion"), valibot.regex(AWS_REGION_REGEX, "Expected AWS region in format like 'us-east-1'"))),
	awsAccountId: valibot.optional(valibot.pipe(require_utils.stringValueSchema("awsAccountId"), valibot.regex(AWS_ACCOUNT_ID_REGEX, "Expected AWS account ID to be a 12-digit number")))
});
/** Schema for subscriber-defined filter. */
const SubscriberFilterSchema = valibot.object({
	name: valibot.pipe(require_utils.stringValueSchema("name"), valibot.minLength(3, "Expected subscriber filter name to be at least 3 characters"), valibot.maxLength(80, "Expected subscriber filter name to be at most 80 characters")),
	description: valibot.optional(valibot.pipe(require_utils.stringValueSchema("description"), valibot.maxLength(250, "Expected subscriber filter description to be at most 250 characters"))),
	subscriberFilter: require_utils.stringValueSchema("subscriberFilter")
});
/** Shared schema for workspace path parameters. */
const WorkspacePathParamsSchema = valibot.object({
	consumerOrgId: require_utils.stringValueSchema("consumerOrgId"),
	projectId: require_utils.stringValueSchema("projectId"),
	workspaceId: require_utils.stringValueSchema("workspaceId")
});
/** Shared schema for events of interest array field. */
const EventsOfInterestArraySchema = valibot.pipe(valibot.array(EventsOfInterestSchema, "Expected eventsOfInterest to be an array of event interest objects"), valibot.minLength(1, "Expected at least one event of interest"));
/** Shared schema for registration body fields. */
const RegistrationBodySchema = valibot.object({
	clientId: valibot.pipe(require_utils.stringValueSchema("clientId"), valibot.minLength(3, "Expected clientId to be at least 1 character"), valibot.maxLength(255, "Expected clientId to be at most 255 characters")),
	name: valibot.pipe(require_utils.stringValueSchema("name"), valibot.minLength(3, "Expected registration name to be at least 3 characters"), valibot.maxLength(255, "Expected registration name to be at most 255 characters")),
	description: valibot.optional(valibot.pipe(require_utils.stringValueSchema("description"), valibot.maxLength(5e3, "Expected description to be at most 5000 characters"))),
	webhookUrl: valibot.optional(valibot.pipe(require_utils.stringValueSchema("webhookUrl"), valibot.maxLength(4e3, "Expected webhook URL to be at most 4000 characters"))),
	eventsOfInterest: EventsOfInterestArraySchema,
	deliveryType: DeliveryTypeSchema,
	runtimeAction: valibot.optional(valibot.pipe(require_utils.stringValueSchema("runtimeAction"), valibot.maxLength(255, "Expected runtime action to be at most 255 characters"))),
	enabled: valibot.optional(require_utils.booleanValueSchema("enabled")),
	destinationMetadata: valibot.optional(DestinationMetadataSchema),
	subscriberFilters: valibot.optional(valibot.pipe(valibot.array(SubscriberFilterSchema, "Expected subscriberFilters to be an array of subscriber filter objects"), valibot.maxLength(1, "Expected at most 1 subscriber filter")))
});
/** Schema for getting all registrations for a workspace. */
const GetAllRegistrationsParamsSchema = WorkspacePathParamsSchema;
/** Schema for getting all registrations for a consumer organization. */
const GetAllRegistrationsByConsumerOrgParamsSchema = valibot.object({ consumerOrgId: require_utils.stringValueSchema("consumerOrgId") });
/** Schema for getting a registration by ID. */
const GetRegistrationByIdParamsSchema = valibot.object({
	...WorkspacePathParamsSchema.entries,
	registrationId: require_utils.stringValueSchema("registrationId")
});
/** Schema for creating a registration. */
const CreateRegistrationParamsSchema = valibot.object({
	...WorkspacePathParamsSchema.entries,
	...RegistrationBodySchema.entries
});
/** Schema for updating a registration. */
const UpdateRegistrationParamsSchema = valibot.object({
	...WorkspacePathParamsSchema.entries,
	registrationId: require_utils.stringValueSchema("registrationId"),
	...RegistrationBodySchema.entries
});
/** Schema for deleting a registration. */
const DeleteRegistrationParamsSchema = valibot.object({
	...WorkspacePathParamsSchema.entries,
	registrationId: require_utils.stringValueSchema("registrationId")
});

//#endregion
//#region source/io-events/api/event-registrations/endpoints.ts
var endpoints_exports = /* @__PURE__ */ require_utils.__exportAll({
	createRegistration: () => createRegistration,
	deleteRegistration: () => deleteRegistration,
	getAllRegistrations: () => getAllRegistrations,
	getAllRegistrationsByConsumerOrg: () => getAllRegistrationsByConsumerOrg,
	getRegistrationById: () => getRegistrationById,
	updateRegistration: () => updateRegistration
});
/**
* Gets all event registrations for a consumer organization (paginated).
* @see https://developer.adobe.com/events/docs/api#operation/getAllRegistrationsForOrg
*
* @param httpClient - The {@link AdobeIoEventsHttpClient} to use to make the request.
* @param params - The parameters to get the registrations with.
* @param fetchOptions - The {@link Options} to use to make the request.
*
* @throws A {@link CommerceSdkValidationError} If the parameters are in the wrong format.
* @throws An {@link HTTPError} If the status code is not 2XX.
*/
async function getAllRegistrationsByConsumerOrg(httpClient, params, fetchOptions) {
	const validatedParams = require_utils.parseOrThrow(GetAllRegistrationsByConsumerOrgParamsSchema, params);
	return httpClient.get(`${validatedParams.consumerOrgId}/registrations`, fetchOptions).json();
}
/**
* Gets all event registrations for a workspace.
* @see https://developer.adobe.com/events/docs/api#operation/getAllRegistrations
*
* @param httpClient - The {@link AdobeIoEventsHttpClient} to use to make the request.
* @param params - The parameters to get the registrations with.
* @param fetchOptions - The {@link Options} to use to make the request.
*
* @throws A {@link CommerceSdkValidationError} If the parameters are in the wrong format.
* @throws An {@link HTTPError} If the status code is not 2XX.
*/
async function getAllRegistrations(httpClient, params, fetchOptions) {
	const validatedParams = require_utils.parseOrThrow(GetAllRegistrationsParamsSchema, params);
	return httpClient.get(`${validatedParams.consumerOrgId}/${validatedParams.projectId}/${validatedParams.workspaceId}/registrations`, fetchOptions).json();
}
/**
* Gets an event registration by ID.
* @see https://developer.adobe.com/events/docs/api#operation/getRegistration
*
* @param httpClient - The {@link AdobeIoEventsHttpClient} to use to make the request.
* @param params - The parameters to get the registration with.
* @param fetchOptions - The {@link Options} to use to make the request.
*
* @throws A {@link CommerceSdkValidationError} If the parameters are in the wrong format.
* @throws An {@link HTTPError} If the status code is not 2XX.
*/
async function getRegistrationById(httpClient, params, fetchOptions) {
	const validatedParams = require_utils.parseOrThrow(GetRegistrationByIdParamsSchema, params);
	return httpClient.get(`${validatedParams.consumerOrgId}/${validatedParams.projectId}/${validatedParams.workspaceId}/registrations/${validatedParams.registrationId}`, fetchOptions).json();
}
/**
* Creates an event registration.
* @see https://developer.adobe.com/events/docs/api#operation/createRegistration
*
* @param httpClient - The {@link AdobeIoEventsHttpClient} to use to make the request.
* @param params - The parameters to create the registration with.
* @param fetchOptions - The {@link Options} to use to make the request.
*
* @throws A {@link CommerceSdkValidationError} If the parameters are in the wrong format.
* @throws An {@link HTTPError} If the status code is not 2XX.
*/
async function createRegistration(httpClient, params, fetchOptions) {
	const validatedParams = require_utils.parseOrThrow(CreateRegistrationParamsSchema, params);
	return httpClient.post(`${validatedParams.consumerOrgId}/${validatedParams.projectId}/${validatedParams.workspaceId}/registrations`, {
		...fetchOptions,
		json: {
			client_id: validatedParams.clientId,
			name: validatedParams.name,
			description: validatedParams.description,
			webhook_url: validatedParams.webhookUrl,
			events_of_interest: validatedParams.eventsOfInterest.map((event) => ({
				provider_id: event.providerId,
				event_code: event.eventCode
			})),
			delivery_type: validatedParams.deliveryType,
			runtime_action: validatedParams.runtimeAction,
			enabled: validatedParams.enabled,
			destination_metadata: validatedParams.destinationMetadata ? {
				aws_region: validatedParams.destinationMetadata.awsRegion,
				aws_account_id: validatedParams.destinationMetadata.awsAccountId
			} : void 0,
			subscriber_filters: validatedParams.subscriberFilters?.map((filter) => ({
				name: filter.name,
				description: filter.description,
				subscriber_filter: filter.subscriberFilter
			}))
		}
	}).json();
}
/**
* Updates an event registration.
* @see https://developer.adobe.com/events/docs/api#operation/updateRegistration
*
* @param httpClient - The {@link AdobeIoEventsHttpClient} to use to make the request.
* @param params - The parameters to update the registration with.
* @param fetchOptions - The {@link Options} to use to make the request.
*
* @throws A {@link CommerceSdkValidationError} If the parameters are in the wrong format.
* @throws An {@link HTTPError} If the status code is not 2XX.
*/
async function updateRegistration(httpClient, params, fetchOptions) {
	const validatedParams = require_utils.parseOrThrow(UpdateRegistrationParamsSchema, params);
	return httpClient.put(`${validatedParams.consumerOrgId}/${validatedParams.projectId}/${validatedParams.workspaceId}/registrations/${validatedParams.registrationId}`, {
		...fetchOptions,
		json: {
			name: validatedParams.name,
			description: validatedParams.description,
			webhook_url: validatedParams.webhookUrl,
			events_of_interest: validatedParams.eventsOfInterest.map((event) => ({
				provider_id: event.providerId,
				event_code: event.eventCode
			})),
			delivery_type: validatedParams.deliveryType,
			runtime_action: validatedParams.runtimeAction,
			enabled: validatedParams.enabled,
			destination_metadata: validatedParams.destinationMetadata ? {
				aws_region: validatedParams.destinationMetadata.awsRegion,
				aws_account_id: validatedParams.destinationMetadata.awsAccountId
			} : void 0,
			subscriber_filters: validatedParams.subscriberFilters?.map((filter) => ({
				name: filter.name,
				description: filter.description,
				subscriber_filter: filter.subscriberFilter
			}))
		}
	}).json();
}
/**
* Deletes an event registration.
* @see https://developer.adobe.com/events/docs/api#operation/deleteRegistration
*
* @param httpClient - The {@link AdobeIoEventsHttpClient} to use to make the request.
* @param params - The parameters to delete the registration with.
* @param fetchOptions - The {@link Options} to use to make the request.
*
* @throws A {@link CommerceSdkValidationError} If the parameters are in the wrong format.
* @throws An {@link HTTPError} If the status code is not 2XX.
*/
async function deleteRegistration(httpClient, params, fetchOptions) {
	const validatedParams = require_utils.parseOrThrow(DeleteRegistrationParamsSchema, params);
	return httpClient.delete(`${validatedParams.consumerOrgId}/${validatedParams.projectId}/${validatedParams.workspaceId}/registrations/${validatedParams.registrationId}`, fetchOptions).then((_res) => {});
}

//#endregion
//#region source/io-events/lib/api-client.ts
/**
* Creates a new API client for the Adobe I/O Events API client.
* @param params - The parameters to build the Adobe I/O Events HTTP client that will communicate with the Adobe I/O Events API.
*/
function createAdobeIoEventsApiClient(params) {
	return _adobe_aio_commerce_lib_api.ApiClient.create(new _adobe_aio_commerce_lib_api.AdobeIoEventsHttpClient(params), {
		...endpoints_exports$1,
		...endpoints_exports$2,
		...shorthands_exports,
		...endpoints_exports
	});
}
/**
* Creates a customized Adobe I/O Events API client.
* @param params - The parameters to build the Adobe I/O Events HTTP client that will communicate with the Adobe I/O Events API.
* @param functions - The API functions to include in the client.
*/
function createCustomAdobeIoEventsApiClient(params, functions) {
	return _adobe_aio_commerce_lib_api.ApiClient.create(new _adobe_aio_commerce_lib_api.AdobeIoEventsHttpClient(params), functions);
}

//#endregion
exports.create3rdPartyCustomEventProvider = create3rdPartyCustomEventProvider;
exports.createAdobeIoEventsApiClient = createAdobeIoEventsApiClient;
exports.createCommerceEventProvider = createCommerceEventProvider;
exports.createCustomAdobeIoEventsApiClient = createCustomAdobeIoEventsApiClient;
exports.createEventMetadataForProvider = createEventMetadataForProvider;
exports.createEventProvider = createEventProvider;
exports.createRegistration = createRegistration;
exports.deleteEventMetadataForProvider = deleteEventMetadataForProvider;
exports.deleteEventProvider = deleteEventProvider;
exports.deleteRegistration = deleteRegistration;
exports.getAll3rdPartyCustomEventProviders = getAll3rdPartyCustomEventProviders;
exports.getAllCommerceEventProviders = getAllCommerceEventProviders;
exports.getAllEventMetadataForProvider = getAllEventMetadataForProvider;
exports.getAllEventProviders = getAllEventProviders;
exports.getAllRegistrations = getAllRegistrations;
exports.getAllRegistrationsByConsumerOrg = getAllRegistrationsByConsumerOrg;
exports.getEventMetadataForEventAndProvider = getEventMetadataForEventAndProvider;
exports.getEventProviderById = getEventProviderById;
exports.getRegistrationById = getRegistrationById;
exports.updateRegistration = updateRegistration;
