// @adobe/aio-commerce-lib-config/dist/es/commands/index.mjs
#!/usr/bin/env node
/**
 * @license
 *
 * Copyright 2026 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */

import { i as validateEncryptionKey, r as generateEncryptionKey } from "../encryption-CSTVRIte.mjs";
import { CommerceSdkValidationError } from "@adobe/aio-commerce-lib-core/error";
import consola from "consola";
import path, { dirname, join, parse } from "path";
import { readFileSync, writeFileSync } from "fs";
import "@adobe/aio-lib-core-config";
import aioIms from "@adobe/aio-lib-ims";
import "dotenv";
import { access, writeFile } from "fs/promises";
import { resolveCommand } from "package-manager-detector";
import { detect, getUserAgent } from "package-manager-detector/detect";

//#region ../../packages-private/scripting-utils/source/error.ts
/**
* This module exports shared error utilities for the AIO Commerce SDK.
* @packageDocumentation
*/
/**
* Stringify an error to a human-friendly string.
* @param error - The error to stringify.
*/
function stringifyError(error) {
	if (error instanceof CommerceSdkValidationError) return error.display();
	return error instanceof Error ? error.message : String(error);
}

//#endregion
//#region ../../packages-private/scripting-utils/source/env.ts
/**
* This module exports shared environment utilities for the AIO Commerce SDK.
* @packageDocumentation
*/
const { context } = aioIms;
/**
* Replaces or creates an environment variable in a .env file
* @param filePath - The path to the .env file
* @param key - The environment variable key to replace or create
* @param value - The new value for the environment variable
*/
function replaceEnvVar(filePath, key, value) {
	const envPath = path.resolve(filePath);
	const envLines = readFileSync(envPath, "utf8").split("\n");
	const formattedValue = value.includes(" ") ? `"${value}"` : value;
	let keyExists = false;
	const updatedLines = envLines.map((line) => {
		if (line.trim().startsWith("#") || !line.includes("=")) return line;
		const [currentKey] = line.split("=");
		if (currentKey === key) {
			keyExists = true;
			return `${key}=${formattedValue}`;
		}
		return line;
	});
	if (!keyExists) updatedLines.push(`${key}=${formattedValue}`);
	writeFileSync(envPath, updatedLines.join("\n"), "utf8");
}

//#endregion
//#region ../../packages-private/scripting-utils/source/filesystem/helpers.ts
/**
* Touch a file, creating it if it doesn't exist.
* @param filePath The path of the file to touch
*/
async function touch(filePath) {
	await writeFile(filePath, "", { flag: "wx" }).catch((e) => {
		if (e.code !== "EEXIST") throw e;
	});
}

//#endregion
//#region ../../packages-private/scripting-utils/source/project.ts
/**
* Find a file by walking up parent directories
* @param name - The file name to search for (or array of file names)
* @param options - Search options
* @returns The path to the file, or undefined if not found
*/
async function findUp(name, options = {}) {
	const names = Array.isArray(name) ? name : [name];
	const cwd = options.cwd || process.cwd();
	const { root } = parse(cwd);
	const stopAt = options.stopAt || root;
	let currentDir = cwd;
	while (true) {
		for (const fileName of names) {
			const filePath = join(currentDir, fileName);
			try {
				await access(filePath);
				return filePath;
			} catch {}
		}
		if (currentDir === stopAt || currentDir === root) return;
		currentDir = dirname(currentDir);
	}
}
/**
* Find the nearest package.json file in the current working directory or its parents
* @param cwd The current working directory
*/
async function findNearestPackageJson(cwd = process.cwd()) {
	const packageJsonPath = await findUp("package.json", { cwd });
	if (!packageJsonPath) return null;
	return packageJsonPath;
}
/**
* Get the root directory of the project
* @param cwd The current working directory
*/
async function getProjectRootDirectory(cwd = process.cwd()) {
	const packageJsonPath = await findNearestPackageJson(cwd);
	if (!packageJsonPath) throw new Error("Could not find a the root directory of the project. `package.json` file not found.");
	return dirname(packageJsonPath);
}
const VALID_PACKAGE_MANAGERS = [
	"npm",
	"pnpm",
	"yarn",
	"bun"
];
/**
* Type guard asserting that a value is a supported `PackageManager`.
* @param value - The value to check
*/
function isValidPackageManager(value) {
	return value !== void 0 && VALID_PACKAGE_MANAGERS.includes(value);
}
/**
* Detect the package manager for a project.
* @param cwd - Directory to start detection from; defaults to `process.cwd()`
*/
async function detectPackageManager(cwd = process.cwd()) {
	const result = await detect({ cwd: await getProjectRootDirectory(cwd) });
	if (isValidPackageManager(result?.name)) return result.name;
	const fromUserAgent = getUserAgent() ?? void 0;
	if (isValidPackageManager(fromUserAgent)) return fromUserAgent;
	return "npm";
}
/**
* Get the exec command that runs a **locally installed** binary from
* `node_modules/.bin` for the given package manager.
* @param packageManager - The detected package manager
*/
function getExecCommand(packageManager) {
	const resolved = resolveCommand(packageManager, "execute-local", []);
	if (!resolved) return "npx";
	return [resolved.command, ...resolved.args].filter(Boolean).join(" ");
}

//#endregion
//#region source/commands/encryption/setup/main.ts
const ENCRYPTION_KEY_ENV_VAR$1 = "AIO_COMMERCE_CONFIG_ENCRYPTION_KEY";
/** Sets up an encryption key for password field at the given environment file path. */
async function run$1(envPath) {
	await touch(envPath);
	process.loadEnvFile(envPath);
	if (!process.env[ENCRYPTION_KEY_ENV_VAR$1]) {
		consola.info("No encryption key found in .env file. Generating new key...");
		replaceEnvVar(envPath, ENCRYPTION_KEY_ENV_VAR$1, generateEncryptionKey());
		consola.success(`Encryption key generated and added to ${envPath}`);
		return;
	}
	consola.info("Encryption key already configured in .env file");
}
/** Run the encryption setup command */
async function exec$1() {
	try {
		const packageJson = await findNearestPackageJson();
		await run$1(join(packageJson ? dirname(packageJson) : process.cwd(), ".env"));
	} catch (error) {
		consola.error(error);
		process.exit(1);
	}
}

//#endregion
//#region source/commands/encryption/validate/main.ts
const ENCRYPTION_KEY_ENV_VAR = "AIO_COMMERCE_CONFIG_ENCRYPTION_KEY";
/** Sets up an encryption key for password field at the given environment file path. */
async function run(envPath) {
	const projectDir = dirname(envPath);
	process.loadEnvFile(envPath);
	const packageExec = getExecCommand(await detectPackageManager(projectDir));
	if (!process.env[ENCRYPTION_KEY_ENV_VAR]) throw new Error([`Password fields found in schema, but encryption key is not set in "${envPath}".`, `  Run ${packageExec} aio-commerce-lib-config encryption setup to generate an encryption key.`].join("\n"));
	validateEncryptionKey(process.env[ENCRYPTION_KEY_ENV_VAR]);
	consola.success("Encryption key is valid");
}
/** Run the encryption validate command */
async function exec() {
	try {
		const packageJson = await findNearestPackageJson();
		await run(join(packageJson ? dirname(packageJson) : process.cwd(), ".env"));
	} catch (error) {
		consola.error(error);
		process.exit(1);
	}
}

//#endregion
//#region source/commands/index.ts
const NAMESPACE = "@adobe/aio-commerce-lib-config";
const [command, subcommand] = process.argv.slice(2);
const USAGE = `
Usage: ${NAMESPACE} <command> [target]

Commands:
  encryption <target>    Encryption related commands
    setup                Ensure an encryption key is configured
    validate             Validate the encryption key

  help                 Show this help message

Examples:
  ${NAMESPACE} validate schema
`;
/**
* Command handlers registry mapping command names to their subcommand handlers
*/
const COMMANDS = { encryption: {
	setup: exec$1,
	validate: exec
} };
/**
* Generic command handler that dispatches to specific command implementations
* @param commandName - The name of the command (for error messages)
* @param target - The target subcommand to execute
* @param handlers - Map of target names to handler functions
*/
async function handleCommand(commandName, target, handlers) {
	if (!target) {
		const availableTargets = Object.keys(handlers).join(", ");
		consola.error(`No ${commandName} target specified`);
		consola.info(`Available targets: ${availableTargets}`);
		process.exit(1);
	}
	const handler = handlers[target];
	if (!handler) {
		const availableTargets = Object.keys(handlers).join(", ");
		consola.error(`Unknown ${commandName} target: ${target}`);
		consola.info(`Available targets: ${availableTargets}`);
		process.exit(1);
	}
	await handler();
}
async function main() {
	if (!command || command === "help" || command === "--help" || command === "-h") {
		consola.log.raw(USAGE);
		process.exit(0);
	}
	try {
		const handlers = COMMANDS[command];
		if (!handlers) {
			consola.error(`Unknown command: ${command}`);
			consola.log.raw(USAGE);
			process.exit(1);
		}
		await handleCommand(command, subcommand, handlers);
	} catch (error) {
		consola.error(stringifyError(error));
		process.exit(1);
	}
}
main();

//#endregion
export {  };
