// @adobe/aio-commerce-lib-webhooks/dist/es/responses/index.mjs
/**
 * @license
 *
 * Copyright 2026 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */

import { ok as ok$1 } from "@adobe/aio-commerce-lib-core/responses";

//#region source/responses/operations/presets.ts
/**
* Creates a success operation response
* The process that triggered the original event continues without any changes.
*
* @example
* ```typescript
* return successOperation();
* ```
*/
const successOperation = () => ({ op: "success" });
/**
* Creates an exception operation response with a message
* Causes Commerce to terminate the process that triggered the original event.
*
* @param message - Exception message
* @param exceptionClass - Optional exception class name
*
* @example
* ```typescript
* return exceptionOperation("The product cannot be added to the cart because it is out of stock");
*
* return exceptionOperation(
*   "Custom error occurred",
*   "Path\\To\\Exception\\Class"
* );
* ```
*/
const exceptionOperation = (message, exceptionClass) => ({
	op: "exception",
	...message && { message },
	...exceptionClass && { type: exceptionClass }
});
/**
* Creates an add operation response
* Causes Commerce to add the provided value to the provided path in the triggered event arguments.
*
* @template TValue - The type of the value to be added
* @param path - Path at which the value should be added
* @param value - Value to be added
* @param instance - Optional DataObject class name
*
* @example
* ```typescript
* return addOperation(
*   "result",
*   { data: { amount: "5", carrier_code: "newshipmethod" } },
*   "Magento\\Quote\\Api\\Data\\ShippingMethodInterface"
* );
* ```
*/
const addOperation = (path, value, instance) => ({
	op: "add",
	path,
	value,
	...instance && { instance }
});
/**
* Creates a replace operation response
* Causes Commerce to replace a value in triggered event arguments for the provided path.
*
* @template TValue - The type of the replacement value
* @param path - Path at which the value should be replaced
* @param value - Replacement value
* @param instance - Optional DataObject class name
*
* @example
* ```typescript
* return replaceOperation("result/shipping_methods/shipping_method_one/amount", 6);
* ```
*/
const replaceOperation = (path, value, instance) => ({
	op: "replace",
	path,
	value,
	...instance && { instance }
});
/**
* Creates a remove operation response
* Causes Commerce to remove a value or node in triggered event arguments by the provided path.
*
* @param path - Path at which the value should be removed
*
* @example
* ```typescript
* return removeOperation("result/key2");
* ```
*/
const removeOperation = (path) => ({
	op: "remove",
	path
});

//#endregion
//#region source/responses/presets.ts
/**
* Creates an HTTP 200 OK response with webhook operation(s)
* Webhook-optimized version of ok() that automatically wraps operations in the response body.
*
* This function shadows the core library's ok() to provide a cleaner API for webhook actions.
* Instead of `ok({ body: operation })`, you can simply use `ok(operation)`.
*
* @template TValue - The type of the value for add/replace operations (defaults to unknown)
* @param operations - Single webhook operation or array of operations
* @returns Success response with operations in body
*
* @example
* ```typescript
* import { ok, successOperation } from "@adobe/aio-commerce-lib-webhooks/responses";
*
* // Single operation
* return ok(successOperation());
*
* // Array of operations
* return ok([
*   addOperation("result", data),
*   removeOperation("result/old_field")
* ]);
* ```
*/
function ok(operations) {
	return ok$1({ body: operations });
}

//#endregion
export { addOperation, exceptionOperation, ok, removeOperation, replaceOperation, successOperation };
