// @adobe/aio-commerce-lib-admin-ui/dist/cjs/index.cjs
/**
 * @license
 *
 * Copyright 2026 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
//#region \0rolldown/runtime.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __exportAll = (all, no_symbols) => {
	let target = {};
	for (var name in all) {
		__defProp(target, name, {
			get: all[name],
			enumerable: true
		});
	}
	if (!no_symbols) {
		__defProp(target, Symbol.toStringTag, { value: "Module" });
	}
	return target;
};
var __copyProps = (to, from, except, desc) => {
	if (from && typeof from === "object" || typeof from === "function") {
		for (var keys = __getOwnPropNames(from), i = 0, n = keys.length, key; i < n; i++) {
			key = keys[i];
			if (!__hasOwnProp.call(to, key) && key !== except) {
				__defProp(to, key, {
					get: ((k) => from[k]).bind(null, key),
					enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
				});
			}
		}
	}
	return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
	value: mod,
	enumerable: true
}) : target, mod));

//#endregion
let _adobe_aio_commerce_lib_api = require("@adobe/aio-commerce-lib-api");
let valibot = require("valibot");
valibot = __toESM(valibot, 1);
let _adobe_aio_commerce_lib_core_error = require("@adobe/aio-commerce-lib-core/error");

//#region ../../packages-private/common-utils/source/valibot/utils.ts
/**
* Parses the input using the provided schema and throws a {@link CommerceSdkValidationError} error if the input is invalid.
* @param schema - The schema to use for parsing.
* @param input - The input to parse.
* @param message - Optional custom error message for the validation error.
*/
function parseOrThrow(schema, input, message) {
	const result = valibot.safeParse(schema, input);
	if (!result.success) throw new _adobe_aio_commerce_lib_core_error.CommerceSdkValidationError(message ?? "Invalid input", { issues: result.issues });
	return result.output;
}

//#endregion
//#region source/api/extensions/schema.ts
/** Parameters for POST /V1/adminuisdk/extension. */
const ExtensionRegistrationParamsSchema = valibot.object({
	extensionName: valibot.pipe(valibot.string(), valibot.minLength(1)),
	extensionTitle: valibot.pipe(valibot.string(), valibot.minLength(1)),
	extensionUrl: valibot.pipe(valibot.string(), valibot.url()),
	extensionWorkspace: valibot.pipe(valibot.string(), valibot.minLength(1))
});
/** Parameters for DELETE /V1/adminuisdk/extension/{workspaceName}/{extensionName}. */
const UnregisterExtensionParamsSchema = valibot.object({
	workspaceName: valibot.pipe(valibot.string(), valibot.minLength(1)),
	extensionName: valibot.pipe(valibot.string(), valibot.minLength(1))
});

//#endregion
//#region source/api/extensions/endpoints.ts
var endpoints_exports = /* @__PURE__ */ __exportAll({
	registerExtension: () => registerExtension,
	unregisterExtension: () => unregisterExtension
});
/**
* Registers an Admin UI extension with Commerce via POST /V1/adminuisdk/extension.
*
* @param httpClient - The {@link AdobeCommerceHttpClient} to use to make the request.
* @param params - The extension registration parameters.
* @param fetchOptions - Optional Ky fetch options.
*
* @throws A `CommerceSdkValidationError` if the parameters are invalid.
* @throws An `HTTPError` if the status code is not 2XX.
*/
async function registerExtension(httpClient, params, fetchOptions) {
	const extension = parseOrThrow(ExtensionRegistrationParamsSchema, params);
	return httpClient.post("adminuisdk/extension", {
		...fetchOptions,
		json: { extension }
	}).json();
}
/**
* Unregisters an Admin UI extension from Commerce via DELETE /V1/adminuisdk/extension/{workspaceName}/{extensionName}.
*
* @param httpClient - The {@link AdobeCommerceHttpClient} to use to make the request.
* @param params - The workspace and extension names.
* @param fetchOptions - Optional Ky fetch options.
*
* @throws A `CommerceSdkValidationError` if the parameters are invalid.
* @throws An `HTTPError` if the status code is not 2XX.
*/
async function unregisterExtension(httpClient, params, fetchOptions) {
	const { workspaceName, extensionName } = parseOrThrow(UnregisterExtensionParamsSchema, params);
	return httpClient.delete(`adminuisdk/extension/${workspaceName}/${extensionName}`, fetchOptions).then((_res) => {});
}

//#endregion
//#region source/lib/api-client.ts
/**
* Creates a new API client for the Admin UI API with all available operations.
*
* @param params - The parameters to build the Commerce HTTP client.
*/
function createAdminUiApiClient(params) {
	return _adobe_aio_commerce_lib_api.ApiClient.create(new _adobe_aio_commerce_lib_api.AdobeCommerceHttpClient(params), { ...endpoints_exports });
}

//#endregion
exports.createAdminUiApiClient = createAdminUiApiClient;
