// @adobe/aio-commerce-lib-app/dist/cjs/actions/config/index.cjs
/**
 * @license
 *
 * Copyright 2026 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
const require_chunk = require('../../chunk-Cozy-1ZF.cjs');
const require_schemas = require('../../schemas-FngDuIeo.cjs');
const require_router = require('../../router-BOR7T6qR.cjs');
let _adobe_aio_commerce_lib_config = require("@adobe/aio-commerce-lib-config");
let valibot = require("valibot");
valibot = require_chunk.__toESM(valibot, 1);
let _adobe_aio_commerce_lib_core_responses = require("@adobe/aio-commerce-lib-core/responses");

//#region source/actions/config/schema.ts
const AcceptedConfigurationValuesSchema = valibot.union([
	valibot.boolean(),
	valibot.string(),
	valibot.array(valibot.string())
]);
/** The set of valid Commerce environments a configuration field can be scoped to. */
const COMMERCE_ENVS = ["paas", "saas"];
/** Query parameters for GET / */
const GetConfigurationQuerySchema = valibot.object({
	scopeId: require_schemas.nonEmptyStringValueSchema("scopeId"),
	commerceEnv: valibot.optional(valibot.picklist(COMMERCE_ENVS))
});
/** Request body for PUT / */
const PutConfigBodySchema = valibot.object({
	scopeId: require_schemas.nonEmptyStringValueSchema("scopeId"),
	config: valibot.array(valibot.object({
		name: require_schemas.nonEmptyStringValueSchema("config[i].name"),
		value: valibot.union([valibot.string(), valibot.array(valibot.string())])
	}))
});
/** Request body for PATCH / */
const PatchConfigBodySchema = valibot.object({
	scopeId: require_schemas.nonEmptyStringValueSchema("scopeId"),
	config: valibot.array(valibot.object({
		name: require_schemas.nonEmptyStringValueSchema("config[i].name"),
		value: valibot.nullable(AcceptedConfigurationValuesSchema)
	}))
});

//#endregion
//#region source/actions/config/router.ts
const MASKED_PASSWORD_VALUE = "*****";
/**
* Filters a business configuration schema to the fields applicable to the
* given Commerce environment. Fields without an `env` property apply to all
* environments and are always included.
*
* @param schema - The business configuration schema to filter.
* @param env - The Commerce environment to filter by.
*/
function filterSchemaByEnv(schema, env) {
	return schema.filter((field) => field.env === void 0 || field.env.includes(env));
}
/**
* Filters password fields from the configuration values.
*
* @param schema - The schema to use to filter the values.
* @param values - The values to filter.
*/
function filterPasswordFields(schema, values) {
	return values.map((item) => {
		if (schema.find((field) => field.name === item.name)?.type === "password") return {
			...item,
			value: MASKED_PASSWORD_VALUE
		};
		return item;
	});
}
/**
* Config action router.
*
* Routes:
* - GET /     Get current configuration values for a given scope
* - POST /    Set configuration (overrides all values for the scope) (deprecated)
* - PATCH /   Partially update configuration (only updates provided fields, allows unsetting)
*/
const router = new require_router.HttpActionRouter().use(require_router.logger({ name: () => "config" }));
/** GET / - Retrieve configuration */
router.get("/", {
	query: GetConfigurationQuerySchema,
	handler: async (req, ctx) => {
		const { logger, rawParams } = ctx;
		const { configSchema: rawConfigSchema } = rawParams;
		const env = req.query.commerceEnv;
		const envFilteredSchema = env ? filterSchemaByEnv(rawConfigSchema, env) : rawConfigSchema;
		logger.debug("Initializing configuration");
		const { configSchema } = await (0, _adobe_aio_commerce_lib_config.initialize)({
			schema: envFilteredSchema,
			params: rawParams
		});
		const { scopeId } = req.query;
		logger.debug(`Retrieving configuration with scope id: ${scopeId}`);
		const appConfiguration = await (0, _adobe_aio_commerce_lib_config.getConfiguration)((0, _adobe_aio_commerce_lib_config.byScopeId)(scopeId), { encryptionKey: rawParams.AIO_COMMERCE_CONFIG_ENCRYPTION_KEY });
		logger.debug("Masking password values...");
		appConfiguration.config = filterPasswordFields(configSchema, appConfiguration.config);
		return (0, _adobe_aio_commerce_lib_core_responses.ok)({ body: {
			schema: configSchema,
			values: appConfiguration
		} });
	}
});
/**
* PUT / - Set configuration (deprecated)
*
* @deprecated Use PATCH instead. This endpoint overwrites all values for the scope
* and does not support partial updates or unset semantics.
*/
router.put("/", {
	body: PutConfigBodySchema,
	handler: async (req, ctx) => {
		const { logger, rawParams } = ctx;
		logger.debug(`Setting configuration with scope id: ${req.body.scopeId}`);
		const { scopeId, config } = req.body;
		const { configSchema } = await (0, _adobe_aio_commerce_lib_config.initialize)({
			schema: rawParams.configSchema,
			params: rawParams
		});
		const result = await (0, _adobe_aio_commerce_lib_config.setConfiguration)({ config: config.filter((item) => item.value !== MASKED_PASSWORD_VALUE) }, (0, _adobe_aio_commerce_lib_config.byScopeId)(scopeId), { encryptionKey: rawParams.AIO_COMMERCE_CONFIG_ENCRYPTION_KEY });
		result.config = filterPasswordFields(configSchema, result.config);
		return (0, _adobe_aio_commerce_lib_core_responses.ok)({
			body: result,
			headers: {
				"Cache-Control": "no-store",
				Deprecation: "Wed, 15 Apr 2026 00:00:00 GMT"
			}
		});
	}
});
/** PATCH / - Partially update configuration */
router.patch("/", {
	body: PatchConfigBodySchema,
	handler: async (req, ctx) => {
		const { logger, rawParams } = ctx;
		logger.debug(`Patching configuration with scope id: ${req.body.scopeId}`);
		const { scopeId, config } = req.body;
		const { configSchema } = await (0, _adobe_aio_commerce_lib_config.initialize)({
			schema: rawParams.configSchema,
			params: rawParams
		});
		const result = await (0, _adobe_aio_commerce_lib_config.setConfiguration)({ config }, (0, _adobe_aio_commerce_lib_config.byScopeId)(scopeId), { encryptionKey: rawParams.AIO_COMMERCE_CONFIG_ENCRYPTION_KEY });
		result.config = filterPasswordFields(configSchema, result.config);
		return (0, _adobe_aio_commerce_lib_core_responses.ok)({
			body: result,
			headers: { "Cache-Control": "no-store" }
		});
	}
});

//#endregion
//#region source/actions/config/index.ts
/**
* Factory to create the route handler for the `config` action.
* @param args - The arguments required to create the runtime action.
*/
const configRuntimeAction = (args) => async (params) => {
	const handler = router.handler();
	const { configSchema } = args;
	return await handler({
		...params,
		configSchema
	});
};

//#endregion
exports.configRuntimeAction = configRuntimeAction;
