// @adobe/aio-commerce-lib-webhooks/dist/es/api/index.mjs
/**
 * @license
 *
 * Copyright 2026 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */

import { AdobeCommerceHttpClient, ApiClient } from "@adobe/aio-commerce-lib-api";
import * as v from "valibot";

//#region source/api/webhooks/schema.ts
function nonEmptyString(fieldName) {
	return v.pipe(v.string(`Expected a string for ${fieldName}`), v.nonEmpty(`${fieldName} must not be empty`));
}
function optionalString(fieldName) {
	return v.optional(v.string(`Expected a string for ${fieldName}`));
}
/** Schema for a webhook field mapping ({ name, source? }). */
const WebhookFieldSchema = v.object({
	name: nonEmptyString("field name"),
	source: optionalString("field source")
});
/** Schema for a conditional webhook rule ({ field, operator, value }). */
const WebhookRuleSchema = v.object({
	field: nonEmptyString("rule field"),
	operator: nonEmptyString("rule operator"),
	value: v.string("Expected a string for rule value")
});
/** Schema for a custom HTTP header ({ name, value }). */
const WebhookHeaderSchema = v.object({
	name: nonEmptyString("header name"),
	value: nonEmptyString("header value")
});
/** Schema for Developer Console OAuth credentials. */
const DeveloperConsoleOAuthSchema = v.object({
	client_id: nonEmptyString("client_id"),
	client_secret: nonEmptyString("client_secret"),
	org_id: nonEmptyString("org_id"),
	environment: optionalString("environment")
});
/**
* Schema for the webhook payload sent to POST /webhooks/subscribe.
* Matches the `webhook` property of the request body.
* Required fields per the Commerce API: webhook_method, webhook_type, batch_name, hook_name, url.
*/
const WebhookSubscribeParamsSchema = v.object({
	webhook_method: nonEmptyString("webhook_method"),
	webhook_type: nonEmptyString("webhook_type"),
	batch_name: nonEmptyString("batch_name"),
	batch_order: v.optional(v.number()),
	hook_name: nonEmptyString("hook_name"),
	url: nonEmptyString("url"),
	priority: v.optional(v.number()),
	required: v.optional(v.boolean()),
	soft_timeout: v.optional(v.number()),
	timeout: v.optional(v.number()),
	method: optionalString("method"),
	fallback_error_message: optionalString("fallback_error_message"),
	ttl: v.optional(v.number()),
	fields: v.optional(v.array(WebhookFieldSchema, "Expected an array of field objects")),
	rules: v.optional(v.array(WebhookRuleSchema, "Expected an array of rule objects")),
	headers: v.optional(v.array(WebhookHeaderSchema, "Expected an array of header objects")),
	developer_console_oauth: v.optional(DeveloperConsoleOAuthSchema)
});
/**
* Schema for the parameters sent to POST /webhooks/unsubscribe.
* Required: webhook_method, webhook_type, batch_name, hook_name.
*/
const WebhookUnsubscribeParamsSchema = v.object({
	webhook_method: nonEmptyString("webhook_method"),
	webhook_type: nonEmptyString("webhook_type"),
	batch_name: nonEmptyString("batch_name"),
	hook_name: nonEmptyString("hook_name")
});

//#endregion
//#region source/api/webhooks/endpoints.ts
/**
* Returns a list of all subscribed webhooks in the Commerce instance.
* @see https://developer.adobe.com/commerce/extensibility/webhooks/api/#get-a-list-of-all-subscribed-webhooks
*
* @param httpClient - The {@link AdobeCommerceHttpClient} to use to make the request.
* @param fetchOptions - The {@link Options} to use to make the request.
*/
function getWebhookList(httpClient, fetchOptions) {
	return httpClient.get("webhooks/list", fetchOptions).json();
}
/**
* Subscribes a webhook in the Commerce instance.
* @see https://developer.adobe.com/commerce/extensibility/webhooks/api/#subscribe-a-webhook
*
* @param httpClient - The {@link AdobeCommerceHttpClient} to use to make the request.
* @param params - The webhook payload (webhook_method, webhook_type, batch_name, hook_name, url, etc.).
* @param fetchOptions - The {@link Options} to use to make the request.
*
* @throws A {@link CommerceSdkValidationError} If the parameters are in the wrong format.
* @throws An {@link HTTPError} If the status code is not 2XX.
*/
function subscribeWebhook(httpClient, params, fetchOptions) {
	const validatedParams = v.parse(WebhookSubscribeParamsSchema, params);
	return httpClient.post("webhooks/subscribe", {
		...fetchOptions,
		json: { webhook: validatedParams }
	}).json();
}
/**
* Unsubscribes a webhook from the Commerce instance.
* @see https://developer.adobe.com/commerce/extensibility/webhooks/api/#unsubscribe-a-webhook
*
* @param httpClient - The {@link AdobeCommerceHttpClient} to use to make the request.
* @param params - The webhook identifiers (webhook_method, webhook_type, batch_name, hook_name).
* @param fetchOptions - The {@link Options} to use to make the request.
*
* @throws A {@link CommerceSdkValidationError} If the parameters are in the wrong format.
* @throws An {@link HTTPError} If the status code is not 2XX.
*/
function unsubscribeWebhook(httpClient, params, fetchOptions) {
	const validatedParams = v.parse(WebhookUnsubscribeParamsSchema, params);
	return httpClient.post("webhooks/unsubscribe", {
		...fetchOptions,
		json: { webhook: validatedParams }
	}).json();
}
/**
* Returns the list of webhooks supported in Adobe Commerce as a Cloud Service (SaaS only).
* @see https://developer.adobe.com/commerce/extensibility/webhooks/api/#get-supported-webhooks-for-saas
*
* @param httpClient - The {@link AdobeCommerceHttpClient} to use to make the request.
* @param fetchOptions - The {@link Options} to use to make the request.
*/
function getSupportedWebhookList(httpClient, fetchOptions) {
	return httpClient.get("webhooks/supportedList", fetchOptions).json();
}

//#endregion
//#region source/lib/api-client.ts
/**
* Creates a new API client for the Commerce Webhooks API.
* @param params - The parameters to build the Commerce HTTP client.
*/
function createCommerceWebhooksApiClient(params) {
	return ApiClient.create(new AdobeCommerceHttpClient(params), {
		getWebhookList,
		subscribeWebhook,
		unsubscribeWebhook,
		getSupportedWebhookList
	});
}
/**
* Creates a customized Commerce Webhooks API client with a user-specified set of endpoint functions.
* @param params - The parameters to build the Commerce HTTP client.
* @param functions - The API functions to include in the client.
*/
function createCustomCommerceWebhooksApiClient(params, functions) {
	return ApiClient.create(new AdobeCommerceHttpClient(params), functions);
}

//#endregion
export { createCommerceWebhooksApiClient, createCustomCommerceWebhooksApiClient, getSupportedWebhookList, getWebhookList, subscribeWebhook, unsubscribeWebhook };
