// @adobe/aio-commerce-lib-api/dist/cjs/utils/index.cjs
/**
 * @license
 *
 * Copyright 2026 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
const require_hooks = require('../hooks-CZhnmKy0.cjs');
let ky = require("ky");

//#region source/utils/http/codes.ts
/**
* The HTTP status code for a successful request.
* @see https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/200
*/
const HTTP_OK = 200;
/**
* The HTTP status code for a bad request.
* @see https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/400
*/
const HTTP_BAD_REQUEST = 400;
/**
* The HTTP status code for an unauthorized request.
* @see https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/401
*/
const HTTP_UNAUTHORIZED = 401;
/**
* The HTTP status code for a forbidden request.
* @see https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/403
*/
const HTTP_FORBIDDEN = 403;
/**
* The HTTP status code for a not found request.
* @see https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/404
*/
const HTTP_NOT_FOUND = 404;
/**
* The HTTP status code for an internal server error.
* @see https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/500
*/
const HTTP_INTERNAL_SERVER_ERROR = 500;

//#endregion
//#region source/utils/http/error.ts
/**
* Unwraps a ky `HTTPError` to produce a human-readable string that includes the
* HTTP status and the message extracted from the response body.
*
* If `error` is not an `HTTPError`, falls back to `error.message` for `Error`
* instances or `String(error)` for anything else.
*
* Tries the following shapes from the response JSON body, in order:
* - `body.message` (string) — if `body.parameters` is a non-empty array, `%1`/`%2`/... placeholders are replaced with the corresponding parameter values
* - `body.error` (string)
* - `body.error.message` (nested object)
* - `body.errors[0].message` (array)
* - raw text body
*
* @example
* // Commerce returns { "message": "Provider already exists" } with status 400
* await unwrapHttpError(err)
* // → "HTTP 400 Bad Request — Provider already exists"
*/
async function unwrapHttpError(error) {
	if (!(error instanceof ky.HTTPError)) return error instanceof Error ? error.message : String(error);
	const { response } = error;
	const statusPrefix = `HTTP ${response.status} ${response.statusText}`;
	let bodyText;
	try {
		bodyText = await response.clone().text();
	} catch {
		return statusPrefix;
	}
	if (!bodyText) return statusPrefix;
	try {
		const message = extractJsonMessage(JSON.parse(bodyText));
		if (message) return `${statusPrefix} — ${message}`;
	} catch {}
	return `${statusPrefix} — ${bodyText}`;
}
function interpolateParameters(message, parameters) {
	return parameters.reduce((result, param, index) => result.replaceAll(`%${index + 1}`, String(param)), message);
}
function extractJsonMessage(json) {
	if (typeof json !== "object" || json === null) return;
	const obj = json;
	if (typeof obj.message === "string") {
		if (Array.isArray(obj.parameters) && obj.parameters.length > 0) return interpolateParameters(obj.message, obj.parameters);
		return obj.message;
	}
	if (typeof obj.error === "string") return obj.error;
	if (typeof obj.error === "object" && obj.error !== null) {
		const nested = obj.error;
		if (typeof nested.message === "string") return nested.message;
	}
	if (Array.isArray(obj.errors) && obj.errors.length > 0) {
		const first = obj.errors[0];
		if (typeof first.message === "string") return first.message;
	}
}

//#endregion
exports.HTTP_BAD_REQUEST = HTTP_BAD_REQUEST;
exports.HTTP_FORBIDDEN = HTTP_FORBIDDEN;
exports.HTTP_INTERNAL_SERVER_ERROR = HTTP_INTERNAL_SERVER_ERROR;
exports.HTTP_NOT_FOUND = HTTP_NOT_FOUND;
exports.HTTP_OK = HTTP_OK;
exports.HTTP_UNAUTHORIZED = HTTP_UNAUTHORIZED;
exports.buildImsAuthBeforeRequestHook = require_hooks.buildImsAuthBeforeRequestHook;
exports.buildIntegrationAuthBeforeRequestHook = require_hooks.buildIntegrationAuthBeforeRequestHook;
exports.isAuthProvider = require_hooks.isAuthProvider;
exports.unwrapHttpError = unwrapHttpError;
