// @adobe/aio-commerce-lib-config/dist/es/index.mjs
/**
 * @license
 *
 * Copyright 2026 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */

import { i as validateEncryptionKey, n as encrypt, r as generateEncryptionKey, t as decrypt } from "./encryption-CSTVRIte.mjs";
import * as v from "valibot";
import { CommerceSdkValidationError } from "@adobe/aio-commerce-lib-core/error";
import { init } from "@adobe/aio-lib-files";
import { init as init$1 } from "@adobe/aio-lib-state";
import stringify, { stringify as stringify$1 } from "safe-stable-stringify";
import AioLogger from "@adobe/aio-lib-core-logging";
import { v4 } from "uuid";
import { AdobeCommerceHttpClient } from "@adobe/aio-commerce-lib-api";

//#region source/modules/schema/config-schema-repository.ts
let __globalSchema = null;
/**
* Sets the global schema in memory.
*
* @param schema - Schema to store in memory.
*/
function setGlobalSchema(schema) {
	__globalSchema = schema;
}
/**
* Gets the global schema from memory and throws an error if not initialized.
*
* @returns The schema stored in memory.
* @throws Error if schema has not been initialized.
*/
function requireGlobalSchema() {
	const schema = __globalSchema;
	if (!schema) throw new Error("Schema not initialized. Call `initialize({ schema })` before using configuration functions.");
	return schema;
}

//#endregion
//#region ../../packages-private/common-utils/source/valibot/utils.ts
/**
* Parses the input using the provided schema and throws a {@link CommerceSdkValidationError} error if the input is invalid.
* @param schema - The schema to use for parsing.
* @param input - The input to parse.
* @param message - Optional custom error message for the validation error.
*/
function parseOrThrow(schema, input, message) {
	const result = v.safeParse(schema, input);
	if (!result.success) throw new CommerceSdkValidationError(message ?? "Invalid input", { issues: result.issues });
	return result.output;
}

//#endregion
//#region source/modules/schema/fields.ts
const MARKDOWN_LINK_REGEX = /\[([^\]]*)\]\(((?:[^)(]|\([^)]*\))*)\)/g;
const SAFE_URL_REGEX = /^https?:\/\//i;
/**
* Returns true if every Markdown link in `text` is well-formed and uses an http(s) URL.
* Plain text with no links always passes.
*/
function validateMarkdownLink(text) {
	for (const match of text.matchAll(MARKDOWN_LINK_REGEX)) if (!SAFE_URL_REGEX.test(match[2].trim())) return false;
	return true;
}
const DEFAULT_BOOLEAN_VALUE = false;
const DEFAULT_STRING_VALUE = "";
const DEFAULT_MULTIPLE_LIST_VALUE = [];
/** The list of supported Commerce environments a configuration field can be scoped to. */
const COMMERCE_ENVS = ["paas", "saas"];
/** Schema for a single Commerce environment a configuration field can be scoped to. */
const CommerceEnvSchema = v.picklist(COMMERCE_ENVS, `Expected one of: ${COMMERCE_ENVS.map((e) => `"${e}"`).join(", ")}`);
/**
* Schema for the optional `env` property used to scope a configuration field to
* specific Commerce environments. When omitted, the field applies to all environments.
*/
const EnvSchema = v.pipe(v.array(CommerceEnvSchema, "Expected an array of commerce environments for the field \"env\""), v.nonEmpty("The \"env\" array must contain at least one commerce environment"));
/**
* Base schema for configuration field options with name, optional label,
* optional description, and optional `env` to scope the field to specific
* Commerce environments (when omitted, the field applies to all environments).
*/
const BaseOptionSchema = v.object({
	name: v.pipe(v.string("Expected a string for the field name"), v.nonEmpty("The field name must not be empty")),
	label: v.optional(v.string("Expected a string for the field label")),
	description: v.optional(v.pipe(v.string("Expected a string for the field description"), v.check(validateMarkdownLink, "Field description contains an invalid markdown URL."))),
	env: v.optional(EnvSchema)
});
/** Schema for a single option in a list field, containing a display label and a value */
const ListOptionSchema = v.object({
	label: v.string("Expected a string for the option label"),
	value: v.string("Expected a string for the option value")
});
/** Entries shared between the single- and multiple-selection list field schemas. */
const ListEntriesCommon = {
	...BaseOptionSchema.entries,
	type: v.literal("list", "Expected the type to be 'list'"),
	options: v.array(ListOptionSchema, "Expected an array of list options")
};
/** Schema for a list field that allows single selection from a list of options */
const SingleListSchema = v.object({
	...ListEntriesCommon,
	selectionMode: v.literal("single", "Expected the selectionMode to be 'single'"),
	default: v.pipe(v.string("Expected a string for the default value"), v.nonEmpty("The default value must not be empty"))
});
/** Schema for a list field that allows multiple selections from a list of options */
const MultipleListSchema = v.object({
	...ListEntriesCommon,
	selectionMode: v.literal("multiple", "Expected the selectionMode to be 'multiple'"),
	default: v.optional(v.array(v.pipe(v.string("Expected a string for each default value"), v.nonEmpty("Each default value must not be empty")), "Expected an array of default values"), DEFAULT_MULTIPLE_LIST_VALUE)
});
/** Schema for list fields supporting either single or multiple selection modes */
const ListSchema = v.variant("selectionMode", [SingleListSchema, MultipleListSchema]);
/** Schema for a text input field that accepts string values */
const TextSchema = v.object({
	...BaseOptionSchema.entries,
	type: v.literal("text", "Expected the type to be 'text'"),
	default: v.optional(v.string("Expected a string for the default value"), DEFAULT_STRING_VALUE)
});
/** Schema for a password input field that accepts string values (typically masked in UI) */
const PasswordSchema = v.object({
	...BaseOptionSchema.entries,
	type: v.literal("password", "Expected the type to be 'password'"),
	default: v.optional(v.literal(DEFAULT_STRING_VALUE, "Password fields do not have a default value"), DEFAULT_STRING_VALUE)
});
/** Schema for an email input field that accepts and validates email addresses */
const EmailSchema = v.object({
	...BaseOptionSchema.entries,
	type: v.literal("email", "Expected the type to be 'email'"),
	default: v.optional(v.union([v.literal(DEFAULT_STRING_VALUE), v.pipe(v.string("Expected a string for the default email value"), v.email("The email must be a valid email address"))]), DEFAULT_STRING_VALUE)
});
/** Schema for a URL input field that accepts and validates URL strings */
const UrlSchema = v.object({
	...BaseOptionSchema.entries,
	type: v.literal("url", "Expected the type to be 'url'"),
	default: v.optional(v.union([v.literal(DEFAULT_STRING_VALUE), v.pipe(v.string("Expected a string for the default URL value"), v.url("The URL must be a valid URL"))]), DEFAULT_STRING_VALUE)
});
/** Schema for a phone number input field that accepts and validates telephone numbers */
const PhoneSchema = v.object({
	...BaseOptionSchema.entries,
	type: v.literal("tel", "Expected the type to be 'tel'"),
	default: v.optional(v.union([v.literal(DEFAULT_STRING_VALUE), v.pipe(v.string("Expected a string for the default phone number value"), v.regex(/^\+?[0-9\s\-()]+$/, "The phone number must contain only numbers and/or country codes"))]), DEFAULT_STRING_VALUE)
});
/** Schema for a boolean toggle field that accepts true/false values */
const BooleanSchema = v.object({
	...BaseOptionSchema.entries,
	type: v.literal("boolean", "Expected the type to be 'boolean'"),
	default: v.optional(v.boolean("Expected a boolean for the default value"), DEFAULT_BOOLEAN_VALUE)
});
/** Entries shared between the single- and multiple-selection dynamic list field schemas. */
const DynamicListEntriesCommon = {
	...BaseOptionSchema.entries,
	type: v.literal("dynamicList", "Expected the type to be 'dynamicList'"),
	options: v.custom((input) => typeof input === "function", "Expected a function for \"options\"")
};
/** Schema for a dynamic list field that allows single selection. */
const SingleDynamicListSchema = v.object({
	...DynamicListEntriesCommon,
	selectionMode: v.literal("single", "Expected the selectionMode to be 'single'"),
	default: v.custom((input) => typeof input === "function", "Expected a function for \"default\"")
});
/** Schema for a dynamic list field that allows multiple selections. */
const MultipleDynamicListSchema = v.object({
	...DynamicListEntriesCommon,
	selectionMode: v.literal("multiple", "Expected the selectionMode to be 'multiple'"),
	default: v.optional(v.custom((input) => typeof input === "function", "Expected a function for \"default\""))
});
/** Schema for dynamic list fields supporting either single or multiple selection modes. */
const DynamicListSchema = v.variant("selectionMode", [SingleDynamicListSchema, MultipleDynamicListSchema]);
/** Schema for a single configuration field. */
const FieldSchema = v.variant("type", [
	ListSchema,
	DynamicListSchema,
	TextSchema,
	PasswordSchema,
	EmailSchema,
	UrlSchema,
	PhoneSchema,
	BooleanSchema
]);
/** Schema for the business configuration schema. */
const SchemaBusinessConfigSchema = v.pipe(v.array(FieldSchema, "Expected an array of configuration fields"), v.minLength(1, "At least one configuration parameter is required"));

//#endregion
//#region source/modules/schema/utils.ts
/**
* Validates a business configuration schema.
*
* @param value - The business configuration schema to validate.
* @returns Validated schema (may contain `dynamicList` fields).
*
* @throws {CommerceSdkValidationError} If the schema is invalid.
*/
function validateBusinessConfigSchema(value) {
	return parseOrThrow(SchemaBusinessConfigSchema, value);
}
/**
* Gets password field names from the schema.
*
* @param schema - The resolved schema to read.
* @returns Set of field names that are of type "password".
*/
function getPasswordFields(schema) {
	return new Set(schema.filter((field) => field.type === "password").map((field) => field.name));
}
/**
* Whether a business configuration schema contains any `dynamicList` fields
* that need runtime resolution before use.
*
* @param schema - The schema to inspect.
*/
function hasDynamicSchema(schema) {
	return schema.some((field) => field.type === "dynamicList");
}
async function resolveDynamicListField(field, params) {
	let options;
	try {
		options = await field.options(params);
	} catch (error) {
		throw new Error(`Failed to resolve options for dynamicList field "${field.name}"`, { cause: error });
	}
	return parseOrThrow(ListSchema, {
		name: field.name,
		label: field.label,
		description: field.description,
		env: field.env,
		type: "list",
		selectionMode: field.selectionMode,
		options,
		default: field.default?.(options)
	}, `Invalid resolved value for dynamicList field "${field.name}"`);
}
/**
* Resolves any dynamic parts of a business configuration schema into a static
* one suitable for validation, storage, and rendering.
*
* @param schema - Schema as the developer defined it; may contain `dynamicList` fields.
* @param params - Runtime action params forwarded to each `dynamicList` factory.
* @returns A new, fully resolved schema. The input is not mutated.
*
* @throws {CommerceSdkValidationError} If a factory returns data that do not
* match the expected schema shape.
*
* @example
* ```ts
* const schema: BusinessConfigSchema = [
*   {
*     name: "paymentMethod",
*     type: "dynamicList",
*     selectionMode: "single",
*     options: async (params) => fetchPaymentMethods(params.SOME_API_KEY),
*     default: (opts) => opts[0].value,
*   },
* ];
*
* const resolved = await resolveBusinessConfigSchema(schema, params);
* ```
*/
async function resolveBusinessConfigSchema(schema, params) {
	return Promise.all(schema.map((field) => field.type === "dynamicList" ? resolveDynamicListField(field, params) : field));
}

//#endregion
//#region source/utils/constants.ts
/**
* Default namespace for configuration storage
* Fixed namespace for simplicity - 1 app : 1 commerce relationship
*/
const DEFAULT_NAMESPACE = "aio-commerce-config";
/** Default custom scope level. */
const DEFAULT_CUSTOM_SCOPE_LEVEL = "base";

//#endregion
//#region source/utils/repository.ts
let __sharedState = null;
let __sharedFiles = null;
let __globalStateOptions = null;
function getOptimalRegion(owRegion) {
	const regionMapping = {
		"us-east-1": "amer",
		"eu-west-1": "emea",
		"ap-northeast-1": "apac",
		"ap-southeast-2": "aus"
	};
	return owRegion && regionMapping[owRegion] ? regionMapping[owRegion] : regionMapping["us-east-1"];
}
/**
* Set global state options for the library. This should be called before any repository functions to ensure the options are applied.
* @param options - The options to configure the Adobe State library.
*/
function setGlobalStateOptions(options) {
	__globalStateOptions = options;
}
/**
* Get the shared state instance (lazy initialization)
* @returns Promise resolving to the shared AdobeState instance
*/
async function getSharedState() {
	if (!__sharedState) {
		const initRegion = __globalStateOptions?.region ?? "auto";
		__sharedState = await init$1({ region: initRegion === "auto" ? getOptimalRegion(process.env.__OW_REGION) : initRegion });
	}
	return __sharedState;
}
/**
* Get the shared files instance (lazy initialization)
* @returns Promise resolving to the shared Files instance
*/
async function getSharedFiles() {
	if (!__sharedFiles) __sharedFiles = await init();
	return __sharedFiles;
}

//#endregion
//#region source/config-utils.ts
/**
* Checks if the given value is a non-empty string.
* @param value The value to check.
* @returns True if the value is a non-empty string, false otherwise.
*/
function isNonEmptyString(value) {
	return typeof value === "string" && value.trim().length > 0;
}
/**
* Finds a scope node by its ID.
* @param tree - The scope tree to search.
* @param id - The ID of the scope node to find.
* @returns The found scope node or null if not found.
*/
function findScopeNodeById(tree, id) {
	const traverse = (node) => {
		if (node.id === id) return node;
		if (node.children) for (const child of node.children) {
			const found = traverse(child);
			if (found) return found;
		}
		return null;
	};
	for (const root of tree) {
		const found = traverse(root);
		if (found) return found;
	}
	return null;
}
/**
* Finds a scope node by its Commerce API ID and level.
*
* `commerce_id` is unique only within a single level (e.g., a website with id=1
* and a store with id=1 can coexist), so the level is required to disambiguate.
*
* @param tree - The scope tree to search.
* @param commerceScopeId - The Commerce API ID of the scope node to find.
* @param level - The level of the scope node to find.
* @returns The found scope node or null if not found.
*/
function findScopeNodeByCommerceScopeId(tree, commerceScopeId, level) {
	const traverse = (node) => {
		if (node.commerce_id === commerceScopeId && node.level === level) return node;
		if (node.children) for (const child of node.children) {
			const found = traverse(child);
			if (found) return found;
		}
		return null;
	};
	for (const root of tree) {
		const found = traverse(root);
		if (found) return found;
	}
	return null;
}
/**
* Derives the scope information from the provided arguments.
* @param code - The code of the scope to find.
* @param level - The level of the scope to find.
* @param tree - The scope tree to search.
* @returns The derived scope information including code, level, id, and path.
*/
function deriveScopeFromCodeAndLevel(code, level, tree) {
	if (!(isNonEmptyString(code) && isNonEmptyString(level))) throw new Error("INVALID_ARGS: expected (code: string, level: string)");
	const trimmedCode = code.trim();
	const trimmedLevel = level.trim();
	const path = findScopePath(tree, trimmedCode, trimmedLevel);
	const node = path.find((scopeNode) => scopeNode.code === trimmedCode && scopeNode.level === trimmedLevel);
	if (path.length === 0 || !node) throw new Error(`INVALID_SCOPE: Unknown scope code='${trimmedCode}' level='${trimmedLevel}'`);
	return {
		scopeCode: node.code,
		scopeLevel: node.level,
		scopeId: node.id,
		scopePath: path
	};
}
/**
* Derives the scope information from a Commerce API ID and level.
* @param commerceScopeId - The Commerce API id of the scope to find.
* @param level - The level of the scope to find.
* @param tree - The scope tree to search.
* @returns The derived scope information including code, level, id, and path.
*/
function deriveScopeFromCommerceScopeId(commerceScopeId, level, tree) {
	if (!(typeof commerceScopeId === "number" && Number.isFinite(commerceScopeId))) throw new Error("INVALID_ARGS: expected (commerceScopeId: number, level: string)");
	if (!isNonEmptyString(level)) throw new Error("INVALID_ARGS: expected (commerceScopeId: number, level: string)");
	const trimmedLevel = level.trim();
	const node = findScopeNodeByCommerceScopeId(tree, commerceScopeId, trimmedLevel);
	if (!node) throw new Error(`INVALID_SCOPE: Unknown scope commerceScopeId='${commerceScopeId}' level='${trimmedLevel}'`);
	const path = findScopePath(tree, node.code, node.level);
	return {
		scopeCode: node.code,
		scopeLevel: node.level,
		scopeId: node.id,
		scopePath: path
	};
}
/**
* Derives the scope information from the provided id.
* @param id - The id of the scope to find.
* @param tree - The scope tree to search for the node.
* @returns The derived scope information including code, level, id, and path.
*/
function deriveScopeFromId(id, tree) {
	if (!isNonEmptyString(id)) throw new Error("INVALID_ARGS: expected (id: string)");
	const trimmedId = id.trim();
	const node = findScopeNodeById(tree, trimmedId);
	if (!node) throw new Error(`INVALID_SCOPE_ID: No scope found for id '${trimmedId}'`);
	const path = findScopePath(tree, node.code, node.level);
	return {
		scopeCode: node.code,
		scopeLevel: node.level,
		scopeId: node.id,
		scopePath: path
	};
}
/**
* Derives scope from code with optional level (applies default if missing)
* @param code - The code of the scope to find.
* @param level - The level of the scope to find (optional, defaults to base level).
* @param tree - The scope tree to search.
* @returns The derived scope information including code, level, id, and path.
*/
function deriveScopeFromCodeWithOptionalLevel(code, level, tree) {
	if (!isNonEmptyString(code)) throw new Error("INVALID_ARGS: expected (code: string)");
	if (level !== void 0 && !isNonEmptyString(level)) throw new Error("INVALID_ARGS: expected (level: string)");
	return deriveScopeFromCodeAndLevel(code, isNonEmptyString(level) ? level.trim() : DEFAULT_CUSTOM_SCOPE_LEVEL, tree);
}
/**
* Derives the scope information from the provided arguments.
* @param args - The arguments containing either (id), (code), or (code, level).
* @param tree - The scope tree to search for the node.
* @returns The derived scope information including code, level, id, and path.
*/
function deriveScopeFromArgs(args, tree) {
	if (args.length === 2) {
		if (typeof args[0] === "number") return deriveScopeFromCommerceScopeId(args[0], args[1], tree);
		return deriveScopeFromCodeAndLevel(args[0], args[1], tree);
	}
	if (args.length === 1) {
		const arg = args[0];
		try {
			return deriveScopeFromId(arg, tree);
		} catch {
			return deriveScopeFromCodeWithOptionalLevel(arg, void 0, tree);
		}
	}
	throw new Error("INVALID_ARGS: expected (id), (code), (code, level), or (commerceScopeId, level)");
}
/**
* Finds a scope path in the tree.
* @param tree - The scope tree to search.
* @param code - The code of the scope to find.
* @param level - The level of the scope to find.
* @returns The found scope path or an empty array if not found.
*/
function findScopePath(tree, code, level) {
	const path = [];
	const traverse = (node, parents) => {
		const extendedPath = parents.concat(node);
		if (node.code === code && node.level === level) {
			path.push(...extendedPath);
			return true;
		}
		if (node.children) {
			for (const child of node.children) if (traverse(child, extendedPath)) return true;
		}
		return false;
	};
	for (const root of tree) if (traverse(root, [])) return path.reverse();
	return [];
}
/**
* Sanitize request config entries to contain only valid name/value pairs.
* A `null` value is allowed and signals that the field should be unset
* (removing the explicit override at the current scope to restore inheritance).
* @param entries - The entries to sanitize.
* @returns The sanitized entries.
*/
function sanitizeRequestEntries(entries) {
	return (Array.isArray(entries) ? entries : []).filter((entry) => {
		if (!entry || typeof entry.name !== "string") return false;
		if (entry.value === null) return true;
		const hasValidValue = ["string", "boolean"].includes(typeof entry.value) || Array.isArray(entry.value) && entry.value.every((item) => typeof item === "string");
		return entry.name.trim().length > 0 && hasValidValue;
	}).map((entry) => ({
		name: String(entry.name).trim(),
		value: entry.value
	}));
}
/**
* Merge existing and requested entries, setting origin to the current scope for requested entries.
* A `null` value in a requested entry removes that field from the persisted config,
* restoring inheritance from the parent scope.
* @param existingScopeEntries - The existing scope entries.
* @param requestedScopeEntries - The requested scope entries.
* @param scopeCode - The code of the scope.
* @param scopeLevel - The level of the scope.
* @returns The merged scope entries.
*/
function mergeScopes(existingScopeEntries, requestedScopeEntries, scopeCode, scopeLevel) {
	const mergedMap = /* @__PURE__ */ new Map();
	for (const existingEntry of existingScopeEntries) mergedMap.set(existingEntry.name, {
		name: existingEntry.name,
		value: existingEntry.value,
		origin: existingEntry.origin || {
			code: scopeCode,
			level: scopeLevel
		}
	});
	for (const requestedEntry of requestedScopeEntries) if (requestedEntry.value === null) mergedMap.delete(requestedEntry.name);
	else mergedMap.set(requestedEntry.name, {
		name: requestedEntry.name,
		value: requestedEntry.value,
		origin: {
			code: scopeCode,
			level: scopeLevel
		}
	});
	return Array.from(mergedMap.values()).map((data) => ({
		name: data.name,
		value: data.value,
		origin: data.origin
	}));
}
/**
* Build default config entries from schema defaults
* @param schema - The schema to build the defaults from.
* @returns The default config entries.
*/
function getSchemaDefaults(schema) {
	return { config: schema.filter((field) => field.default !== void 0).map((field) => ({
		name: field.name,
		value: field.default,
		origin: {
			code: "default",
			level: "system"
		}
	})) };
}
/**
* Merges config entries into the target map if not already present
* @param merged - The target map to merge into
* @param entries - The config entries to merge
* @param origin - The origin information for the entries
*/
function mergeConfigEntries(merged, entries, origin) {
	for (const entry of entries) if (!merged.has(entry.name)) merged.set(entry.name, {
		name: entry.name,
		value: entry.value,
		origin
	});
}
/**
* Loads and merges config from the inheritance path
* @param merged - The target map to merge into
* @param scopePath - The path of scope nodes to load from
* @param loadScopeConfigFn - Function to load config for a scope
*/
async function mergeConfigFromPath(merged, scopePath, loadScopeConfigFn) {
	for (const node of scopePath) {
		const persisted = await loadScopeConfigFn(node.code);
		if (persisted?.config && Array.isArray(persisted.config)) mergeConfigEntries(merged, persisted.config, {
			code: node.code,
			level: node.level
		});
	}
}
/**
* Checks if global scope is in the path and loads it if not
* @param merged - The target map to merge into
* @param scopePath - The path of scope nodes
* @param loadScopeConfigFn - Function to load config for a scope
*/
async function mergeGlobalConfigIfNeeded(merged, scopePath, loadScopeConfigFn) {
	if (!scopePath.some((node) => node.code === "global" && node.level === "global")) {
		const globalConfig = await loadScopeConfigFn("global");
		if (globalConfig?.config && Array.isArray(globalConfig.config)) mergeConfigEntries(merged, globalConfig.config, {
			code: "global",
			level: "global"
		});
	}
}
/**
* Merges the current config data into the map
* @param merged - The target map to merge into
* @param configData - The current config data
* @param scopeCode - The code of the scope
* @param scopeLevel - The level of the scope
*/
function mergeCurrentConfigData(merged, configData, scopeCode, scopeLevel) {
	if (configData?.config && Array.isArray(configData.config)) {
		for (const entry of configData.config) if (!merged.has(entry.name)) merged.set(entry.name, {
			name: entry.name,
			value: entry.value,
			origin: entry.origin || {
				code: configData.scope?.code || scopeCode,
				level: configData.scope?.level || scopeLevel
			}
		});
	}
}
/**
* Applies schema defaults to the merged config
* @param merged - The target map to apply defaults to
* @param defaultMap - The map of default values
*/
function applySchemaDefaults(merged, defaultMap) {
	for (const [name, def] of defaultMap.entries()) if (!merged.has(name)) merged.set(name, {
		name,
		value: def,
		origin: {
			code: "default",
			level: "system"
		}
	});
}
/**
* Merge inherited configs along the scope path and ensure schema defaults exist
* @param params - The parameters object
* @param params.loadScopeConfigFn - The function to load the scope config.
* @param params.getSchemaFn - The function to get the schema.
* @param params.configData - The config data to merge.
* @param params.scopeCode - The code of the scope.
* @param params.scopeLevel - The level of the scope.
* @param params.scopePath - The path of the scope.
* @returns The merged config data. Returns null if the config data is not valid.
* @throws An error if the config data is not valid.
*/
async function mergeWithSchemaDefaults({ loadScopeConfigFn, getSchemaFn, configData, scopeCode, scopeLevel, scopePath }) {
	const schema = await getSchemaFn();
	const defaultMap = /* @__PURE__ */ new Map();
	for (const field of schema) if (field.default !== void 0) defaultMap.set(field.name, field.default);
	const merged = /* @__PURE__ */ new Map();
	await mergeConfigFromPath(merged, scopePath, loadScopeConfigFn);
	await mergeGlobalConfigIfNeeded(merged, scopePath, loadScopeConfigFn);
	mergeCurrentConfigData(merged, configData, scopeCode, scopeLevel);
	applySchemaDefaults(merged, defaultMap);
	configData.config = Array.from(merged.entries()).map(([name, data]) => ({
		name,
		value: data.value,
		origin: data.origin
	}));
	return configData;
}
/**
* Creates a scope selector that identifies a scope by its unique ID.
*
* This is the most direct way to identify a scope when you already know its ID.
* Scope IDs are unique identifiers assigned to each scope in the scope tree.
*
* @param scopeId - The unique identifier of the scope.
* @returns A selector object that identifies the scope by ID.
*
* @example
* ```typescript
* import { getConfiguration, byScopeId } from "@adobe/aio-commerce-lib-config";
*
* // Get configuration for a scope by its ID
* const config = await getConfiguration(byScopeId("scope-123"));
* ```
*/
function byScopeId(scopeId) {
	return { by: {
		_tag: "scopeId",
		scopeId
	} };
}
/**
* Creates a scope selector that identifies a scope by its code and level.
*
* This selector is useful when you know both the scope code and its level in
* the hierarchy (e.g., "website" level "website", "store" level "store").
*
* @param code - The code identifier of the scope (e.g., "us-east", "main_store").
* @param level - The level of the scope in the hierarchy (e.g., "website", "store", "store_view").
* @returns A selector object that identifies the scope by code and level.
*
* @example
* ```typescript
* import { getConfiguration, byCodeAndLevel } from "@adobe/aio-commerce-lib-config";
*
* // Get configuration for a website scope
* const config = await getConfiguration(byCodeAndLevel("us-east", "website"));
*
* // Get configuration for a store scope
* const storeConfig = await getConfiguration(byCodeAndLevel("main_store", "store"));
* ```
*/
function byCodeAndLevel(code, level) {
	return { by: {
		_tag: "codeAndLevel",
		code,
		level
	} };
}
/**
* Creates a scope selector that identifies a scope by its code only.
*
* This selector uses the scope code and will resolve to the default level for that code.
* Use this when you want to let the system determine the appropriate level based on
* the code alone.
*
* @param code - The code identifier of the scope.
* @returns A selector object that identifies the scope by code.
*
* @example
* ```typescript
* import { getConfiguration, byCode } from "@adobe/aio-commerce-lib-config";
*
* // Get configuration by code (uses default level)
* const config = await getConfiguration(byCode("us-east"));
* ```
*/
function byCode(code) {
	return { by: {
		_tag: "code",
		code
	} };
}
/**
* Creates a scope selector that identifies a website by its Commerce API ID.
*
* Websites are returned by the Commerce REST endpoint `/V1/store/websites`. The
* numeric ID is matched against the `commerce_id` of website-level scopes in
* the scope tree.
*
* @param commerceScopeId - The Commerce API numeric ID of the website.
* @returns A selector that identifies the website scope.
*
* @example
* ```typescript
* import { getConfiguration, byWebsiteId } from "@adobe/aio-commerce-lib-config";
*
* const config = await getConfiguration(byWebsiteId(1));
* ```
*/
function byWebsiteId(commerceScopeId) {
	return { by: {
		_tag: "commerceScopeId",
		commerceScopeId,
		level: "website"
	} };
}
/**
* Creates a scope selector that identifies a store (store group) by its Commerce API ID.
*
* Store groups are returned by the Commerce REST endpoint `/V1/store/storeGroups`.
* In the scope tree they live at the `"store"` level. The numeric ID is matched
* against the `commerce_id` of store-level scopes.
*
* @param commerceScopeId - The Commerce API numeric ID of the store group.
* @returns A selector that identifies the store scope.
*
* @example
* ```typescript
* import { getConfiguration, byStoreId } from "@adobe/aio-commerce-lib-config";
*
* const config = await getConfiguration(byStoreId(1));
* ```
*/
function byStoreId(commerceScopeId) {
	return { by: {
		_tag: "commerceScopeId",
		commerceScopeId,
		level: "store"
	} };
}
/**
* Creates a scope selector that identifies a store view by its Commerce API ID.
*
* Store views are returned by the Commerce REST endpoint `/V1/store/storeViews`.
* The numeric ID is matched against the `commerce_id` of store_view-level scopes.
*
* @param commerceScopeId - The Commerce API numeric ID of the store view.
* @returns A selector that identifies the store view scope.
*
* @example
* ```typescript
* import { getConfiguration, byStoreViewId } from "@adobe/aio-commerce-lib-config";
*
* const config = await getConfiguration(byStoreViewId(2));
* ```
*/
function byStoreViewId(commerceScopeId) {
	return { by: {
		_tag: "commerceScopeId",
		commerceScopeId,
		level: "store_view"
	} };
}

//#endregion
//#region source/utils/logger.ts
/**
* Get a logger instance for a given module name
* @param moduleName - The name of the module to create a logger for
* @returns Logger instance
*/
function getLogger(moduleName) {
	return AioLogger(moduleName, { level: process.env.LOG_LEVEL ?? "info" });
}

//#endregion
//#region source/utils/uuid.ts
/** Generate a UUID v4 */
function generateUUID() {
	return v4();
}

//#endregion
//#region source/modules/scope-tree/scope-tree-repository.ts
/**
* Gets cached scope tree from state store.
*
* @param namespace - Namespace identifier for the scope tree.
* @returns Promise resolving to cached scope tree or null if not found.
*/
async function getCachedScopeTree(namespace) {
	try {
		const cached = await (await getSharedState()).get(`${namespace}:scope-tree`);
		if (cached?.value) return JSON.parse(cached.value).data || null;
		return null;
	} catch {
		return null;
	}
}
/**
* Caches scope tree in state store with TTL.
*
* @param namespace - Namespace identifier for the scope tree.
* @param data - Scope tree data to cache.
* @param ttlSeconds - Time to live in seconds.
*/
async function setCachedScopeTree(namespace, data, ttlSeconds) {
	const logger = getLogger("@adobe/aio-commerce-lib-config:scope-tree-repository");
	try {
		await (await getSharedState()).put(`${namespace}:scope-tree`, stringify$1({ data }), { ttl: ttlSeconds });
	} catch (error) {
		logger.debug("Error caching scope tree:", error instanceof Error ? error.message : String(error));
	}
}
/**
* Deletes cached scope tree from state store.
*
* @param namespace - Namespace identifier for the scope tree.
*/
async function deleteCachedScopeTree(namespace) {
	const logger = getLogger("@adobe/aio-commerce-lib-config:scope-tree-repository");
	try {
		await (await getSharedState()).delete(`${namespace}:scope-tree`);
	} catch (error) {
		logger.debug("Failed to delete cached scope tree:", error instanceof Error ? error.message : String(error));
	}
}
/**
* Gets persisted scope tree from files.
*
* @param namespace - Namespace identifier for the scope tree.
* @returns Promise resolving to scope tree, creating initial tree if not found.
*/
async function getPersistedScopeTree(namespace) {
	const logger = getLogger("@adobe/aio-commerce-lib-config:scope-tree-repository");
	try {
		const files = await getSharedFiles();
		const filePath = generateScopeFilePath(namespace);
		try {
			const content = await files.read(filePath);
			return JSON.parse(content.toString()).scopes;
		} catch {
			const initialTree = createInitialScopeTree();
			await saveScopeTree(namespace, initialTree);
			return initialTree;
		}
	} catch (error) {
		logger.debug("Error getting scope tree from files:", error instanceof Error ? error.message : String(error));
		return createInitialScopeTree();
	}
}
/**
* Saves scope tree to files.
*
* @param namespace - Namespace identifier for the scope tree.
* @param scopes - Scope tree to save.
*
* @throws {Error} If saving to files fails.
*/
async function saveScopeTree(namespace, scopes) {
	const logger = getLogger("@adobe/aio-commerce-lib-config:scope-tree-repository");
	try {
		const files = await getSharedFiles();
		const filePath = generateScopeFilePath(namespace);
		const data = {
			scopes,
			lastUpdated: (/* @__PURE__ */ new Date()).toISOString(),
			version: "1.0"
		};
		await files.write(filePath, stringify$1(data, null, 2));
	} catch (error) {
		logger.debug("Error saving scope tree to files:", error instanceof Error ? error.message : String(error));
		throw error;
	}
}
/**
* Creates initial scope tree with global scope.
*
* @returns Initial scope tree containing only global scope.
*/
function createInitialScopeTree() {
	return [{
		id: generateUUID(),
		code: "global",
		label: "Global",
		level: "global",
		is_editable: true,
		is_removable: false,
		is_final: true
	}];
}
/**
* Generates file path for scope tree storage.
*
* @param namespace - Namespace identifier.
* @returns File path string.
*/
function generateScopeFilePath(namespace) {
	return `${namespace}/scope-tree.json`;
}

//#endregion
//#region source/modules/configuration/configuration-repository.ts
/**
* Gets cached configuration payload from state store.
*
* @param scopeCode - Scope code identifier.
* @returns Promise resolving to cached configuration payload or null if not found.
*/
async function getCachedConfig(scopeCode) {
	try {
		const state = await getSharedState();
		const key = getConfigStateKey(scopeCode);
		const result = await state.get(key);
		if (result.value) return JSON.parse(result.value).data || null;
		return null;
	} catch {
		return null;
	}
}
/**
* Caches configuration payload in state store.
*
* @param scopeCode - Scope code identifier.
* @param payload - Configuration payload as JSON string.
* @param ttlSeconds - Time to live for the cached configuration value.
*/
async function setCachedConfig(scopeCode, payload, ttlSeconds) {
	const logger = getLogger("@adobe/aio-commerce-lib-config:configuration-repository");
	try {
		const state = await getSharedState();
		const key = getConfigStateKey(scopeCode);
		await state.put(key, stringify({ data: payload }), { ttl: ttlSeconds });
	} catch (error) {
		logger.debug("Failed to cache configuration:", error instanceof Error ? error.message : String(error));
	}
}
/**
* Gets persisted configuration payload from files.
*
* @param scopeCode - Scope code identifier.
* @returns Promise resolving to configuration payload as string or null if not found.
*/
async function getPersistedConfig(scopeCode) {
	try {
		const files = await getSharedFiles();
		const filePath = getConfigFilePath(scopeCode);
		if (!(await files.list("scope/")).find((file) => file.name === filePath)) return null;
		const content = await files.read(filePath);
		return content ? content.toString("utf8") : null;
	} catch {
		return null;
	}
}
/**
* Saves configuration payload to files.
*
* @param scopeCode - Scope code identifier.
* @param payload - Configuration payload as JSON string.
*/
async function saveConfig(scopeCode, payload) {
	const files = await getSharedFiles();
	const filePath = getConfigFilePath(scopeCode);
	await files.write(filePath, payload);
}
/**
* Persists configuration with caching strategy (files + state cache).
*
* @param scopeCode - The scope code to persist configuration for.
* @param payload - The configuration payload object.
* @param ttlSeconds - Time to live for the cached configuration value.
*/
async function persistConfig(scopeCode, payload, ttlSeconds) {
	const payloadString = stringify(payload);
	const logger = getLogger("@adobe/aio-commerce-lib-config:configuration-repository");
	await saveConfig(scopeCode, payloadString);
	try {
		await setCachedConfig(scopeCode, payloadString, ttlSeconds);
	} catch (e) {
		logger.debug("Failed to cache configuration in state", {
			scopeCode,
			error: e instanceof Error ? e.message : String(e)
		});
	}
}
/**
* Tries to load configuration from state cache.
*
* @param scopeCode - The scope code to load configuration for.
* @returns Promise resolving to parsed configuration or null if not found.
*/
async function loadFromStateCache(scopeCode) {
	const logger = getLogger("@adobe/aio-commerce-lib-config:configuration-repository");
	try {
		const statePayload = await getCachedConfig(scopeCode);
		if (statePayload) return JSON.parse(statePayload);
	} catch (err) {
		logger.debug("Failed to load configuration from state cache", {
			scopeCode,
			error: err instanceof Error ? err.message : String(err)
		});
	}
	return null;
}
/**
* Tries to load configuration from persisted files.
*
* @param scopeCode - The scope code to load configuration for.
* @param ttlSeconds - Time to live for the cached configuration value.

* @returns Promise resolving to parsed configuration or null if not found.
*/
async function loadFromPersistedFiles(scopeCode, ttlSeconds) {
	const logger = getLogger("@adobe/aio-commerce-lib-config:configuration-repository");
	try {
		const filePayload = await getPersistedConfig(scopeCode);
		if (!filePayload) return null;
		try {
			await setCachedConfig(scopeCode, filePayload, ttlSeconds);
		} catch (err) {
			logger.debug("Failed to cache configuration in state", {
				scopeCode,
				error: err instanceof Error ? err.message : String(err)
			});
		}
		return JSON.parse(filePayload);
	} catch (err) {
		if (isNotFoundError(err)) logger.debug(`No persisted configuration file found for scope ${scopeCode}.`);
		else logger.debug(`Failed to retrieve configuration from files for scope ${scopeCode}:`, err instanceof Error ? err.message : String(err));
	}
	return null;
}
/**
* Loads configuration with smart fallback strategy (state -> files -> cache).
*
* @param scopeCode - The scope code to load configuration for.
* @param ttlSeconds - Time to live for the cached configuration value.

* @returns Promise resolving to configuration payload or null if not found.
*/
async function loadConfig(scopeCode, ttlSeconds) {
	const fromState = await loadFromStateCache(scopeCode);
	if (fromState) return fromState;
	const fromFiles = await loadFromPersistedFiles(scopeCode, ttlSeconds);
	if (fromFiles) return fromFiles;
	return null;
}
/**
* Gets the state key for a given scope code.
*
* @param scopeCode - The scope code to get the state key for.
* @returns State key string.
*/
function getConfigStateKey(scopeCode) {
	return `configuration.${scopeCode}`;
}
/**
* Gets the file path for a given scope code.
*
* @param scopeCode - The scope code to get the file path for.
* @returns File path string.
*/
function getConfigFilePath(scopeCode) {
	return `scope/${scopeCode.toLowerCase()}/configuration.json`;
}
/**
* Checks if error is a not-found error.
*
* @param err - Error to check.
* @returns True if error is a not-found error (404 or ENOENT), false otherwise.
*/
function isNotFoundError(err) {
	return typeof err === "object" && err !== null && ("statusCode" in err && err.statusCode === 404 || "code" in err && err.code === "ENOENT");
}

//#endregion
//#region source/modules/configuration/get-config.ts
/**
* Fetches and merges configuration for a scope without decrypting password fields.
*
* @internal
*/
async function fetchRawConfiguration(context, ...args) {
	const schema = requireGlobalSchema();
	const { scopeCode, scopeLevel, scopeId, scopePath } = deriveScopeFromArgs(args, await getPersistedScopeTree(context.namespace));
	let configData = await loadConfig(scopeCode, context.cacheTimeout);
	if (!configData) {
		const defaults = getSchemaDefaults(schema);
		configData = {
			scope: {
				id: scopeId,
				code: scopeCode,
				level: scopeLevel
			},
			config: defaults.config
		};
	} else if (!Array.isArray(configData.config)) configData.config = [];
	try {
		configData = await mergeWithSchemaDefaults({
			configData,
			scopeCode,
			scopeLevel,
			scopePath,
			loadScopeConfigFn: (code) => loadConfig(code, context.cacheTimeout),
			getSchemaFn: async () => Promise.resolve(schema)
		});
	} catch {}
	return {
		configData,
		passwordFields: getPasswordFields(schema)
	};
}
/**
* Gets configuration for a scope identified by code and level or id.
*
* This function retrieves all configuration values for a specific scope, including
* inherited values from parent scopes and schema defaults. The configuration is
* merged according to the scope hierarchy, with values from more specific scopes
* taking precedence over parent scopes. Password-type fields are automatically
* decrypted before being returned.
*
* @param context - Configuration context containing namespace and cache timeout.
* @param args - Scope identifier: `(id: string)`, `(code: string, level: string)`, or `(commerceScopeId: number, level: string)`.
* @returns Promise resolving to configuration response with scope information and config values.
*
* @throws {Error} If the scope arguments are invalid or the scope is not found.
*/
async function getConfiguration$1(context, ...args) {
	const { configData, passwordFields } = await fetchRawConfiguration(context, ...args);
	return decryptPasswordFields(configData, passwordFields, context.encryptionKey);
}
/**
* Decrypts password fields in the configuration data.
*
* @param configData - Configuration data to process.
* @param passwordFields - Set of field names that should be decrypted.
* @returns Configuration data with password fields decrypted.
*/
function decryptPasswordFields(configData, passwordFields, encryptionKey) {
	if (!Array.isArray(configData.config)) return configData;
	return {
		...configData,
		config: configData.config.map((entry) => {
			if (passwordFields.has(entry.name) && typeof entry.value === "string" && entry.value.length > 0) {
				if (!encryptionKey) throw new Error("Encryption key has not been given");
				return {
					...entry,
					value: decrypt(entry.value, encryptionKey)
				};
			}
			return entry;
		})
	};
}

//#endregion
//#region source/modules/configuration/get-config-by-key.ts
/**
* Gets a specific configuration value by key for a scope identified by code and level or id.
*
* This function retrieves a single configuration value for a specific scope. It's useful
* when you only need one configuration field rather than the entire configuration object.
* The value includes origin information indicating where it was inherited from.
*
* @param context - Configuration context containing namespace and cache timeout.
* @param configKey - The name of the configuration field to retrieve.
* @param args - Scope identifier: `(id: string)`, `(code: string, level: string)`, or `(commerceScopeId: number, level: string)`.
* @returns Promise resolving to configuration response with scope information and single config value (or null if not found).
*
* @throws {Error} If the scope arguments are invalid or the scope is not found.
* @throws {Error} If the requested key is a password field and no encryption key is provided.
*/
async function getConfigurationByKey$1(context, configKey, ...args) {
	const { configData, passwordFields } = await fetchRawConfiguration(context, ...args);
	let configValue = configData.config.find((item) => item.name === configKey) || null;
	if (configValue && passwordFields.has(configKey) && typeof configValue.value === "string" && configValue.value.length > 0) {
		if (!context.encryptionKey) throw new Error("Encryption key has not been given");
		configValue = {
			...configValue,
			value: decrypt(configValue.value, context.encryptionKey)
		};
	}
	return {
		scope: configData.scope,
		config: configValue
	};
}

//#endregion
//#region source/modules/configuration/set-config.ts
/**
* Sets configuration values for a scope identified by code and level or id.
*
* This function updates configuration values for a specific scope. The provided values
* are merged with existing configuration, and the origin is set to the current scope.
* Configuration values are inherited from parent scopes unless explicitly overridden.
* Password-type fields are automatically encrypted before storage.
*
* @param context - Configuration context containing namespace and cache timeout.
* @param request - Configuration set request containing the config values to set.
* @param args - Scope identifier: `(id: string)`, `(code: string, level: string)`, or `(commerceScopeId: number, level: string)`.
* @returns Promise resolving to configuration response with updated scope and config values.
*
* @throws {Error} If the scope arguments are invalid or the scope is not found.
*/
async function setConfiguration$1(context, request, ...args) {
	const schema = requireGlobalSchema();
	const { scopeCode, scopeLevel, scopeId } = deriveScopeFromArgs(args, await getPersistedScopeTree(context.namespace));
	const encryptedEntries = encryptPasswordFields(sanitizeRequestEntries(request?.config), getPasswordFields(schema), context.encryptionKey);
	const existingPersisted = await loadConfig(scopeCode, context.cacheTimeout);
	const mergedScopeConfig = mergeScopes(Array.isArray(existingPersisted?.config) ? existingPersisted?.config ?? [] : [], encryptedEntries, scopeCode, scopeLevel);
	const payload = {
		scope: {
			id: scopeId,
			code: scopeCode,
			level: scopeLevel
		},
		config: mergedScopeConfig
	};
	await persistConfig(scopeCode, payload, context.cacheTimeout);
	return {
		message: "Configuration values updated successfully",
		timestamp: (/* @__PURE__ */ new Date()).toISOString(),
		scope: {
			id: String(scopeId),
			code: scopeCode,
			level: scopeLevel
		},
		config: mergedScopeConfig.map(({ name, value }) => ({
			name,
			value
		}))
	};
}
/**
* Encrypts password fields in the configuration entries.
* Entries with a `null` value (unset) are passed through unchanged.
*
* @param entries - Configuration entries to process.
* @param passwordFields - Set of field names that should be encrypted.
* @returns Configuration entries with password fields encrypted.
*/
function encryptPasswordFields(entries, passwordFields, encryptionKey) {
	return entries.map((entry) => {
		if (entry.value !== null && passwordFields.has(entry.name) && typeof entry.value === "string" && entry.value.length > 0) {
			if (!encryptionKey) throw new Error("Encryption key has not been given");
			return {
				...entry,
				value: encrypt(entry.value, encryptionKey)
			};
		}
		return entry;
	});
}

//#endregion
//#region source/api/commerce.ts
/**
* Fetch all scope data from Commerce REST API endpoints
* @param httpClient - The Adobe Commerce HTTP client instance
* @returns Complete Commerce scope data with websites, store groups, and store views
*/
async function getAllScopeData(httpClient) {
	try {
		const [websites, storeGroups, storeViews] = await Promise.all([
			getWebsites(httpClient),
			getStoreGroups(httpClient),
			getStoreViews(httpClient)
		]);
		return {
			websites,
			storeGroups,
			storeViews
		};
	} catch (error) {
		throw new Error(`Failed to fetch Commerce scope data: ${error}`);
	}
}
/**
* Get all websites using /V1/store/websites endpoint
* @param httpClient - The Adobe Commerce HTTP client instance
* @see https://adobe-commerce.redoc.ly/2.4.8-admin/tag/storewebsites
*/
async function getWebsites(httpClient) {
	try {
		return await httpClient.get("store/websites").json();
	} catch (error) {
		throw new Error(`Failed to fetch websites: ${error}`);
	}
}
/**
* Get all store groups using /V1/store/storeGroups endpoint
* @param httpClient - The Adobe Commerce HTTP client instance
* @see https://adobe-commerce.redoc.ly/2.4.8-admin/tag/storestoreGroups
*/
async function getStoreGroups(httpClient) {
	try {
		return await httpClient.get("store/storeGroups").json();
	} catch (error) {
		throw new Error(`Failed to fetch store groups: ${error}`);
	}
}
/**
* Get all store views using /V1/store/storeViews endpoint
* @param httpClient - The Adobe Commerce HTTP client instance
* @see https://adobe-commerce.redoc.ly/2.4.8-admin/tag/storestoreViews
*/
async function getStoreViews(httpClient) {
	try {
		return await httpClient.get("store/storeViews").json();
	} catch (error) {
		throw new Error(`Failed to fetch store views: ${error}`);
	}
}

//#endregion
//#region source/modules/scope-tree/merge-scopes.ts
/**
* Merges fresh Commerce API data with existing scope tree while preserving existing UUIDs.
*
* This function transforms Commerce API scope data (websites, store groups, store views)
* into scope nodes while preserving UUIDs from the existing tree when possible. This ensures
* that scope IDs remain stable across sync operations. UUIDs are matched by `commerce_id`
* and `level` combination.
*
* @param freshData - Fresh data from Commerce API containing websites, store groups, and store views.
* @param existingTree - Existing scope tree with UUIDs to preserve.
* @returns Updated Commerce scopes with preserved UUIDs.
*
* @example
* ```typescript
* import { mergeCommerceScopes } from "./modules/scope-tree";
* import type { CommerceScopeData } from "../../types";
* import type { ScopeTree } from "./types";
*
* const freshData: CommerceScopeData = {
*   websites: [{ id: 1, code: "main", name: "Main Website", default_group_id: 1 }],
*   storeGroups: [{ id: 1, website_id: 1, root_category_id: 2, default_store_id: 1, name: "Main Store", code: "main_store" }],
*   storeViews: [{ id: 1, code: "default", name: "Default Store View", website_id: 1, store_group_id: 1, is_active: true }],
* };
*
* const existingTree: ScopeTree = [
*   {
*     id: "existing-uuid-123",
*     code: "commerce",
*     children: [
*       {
*         id: "existing-website-uuid",
*         commerce_id: 1,
*         code: "main",
*         level: "website",
*         // ... other properties
*       },
*     ],
*   },
* ];
*
* // Returns scopes with preserved UUIDs where commerce_id matches
* const updatedScopes = mergeCommerceScopes(freshData, existingTree);
* // existing-website-uuid is preserved because commerce_id: 1 matches
* ```
*/
function mergeCommerceScopes(freshData, existingTree) {
	const existingCommerceScopes = existingTree.find((node) => node.code === "commerce")?.children || [];
	const { websites, storeGroups, storeViews } = freshData;
	return websites.map((website) => {
		const existingWebsite = findScopeByCommerceId(existingCommerceScopes, website.id, "website");
		return {
			id: existingWebsite?.id || generateUUID(),
			commerce_id: website.id,
			code: website.code,
			label: website.name,
			level: "website",
			is_editable: true,
			is_final: true,
			is_removable: false,
			children: buildStoreGroups(storeGroups.filter((sg) => sg.website_id === website.id), storeViews, existingWebsite?.children || [])
		};
	});
}
/**
* Builds store groups while preserving UUIDs.
*
* @param storeGroups - Store groups from Commerce API.
* @param allStoreViews - All store views from Commerce API.
* @param existingStoreGroups - Existing store group nodes with UUIDs to preserve.
* @returns Array of store group scope nodes with preserved UUIDs.
*/
function buildStoreGroups(storeGroups, allStoreViews, existingStoreGroups) {
	return storeGroups.map((storeGroup) => {
		const existing = findScopeByCommerceId(existingStoreGroups, storeGroup.id, "store");
		return {
			id: existing?.id || generateUUID(),
			commerce_id: storeGroup.id,
			code: storeGroup.code,
			label: storeGroup.name,
			level: "store",
			is_editable: false,
			is_final: true,
			is_removable: false,
			children: buildStoreViews(allStoreViews.filter((sv) => sv.store_group_id === storeGroup.id), existing?.children || [])
		};
	});
}
/**
* Builds store views while preserving UUIDs.
*
* @param storeViews - Store views from Commerce API.
* @param existingStoreViews - Existing store view nodes with UUIDs to preserve.
* @returns Array of store view scope nodes with preserved UUIDs.
*/
function buildStoreViews(storeViews, existingStoreViews) {
	return storeViews.map((storeView) => {
		return {
			id: findScopeByCommerceId(existingStoreViews, storeView.id, "store_view")?.id || generateUUID(),
			commerce_id: storeView.id,
			code: storeView.code,
			label: storeView.name,
			level: "store_view",
			is_editable: true,
			is_final: true,
			is_removable: false
		};
	});
}
/**
* Finds existing scope by commerce_id and level to preserve UUID.
*
* @param scopes - Array of scope nodes to search.
* @param commerceId - Commerce API ID to match.
* @param level - Scope level to match.
* @returns Matching scope node or undefined if not found.
*/
function findScopeByCommerceId(scopes, commerceId, level) {
	return scopes.find((scope) => scope.commerce_id === commerceId && scope.level === level);
}
/**
* Builds an updated complete scope tree by merging updated Commerce scopes with existing custom scopes.
*
* This function takes updated Commerce scopes and merges them into the existing complete scope tree,
* preserving all system scopes (global) and custom scopes while updating only the Commerce scopes.
* If a "commerce" node doesn't exist, it will be created.
*
* @param updatedCommerceScopes - Fresh Commerce scopes with preserved UUIDs.
* @param existingTree - Existing complete scope tree.
* @returns Complete merged scope tree with updated Commerce scopes.
*
* @example
* ```typescript
* import { buildUpdatedScopeTree } from "./modules/scope-tree";
* import type { ScopeNode, ScopeTree } from "./types";
*
* const updatedCommerceScopes: ScopeNode[] = [
*   {
*     id: "website-1",
*     code: "main",
*     level: "website",
*     commerce_id: 1,
*     label: "Main Website",
*     is_editable: true,
*     is_final: true,
*     is_removable: false,
*   },
* ];
*
* const existingTree: ScopeTree = [
*   {
*     id: "global-1",
*     code: "global",
*     level: "global",
*     label: "Global",
*     is_editable: false,
*     is_final: false,
*     is_removable: false,
*   },
*   {
*     id: "commerce-1",
*     code: "commerce",
*     level: "commerce",
*     label: "Commerce",
*     is_editable: false,
*     is_final: true,
*     is_removable: false,
*     children: [],
*   },
*   {
*     id: "custom-1",
*     code: "region_us",
*     level: "custom",
*     label: "US Region",
*     is_editable: true,
*     is_final: false,
*     is_removable: true,
*   },
* ];
*
* // Returns tree with updated Commerce scopes, preserving global and custom scopes
* const mergedTree = buildUpdatedScopeTree(updatedCommerceScopes, existingTree);
* // mergedTree[0] = global scope (unchanged)
* // mergedTree[1] = commerce scope (updated with new children)
* // mergedTree[2] = custom scope (unchanged)
* ```
*/
function buildUpdatedScopeTree(updatedCommerceScopes, existingTree) {
	if (!existingTree.some((node) => node.code === "commerce")) return [...existingTree, {
		id: generateUUID(),
		code: "commerce",
		label: "Commerce",
		level: "commerce",
		is_editable: false,
		is_removable: false,
		is_final: true,
		children: updatedCommerceScopes
	}];
	return existingTree.map((node) => {
		if (node.code === "commerce") return {
			...node,
			children: updatedCommerceScopes
		};
		return node;
	});
}

//#endregion
//#region source/modules/scope-tree/get-scope-tree.ts
const logger = AioLogger("@adobe/aio-commerce-lib-config:get-scope-tree", { level: process.env.LOG_LEVEL ?? "info" });
function hasCommerceConfig(ctx) {
	return !!ctx.commerceConfig;
}
/**
* Gets the scope tree and optionally refreshes Commerce scopes from Commerce API.
*
* The scope tree represents the hierarchical structure of configuration scopes available
* in your Adobe Commerce instance. This includes both system scopes (global, website, store)
* and custom scopes that may have been defined.
*
* @param context - Scope tree context containing namespace, cache timeout, and optional Commerce config.
* @param options - Options for scope tree retrieval, including whether to fetch fresh data.
* @returns Promise resolving to scope tree with metadata about data freshness and any fallback information.
*/
async function getScopeTree$1(context, options = {}) {
	const { remoteFetch = false } = options;
	if (remoteFetch && hasCommerceConfig(context)) return await buildTreeWithUpdatedCommerceScopes(context);
	const cached = await getCachedScopeTree(context.namespace);
	if (cached) return {
		scopeTree: cached,
		isCachedData: true
	};
	const persistedTree = await getPersistedScopeTree(context.namespace);
	await setCachedScopeTree(context.namespace, persistedTree, context.cacheTimeout);
	return {
		scopeTree: persistedTree,
		isCachedData: true
	};
}
/**
* Builds scope tree with updated Commerce data merged with existing scopes.
*
* @param context - Scope tree context with Commerce configuration for fetching fresh data.
* @returns Promise resolving to scope tree result with fresh Commerce data or fallback data.
*/
async function buildTreeWithUpdatedCommerceScopes(context) {
	try {
		const updatedCommerceScopeData = await getAllScopeData(initializeCommerceClient(context.commerceConfig));
		const existingTree = await getPersistedScopeTree(context.namespace);
		const finalTree = buildUpdatedScopeTree(mergeCommerceScopes(updatedCommerceScopeData, existingTree), existingTree);
		await saveScopeTree(context.namespace, finalTree);
		await setCachedScopeTree(context.namespace, finalTree, context.cacheTimeout);
		return {
			scopeTree: finalTree,
			isCachedData: false
		};
	} catch (error) {
		const errorMessage = error instanceof Error ? error.message : "";
		logger.debug("Failed to fetch fresh Commerce data, using fallback: ", error instanceof Error ? error.message : String(error));
		const cachedFlattenedTree = await getCachedScopeTree(context.namespace);
		if (cachedFlattenedTree) return {
			scopeTree: cachedFlattenedTree,
			isCachedData: true,
			fallbackError: errorMessage
		};
		const existingTree = await getPersistedScopeTree(context.namespace);
		await setCachedScopeTree(context.namespace, existingTree, context.cacheTimeout);
		return {
			scopeTree: existingTree,
			isCachedData: true,
			fallbackError: errorMessage
		};
	}
}
/**
* Initializes a Commerce HTTP client from the provided configuration.
*
* @param commerceConfig - Commerce HTTP client configuration parameters.
* @returns Initialized Commerce HTTP client instance.
*
* @throws {Error} If the client initialization fails.
*/
function initializeCommerceClient(commerceConfig) {
	try {
		return new AdobeCommerceHttpClient(commerceConfig);
	} catch (error) {
		throw new Error(`Failed to initialize Commerce client: ${error}`);
	}
}

//#endregion
//#region source/modules/scope-tree/custom-scope-tree-validation.ts
const COMMERCE_SCOPE_CODE = "commerce";
const GLOBAL_SCOPE_CODE = "global";
/**
* Creates validation errors with consistent property.
*
* @param message - Error message.
* @returns Error object with isValidationError property set to true.
*/
function createValidationError(message) {
	const error = new Error(message);
	return Object.assign(error, { isValidationError: true });
}
/**
* Validates and normalizes the complete request structure, field types, and business rules.
*
* @param request - Custom scope tree request to validate.
* @returns Array of validated and normalized custom scope inputs.
*
* @throws {Error} If the request is invalid (missing scopes array, invalid fields, duplicate codes, etc.).
*/
function validateCustomScopeRequest(request) {
	if (!(request.scopes && Array.isArray(request.scopes))) throw createValidationError("Request must include a scopes array");
	const validatedScopes = [];
	for (const scope of request.scopes) {
		const validatedScope = validateAndNormalizeSingleScope(scope);
		validatedScopes.push(validatedScope);
	}
	checkForDuplicateCodeLevelCombinations(validatedScopes);
	return validatedScopes;
}
/**
* Validates field types for a scope.
*
* @param scope - Custom scope input to validate.
*
* @throws {Error} If any field type is invalid.
*/
function validateFieldTypes(scope) {
	if (typeof scope.code !== "string" || scope.code.trim() === "") throw createValidationError(`Field 'code' must be a non-empty string`);
	if (typeof scope.label !== "string" || scope.label.trim() === "") throw createValidationError(`Field 'label' must be a non-empty string`);
	if (isProvidedButInvalidString(scope.level)) throw createValidationError(`Field 'level' must be a non-empty string if provided`);
	if (typeof scope.is_editable !== "boolean") throw createValidationError(`Field 'is_editable' must be a boolean`);
	if (typeof scope.is_final !== "boolean") throw createValidationError(`Field 'is_final' must be a boolean`);
}
/**
* Validates and normalizes a single scope recursively including structure, types, and business rules.
*
* @param scope - Custom scope input to validate and normalize.
* @returns Normalized scope with default values applied.
*
* @throws {Error} If the scope is invalid (missing required fields, invalid types, reserved codes, etc.).
*/
function validateAndNormalizeSingleScope(scope) {
	if (!scope || typeof scope !== "object") throw createValidationError("Scope must be an object");
	for (const field of [
		"code",
		"label",
		"is_editable",
		"is_final"
	]) if (scope[field] === void 0 || scope[field] === null) throw createValidationError(`Missing required field '${field}'`);
	validateFieldTypes(scope);
	if (isProvidedButInvalidString(scope.id)) throw createValidationError(`Field 'id' must be a non-empty string if provided`);
	const trimmedCode = scope.code.trim();
	const normalizedLevel = scope.level?.trim() || "base";
	if (trimmedCode === "commerce" || trimmedCode === "global") throw createValidationError(`Scope code '${trimmedCode}' is reserved and cannot be used in custom scopes`);
	const normalizedScope = {
		...scope,
		code: trimmedCode,
		label: scope.label.trim(),
		level: normalizedLevel
	};
	if (scope.id?.trim()) normalizedScope.id = scope.id.trim();
	if (scope.children !== void 0) {
		if (!Array.isArray(scope.children)) throw createValidationError(`Field 'children' must be an array if provided`);
		normalizedScope.children = scope.children.map((child) => validateAndNormalizeSingleScope(child));
	}
	return normalizedScope;
}
/**
* Checks for duplicate code-level combinations across all scopes (including nested children).
*
* @param scopes - Array of custom scope inputs to check.
*
* @throws {Error} If duplicate code-level combinations are found.
*/
function checkForDuplicateCodeLevelCombinations(scopes) {
	const combinations = /* @__PURE__ */ new Set();
	function collectCombinations(scopeList) {
		for (const scope of scopeList) {
			const combination = `${scope.code}:${scope.level}`;
			if (combinations.has(combination)) throw createValidationError(`Duplicate code-level combination '${combination}' found in custom scope tree. Custom scope code must be unique.`);
			combinations.add(combination);
			if (scope.children && scope.children.length > 0) collectCombinations(scope.children);
		}
	}
	collectCombinations(scopes);
}
/**
* Checks if an optional string field is provided but invalid.
*
* @param value - String value to check.
* @returns True if value is provided but empty after trimming, false otherwise.
*/
function isProvidedButInvalidString(value) {
	return value !== void 0 && value.trim() === "";
}

//#endregion
//#region source/modules/scope-tree/set-custom-scope-tree.ts
/**
* Sets the custom scope tree, replacing all existing custom scopes with the provided ones.
*
* Custom scopes allow you to define additional configuration scopes beyond the standard
* Commerce scopes (global, website, store, store_view). This function replaces all
* custom scopes, preserving system scopes (global and commerce).
*
* **Important:** If a scope with the same `code` and `level` already exists, its ID
* will be preserved. If `id` is provided in the input, it will be used; otherwise, a
* new UUID will be generated for new scopes.
*
* @param context - Scope tree context containing namespace and cache timeout.
* @param request - Custom scope tree request containing the scopes to set.
* @returns Promise resolving to response with updated custom scopes.
*
* @throws {Error} If the request validation fails (invalid scope structure, missing required fields, duplicate codes, etc.).
*
* @example
* ```typescript
* import { setCustomScopeTree } from "./modules/scope-tree";
*
* const context = { namespace: "my-app", cacheTimeout: 300000 };
*
* // Replace all custom scopes with new ones
* const result = await setCustomScopeTree(context, {
*   scopes: [
*     {
*       code: "region_us",
*       label: "US Region",
*       level: "custom",
*       is_editable: true,
*       is_final: false,
*       children: [
*         {
*           code: "region_us_west",
*           label: "US West",
*           level: "custom",
*           is_editable: true,
*           is_final: true,
*         },
*       ],
*     },
*   ],
* });
*
* // Existing scopes with matching code+level will preserve their IDs
* // New scopes will get new UUIDs assigned
* ```
*/
async function setCustomScopeTree$1(context, request) {
	const validatedScopes = validateCustomScopeRequest(request);
	const completeCurrentTree = await getPersistedScopeTree(context.namespace);
	const processedCustomScopes = processCustomScopes(validatedScopes, getExistingCustomScopes(completeCurrentTree));
	const updatedCompleteTree = buildUpdatedCompleteTree(completeCurrentTree, processedCustomScopes);
	await saveScopeTree(context.namespace, updatedCompleteTree);
	await deleteCachedScopeTree(context.namespace);
	return {
		message: "Custom scope tree updated successfully",
		timestamp: (/* @__PURE__ */ new Date()).toISOString(),
		scopes: convertToResponseFormat(processedCustomScopes)
	};
}
/**
* Extracts existing custom scopes from the complete tree.
*
* @param completeTree - Complete scope tree including system and custom scopes.
* @returns Array of custom scope nodes (excluding global and commerce scopes).
*/
function getExistingCustomScopes(completeTree) {
	return completeTree.filter((scope) => scope.code !== "global" && scope.code !== "commerce");
}
/**
* Builds updated complete tree by preserving system scopes and replacing custom scopes.
*
* @param currentCompleteTree - Current complete scope tree.
* @param newCustomScopes - New custom scopes to replace existing ones.
* @returns Updated complete scope tree with new custom scopes.
*/
function buildUpdatedCompleteTree(currentCompleteTree, newCustomScopes) {
	return [...currentCompleteTree.filter((scope) => scope.code === "global" || scope.code === "commerce"), ...newCustomScopes];
}
/**
* Processes custom scopes, preserving existing IDs where possible.
*
* @param inputScopes - Custom scope inputs to process.
* @param existingCustomScopes - Existing custom scopes to match against.
* @returns Array of processed scope nodes with preserved IDs.
*/
function processCustomScopes(inputScopes, existingCustomScopes) {
	return inputScopes.map((scope) => processSingleScope(scope, existingCustomScopes));
}
/**
* Processes a single scope recursively, preserving ID if found.
*
* @param inputScope - Custom scope input to process.
* @param existingCustomScopes - Existing custom scopes to match against.
* @returns Processed scope node with preserved or generated ID.
*/
function processSingleScope(inputScope, existingCustomScopes) {
	const baseCustomScope = {
		id: findExistingId(inputScope, existingCustomScopes) || generateUUID(),
		code: inputScope.code,
		label: inputScope.label,
		level: inputScope.level ?? "base",
		is_editable: inputScope.is_editable,
		is_final: inputScope.is_final,
		is_removable: true
	};
	if (inputScope.children && inputScope.children.length > 0) return {
		...baseCustomScope,
		children: inputScope.children.map((child) => processSingleScope(child, existingCustomScopes))
	};
	return baseCustomScope;
}
/**
* Finds existing ID for a scope based on code and level match.
*
* @param inputScope - Custom scope input to find ID for.
* @param existingScopes - Existing scope nodes to search.
* @returns Existing scope ID if found, undefined otherwise.
*/
function findExistingId(inputScope, existingScopes) {
	if (inputScope.id) return inputScope.id;
	return findInScopeTree(inputScope.code, inputScope.level ?? "base", existingScopes);
}
/**
* Recursively searches for a scope by code and level in a tree.
*
* @param targetCode - Scope code to find.
* @param targetLevel - Scope level to find.
* @param scopes - Array of scope nodes to search.
* @returns Scope ID if found, undefined otherwise.
*/
function findInScopeTree(targetCode, targetLevel, scopes) {
	for (const scope of scopes) {
		if (scope.code === targetCode && scope.level === targetLevel) return scope.id;
		if (scope.children) {
			const foundId = findInScopeTree(targetCode, targetLevel, scope.children);
			if (foundId) return foundId;
		}
	}
}
/**
* Converts processed scopes to response format.
*
* @param scopes - Processed scope nodes to convert.
* @returns Array of custom scope outputs.
*/
function convertToResponseFormat(scopes) {
	return scopes.map(convertSingleScopeToOutput);
}
/**
* Converts a single scope to response format recursively.
*
* @param scope - Scope node to convert.
* @returns Custom scope output.
*/
function convertSingleScopeToOutput(scope) {
	const baseScope = {
		id: scope.id,
		code: scope.code,
		label: scope.label,
		level: scope.level,
		is_editable: scope.is_editable,
		is_final: scope.is_final
	};
	if (scope.children && scope.children.length > 0) return {
		...baseScope,
		children: scope.children.map(convertSingleScopeToOutput)
	};
	return baseScope;
}

//#endregion
//#region source/config-manager.ts
function initializeStatic(configSchema) {
	setGlobalSchema(configSchema);
	return { configSchema };
}
function hasSchemaResolutionOptions(options) {
	return Boolean(options.schema && "params" in options && options.params);
}
function initialize(options) {
	if (options.libStateOptions) setGlobalStateOptions(options.libStateOptions);
	if (!options.schema) return { configSchema: requireGlobalSchema() };
	const validated = validateBusinessConfigSchema(options.schema);
	if (hasSchemaResolutionOptions(options)) return resolveBusinessConfigSchema(validated, options.params).then(initializeStatic);
	if (hasDynamicSchema(validated)) throw new Error("Dynamic list options require runtime params. Call `initialize({ schema, params })` to resolve them.");
	return initializeStatic(validated);
}
async function getScopeTree(params, options) {
	const context = {
		namespace: DEFAULT_NAMESPACE,
		cacheTimeout: options?.cacheTimeout ?? 300
	};
	if (params?.refreshData === true) return getScopeTree$1({
		...context,
		commerceConfig: params.commerceConfig
	}, { remoteFetch: true });
	return getScopeTree$1(context, { remoteFetch: false });
}
/**
* Syncs Commerce scopes by forcing a fresh fetch from Commerce API and updating the cache.
*
* This function is useful when you need to ensure your scope tree is up-to-date with
* the latest changes from your Commerce instance. It will fetch fresh data and update
* both the cache and persistent storage.
*
* @param commerceConfig - The Commerce HTTP client configuration required for API calls.
* @param options - Optional library configuration options for cache timeout.
* @returns Promise resolving to sync result with updated scope tree and sync status.
*
* @example
* ```typescript
* import { syncCommerceScopes } from "@adobe/aio-commerce-lib-config";
* import type { CommerceHttpClientParams } from "@adobe/aio-commerce-lib-api";
*
* const commerceConfig: CommerceHttpClientParams = {
*   url: "https://your-commerce-instance.com",
*   // ... other auth config
* };
*
* const result = await syncCommerceScopes(commerceConfig);
*
* if (result.synced) {
*   console.log("Successfully synced scope tree");
*   console.log(result.scopeTree); // Updated scope tree
* } else {
*   console.log("Used cached data");
* }
*
* if (result.error) {
*   console.warn("Sync completed with errors:", result.error);
* }
* ```
*/
async function syncCommerceScopes(commerceConfig, options) {
	try {
		const result = await getScopeTree({
			refreshData: true,
			commerceConfig
		}, options);
		const syncResult = {
			scopeTree: result.scopeTree,
			synced: !result.isCachedData
		};
		if (result.fallbackError) syncResult.error = result.fallbackError;
		return syncResult;
	} catch (error) {
		throw new Error(`Failed to sync Commerce scopes: ${error instanceof Error ? error.message : "Unknown error"}`);
	}
}
/**
* Removes the commerce scope from the persisted scope tree.
*
* @returns Promise resolving to an object with `unsynced` boolean indicating whether the
* scope was found and removed, or `false` if it was already not present.
*
* @example
* ```typescript
* import { unsyncCommerceScopes } from "@adobe/aio-commerce-lib-config";
*
* const result = await unsyncCommerceScopes();
* if (result.unsynced) {
*   console.log("Commerce scope removed successfully");
* } else {
*   console.log("Commerce scope not found");
* }
* ```
*/
async function unsyncCommerceScopes() {
	const COMMERCE_SCOPE_CODE = "commerce";
	const scopeTree = await getPersistedScopeTree(DEFAULT_NAMESPACE);
	if (!scopeTree.some((scope) => scope.code === COMMERCE_SCOPE_CODE)) return { unsynced: false };
	await saveScopeTree(DEFAULT_NAMESPACE, scopeTree.filter((scope) => scope.code !== COMMERCE_SCOPE_CODE));
	return { unsynced: true };
}
/**
* Gets configuration for a scope.
*
* This function retrieves all configuration values for a specific scope, including
* inherited values from parent scopes and schema defaults. The configuration is
* merged according to the scope hierarchy.
*
* @param selector - Scope selector specifying how to identify the scope.
* @param options - Optional library configuration options for cache timeout.
* @returns Promise resolving to configuration response with scope information and config values.
*
* @example
* ```typescript
* import {
*   getConfiguration,
*   byScopeId,
*   byCodeAndLevel,
*   byCode,
*   byWebsiteId,
*   byStoreId,
*   byStoreViewId,
* } from "@adobe/aio-commerce-lib-config";
*
* // Get configuration by scope ID
* const config1 = await getConfiguration(byScopeId("scope-123"));
* console.log(config1.scope); // { id, code, level }
* console.log(config1.config); // Array of config values with origins
*
* // Get configuration by code and level
* const config2 = await getConfiguration(byCodeAndLevel("website", "website"));
* config2.config.forEach((item) => {
*   console.log(`${item.name}: ${item.value} (from ${item.origin.code})`);
* });
*
* // Get configuration by code (uses default level)
* const config3 = await getConfiguration(byCode("website"));
*
* // Get configuration by Commerce API ID of a website, store, or store view
* const websiteCfg = await getConfiguration(byWebsiteId(1));
* const storeCfg = await getConfiguration(byStoreId(1));
* const storeViewCfg = await getConfiguration(byStoreViewId(1));
* ```
*/
async function getConfiguration(selector, options) {
	const context = {
		namespace: DEFAULT_NAMESPACE,
		cacheTimeout: options?.cacheTimeout ?? 300,
		encryptionKey: options?.encryptionKey
	};
	if (selector.by._tag === "scopeId") return await getConfiguration$1(context, selector.by.scopeId);
	if (selector.by._tag === "codeAndLevel") return await getConfiguration$1(context, selector.by.code, selector.by.level);
	if (selector.by._tag === "commerceScopeId") return await getConfiguration$1(context, selector.by.commerceScopeId, selector.by.level);
	return await getConfiguration$1(context, selector.by.code);
}
/**
* Gets a specific configuration value by key for a scope.
*
* This function retrieves a single configuration value for a specific scope. It's useful
* when you only need one configuration field rather than the entire configuration object.
*
* @param configKey - The name of the configuration field to retrieve.
* @param selector - Scope selector specifying how to identify the scope.
* @param options - Optional library configuration options for cache timeout.
* @returns Promise resolving to configuration response with scope information and single config value (or null if not found).
*
* @example
* ```typescript
* import { getConfigurationByKey, byScopeId, byCodeAndLevel, byCode } from "@adobe/aio-commerce-lib-config";
*
* // Get a specific config value by scope ID
* const result1 = await getConfigurationByKey("api_key", byScopeId("scope-123"));
*
* if (result1.config) {
*   console.log(`API Key: ${result1.config.value}`);
*   console.log(`Origin: ${result1.config.origin.code}`);
* }
*
* // Get a specific config value by code and level
* const result2 = await getConfigurationByKey("enable_feature", byCodeAndLevel("website", "website"));
*
* // Get a specific config value by code
* const result3 = await getConfigurationByKey("api_key", byCode("website"));
* ```
*/
async function getConfigurationByKey(configKey, selector, options) {
	const context = {
		namespace: DEFAULT_NAMESPACE,
		cacheTimeout: options?.cacheTimeout ?? 300,
		encryptionKey: options?.encryptionKey
	};
	if (selector.by._tag === "scopeId") return await getConfigurationByKey$1(context, configKey, selector.by.scopeId);
	if (selector.by._tag === "codeAndLevel") return await getConfigurationByKey$1(context, configKey, selector.by.code, selector.by.level);
	if (selector.by._tag === "commerceScopeId") return await getConfigurationByKey$1(context, configKey, selector.by.commerceScopeId, selector.by.level);
	return await getConfigurationByKey$1(context, configKey, selector.by.code);
}
/**
* Sets configuration values for a scope.
*
* This function updates configuration values for a specific scope. The provided values
* are merged with existing configuration, and the origin is set to the current scope.
* Configuration values are inherited from parent scopes unless explicitly overridden.
*
* @param request - Configuration set request containing the config values to set.
* @param selector - Scope selector specifying how to identify the scope.
* @param options - Optional library configuration options for cache timeout.
* @returns Promise resolving to configuration response with updated scope and config values.
*
* @example
* ```typescript
* import { setConfiguration, byScopeId, byCodeAndLevel } from "@adobe/aio-commerce-lib-config";
*
* // Set configuration by scope ID
* const result1 = await setConfiguration(
*   {
*     config: [
*       { name: "api_key", value: "your-api-key-here" },
*       { name: "enable_feature", value: true },
*     ],
*   },
*   byScopeId("scope-123")
* );
*
* // Set configuration by code and level
* const result2 = await setConfiguration(
*   {
*     config: [
*       { name: "timeout", value: 5000 },
*       { name: "retry_count", value: 3 },
*     ],
*   },
*   byCodeAndLevel("website", "website")
* );
*
* console.log(result2.message); // "Configuration values updated successfully"
* console.log(result2.scope); // Updated scope information
* console.log(result2.config); // Array of updated config values
* ```
*/
async function setConfiguration(request, selector, options) {
	const context = {
		namespace: DEFAULT_NAMESPACE,
		cacheTimeout: options?.cacheTimeout ?? 300,
		encryptionKey: options?.encryptionKey
	};
	if (selector.by._tag === "scopeId") return await setConfiguration$1(context, request, selector.by.scopeId);
	if (selector.by._tag === "codeAndLevel") return await setConfiguration$1(context, request, selector.by.code, selector.by.level);
	if (selector.by._tag === "commerceScopeId") return await setConfiguration$1(context, request, selector.by.commerceScopeId, selector.by.level);
	return await setConfiguration$1(context, request, selector.by.code);
}
/**
* Sets the custom scope tree, replacing all existing custom scopes with the provided ones.
*
* Custom scopes allow you to define additional configuration scopes beyond the standard
* Commerce scopes (global, website, store, store_view). This function replaces all
* custom scopes, preserving system scopes (global and commerce).
*
* @param request - Custom scope tree request containing the scopes to set.
* @param options - Optional library configuration options for cache timeout.
* @returns Promise resolving to response with updated custom scopes.
*
* @example
* ```typescript
* import { setCustomScopeTree } from "@adobe/aio-commerce-lib-config";
*
* // Set custom scopes
* const result = await setCustomScopeTree(
*   {
*     scopes: [
*       {
*         code: "region_us",
*         label: "US Region",
*         level: "custom",
*         is_editable: true,
*         is_final: false,
*         children: [
*           {
*             code: "region_us_west",
*             label: "US West",
*             level: "custom",
*             is_editable: true,
*             is_final: true,
*           },
*         ],
*       },
*     ],
*   }
* );
*
* console.log(result.message); // "Custom scope tree updated successfully"
* console.log(result.scopes); // Array of created/updated custom scopes
* ```
*
* @example
* ```typescript
* import { setCustomScopeTree } from "@adobe/aio-commerce-lib-config";
*
* // Update existing custom scope (preserves ID if code and level match)
* const result = await setCustomScopeTree(
*   {
*     scopes: [
*       {
*         id: "existing-scope-id", // Preserve existing ID
*         code: "region_eu",
*         label: "European Region",
*         level: "custom",
*         is_editable: true,
*         is_final: false,
*       },
*     ],
*   }
* );
* ```
*/
async function setCustomScopeTree(request, options) {
	return await setCustomScopeTree$1({
		namespace: DEFAULT_NAMESPACE,
		cacheTimeout: options?.cacheTimeout ?? 300
	}, request);
}

//#endregion
//#region source/modules/schema/index.ts
/** The schema used to validate the business configuration settings. */
const SchemaBusinessConfig = v.object({ schema: v.optional(SchemaBusinessConfigSchema, []) });

//#endregion
export { SchemaBusinessConfig, byCode, byCodeAndLevel, byScopeId, byStoreId, byStoreViewId, byWebsiteId, generateEncryptionKey, getConfiguration, getConfigurationByKey, getScopeTree, hasDynamicSchema, initialize, resolveBusinessConfigSchema, setConfiguration, setCustomScopeTree, syncCommerceScopes, unsyncCommerceScopes, validateEncryptionKey };
