// @adobe/aio-commerce-lib-events/dist/es/io-events/index.mjs
/**
 * @license
 *
 * Copyright 2026 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */

import { i as booleanValueSchema, o as stringValueSchema, s as __exportAll, t as parseOrThrow } from "../utils-BMSXSLE3.mjs";
import * as v from "valibot";
import { AdobeIoEventsHttpClient, ApiClient } from "@adobe/aio-commerce-lib-api";

//#region source/io-events/api/event-metadata/schema.ts
const BASE64_ENCODED_STRING_REGEX = /^[A-Za-z0-9+/]*={0,2}$/;
const BASE64_ENCODED_STRING_LENGTH_REMAINDER = 4;
function isBase64EncodedString(value) {
	return BASE64_ENCODED_STRING_REGEX.test(value) && value.length % BASE64_ENCODED_STRING_LENGTH_REMAINDER === 0;
}
function encodeToBase64(value) {
	return Buffer.from(value).toString("base64");
}
function tryDecodeFromBase64(value) {
	if (isBase64EncodedString(value)) return Buffer.from(value, "base64").toString("utf-8");
	return value;
}
function sampleEventTemplateSchema(fieldName) {
	return v.pipe(v.union([
		v.pipe(stringValueSchema(fieldName), v.transform(tryDecodeFromBase64), v.parseJson(void 0, `Expected valid JSON string for property '${fieldName}'`), v.record(v.string(), v.unknown())),
		v.record(v.string(), v.unknown()),
		v.array(v.unknown())
	]), v.stringifyJson(void 0, `Expected valid JSON data for property '${fieldName}'`), v.transform(encodeToBase64));
}
/** The schema of the parameters received by the GET `providers/:id/eventmetadata` Adobe I/O Events API endpoint. */
const GetAllEventMetadataForProviderSchema = v.object({ providerId: stringValueSchema("providerId") });
/** The schema of the parameters received by the GET `providers/:id/eventmetadata/:code` Adobe I/O Events API endpoint. */
const GetEventMetadataForEventAndProviderSchema = v.object({
	eventCode: stringValueSchema("eventCode"),
	providerId: stringValueSchema("providerId")
});
/** The schema of the parameters received by the POST `providers/:id/eventmetadata` Adobe I/O Events API endpoint. */
const CreateEventMetadataForProviderSchema = v.object({
	consumerOrgId: stringValueSchema("consumerOrgId"),
	projectId: stringValueSchema("projectId"),
	workspaceId: stringValueSchema("workspaceId"),
	providerId: stringValueSchema("providerId"),
	label: stringValueSchema("label"),
	description: stringValueSchema("description"),
	eventCode: stringValueSchema("eventCode"),
	sampleEventTemplate: v.optional(sampleEventTemplateSchema("sampleEventTemplate"))
});
/** The schema of the parameters received by the DELETE `providers/:id/eventmetadata/:code` Adobe I/O Events API endpoint. */
const DeleteEventMetadataForProviderSchema = v.object({
	consumerOrgId: stringValueSchema("consumerOrgId"),
	projectId: stringValueSchema("projectId"),
	workspaceId: stringValueSchema("workspaceId"),
	providerId: stringValueSchema("providerId"),
	eventCode: stringValueSchema("eventCode")
});

//#endregion
//#region source/io-events/api/event-metadata/endpoints.ts
var endpoints_exports$2 = /* @__PURE__ */ __exportAll({
	createEventMetadataForProvider: () => createEventMetadataForProvider,
	deleteEventMetadataForProvider: () => deleteEventMetadataForProvider,
	getAllEventMetadataForProvider: () => getAllEventMetadataForProvider,
	getEventMetadataForEventAndProvider: () => getEventMetadataForEventAndProvider
});
/**
* Gets all event metadata for a specific provider.
* @see https://developer.adobe.com/events/docs/api#operation/getByProviderId
*
* @param httpClient - The {@link AdobeIoEventsHttpClient} to use to make the request.
* @param params - The parameters to get the event metadata with.
* @param fetchOptions - The {@link Options} to use to make the request.
*
* @throws A {@link CommerceSdkValidationError} If the parameters are in the wrong format.
* @throws An {@link HTTPError} If the status code is not 2XX.
*/
async function getAllEventMetadataForProvider(httpClient, params, fetchOptions) {
	const validatedParams = parseOrThrow(GetAllEventMetadataForProviderSchema, params);
	return httpClient.get(`providers/${validatedParams.providerId}/eventmetadata`, fetchOptions).json();
}
/**
* Gets event metadata for a specific event code and provider.
* @see https://developer.adobe.com/events/docs/api#operation/getByProviderIdAndEventCode
*
* @param httpClient - The {@link AdobeIoEventsHttpClient} to use to make the request.
* @param params - The parameters to get the event metadata with.
* @param fetchOptions - The {@link Options} to use to make the request.
*
* @throws A {@link CommerceSdkValidationError} If the parameters are in the wrong format.
* @throws An {@link HTTPError} If the status code is not 2XX.
*/
async function getEventMetadataForEventAndProvider(httpClient, params, fetchOptions) {
	const validatedParams = parseOrThrow(GetEventMetadataForEventAndProviderSchema, params);
	return httpClient.get(`providers/${validatedParams.providerId}/eventmetadata/${validatedParams.eventCode}`, fetchOptions).json();
}
/**
* Creates event metadata for a provider.
* @see https://developer.adobe.com/events/docs/api#operation/postEventMetadata
*
* @param httpClient - The {@link AdobeIoEventsHttpClient} to use to make the request.
* @param params - The parameters to create the event metadata with.
* @param fetchOptions - The {@link Options} to use to make the request.
*
* @throws A {@link CommerceSdkValidationError} If the parameters are in the wrong format.
* @throws An {@link HTTPError} If the status code is not 2XX.
*/
async function createEventMetadataForProvider(httpClient, params, fetchOptions) {
	const validatedParams = parseOrThrow(CreateEventMetadataForProviderSchema, params);
	return httpClient.post(`${validatedParams.consumerOrgId}/${validatedParams.projectId}/${validatedParams.workspaceId}/providers/${validatedParams.providerId}/eventmetadata`, {
		...fetchOptions,
		json: {
			label: validatedParams.label,
			description: validatedParams.description,
			event_code: validatedParams.eventCode,
			sample_event_template: validatedParams.sampleEventTemplate
		}
	}).json();
}
/**
* Deletes event metadata for a specific event code and provider.
* @see https://developer.adobe.com/events/docs/api#operation/deleteEventMetadata
*
* @param httpClient - The {@link AdobeIoEventsHttpClient} to use to make the request.
* @param params - The parameters to delete the event metadata with.
* @param fetchOptions - The {@link Options} to use to make the request.
*
* @throws A {@link CommerceSdkValidationError} If the parameters are in the wrong format.
* @throws An {@link HTTPError} If the status code is not 2XX.
*/
async function deleteEventMetadataForProvider(httpClient, params, fetchOptions) {
	const validatedParams = parseOrThrow(DeleteEventMetadataForProviderSchema, params);
	return httpClient.delete(`${validatedParams.consumerOrgId}/${validatedParams.projectId}/${validatedParams.workspaceId}/providers/${validatedParams.providerId}/eventmetadata/${validatedParams.eventCode}`, fetchOptions).then((_res) => {});
}

//#endregion
//#region source/utils/query-params.ts
/**
* Sets a query parameter if the value is truthy.
* @param queryParams The URLSearchParams instance to set the query parameter on.
* @param key The key of the query parameter to set.
* @param value The value of the query parameter to set.
*/
function setQueryParamIfTruthy(queryParams, key, value) {
	if (value) queryParams.set(key, `${value}`);
}
/**
* Sets an array query parameter if the values are truthy.
* @param queryParams The URLSearchParams instance to set the query parameter on.
* @param key The key of the query parameter to set.
* @param values The values of the query parameter to set.
*/
function setArrayQueryParam(queryParams, key, values) {
	if (values !== void 0 && values.length > 0) for (const value of values) queryParams.set(key, `${value}`);
}

//#endregion
//#region source/io-events/lib/schema.ts
/**
* The provider types that this library is able to handle.
* We deliberately don't support any other provider types.
*/
const VALID_EVENT_PROVIDER_TYPES = ["dx_commerce_events", "3rd_party_custom_events"];
/** The data residency regions that this I/O Events API supports. */
const VALID_DATA_RESIDENCY_REGIONS = ["va6", "irl1"];
const EventProviderTypeSchema = v.picklist(VALID_EVENT_PROVIDER_TYPES);
const DataResidencyRegionSchema = v.picklist(VALID_DATA_RESIDENCY_REGIONS);

//#endregion
//#region source/io-events/api/event-providers/schema.ts
const EventProviderListAllParamsSchema = v.object({
	consumerOrgId: stringValueSchema("consumerOrgId"),
	withEventMetadata: v.optional(booleanValueSchema("withEventMetadata")),
	filterBy: v.optional(v.object({
		instanceId: v.optional(stringValueSchema("instanceId")),
		providerTypes: v.optional(v.union([v.array(EventProviderTypeSchema, "Expected an array of event provider types")]))
	}))
});
const EventProviderGetByIdParamsSchema = v.object({
	providerId: stringValueSchema("providerId"),
	withEventMetadata: v.optional(booleanValueSchema("withEventMetadata"))
});
const EventProviderCreateParamsSchema = v.object({
	consumerOrgId: stringValueSchema("consumerOrgId"),
	projectId: stringValueSchema("projectId"),
	workspaceId: stringValueSchema("workspaceId"),
	label: stringValueSchema("label"),
	description: v.optional(stringValueSchema("description")),
	docsUrl: v.optional(stringValueSchema("docsUrl")),
	instanceId: v.optional(stringValueSchema("instanceId")),
	providerType: v.optional(EventProviderTypeSchema),
	dataResidencyRegion: v.optional(DataResidencyRegionSchema)
});
const EventProviderDeleteParamsSchema = v.object({
	consumerOrgId: stringValueSchema("consumerOrgId"),
	projectId: stringValueSchema("projectId"),
	workspaceId: stringValueSchema("workspaceId"),
	providerId: stringValueSchema("providerId")
});

//#endregion
//#region source/io-events/api/event-providers/endpoints.ts
var endpoints_exports$1 = /* @__PURE__ */ __exportAll({
	createEventProvider: () => createEventProvider,
	deleteEventProvider: () => deleteEventProvider,
	getAllEventProviders: () => getAllEventProviders,
	getEventProviderById: () => getEventProviderById
});
/**
* Lists all event providers for the given consumer organization ID.
* @see https://developer.adobe.com/events/docs/api#operation/getProvidersByConsumerOrgId
*
* @param httpClient - The {@link AdobeIoEventsHttpClient} to use to make the request.
* @param params - The parameters to list the event providers with.
* @param fetchOptions - The {@link Options} to use to make the request.
*/
async function getAllEventProviders(httpClient, params, fetchOptions) {
	const validatedParams = parseOrThrow(EventProviderListAllParamsSchema, params);
	const queryParams = new URLSearchParams();
	const { providerTypes = [], instanceId } = validatedParams.filterBy ?? {};
	setArrayQueryParam(queryParams, "providerMetadataIds", providerTypes);
	setQueryParamIfTruthy(queryParams, "instanceId", instanceId);
	setQueryParamIfTruthy(queryParams, "eventmetadata", validatedParams.withEventMetadata);
	return httpClient.get(`${validatedParams.consumerOrgId}/providers`, {
		...fetchOptions,
		searchParams: queryParams
	}).json();
}
/**
* Gets an event provider by ID.
* @see https://developer.adobe.com/events/docs/api#operation/getProvidersById
*
* @param httpClient - The {@link AdobeIoEventsHttpClient} to use to make the request.
* @param params - The parameters to get the event provider by.
* @param fetchOptions - The {@link Options} to use to make the request.
*
* @throws A {@link CommerceSdkValidationError} If the parameters are in the wrong format.
* @throws An {@link HTTPError} If the status code is not 2XX.
*/
async function getEventProviderById(httpClient, params, fetchOptions) {
	const validatedParams = parseOrThrow(EventProviderGetByIdParamsSchema, params);
	const searchParams = new URLSearchParams();
	setQueryParamIfTruthy(searchParams, "eventmetadata", validatedParams.withEventMetadata);
	return httpClient.get(`providers/${params.providerId}`, {
		...fetchOptions,
		searchParams
	}).json();
}
/**
* Creates an event provider.
* @see https://developer.adobe.com/events/docs/api#operation/createProvider
*
* @param httpClient - The {@link AdobeIoEventsHttpClient} to use to make the request.
* @param params - The parameters to create the event provider with.
* @param fetchOptions - The {@link Options} to use to make the request.
*
* @throws A {@link CommerceSdkValidationError} If the parameters are in the wrong format.
* @throws An {@link HTTPError} If the status code is not 2XX.
*/
async function createEventProvider(httpClient, params, fetchOptions) {
	const validatedParams = parseOrThrow(EventProviderCreateParamsSchema, params);
	return httpClient.post(`${validatedParams.consumerOrgId}/${validatedParams.projectId}/${validatedParams.workspaceId}/providers`, {
		...fetchOptions,
		json: {
			...validatedParams,
			docs_url: validatedParams.docsUrl,
			provider_metadata: validatedParams.providerType,
			instance_id: validatedParams.instanceId,
			data_residency_region: validatedParams.dataResidencyRegion
		}
	}).json();
}
/**
* Deletes an event provider.
* @see https://developer.adobe.com/events/docs/api#operation/deleteProvider
*
* @param httpClient - The {@link AdobeIoEventsHttpClient} to use to make the request.
* @param params - The parameters to delete the event provider with.
* @param fetchOptions - The {@link Options} to use to make the request.
*
* @throws A {@link CommerceSdkValidationError} If the parameters are in the wrong format.
* @throws An {@link HTTPError} If the status code is not 2XX.
*/
async function deleteEventProvider(httpClient, params, fetchOptions) {
	const validatedParams = parseOrThrow(EventProviderDeleteParamsSchema, params);
	return httpClient.delete(`${validatedParams.consumerOrgId}/${validatedParams.projectId}/${validatedParams.workspaceId}/providers/${validatedParams.providerId}`, fetchOptions).then((_res) => {});
}

//#endregion
//#region source/io-events/api/event-providers/shorthands.ts
var shorthands_exports = /* @__PURE__ */ __exportAll({
	create3rdPartyCustomEventProvider: () => create3rdPartyCustomEventProvider,
	createCommerceEventProvider: () => createCommerceEventProvider,
	getAll3rdPartyCustomEventProviders: () => getAll3rdPartyCustomEventProviders,
	getAllCommerceEventProviders: () => getAllCommerceEventProviders
});
/**
* Lists all Commerce (of type `dx_commerce_events`) event providers for the given consumer organization ID.
* @see https://developer.adobe.com/events/docs/api#operation/getProvidersByConsumerOrgId
*
* @param httpClient - The {@link AdobeIoEventsHttpClient} to use to make the request.
* @param params - The parameters to list the event providers with.
* @param fetchOptions - The {@link Options} to use to make the request.
*
* @throws A {@link CommerceSdkValidationError} If the parameters are in the wrong format.
* @throws An {@link HTTPError} If the status code is not 2XX.
*/
async function getAllCommerceEventProviders(httpClient, params, fetchOptions) {
	return getAllEventProviders(httpClient, {
		...params,
		filterBy: {
			...params.filterBy,
			providerTypes: ["dx_commerce_events"]
		}
	}, fetchOptions);
}
/**
* Lists all 3rd Party Custom Events (of type `3rd_party_custom_events`) event providers for the given consumer organization ID.
* @see https://developer.adobe.com/events/docs/api#operation/getProvidersByConsumerOrgId
*
* @param httpClient - The {@link AdobeIoEventsHttpClient} to use to make the request.
* @param params - The parameters to list the event providers with.
* @param fetchOptions - The {@link Options} to use to make the request.
*
* @throws A {@link CommerceSdkValidationError} If the parameters are in the wrong format.
* @throws An {@link HTTPError} If the status code is not 2XX.
*/
async function getAll3rdPartyCustomEventProviders(httpClient, params, fetchOptions) {
	return getAllEventProviders(httpClient, {
		...params,
		filterBy: {
			...params.filterBy,
			providerTypes: ["3rd_party_custom_events"]
		}
	}, fetchOptions);
}
/**
* Creates a Commerce (of type `dx_commerce_events`) event provider.
* @see https://developer.adobe.com/events/docs/api#operation/createProvider
*
* @param httpClient - The {@link AdobeIoEventsHttpClient} to use to make the request.
* @param params - The parameters to create the event provider with.
* @param fetchOptions - The {@link Options} to use to make the request.
*
* @throws A {@link CommerceSdkValidationError} If the parameters are in the wrong format.
* @throws An {@link HTTPError} If the status code is not 2XX.
*/
async function createCommerceEventProvider(httpClient, params, fetchOptions) {
	return createEventProvider(httpClient, {
		...params,
		providerType: "dx_commerce_events"
	}, fetchOptions);
}
/**
* Creates a 3rd Party Custom Events (of type `3rd_party_custom_events`) event provider.
* @see https://developer.adobe.com/events/docs/api#operation/createProvider
*
* @param httpClient - The {@link AdobeIoEventsHttpClient} to use to make the request.
* @param params - The parameters to create the event provider with.
* @param fetchOptions - The {@link Options} to use to make the request.
*
* @throws A {@link CommerceSdkValidationError} If the parameters are in the wrong format.
* @throws An {@link HTTPError} If the status code is not 2XX.
*/
async function create3rdPartyCustomEventProvider(httpClient, params, fetchOptions) {
	return createEventProvider(httpClient, {
		...params,
		providerType: "3rd_party_custom_events"
	}, fetchOptions);
}

//#endregion
//#region source/io-events/api/event-registrations/schema.ts
/** Valid delivery types for event registrations. */
const VALID_DELIVERY_TYPES = [
	"webhook",
	"webhook_batch",
	"journal",
	"aws_eventbridge"
];
/** Regex pattern for AWS region validation. */
const AWS_REGION_REGEX = /^(us|ca|eu|af|ap|cn|me|sa|il)-[a-z]+-\d$/;
/** Regex pattern for AWS account ID validation (12 digits). */
const AWS_ACCOUNT_ID_REGEX = /^\d{12}$/;
/** Schema for delivery type validation. */
const DeliveryTypeSchema = v.picklist(VALID_DELIVERY_TYPES, `Expected delivery type to be one of: ${VALID_DELIVERY_TYPES.join(", ")}`);
/** Schema for events of interest. */
const EventsOfInterestSchema = v.object({
	providerId: stringValueSchema("providerId"),
	eventCode: stringValueSchema("eventCode")
});
/** Schema for AWS EventBridge destination metadata. */
const DestinationMetadataSchema = v.object({
	awsRegion: v.optional(v.pipe(stringValueSchema("awsRegion"), v.regex(AWS_REGION_REGEX, "Expected AWS region in format like 'us-east-1'"))),
	awsAccountId: v.optional(v.pipe(stringValueSchema("awsAccountId"), v.regex(AWS_ACCOUNT_ID_REGEX, "Expected AWS account ID to be a 12-digit number")))
});
/** Schema for subscriber-defined filter. */
const SubscriberFilterSchema = v.object({
	name: v.pipe(stringValueSchema("name"), v.minLength(3, "Expected subscriber filter name to be at least 3 characters"), v.maxLength(80, "Expected subscriber filter name to be at most 80 characters")),
	description: v.optional(v.pipe(stringValueSchema("description"), v.maxLength(250, "Expected subscriber filter description to be at most 250 characters"))),
	subscriberFilter: stringValueSchema("subscriberFilter")
});
/** Shared schema for workspace path parameters. */
const WorkspacePathParamsSchema = v.object({
	consumerOrgId: stringValueSchema("consumerOrgId"),
	projectId: stringValueSchema("projectId"),
	workspaceId: stringValueSchema("workspaceId")
});
/** Shared schema for events of interest array field. */
const EventsOfInterestArraySchema = v.pipe(v.array(EventsOfInterestSchema, "Expected eventsOfInterest to be an array of event interest objects"), v.minLength(1, "Expected at least one event of interest"));
/** Shared schema for registration body fields. */
const RegistrationBodySchema = v.object({
	clientId: v.pipe(stringValueSchema("clientId"), v.minLength(3, "Expected clientId to be at least 1 character"), v.maxLength(255, "Expected clientId to be at most 255 characters")),
	name: v.pipe(stringValueSchema("name"), v.minLength(3, "Expected registration name to be at least 3 characters"), v.maxLength(255, "Expected registration name to be at most 255 characters")),
	description: v.optional(v.pipe(stringValueSchema("description"), v.maxLength(5e3, "Expected description to be at most 5000 characters"))),
	webhookUrl: v.optional(v.pipe(stringValueSchema("webhookUrl"), v.maxLength(4e3, "Expected webhook URL to be at most 4000 characters"))),
	eventsOfInterest: EventsOfInterestArraySchema,
	deliveryType: DeliveryTypeSchema,
	runtimeAction: v.optional(v.pipe(stringValueSchema("runtimeAction"), v.maxLength(255, "Expected runtime action to be at most 255 characters"))),
	enabled: v.optional(booleanValueSchema("enabled")),
	destinationMetadata: v.optional(DestinationMetadataSchema),
	subscriberFilters: v.optional(v.pipe(v.array(SubscriberFilterSchema, "Expected subscriberFilters to be an array of subscriber filter objects"), v.maxLength(1, "Expected at most 1 subscriber filter")))
});
/** Schema for getting all registrations for a workspace. */
const GetAllRegistrationsParamsSchema = WorkspacePathParamsSchema;
/** Schema for getting all registrations for a consumer organization. */
const GetAllRegistrationsByConsumerOrgParamsSchema = v.object({ consumerOrgId: stringValueSchema("consumerOrgId") });
/** Schema for getting a registration by ID. */
const GetRegistrationByIdParamsSchema = v.object({
	...WorkspacePathParamsSchema.entries,
	registrationId: stringValueSchema("registrationId")
});
/** Schema for creating a registration. */
const CreateRegistrationParamsSchema = v.object({
	...WorkspacePathParamsSchema.entries,
	...RegistrationBodySchema.entries
});
/** Schema for updating a registration. */
const UpdateRegistrationParamsSchema = v.object({
	...WorkspacePathParamsSchema.entries,
	registrationId: stringValueSchema("registrationId"),
	...RegistrationBodySchema.entries
});
/** Schema for deleting a registration. */
const DeleteRegistrationParamsSchema = v.object({
	...WorkspacePathParamsSchema.entries,
	registrationId: stringValueSchema("registrationId")
});

//#endregion
//#region source/io-events/api/event-registrations/endpoints.ts
var endpoints_exports = /* @__PURE__ */ __exportAll({
	createRegistration: () => createRegistration,
	deleteRegistration: () => deleteRegistration,
	getAllRegistrations: () => getAllRegistrations,
	getAllRegistrationsByConsumerOrg: () => getAllRegistrationsByConsumerOrg,
	getRegistrationById: () => getRegistrationById,
	updateRegistration: () => updateRegistration
});
/**
* Gets all event registrations for a consumer organization (paginated).
* @see https://developer.adobe.com/events/docs/api#operation/getAllRegistrationsForOrg
*
* @param httpClient - The {@link AdobeIoEventsHttpClient} to use to make the request.
* @param params - The parameters to get the registrations with.
* @param fetchOptions - The {@link Options} to use to make the request.
*
* @throws A {@link CommerceSdkValidationError} If the parameters are in the wrong format.
* @throws An {@link HTTPError} If the status code is not 2XX.
*/
async function getAllRegistrationsByConsumerOrg(httpClient, params, fetchOptions) {
	const validatedParams = parseOrThrow(GetAllRegistrationsByConsumerOrgParamsSchema, params);
	return httpClient.get(`${validatedParams.consumerOrgId}/registrations`, fetchOptions).json();
}
/**
* Gets all event registrations for a workspace.
* @see https://developer.adobe.com/events/docs/api#operation/getAllRegistrations
*
* @param httpClient - The {@link AdobeIoEventsHttpClient} to use to make the request.
* @param params - The parameters to get the registrations with.
* @param fetchOptions - The {@link Options} to use to make the request.
*
* @throws A {@link CommerceSdkValidationError} If the parameters are in the wrong format.
* @throws An {@link HTTPError} If the status code is not 2XX.
*/
async function getAllRegistrations(httpClient, params, fetchOptions) {
	const validatedParams = parseOrThrow(GetAllRegistrationsParamsSchema, params);
	return httpClient.get(`${validatedParams.consumerOrgId}/${validatedParams.projectId}/${validatedParams.workspaceId}/registrations`, fetchOptions).json();
}
/**
* Gets an event registration by ID.
* @see https://developer.adobe.com/events/docs/api#operation/getRegistration
*
* @param httpClient - The {@link AdobeIoEventsHttpClient} to use to make the request.
* @param params - The parameters to get the registration with.
* @param fetchOptions - The {@link Options} to use to make the request.
*
* @throws A {@link CommerceSdkValidationError} If the parameters are in the wrong format.
* @throws An {@link HTTPError} If the status code is not 2XX.
*/
async function getRegistrationById(httpClient, params, fetchOptions) {
	const validatedParams = parseOrThrow(GetRegistrationByIdParamsSchema, params);
	return httpClient.get(`${validatedParams.consumerOrgId}/${validatedParams.projectId}/${validatedParams.workspaceId}/registrations/${validatedParams.registrationId}`, fetchOptions).json();
}
/**
* Creates an event registration.
* @see https://developer.adobe.com/events/docs/api#operation/createRegistration
*
* @param httpClient - The {@link AdobeIoEventsHttpClient} to use to make the request.
* @param params - The parameters to create the registration with.
* @param fetchOptions - The {@link Options} to use to make the request.
*
* @throws A {@link CommerceSdkValidationError} If the parameters are in the wrong format.
* @throws An {@link HTTPError} If the status code is not 2XX.
*/
async function createRegistration(httpClient, params, fetchOptions) {
	const validatedParams = parseOrThrow(CreateRegistrationParamsSchema, params);
	return httpClient.post(`${validatedParams.consumerOrgId}/${validatedParams.projectId}/${validatedParams.workspaceId}/registrations`, {
		...fetchOptions,
		json: {
			client_id: validatedParams.clientId,
			name: validatedParams.name,
			description: validatedParams.description,
			webhook_url: validatedParams.webhookUrl,
			events_of_interest: validatedParams.eventsOfInterest.map((event) => ({
				provider_id: event.providerId,
				event_code: event.eventCode
			})),
			delivery_type: validatedParams.deliveryType,
			runtime_action: validatedParams.runtimeAction,
			enabled: validatedParams.enabled,
			destination_metadata: validatedParams.destinationMetadata ? {
				aws_region: validatedParams.destinationMetadata.awsRegion,
				aws_account_id: validatedParams.destinationMetadata.awsAccountId
			} : void 0,
			subscriber_filters: validatedParams.subscriberFilters?.map((filter) => ({
				name: filter.name,
				description: filter.description,
				subscriber_filter: filter.subscriberFilter
			}))
		}
	}).json();
}
/**
* Updates an event registration.
* @see https://developer.adobe.com/events/docs/api#operation/updateRegistration
*
* @param httpClient - The {@link AdobeIoEventsHttpClient} to use to make the request.
* @param params - The parameters to update the registration with.
* @param fetchOptions - The {@link Options} to use to make the request.
*
* @throws A {@link CommerceSdkValidationError} If the parameters are in the wrong format.
* @throws An {@link HTTPError} If the status code is not 2XX.
*/
async function updateRegistration(httpClient, params, fetchOptions) {
	const validatedParams = parseOrThrow(UpdateRegistrationParamsSchema, params);
	return httpClient.put(`${validatedParams.consumerOrgId}/${validatedParams.projectId}/${validatedParams.workspaceId}/registrations/${validatedParams.registrationId}`, {
		...fetchOptions,
		json: {
			name: validatedParams.name,
			description: validatedParams.description,
			webhook_url: validatedParams.webhookUrl,
			events_of_interest: validatedParams.eventsOfInterest.map((event) => ({
				provider_id: event.providerId,
				event_code: event.eventCode
			})),
			delivery_type: validatedParams.deliveryType,
			runtime_action: validatedParams.runtimeAction,
			enabled: validatedParams.enabled,
			destination_metadata: validatedParams.destinationMetadata ? {
				aws_region: validatedParams.destinationMetadata.awsRegion,
				aws_account_id: validatedParams.destinationMetadata.awsAccountId
			} : void 0,
			subscriber_filters: validatedParams.subscriberFilters?.map((filter) => ({
				name: filter.name,
				description: filter.description,
				subscriber_filter: filter.subscriberFilter
			}))
		}
	}).json();
}
/**
* Deletes an event registration.
* @see https://developer.adobe.com/events/docs/api#operation/deleteRegistration
*
* @param httpClient - The {@link AdobeIoEventsHttpClient} to use to make the request.
* @param params - The parameters to delete the registration with.
* @param fetchOptions - The {@link Options} to use to make the request.
*
* @throws A {@link CommerceSdkValidationError} If the parameters are in the wrong format.
* @throws An {@link HTTPError} If the status code is not 2XX.
*/
async function deleteRegistration(httpClient, params, fetchOptions) {
	const validatedParams = parseOrThrow(DeleteRegistrationParamsSchema, params);
	return httpClient.delete(`${validatedParams.consumerOrgId}/${validatedParams.projectId}/${validatedParams.workspaceId}/registrations/${validatedParams.registrationId}`, fetchOptions).then((_res) => {});
}

//#endregion
//#region source/io-events/lib/api-client.ts
/**
* Creates a new API client for the Adobe I/O Events API client.
* @param params - The parameters to build the Adobe I/O Events HTTP client that will communicate with the Adobe I/O Events API.
*/
function createAdobeIoEventsApiClient(params) {
	return ApiClient.create(new AdobeIoEventsHttpClient(params), {
		...endpoints_exports$1,
		...endpoints_exports$2,
		...shorthands_exports,
		...endpoints_exports
	});
}
/**
* Creates a customized Adobe I/O Events API client.
* @param params - The parameters to build the Adobe I/O Events HTTP client that will communicate with the Adobe I/O Events API.
* @param functions - The API functions to include in the client.
*/
function createCustomAdobeIoEventsApiClient(params, functions) {
	return ApiClient.create(new AdobeIoEventsHttpClient(params), functions);
}

//#endregion
export { create3rdPartyCustomEventProvider, createAdobeIoEventsApiClient, createCommerceEventProvider, createCustomAdobeIoEventsApiClient, createEventMetadataForProvider, createEventProvider, createRegistration, deleteEventMetadataForProvider, deleteEventProvider, deleteRegistration, getAll3rdPartyCustomEventProviders, getAllCommerceEventProviders, getAllEventMetadataForProvider, getAllEventProviders, getAllRegistrations, getAllRegistrationsByConsumerOrg, getEventMetadataForEventAndProvider, getEventProviderById, getRegistrationById, updateRegistration };
