// @adobe/aio-commerce-lib-app/dist/es/actions/app-config/index.mjs
/**
 * @license
 *
 * Copyright 2026 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */

import { i as requiresInstallationFromDomains, o as getConfigDomains, t as validateCommerceAppConfig, u as hasBusinessConfigSchema } from "../../validate-Dfajrp99.mjs";
import { n as logger, t as HttpActionRouter } from "../../router-D58JJHHA.mjs";
import { resolveBusinessConfigSchema } from "@adobe/aio-commerce-lib-config";
import * as v from "valibot";
import { internalServerError, ok } from "@adobe/aio-commerce-lib-core/responses";
import { createHash } from "crypto";

//#region source/actions/app-config/openapi.ts
/**
* Loads the committed OpenAPI spec via dynamic import so it
* lands in its own chunk and is not parsed on every cold start.
*/
const importOpenApi = () => import("../../openapi-B6uP0SRn.mjs");
/** Matches component schema `$ref` strings, capturing the schema name. */
const SCHEMA_REF_PATTERN = /"#\/components\/schemas\/(\w+)"/g;
/**
* Collects the names of every component schema referenced via `$ref` in a value
* Scans the serialized JSON, so it catches refs at any depth.
*
* @param value - Any part of the spec to scan.
*/
function collectSchemaRefs(value) {
	return [...(JSON.stringify(value) ?? "").matchAll(SCHEMA_REF_PATTERN)].map((match) => match[1]);
}
/**
* Deletes component schemas not reachable from the remaining paths, in place.
*
* Follows `$ref`s transitively from `spec.paths`; the visited set handles
* recursive schemas (e.g. `ScopeNode`).
*
* @param spec - The spec to prune, after paths have been stripped.
*/
function pruneUnusedSchemas(spec) {
	const schemas = spec.components.schemas;
	const reachable = new Set(collectSchemaRefs(spec.paths));
	const pending = [...reachable];
	while (pending.length > 0) {
		const name = pending.pop();
		if (name === void 0) continue;
		for (const ref of collectSchemaRefs(schemas[name])) if (!reachable.has(ref)) {
			reachable.add(ref);
			pending.push(ref);
		}
	}
	for (const name of Object.keys(schemas)) if (!reachable.has(name)) delete schemas[name];
}
/**
* Removes top-level tags no longer referenced by any remaining path operation, in place.
* @param spec - The spec to prune, after paths have been stripped.
*/
function pruneUnusedTags(spec) {
	const usedTags = /* @__PURE__ */ new Set();
	for (const pathItem of Object.values(spec.paths)) for (const operation of Object.values(pathItem)) for (const tag of operation.tags ?? []) usedTags.add(tag);
	for (let i = spec.tags.length - 1; i >= 0; i--) if (!usedTags.has(spec.tags[i].name)) spec.tags.splice(i, 1);
}
/**
* Computes a short hash identifying the spec served for the given config domains.
* Changes whenever the served spec would: with the OpenAPI spec version (`info.version`),
* the package version, or the domains that drive pruning.
*
* @param domains - The active config domains.
*/
function getOpenApiCacheKey(domains) {
	const input = `1.7.0:${[...domains].sort().join(",")}:1.0.0`;
	return createHash("sha256").update(input).digest("hex").slice(0, 8);
}
/** Returns the server URL to be set in the OpenAPI spec, based on the current namespace. */
function getServerUrl() {
	return `https://${process.env.__OW_NAMESPACE}.adobeioruntime.net/api/v1/web/app-management`;
}
/**
* Builds the OpenAPI spec served by `GET /openapi.json`, tailored to the app's
* capabilities: unused paths are stripped, along with the schemas they leave
* unreferenced.
*
* Works on a fresh copy so the shared module import is never mutated. The
* committed `docs/openapi.json` always describes the full surface.
*
* @param domains - The active config domains that drive path trimming.
* @param logger - Logger used to report which paths get stripped.
*/
async function buildOpenApiSpec(domains, logger) {
	const { default: openAPISpec } = await importOpenApi();
	const spec = structuredClone(openAPISpec);
	const stripPath = (path) => {
		if (spec.paths[path]) {
			logger.debug(`Stripping OpenAPI spec path: ${path}`);
			delete spec.paths[path];
		}
	};
	if (!domains.has("businessConfig.schema")) {
		logger.debug("Application doesn't define business configuration, stripping references...");
		stripPath("/config");
		stripPath("/scope-tree");
		stripPath("/scope-tree/commerce");
	}
	if (!domains.has("adminUiSdk")) {
		logger.debug("Application doesn't define Admin UI SDK registration, stripping references...");
		stripPath("/registration");
	}
	if (!requiresInstallationFromDomains(domains)) {
		logger.debug("Application doesn't require installation, stripping references...");
		stripPath("/installation");
		stripPath("/installation/validation");
		stripPath("/installation/uninstallation");
	}
	pruneUnusedSchemas(spec);
	pruneUnusedTags(spec);
	spec.servers[0].url = getServerUrl();
	return spec;
}

//#endregion
//#region source/actions/app-config/router.ts
/**
* App Config action router.
*
* Routes:
* - GET /               Retrieve the Commerce App configuration.
* - GET /openapi.json   Returns the OpenAPI spec for all SDK actions
*/
const router = new HttpActionRouter().use(logger({ name: () => "app-config" }));
/** GET / - Get app config */
router.get("/", { handler: async (_req, { logger, rawParams }) => {
	const rawAppConfig = rawParams.appConfig;
	if (!rawAppConfig) return internalServerError("The app config is missing. Does the action receive it as a parameter?");
	let appConfig = rawAppConfig;
	if (hasBusinessConfigSchema(rawAppConfig)) {
		logger.debug("Resolving business config schema...");
		const schema = await resolveBusinessConfigSchema(rawAppConfig.businessConfig.schema, rawParams);
		appConfig = {
			...rawAppConfig,
			businessConfig: {
				...rawAppConfig.businessConfig,
				schema
			}
		};
	}
	logger.debug("Validating app config...");
	const config = validateCommerceAppConfig(appConfig);
	logger.debug("Successfully validated the app config");
	const domains = getConfigDomains(rawParams.appConfig);
	const openApiSpecUrl = `${getServerUrl()}/app-config/openapi.json?ck=${getOpenApiCacheKey(domains)}`;
	return ok({ body: {
		...config,
		openApiSpecUrl
	} });
} });
/**
* GET /openapi.json - Returns the OpenAPI spec for all SDK actions
* @internal - Do not add to OpenAPI Spec.
*/
router.get("/openapi.json", {
	query: v.object({ ck: v.optional(v.string()) }),
	handler: async (req, { logger, rawParams }) => {
		const { ck } = req.query;
		const domains = getConfigDomains(rawParams.appConfig);
		if (ck && getOpenApiCacheKey(domains) === ck) {
			logger.debug(`Received request for OpenAPI spec with cache key query param: ${ck}`);
			return ok({
				body: await buildOpenApiSpec(domains, logger),
				headers: { "Cache-Control": "public, max-age=31536000, immutable" }
			});
		}
		return ok({ body: await buildOpenApiSpec(domains, logger) });
	}
});

//#endregion
//#region source/actions/app-config/index.ts
/**
* Factory to create the route handler for the `app-config` action.
* @param args - The arguments required to create the runtime action.
*/
const appConfigRuntimeAction = (args) => async (params) => {
	const handler = router.handler();
	const { appConfig } = args;
	return await handler({
		...params,
		appConfig
	});
};

//#endregion
export { appConfigRuntimeAction };
