// @adobe/aio-commerce-lib-app/dist/cjs/actions/scope-tree/index.cjs
/**
 * @license
 *
 * Copyright 2026 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
const require_chunk = require('../../chunk-Cozy-1ZF.cjs');
const require_schemas = require('../../schemas-FngDuIeo.cjs');
const require_logging = require('../../logging-BVz3weoC.cjs');
const require_router = require('../../router-BOR7T6qR.cjs');
let _adobe_aio_commerce_lib_config = require("@adobe/aio-commerce-lib-config");
let valibot = require("valibot");
valibot = require_chunk.__toESM(valibot, 1);
let _adobe_aio_commerce_lib_api = require("@adobe/aio-commerce-lib-api");
let _adobe_aio_commerce_lib_core_responses = require("@adobe/aio-commerce-lib-core/responses");

//#region source/actions/scope-tree/schema.ts
const CustomScopeInputSchema = valibot.object({
	id: valibot.optional(valibot.string()),
	code: valibot.string(),
	label: valibot.string(),
	level: valibot.optional(valibot.string()),
	is_editable: valibot.boolean(),
	is_final: valibot.boolean(),
	children: valibot.optional(valibot.array(valibot.lazy(() => CustomScopeInputSchema)))
});
/** Request body for PUT / */
const SetCustomScopeTreeBodySchema = valibot.object({ scopes: valibot.array(CustomScopeInputSchema) });
/** Request body for POST /commerce */
const SyncCommerceScopesBodySchema = valibot.object({
	commerceBaseUrl: require_schemas.nonEmptyStringValueSchema("commerceBaseUrl"),
	commerceEnv: valibot.optional(require_schemas.nonEmptyStringValueSchema("commerceEnv"))
});

//#endregion
//#region source/actions/scope-tree/router.ts
/**
* Scope Tree action router.
*
* Routes:
* - GET /              Retrieve the scope tree.
* - PUT /              Set custom scope tree.
* - POST /commerce     Sync commerce scopes
* - DELETE /commerce   Unsync commerce scopes
*/
const router = new require_router.HttpActionRouter().use(require_router.logger({ name: () => "scope-tree" }));
/** GET / - Get scope tree */
router.get("/", { handler: async (_req, ctx) => {
	const { logger } = ctx;
	const result = await (0, _adobe_aio_commerce_lib_config.getScopeTree)();
	logger.debug(`Successfully retrieved scope tree (cached: ${result.isCachedData}): ${require_logging.inspect(result.scopeTree)}`);
	if (result.isCachedData) return (0, _adobe_aio_commerce_lib_core_responses.nonAuthoritativeInformation)({
		headers: { "x-cache": "hit" },
		body: { scopes: result.scopeTree }
	});
	return (0, _adobe_aio_commerce_lib_core_responses.ok)({ body: { scopes: result.scopeTree } });
} });
/** PUT / - Set custom scope tree */
router.put("/", {
	body: SetCustomScopeTreeBodySchema,
	handler: async (req, ctx) => {
		const { logger } = ctx;
		logger.debug(`Setting custom scope tree with ${req.body.scopes?.length || 0} scopes`);
		const request = { scopes: req.body.scopes };
		let result;
		logger.debug(`Setting custom scope tree: ${require_logging.inspect(request)}`);
		try {
			result = await (0, _adobe_aio_commerce_lib_config.setCustomScopeTree)(request);
		} catch (err) {
			if (err instanceof Error && "isValidationError" in err) {
				logger.debug(`Custom scope tree validation failed: ${err.message}`);
				return (0, _adobe_aio_commerce_lib_core_responses.badRequest)({ body: { message: err.message } });
			}
			throw err;
		}
		logger.debug(`Successfully set custom scope tree: ${require_logging.inspect(result)}`);
		return (0, _adobe_aio_commerce_lib_core_responses.ok)({
			body: { result },
			headers: { "Cache-Control": "no-store" }
		});
	}
});
/** POST /commerce - Sync commerce scopes */
router.post("/commerce", {
	body: SyncCommerceScopesBodySchema,
	handler: async (req, ctx) => {
		const { logger } = ctx;
		logger.debug("Syncing commerce scopes...");
		const { commerceBaseUrl, commerceEnv } = req.body;
		const result = await (0, _adobe_aio_commerce_lib_config.syncCommerceScopes)((0, _adobe_aio_commerce_lib_api.resolveCommerceHttpClientParams)({
			...ctx.rawParams,
			AIO_COMMERCE_API_BASE_URL: commerceBaseUrl,
			...commerceEnv && { AIO_COMMERCE_API_FLAVOR: commerceEnv }
		}, { tryForwardAuthProvider: true }));
		if (result.error) {
			logger.error(`Error syncing commerce scopes: ${require_logging.inspect(result.error)}`);
			return (0, _adobe_aio_commerce_lib_core_responses.internalServerError)({ body: {
				message: "An internal server error occurred",
				error: result.error
			} });
		}
		if (!result.synced) {
			logger.debug(`Commerce scopes not synced (cached): ${require_logging.inspect(result.scopeTree)}`);
			return (0, _adobe_aio_commerce_lib_core_responses.nonAuthoritativeInformation)({
				headers: { "x-cache": "hit" },
				body: {
					scopes: result.scopeTree,
					synced: false
				}
			});
		}
		logger.debug(`Successfully synced commerce scopes: ${require_logging.inspect(result.scopeTree)}`);
		return (0, _adobe_aio_commerce_lib_core_responses.ok)({ body: {
			scopes: result.scopeTree,
			synced: true
		} });
	}
});
/** DELETE /commerce - Unsync commerce scopes */
router.delete("/commerce", { handler: async (_req, ctx) => {
	const { logger } = ctx;
	logger.debug("Unsyncing commerce scopes...");
	const { unsynced } = await (0, _adobe_aio_commerce_lib_config.unsyncCommerceScopes)();
	if (unsynced) {
		const message = "Commerce scopes unsynced successfully";
		logger.debug(message);
		return (0, _adobe_aio_commerce_lib_core_responses.ok)(message);
	}
	const message = "No commerce scopes to unsync";
	logger.debug(message);
	return (0, _adobe_aio_commerce_lib_core_responses.ok)(message);
} });

//#endregion
//#region source/actions/scope-tree/index.ts
/** The handler method for the `scope-tree` action. */
const scopeTreeRuntimeAction = router.handler();

//#endregion
exports.scopeTreeRuntimeAction = scopeTreeRuntimeAction;
