// @adobe/aio-commerce-lib-events/dist/cjs/commerce/index.cjs
/**
 * @license
 *
 * Copyright 2026 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
const require_utils = require('../utils-Da725lXV.cjs');
let _adobe_aio_commerce_lib_core_error = require("@adobe/aio-commerce-lib-core/error");
let valibot = require("valibot");
valibot = require_utils.__toESM(valibot, 1);
let _adobe_aio_commerce_lib_api = require("@adobe/aio-commerce-lib-api");

//#region source/commerce/lib/schema.ts
/** The schema of the workspace configuration in the Developer Console. */
function workspaceConfigurationSchema(propertyName) {
	return valibot.union([
		valibot.pipe(require_utils.stringValueSchema(propertyName), valibot.empty()),
		valibot.pipe(require_utils.stringValueSchema(propertyName), valibot.parseJson(void 0, `Expected valid JSON string for property '${propertyName}'`), valibot.record(valibot.string(), valibot.unknown()), valibot.stringifyJson()),
		valibot.pipe(valibot.record(valibot.string(), valibot.unknown()), valibot.stringifyJson(void 0, `Expected valid JSON data for property '${propertyName}'`))
	]);
}

//#endregion
//#region source/commerce/api/event-providers/schema.ts
const EventProviderGetByIdParamsSchema = valibot.object({ provider_id: require_utils.stringValueSchema("provider_id") });
const EventProviderCreateParamsSchema = valibot.object({
	provider_id: require_utils.stringValueSchema("provider_id"),
	instance_id: require_utils.stringValueSchema("instance_id"),
	label: valibot.optional(require_utils.stringValueSchema("label")),
	description: valibot.optional(require_utils.stringValueSchema("description")),
	workspace_configuration: valibot.optional(workspaceConfigurationSchema("workspace_configuration"))
});
const EventProviderDeleteParamsSchema = valibot.object({ provider_id: require_utils.stringValueSchema("provider_id") });

//#endregion
//#region source/commerce/api/event-providers/endpoints.ts
var endpoints_exports$1 = /* @__PURE__ */ require_utils.__exportAll({
	createEventProvider: () => createEventProvider,
	deleteEventProvider: () => deleteEventProvider,
	getAllEventProviders: () => getAllEventProviders,
	getEventProviderById: () => getEventProviderById
});
/**
* Lists all event providers of the Commerce instance bound to the given {@link AdobeCommerceHttpClient}.
* @see https://developer.adobe.com/commerce/extensibility/events/api/#get-list-of-all-event-providers
*
* @param httpClient - The {@link AdobeCommerceHttpClient} to use to make the request.
* @param fetchOptions - The {@link Options} to use to make the request.
*
* @throws An {@link HTTPError} If the status code is not 2XX.
*/
async function getAllEventProviders(httpClient, fetchOptions) {
	return httpClient.get("eventing/eventProvider", fetchOptions).json();
}
/**
* Gets the info of the event provider with the given ID of the Commerce instance bound to the given {@link AdobeCommerceHttpClient}.
* @see https://developer.adobe.com/commerce/extensibility/events/api/#get-event-provider-by-id
*
* @param httpClient - The {@link AdobeCommerceHttpClient} to use to make the request.
* @param params - The parameters to get the event provider by.
* @param fetchOptions - The {@link Options} to use to make the request.
*
* @throws A {@link CommerceSdkValidationError} If the parameters are in the wrong format.
* @throws An {@link HTTPError} If the status code is not 2XX.
*/
async function getEventProviderById(httpClient, params, fetchOptions) {
	const validatedParams = require_utils.parseOrThrow(EventProviderGetByIdParamsSchema, params);
	return httpClient.get(`eventing/eventProvider/${validatedParams.provider_id}`, fetchOptions).json();
}
/**
* Creates an event provider in the Commerce instance bound to the given {@link AdobeCommerceHttpClient}.
* @see https://developer.adobe.com/commerce/extensibility/events/api/#create-an-event-provider
*
* @param httpClient - The {@link AdobeCommerceHttpClient} to use to make the request.
* @param params - The parameters to create the event provider with.
* @param fetchOptions - The {@link Options} to use to make the request.
*
* @throws A {@link CommerceSdkValidationError} If the parameters are in the wrong format.
* @throws An {@link HTTPError} If the status code is not 2XX.
*/
async function createEventProvider(httpClient, params, fetchOptions) {
	const eventProvider = require_utils.parseOrThrow(EventProviderCreateParamsSchema, params);
	return httpClient.post("eventing/eventProvider", {
		...fetchOptions,
		json: { eventProvider }
	}).json();
}
/**
* Deletes the event provider with the given ID from the Commerce instance.
* All event subscriptions that use this provider must be deleted first.
* @see https://developer.adobe.com/commerce/extensibility/events/api/#delete-event-provider
*
* @param httpClient - The {@link AdobeCommerceHttpClient} to use to make the request.
* @param params - The parameters to delete the event provider with.
* @param fetchOptions - The {@link Options} to use to make the request.
*
* @throws A {@link CommerceSdkValidationError} If the parameters are in the wrong format.
* @throws An {@link HTTPError} If the status code is not 2XX.
*/
async function deleteEventProvider(httpClient, params, fetchOptions) {
	const validatedParams = require_utils.parseOrThrow(EventProviderDeleteParamsSchema, params);
	return httpClient.delete(`eventing/eventProvider/${validatedParams.provider_id}`, fetchOptions).then((_res) => {});
}

//#endregion
//#region source/commerce/api/event-subscriptions/schema.ts
/**
* Regex for field names according to XSD fieldName pattern.
* Field name can either contain only [a-zA-Z0-9_\-\.\[\]] or be set to *.
*/
const FIELD_NAME_REGEX = /^([a-zA-Z0-9_\-.[\]]+|\*)$/;
/**
* Schema for field names.
* Validates that the field name matches the XSD fieldName pattern:
* can either contain only [a-zA-Z0-9_\-\.\[\]] or be set to *.
*/
function fieldNameSchema(propertyName) {
	return valibot.pipe(require_utils.nonEmptyStringValueSchema(propertyName), valibot.regex(FIELD_NAME_REGEX, "Field name must contain only letters (a-z, A-Z), numbers (0-9), underscores (_), dashes (-), dots (.), and square brackets ([, ]), or be exactly \"*\""));
}
function fieldsSchema(propertyName) {
	return valibot.array(valibot.object({
		name: fieldNameSchema(`${propertyName}[i].name`),
		source: valibot.optional(require_utils.stringValueSchema(`${propertyName}[i].source`))
	}), `Expected an array of objects with a 'name' property for the property "${propertyName}"`);
}
/**
* Schema for rule operator values.
* Valid operators for Commerce event filtering rules.
*/
const OPERATORS = [
	"greaterThan",
	"lessThan",
	"equal",
	"regex",
	"in",
	"onChange"
];
const ruleOperatorSchema = valibot.union(OPERATORS.map((op) => valibot.literal(op)), `Operator must be one of: ${OPERATORS.join(", ")}`);
function rulesSchema(propertyName) {
	return valibot.array(valibot.object({
		field: fieldNameSchema(`${propertyName}[i].field`),
		operator: ruleOperatorSchema,
		value: require_utils.stringValueSchema(`${propertyName}[i].value`)
	}), `Expected an array of objects with 'field', 'operator', and 'value' properties for the property "${propertyName}"`);
}
const EventSubscriptionCreateParamsSchema = valibot.object({
	name: require_utils.stringValueSchema("name"),
	provider_id: valibot.optional(require_utils.stringValueSchema("provider_id")),
	parent: valibot.optional(require_utils.stringValueSchema("parent")),
	fields: fieldsSchema("fields"),
	rules: valibot.optional(rulesSchema("rules")),
	destination: valibot.optional(require_utils.stringValueSchema("destination")),
	hipaa_audit_required: valibot.optional(require_utils.booleanValueSchema("hipaa_audit_required")),
	priority: valibot.optional(require_utils.booleanValueSchema("priority")),
	force: valibot.optional(require_utils.booleanValueSchema("force"))
});
const EventSubscriptionDeleteParamsSchema = valibot.object({ name: require_utils.stringValueSchema("name") });

//#endregion
//#region source/commerce/api/event-subscriptions/endpoints.ts
var endpoints_exports = /* @__PURE__ */ require_utils.__exportAll({
	createEventSubscription: () => createEventSubscription,
	deleteEventSubscription: () => deleteEventSubscription,
	getAllEventSubscriptions: () => getAllEventSubscriptions
});
/**
* Gets all event subscriptions in the Commerce instance bound to the given {@link AdobeCommerceHttpClient}.
* @see https://developer.adobe.com/commerce/extensibility/events/api/#get-a-list-of-all-subscribed-events
*
* @param httpClient - The {@link AdobeCommerceHttpClient} to use to make the request.
* @param fetchOptions - The {@link Options} to use to make the request.
*/
async function getAllEventSubscriptions(httpClient, fetchOptions) {
	return httpClient.get("eventing/getEventSubscriptions", fetchOptions).json();
}
/**
* Creates an event subscription in the Commerce instance bound to the given {@link AdobeCommerceHttpClient}.
* @see https://developer.adobe.com/commerce/extensibility/events/api/#subscribe-to-events
*
* @param httpClient - The {@link AdobeCommerceHttpClient} to use to make the request.
* @param params - The parameters to create the event subscription with.
* @param fetchOptions - The {@link Options} to use to make the request.
*
* @throws A {@link CommerceSdkValidationError} If the parameters are in the wrong format.
* @throws An {@link HTTPError} If the status code is not 2XX.
*/
async function createEventSubscription(httpClient, params, fetchOptions) {
	const { force, ...event } = require_utils.parseOrThrow(EventSubscriptionCreateParamsSchema, params);
	return httpClient.post("eventing/eventSubscribe", {
		...fetchOptions,
		json: {
			force,
			event
		}
	}).json().then((_res) => {});
}
/**
* Unsubscribes from an event in the Commerce instance bound to the given {@link AdobeCommerceHttpClient}.
* @see https://developer.adobe.com/commerce/extensibility/events/api/#unsubscribe-from-events
*
* @param httpClient - The {@link AdobeCommerceHttpClient} to use to make the request.
* @param params - The parameters to delete the event subscription with.
* @param fetchOptions - The {@link Options} to use to make the request.
*
* @throws A {@link CommerceSdkValidationError} If the parameters are in the wrong format.
* @throws An {@link HTTPError} If the status code is not 2XX.
*/
async function deleteEventSubscription(httpClient, params, fetchOptions) {
	const validatedParams = require_utils.parseOrThrow(EventSubscriptionDeleteParamsSchema, params);
	return httpClient.post(`eventing/eventUnsubscribe/${validatedParams.name}`, fetchOptions).then((_res) => {});
}

//#endregion
//#region source/commerce/api/eventing-configuration/schema.ts
/** The schema of the parameters received by the `updateConfiguration` Commerce Eventing API endpoint. */
const UpdateEventingConfigurationParamsSchema = valibot.partial(valibot.object({
	enabled: require_utils.booleanValueSchema("enabled"),
	provider_id: require_utils.alphaNumericOrUnderscoreOrHyphenSchema("provider_id"),
	instance_id: require_utils.alphaNumericOrUnderscoreOrHyphenSchema("instance_id"),
	merchant_id: require_utils.alphaNumericOrUnderscoreSchema("merchant_id"),
	environment_id: require_utils.alphaNumericOrUnderscoreSchema("environment_id"),
	workspace_configuration: workspaceConfigurationSchema("workspace_configuration")
}));

//#endregion
//#region source/commerce/api/eventing-configuration/endpoints.ts
/**
* Updates the configuration of the Commerce Eventing API.
* @see https://developer.adobe.com/commerce/extensibility/events/api/#configure-commerce-eventing
*
* @param httpClient - The {@link AdobeCommerceHttpClient} to use to make the request.
* @param params - The parameters to update the configuration with.
* @param fetchOptions - The {@link Options} to use to make the request.
*
* @throws A {@link CommerceSdkValidationError} If the parameters are in the wrong format.
* @throws An {@link HTTPError} If the status code is not 2XX.
*/
async function updateEventingConfiguration(httpClient, params, fetchOptions) {
	const validatedParams = require_utils.parseOrThrow(UpdateEventingConfigurationParamsSchema, params);
	return httpClient.put("eventing/updateConfiguration", {
		...fetchOptions,
		json: { config: validatedParams }
	}).json();
}

//#endregion
//#region source/commerce/lib/api-client.ts
/**
* Creates a new API client for the Commerce Events API client.
* @param params - The parameters to build the Commerce HTTP client that will communicate with the Commerce Events API.
*/
function createCommerceEventsApiClient(params) {
	return _adobe_aio_commerce_lib_api.ApiClient.create(new _adobe_aio_commerce_lib_api.AdobeCommerceHttpClient(params), {
		...endpoints_exports$1,
		...endpoints_exports,
		updateEventingConfiguration
	});
}
/**
* Creates a customized Commerce Events API client.
* @param params - The parameters to build the Commerce HTTP client that will communicate with the Commerce Events API.
* @param functions - The API functions to include in the client.
*/
function createCustomCommerceEventsApiClient(params, functions) {
	return _adobe_aio_commerce_lib_api.ApiClient.create(new _adobe_aio_commerce_lib_api.AdobeCommerceHttpClient(params), functions);
}

//#endregion
exports.createCommerceEventsApiClient = createCommerceEventsApiClient;
exports.createCustomCommerceEventsApiClient = createCustomCommerceEventsApiClient;
exports.createEventProvider = createEventProvider;
exports.createEventSubscription = createEventSubscription;
exports.deleteEventProvider = deleteEventProvider;
exports.deleteEventSubscription = deleteEventSubscription;
exports.getAllEventProviders = getAllEventProviders;
exports.getAllEventSubscriptions = getAllEventSubscriptions;
exports.getEventProviderById = getEventProviderById;
exports.updateEventingConfiguration = updateEventingConfiguration;
